<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>WhiteboxTools User Manual</title>
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" rel="stylesheet" type="text/css">
        <link href="https://fonts.googleapis.com/css?family=Source+Code+Pro:500" rel="stylesheet" type="text/css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
        

        
    </head>
    <body class="light">
        <!-- Provide site root to javascript -->
        <script type="text/javascript">var path_to_root = "";</script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script type="text/javascript">
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script type="text/javascript">
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { } 
            if (theme === null || theme === undefined) { theme = 'light'; }
            document.body.className = theme;
            document.querySelector('html').className = theme + ' js';
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script type="text/javascript">
            var html = document.querySelector('html');
            var sidebar = 'hidden';
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <ol class="chapter"><li class="affix"><a href="preface.html">Preface</a></li><li><a href="intro.html"><strong aria-hidden="true">1.</strong> Introduction</a></li><li><a href="install.html"><strong aria-hidden="true">2.</strong> Setting Up WhiteboxTools</a></li><li><a href="using_whiteboxtools.html"><strong aria-hidden="true">3.</strong> Using WhiteboxTools</a></li><li><ol class="section"><li><a href="command_prompt.html"><strong aria-hidden="true">3.1.</strong> Command-line interface</a></li><li><a href="python_scripting/scripting.html"><strong aria-hidden="true">3.2.</strong> Interfacing with Python</a></li><li><ol class="section"><li><a href="python_scripting/using_whitebox_tools.html"><strong aria-hidden="true">3.2.1.</strong> Using whitebox_tools.py</a></li><li><a href="python_scripting/tool_output.html"><strong aria-hidden="true">3.2.2.</strong> Handling tool output</a></li><li><a href="python_scripting/additional_functions.html"><strong aria-hidden="true">3.2.3.</strong> Additional functions</a></li><li><a href="python_scripting/example.html"><strong aria-hidden="true">3.2.4.</strong> An example Python project</a></li></ol></li><li><a href="whitebox_tools_runner.html"><strong aria-hidden="true">3.3.</strong> WhiteboxTools Runner</a></li><li><a href="qgis_plugin.html"><strong aria-hidden="true">3.4.</strong> WhiteboxTools QGIS plugin</a></li></ol></li><li><a href="available_tools/index.html"><strong aria-hidden="true">4.</strong> Tools Reference</a></li><li><ol class="section"><li><a href="available_tools/data_tools.html"><strong aria-hidden="true">4.1.</strong> Data tools</a></li><li><a href="available_tools/geomorphometric_analysis.html"><strong aria-hidden="true">4.2.</strong> Geomorphometric analysis</a></li><li><a href="available_tools/gis_analysis.html"><strong aria-hidden="true">4.3.</strong> GIS analysis</a></li><li><ol class="section"><li><a href="available_tools/gis_analysis_distance_tools.html"><strong aria-hidden="true">4.3.1.</strong> Distance tools</a></li><li><a href="available_tools/gis_analysis_overlay_tools.html"><strong aria-hidden="true">4.3.2.</strong> Overlay tools</a></li><li><a href="available_tools/gis_analysis_patch_shape_tools.html"><strong aria-hidden="true">4.3.3.</strong> Patch shape tools</a></li></ol></li><li><a href="available_tools/hydrological_analysis.html"><strong aria-hidden="true">4.4.</strong> Hydrological analysis</a></li><li><a href="available_tools/image_processing_tools.html"><strong aria-hidden="true">4.5.</strong> Image processing tools</a></li><li><ol class="section"><li><a href="available_tools/image_processing_tools_filters.html"><strong aria-hidden="true">4.5.1.</strong> Filters</a></li><li><a href="available_tools/image_processing_tools_image_enhancement.html"><strong aria-hidden="true">4.5.2.</strong> Image enchancement</a></li></ol></li><li><a href="available_tools/lidar_tools.html"><strong aria-hidden="true">4.6.</strong> LiDAR tools</a></li><li><a href="available_tools/mathand_stats_tools.html"><strong aria-hidden="true">4.7.</strong> Mathematical and statistical analysis</a></li><li><a href="available_tools/stream_network_analysis.html"><strong aria-hidden="true">4.8.</strong> Stream network analysis</a></li></ol></li><li><a href="tutorials/index.html"><strong aria-hidden="true">5.</strong> Tutorials</a></li><li><ol class="section"><li><a href="tutorials/lidar.html"><strong aria-hidden="true">5.1.</strong> Processing LiDAR data</a></li></ol></li><li><a href="supported_formats.html"><strong aria-hidden="true">6.</strong> Supported Data Formats</a></li><li><a href="contributing.html"><strong aria-hidden="true">7.</strong> Contributing and Reporting Bugs</a></li><li><a href="limitations.html"><strong aria-hidden="true">8.</strong> Limitations</a></li><li><a href="license.html"><strong aria-hidden="true">9.</strong> License</a></li><li><a href="faq.html"><strong aria-hidden="true">10.</strong> Frequently Asked Questions</a></li><li><a href="tool_index.html"><strong aria-hidden="true">11.</strong> Tool Index</a></li></ol>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                
                <div id="menu-bar" class="menu-bar">
                    <div id="menu-bar-sticky-container">
                        <div class="left-buttons">
                            <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                                <i class="fa fa-bars"></i>
                            </button>
                            <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                                <i class="fa fa-paint-brush"></i>
                            </button>
                            <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                                <li role="none"><button role="menuitem" class="theme" id="light">Light <span class="default">(default)</span></button></li>
                                <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                                <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                                <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                                <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                            </ul>
                            
                            <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                                <i class="fa fa-search"></i>
                            </button>
                            
                        </div>

                        <h1 class="menu-title">WhiteboxTools User Manual</h1> 

                        <div class="right-buttons">
                            <a href="print.html" title="Print this book" aria-label="Print this book">
                                <i id="print-button" class="fa fa-print"></i>
                            </a>
                        </div>
                    </div>
                </div>

                
                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" name="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>
                

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script type="text/javascript">
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <!-- <center>![](img/WhiteboxToolsLogoBlue.png)</center> -->
<center><img src="img/WhiteboxToolsLogoBlue.png" alt="drawing" height="120"/></center>
<!-- ![](img/WhiteboxToolsLogo_vert.svg) -->
<p>WhiteboxTools Version 0.13  <br />
Dr. John B. Lindsay © 2017-2019  <br />
Geomorphometry and Hydrogeomatics Research Group  <br />
University of Guelph  <br />
Guelph, Canada <br />
January 4, 2019
<br />
<br />
<br />
<br />
Sponsored by:</p>
<!-- ![](img/GHRGLogoSm.png =100x20) -->
<img src="img/GHRGLogoSm.png" alt="drawing" height="70"/><!-- ![](img/WhiteboxToolsLogo_vert.svg) -->
<!-- <img src="img/WhiteboxToolsLogo_vert.svg" alt="drawing" height="190"/> -->
<img src="img/WhiteboxToolsLogo_vert.svg" alt="drawing" height="150" align="right"/>
<a class="header" href="#introduction" id="introduction"><h1>Introduction</h1></a>
<p><strong><em>WhiteboxTools</em></strong> is an advanced geospatial data analysis platform created by Prof. <a href="http://www.uoguelph.ca/%7Ehydrogeo/index.html">John Lindsay</a> at the <a href="http://www.uoguelph.ca">University of Guelph's</a> <em>Geomorphometry and Hydrogeomatics Research Group</em> (GHRG). The project began in January 2017 and quickly evolved in terms of its analytical capabilities.</p>
<blockquote>
<p><strong>Project Highlights</strong></p>
<ul>
<li>Contains more than 395 tools for processing various types of geospatial data.</li>
<li>Many tools operate in parallel, taking full advantage of your multi-core processor.</li>
<li>Written in the safe and cross-platform systems programming language Rust and compiled to highly efficient native code.</li>
<li>Small stand-alone application with no external dependencies, making installation as easy as downloading the 8Mb zip file and decompressing it.</li>
<li>Simple yet powerful Python scripting interface that allows users to develop custom scripted workflows.</li>
<li><a href="python_scripting/scripting.html">Embed WhiteboxTools</a> functions into hetergenous scripting environments along with ArcPy, GDAL, and other geoprocessing libraries.</li>
<li>Serves as an analytical back-end for other GIS and remote sensing software (e.g. the <a href="./qgis_plugin.html">QGIS Whitebox for Processing</a> plugin).</li>
<li>Permissive <a href="./license.html">MIT open-source license</a> allows for ready integration with other software.</li>
<li>Transparent software philosopy allows for easy source code inspection and rapid innovation and development.</li>
</ul>
</blockquote>
<a class="header" href="#what-can-whiteboxtools-do" id="what-can-whiteboxtools-do"><h2>What can WhiteboxTools do?</h2></a>
<p><em>WhiteboxTools</em> can be used to perform common geographical information systems (GIS) analysis operations, such as cost-distance analysis, distance buffering, and raster reclassification. Remote sensing and image processing tasks include image enhancement (e.g. panchromatic sharpening, contrast adjustments), image mosaicking, numerous filtering operations, simple classification (k-means clustering), and common image transformations. <em>WhiteboxTools</em> also contains advanced tooling for spatial hydrological analysis (e.g. flow-accumulation, watershed delineation, stream network analysis, sink removal), terrain analysis (e.g. common terrain indices such as slope, curvatures, wetness index, hillshading; hypsometric analysis; multi-scale topographic position analysis), and LiDAR data processing. LiDAR point clouds can be interrogated (LidarInfo, LidarHistogram), segmented, tiled and joined, analyzed for outliers, interpolated to rasters (DEMs, intensity images), and ground-points can be classified or filtered. <em>WhiteboxTools</em> is not a cartographic or spatial data visualization package; instead it is meant to serve as an analytical backend for other data visualization software, mainly GIS.</p>
<a class="header" href="#whiteboxtools-vs-whitebox-geospatial-analysis-tools-gat" id="whiteboxtools-vs-whitebox-geospatial-analysis-tools-gat"><h2>WhiteboxTools vs. Whitebox Geospatial Analysis Tools (GAT)</h2></a>
<p>Although <em>WhiteboxTools</em> was first developed with to serve as a source of plugin tools for the <a href="http://www.uoguelph.ca/%7Ehydrogeo/Whitebox/"><em>Whitebox Geospatial Analysis Tools</em></a> (GAT) open-source GIS project, the tools contained in the library are stand-alone and can run outside of the larger <em>Whitebox GAT</em> project. See <a href="#interacting-with-whiteboxtools-from-the-command-prompt">Interacting With <em>WhiteboxTools</em> From the Command Prompt</a> for further details. There have been a large number of requests to call <em>Whitebox GAT</em> tools and functionality from outside of the <em>Whitebox GAT</em> user-interface (e.g. from Python automation scripts). <em>WhiteboxTools</em> is intended to meet these usage requirements. For example, a <a href="https://plugins.bruy.me/plugins/plugins.xml"><em>WhiteboxTools</em> plug-in for QGIS</a> is available.</p>
<blockquote>
<p>In this manual, <strong><em>WhiteboxTools</em></strong> refers to the standalone geospatial analysis library, a collection of tools contained within a compiled binary executable command-line program and the associated Python scripts that are distributed alongside the binary file (e.g. <em>whitebox_tools.py</em> and <em>wb_runner.py</em>). <strong><em>Whitebox Geospatial Analysis Tools</em></strong> and <strong><em>Whitebox GAT</em></strong> refer to the GIS software, which includes a user-interface (front-end), point-and-click tool interfaces, and cartographic data visualization capabilities. Importantly, <strong><em>WhiteboxTools</em></strong> and <strong><em>Whitebox GAT</em></strong> are related but separate projects.</p>
</blockquote>
<a class="header" href="#why-is-it-named-whiteboxtools" id="why-is-it-named-whiteboxtools"><h2>Why is it named <em>WhiteboxTools</em>?</h2></a>
<p>The project name <em>WhiteboxTools</em> clearly takes it inspiration from the related project Whitebox GAT. However, the name <em>Whitebox</em> is intended to convey opposition to a 'black box' system, one for which only the inputs and outputs may be observed and the internal workings may not be scrutinized. <em>WhiteboxTools</em> is inspired by the concept of <em>open-access software</em>, the tenants of which were described by Lindsay (2014)<sup class="footnote-reference"><a href="#1">1</a></sup>. Open-<em>access</em> software can be viewed as a complementary extension to the traditional open-<em>source</em> software (OSS) model of development. The concept of open access has been previously defined in the context of publishing scholarly literature in a way that removes financial, legal, and technical access barriers to knowledge transfer. Lindsay (2014) argued that the stated goals of reducing barriers associated with knowledge transfer applies equally to the software used in research. Open-access software is distinct from other OSS in that it has an explicitly stated design goal of reducing barriers to the transfer of knowledge to the user community. Direct insight into the workings of algorithm design and implementation allows for educational opportunities and increases the potential for rapid innovation, experimentation with algorithms, and community-directed development. This is particularly important in geomatics because many geospatial algorithms are complex and are strongly affected by implementation details. Also, there are often multiple competing algorithms for accomplishing the same task and the choice of one method over another can greatly impact the outcome of a workflow.</p>
<p>All OSS allow users the opportunity to download the source code and inspect the software’s internal workings. However, traditional OSS often does not lend itself to end-user source code inspection. Open-access software, by comparison, is <em>designed from the project's inception in a way that reduces the barriers that typically discourage end-users from examining the algorithm and implementation details associated with specific software artifacts</em>. WhiteboxTools attempts to address some of the barriers to knowledge transfer by allowing users to view the source code associated with each tool directly (e.g. <code>--viewcode=ExtendVectorLines</code>). This functionality removes the need to download separate, and often large, project source code files and eliminates the requisite familiarity with the project to identify the code sections related to the operation of the tool of interest. The <code>viewcode</code> flag is the embodiment of a design philosophy that is intended to empower the user community. Each tool included in the library has been written in a way to isolate the major functionality within a single file, thereby easing the task of interpreting the code (traditional coding style would split complex code among numerous files). This design goal is also why the developers have chosen to exclude external libraries commonly found in other similarly software (e.g. GDAL), thereby simplifying the installation process and code interpretation. This approach has the potential to encourage further community involvement and experimentation with geospatial analysis techniques.</p>
<div class="footnote-definition" id="1"><sup class="footnote-definition-label">1</sup>
<p>Lindsay, JB. 2014. <a href="https://www.researchgate.net/publication/271205138_The_Whitebox_Geospatial_Analysis_Tools_project_and_open-access_GIS">The Whitebox Geospatial Analysis Tools project and open-access GIS</a>. Proceedings of the GIS Research UK 22nd Annual Conference, The University of Glasgow, 16-18 April, DOI: 10.13140/RG.2.1.1010.8962.</p>
</div>
<a class="header" href="#setting-up-whiteboxtool" id="setting-up-whiteboxtool"><h1>Setting Up <em>WhiteboxTool</em></h1></a>
<p><em>WhiteboxTools</em> is a stand-alone executable command-line program with no actual installation. Simply <a href="http://www.uoguelph.ca/%7Ehydrogeo/WhiteboxTools/index.html">download the appropriate file for your system</a> and decompress the folder. Pre-compiled binaries can be downloaded from the <a href="http://www.uoguelph.ca/%7Ehydrogeo/WhiteboxTools/download.html"><em>Geomorphometry and Hydrogeomatics Research Group</em></a> software web site for various supported operating systems. Depending on your operating system, you may need to grant the <em>WhiteboxTools</em> executable file execution privileges before running it.</p>
<p>If you intend to use WhiteboxTools from the <a href="command_prompt.html">command prompt</a> (i.e. a terminal application), you may wish to add the directory containing the WhiteboxTools executable file to your system <strong>PATH</strong> variable. Doing so will greatly simplify usage of the library from your terminal. Instructions for doing this depend on your operating system and can be found on the Internet.</p>
<p>If you intend to use the <a href="python_scripting/scripting.html">Python programming interface</a> for <em>WhiteboxTools</em> you will need to have <a href="https://www.python.org/downloads/">Python 3</a> installed. Please note that the Python interface will not work correctly with Python 2. If your system has Python 2 as the default Python version, it is possible to install Python 3 alongside your current version. You may need to use the <code>python3</code> command in place of the usual <code>python</code> if this is the case.</p>
<a class="header" href="#building-whiteboxtools-from-source-code" id="building-whiteboxtools-from-source-code"><h2>Building WhiteboxTools From Source Code</h2></a>
<p>Most users rely on the pre-compiled versions of WhiteboxTools and will never need to compile the software from its source code. There are two circumstances under which you may find that you need to build your own binary executable: 1) if you need a binary compiled for a platform other than the project-supported operating systems, and 2) if you require a bleeding-edge feature or bug-fix only available in the development branch. It is likely that <em>WhiteboxTools</em> will work on a wider variety of operating systems and architectures than those of the distributed pre-compiled binaries. If you do not find your operating system/architecture in the list of available <em>WhiteboxTool</em> binaries, then compilation from source code will be necessary. WhiteboxTools can be compiled from the source code with the following steps:</p>
<ol>
<li>
<p>Install the Rust compiler; Rustup is recommended for this purpose. Further instruction can be found at this <a href="https://www.rust-lang.org/en-US/install.html">link</a>.</p>
</li>
<li>
<p>Download the <em>WhiteboxTools</em> <a href="https://github.com/jblindsay/whitebox-tools">source code</a>. To download the code, click the green Clone or download button on the GitHub repository site.</p>
</li>
<li>
<p>Decompress the zipped download file.</p>
</li>
<li>
<p>Open a terminal (command prompt) window and change the working directory to the whitebox_tools sub-folder, which is contained within the decompressed downloaded Whitebox GAT folder:</p>
</li>
</ol>
<pre><code>&gt;&gt; cd /path/to/folder/whitebox_tools/
</code></pre>
<ol start="5">
<li>Finally, use the rust package manager Cargo, which will be installed alongside Rust, to compile the executable:</li>
</ol>
<pre><code>&gt;&gt; cargo build --release
</code></pre>
<p>Depending on your system, the compilation may take several minutes. When completed, the compiled binary executable file will be contained within the <em>whitebox_tools/target/release/ folder</em>. Type <em>./whitebox_tools --help</em> at the command prompt (after changing the directory to the containing folder) for information on how to run the executable from the terminal.</p>
<blockquote>
<p>The '&gt;&gt;' is shorthand used in this document to denote the command prompt and is not intended to be typed.</p>
</blockquote>
<p>Be sure to follow the instructions for installing Rust carefully. In particular, if you are installing on Microsoft Windows, you must have a linker installed prior to installing the Rust compiler (<em>rustc</em>). The Rust webpage recommends either the <strong>MS Visual C++ 2015 Build Tools</strong> or the GNU equivalent and offers details for each installation approach. You should also consider using <strong>RustUp</strong> to install the Rust compiler.</p>
<a class="header" href="#using-whiteboxtools" id="using-whiteboxtools"><h1>Using <em>WhiteboxTools</em></h1></a>
<p>Users interact with the <em>WhiteboxTools</em> platform either from a <a href="./command_prompt.html">command prompt</a> (i.e. terminal), through the <a href="python_scripting/scripting.html">Python scripting interface</a>, or from one of the available graphical user interfaces (GUI) applications.</p>
<a class="header" href="#command-line-interface" id="command-line-interface"><h1>Command-Line Interface</h1></a>
<p><em>WhiteboxTools</em> is a command-line program and can be run either by calling it from a terminal application with appropriate commands and arguments, or, more conveniently, by calling it from a script. The following commands are recognized by the <em>WhiteboxTools</em> library:</p>
<table><thead><tr><th><strong>Command</strong>        </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--cd, --wd       </td><td>  Changes the working directory; used in conjunction with --run flag.</td></tr>
<tr><td>-h, --help        </td><td>  Prints help information.</td></tr>
<tr><td>-l, --license     </td><td>  Prints the whitebox-tools license.</td></tr>
<tr><td>--listtools       </td><td>  Lists all available tools, with tool descriptions. Keywords may also be used, --listtools slope.</td></tr>
<tr><td>-r, --run         </td><td>  Runs a tool; used in conjunction with --cd flag; -r=&quot;LidarInfo&quot;.</td></tr>
<tr><td>--toolbox         </td><td>  Prints the toolbox associated with a tool; --toolbox=Slope.</td></tr>
<tr><td>--toolhelp        </td><td>  Prints the help associated with a tool; --toolhelp=&quot;LidarInfo&quot;.</td></tr>
<tr><td>--toolparameters  </td><td>  Prints the parameters (in json form) for a specific tool; e.g. --toolparameters=&quot;FeaturePreservingDenoise&quot;.</td></tr>
<tr><td>-v                 </td><td>  Verbose mode. Without this flag, tool outputs will not be printed.</td></tr>
<tr><td>--viewcode        </td><td>  Opens the source code of a tool in a web browser; --viewcode=&quot;LidarInfo&quot;.</td></tr>
<tr><td>--version         </td><td>  Prints the version information.</td></tr>
</tbody></table>
<p>Generally, the Unix convention is that single-letter arguments (options) use a single hyphen (e.g. -h) while word-arguments (longer, more descriptive argument names) use double hyphens (e.g. --help). The same rule is used for passing arguments to tools as well. Use the <em>--toolhelp</em> argument to print information about a specific tool (e.g. --toolhelp=Clump).</p>
<blockquote>
<p>Tool names can be specified either using the snake_case or CamelCase convention (e.g. <em>lidar_info</em> or <em>LidarInfo</em>).</p>
</blockquote>
<p>The following is an example of calling the <em>WhiteboxTools</em> binary executable file directly from the command prompt:</p>
<pre><code>&gt;&gt;./whitebox_tools --wd='/Users/johnlindsay/Documents/data/' ^
--run=DevFromMeanElev --input='DEM clipped.tif' ^
--output='DEV raster.tif' -v
</code></pre>
<p>Notice the quotation marks (single or double) used around directories and filenames, and string tool arguments in general. After the <em>--run</em> flag, used to call a tool, a series of tool-specific
flags are provided to indicate the values of various input parameters. Note
that the order of these flags is unimportant. Use the '-v' flag (run in verbose mode) to force the tool to print output to the command prompt. Please note that the whitebox_tools executable file must have permission to be executed; on some systems, this may require setting special permissions. Also, the above example uses the forward slash character (/), the directory path separator used on unix based systems. On Windows, users should use the back slash character (\) instead. Also, it is sometimes necessary to break (^) commands across multiple lines, as above, in order to better fit with the documents format. Actual command prompts (&gt;&gt;) should be contained to a single line.</p>
<a class="header" href="#interfacing-with-python" id="interfacing-with-python"><h1>Interfacing With Python</h1></a>
<p>By combining the <em>WhiteboxTools</em> library with a high-level scripting language, such as Python, users are capable of creating powerful stand-alone geospatial applications and workflow automation scripts. In fact, <em>WhiteboxTools</em> functionality can be called from many different programming languages. However, given the prevalent use of the Python language in the geospatial fields, the library is distributed with several resources specifically aimed at Python scripting. This section focuses on how Python programming can be used to interact with the <em>WhiteboxTools</em> library.</p>
<blockquote>
<p>Note that all of the following material assumes the user system is configured with Python 3. The code snippets below are not guaranteed to work with older versions of the language.</p>
</blockquote>
<a class="header" href="#using-the-whitebox_toolspy-script" id="using-the-whitebox_toolspy-script"><h1>Using the <em>whitebox_tools.py</em> script</h1></a>
<p>Interacting with <em>WhiteboxTools</em> from Python scripts is easy. To begin, each script must start by importing the <em>WhiteboxTools</em> class, contained with the <em>whitebox_tools.py</em> script; a new <code>WhiteboxTools</code> object can then be created:</p>
<pre><code class="language-python">from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
</code></pre>
<p>Depending on the relative location of the <em>WhiteboxTools</em> directory and the script file that you are importing to, the import statement may need to be altered slightly. In the above script, it is assumed that the folder containing the <em>WhiteboxTools</em> files (including the <em>whitebox_tools</em> Python script) is named <code>WBT</code> (Line 1) and that the calling script that is importing WhiteboxTools is located in the parent directory of <code>WBT</code>. See <a href="#an-example-whiteboxtools-python-project">An Example WhiteboxTools Python Project</a> for more details on project set-up. The use of <code>wbt</code> to designate the WhiteboxTools object variable in the above script (Line 3) is just the convention used in this manual and other project resources. In fact, any variable name can be used for this purpose.</p>
<p>The <code>WhiteboxTools</code> class expects to find the <em>WhiteboxTools</em> executable file (<em>whitebox_tools.exe</em> on Windows and <em>whitebox_tools</em> on other platforms) within the same directory (<code>WBT</code>) as the <em>whitebox_tools.py</em> script. If the binary file is located in a separate directory, you will need to set the executable directory as follows:</p>
<pre><code class="language-python">wbt.set_whitebox_dir('/local/path/to/whitebox/binary/')
# Or alternatively...
wbt.exe_path = '/local/path/to/whitebox/binary/'
</code></pre>
<p>Individual tools can be called using the convenience methods provided in the <code>WhiteboxTools</code> class:</p>
<pre><code class="language-python"># This line performs a 5 x 5 mean filter on 'inFile.tif':
wbt.mean_filter('/file/path/inFile.tif', '/file/path/outFile.tif', 5, 5)
</code></pre>
<p>Each tool has a cooresponding convenience method. The listing of tools in this manual includes information about each tool's Python convienience method, including default parameter values. Parameters with default values may be optionally left off of function calls. In addition to the convenience methods, tools can be called using the <code>run_tool()</code> method, specifying the tool name and a list of tool arguments.</p>
<pre><code class="language-python">source = &quot;source.tif&quot;
cost = &quot;cost.tif&quot;
out_accum = &quot;accum.tif&quot;
out_backlink = &quot;backlink.tif&quot;
args = []
args.append(&quot;--source='{}'&quot;.format(source))
args.append(&quot;--cost='{}'&quot;.format(cost))
args.append(&quot;--out_accum='{}'&quot;.format(out_accum))
args.append(&quot;--out_backlink='{}'&quot;.format(out_backlink))
self.run_tool('cost_distance', args)
</code></pre>
<p>Each of the tool-specific convenience methods collect their parameters into a properly formated list and then ultimately call the <code>run_tools()</code> method. Notice that while internally <em>whitebox_tools.exe</em> uses CamelCase (e.g. MeanFilter) to denote tool names, the Python interface of <em>whitebox_tools.py</em> uses snake_case (e.g. mean_filter), according to Python style conventions. The only exceptions are tools with names that clash with Python keywords (e.g. <code>And()</code>, <code>Not()</code>, and <code>Or()</code>).</p>
<p>The return value can be used to check for errors during operation:</p>
<pre><code class="language-python">if wbt.ruggedness_index('/path/DEM.tif', '/path/ruggedness.tif') != 0:
    # Non-zero returns indicate an error.
    print('ERROR running ruggedness_index')
</code></pre>
<p>If your data files tend to be burried deeply in layers of sub-directories, specifying complete file names as input parameters can be tedius. In this case, the best option is setting the working directory before calling tools:</p>
<pre><code class="language-python">from whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
wbt.work_dir = &quot;/path/to/data/&quot; # Sets the Whitebox working directory

# Because the working directory has been set, file arguments can be
# specified simply using file names, without paths.
wbt.d_inf_flow_accumulation(&quot;DEM.tif&quot;, &quot;output.tif&quot;, log=True)
</code></pre>
<p>An advanced text editor, such as VS Code or Atom, can provide hints and autocompletion for available tool convenience methods and their parameters, including default values (Figure 1).</p>
<p><img src="../img/wbt_auotcomplete.png" alt="Autocompletion in Atom text editor makes calling WhiteboxTools functions easier." /></p>
<p>Sometimes it can be useful to print a complete list of available tools:</p>
<pre><code class="language-python">print(wbt.list_tools()) # List all tools in WhiteboxTools
</code></pre>
<p>The <code>list_tools()</code> method also takes an optional keywords list to search for tools:</p>
<pre><code class="language-python"># Lists tools with 'lidar' or 'LAS' in tool name or description.
print(wbt.list_tools(['lidar', 'LAS']))
</code></pre>
<p>To retrieve more detailed information for a specific tool, use the <code>tool_help()</code> method:</p>
<pre><code class="language-python">print(wbt.tool_help(&quot;elev_percentile&quot;))
</code></pre>
<p><code>tool_help()</code> prints tool details including a description, tool parameters (and their flags), and example usage at the command line prompt. The above statement prints this report:</p>
<pre><code>
ElevPercentile
Description:
Calculates the elevation percentile raster from a DEM.
Toolbox: Geomorphometric Analysis
Parameters:

Flag               Description
-----------------  -----------
-i, --input, --dem Input raster DEM file.
-o, --output       Output raster file.
--filterx          Size of the filter kernel in the x-direction.
--filtery          Size of the filter kernel in the y-direction.
--sig_digits       Number of significant digits.

Example usage:
&gt;&gt;./whitebox_tools -r=ElevPercentile -v --wd=&quot;/path/to/data/&quot; --dem=DEM.tif
&gt;&gt;-o=output.tif --filterx=25


</code></pre>
<blockquote>
<p><strong>A note on default parameter values</strong></p>
<p>Each tool contains one or more parameters with default values. These will always be listed after any input parameters that do not have default values. You do not need to specify a parameter with a default value if you accept the default. That is, unless you intend to specify an input value different from the default, you may leave these parameters off of the function call. However, be mindful of the fact that Python assigns values to parameters based on order, unless parameter names are specified.</p>
<p>Consider the Hillshade tool as an example. The User Manual gives the following function definition for the tool:</p>
<pre><code class="language-{.python}">hillshade(
dem,
output,
azimuth=315.0,
altitude=30.0,
zfactor=1.0,
callback=default_callback)
</code></pre>
<p>The <code>dem</code> and <code>output</code> parameters do not have default values and must be specified every time you call this function. Each of the remaining parameters have default values and can, optionally, be left off of calls to the <code>hillshade</code> function. As an example, say I want to accept the default values for all the parameters except <code>altitude</code>. I would then need to use the named-parameter form of the function call:</p>
<pre><code class="language-python">wbt.hillshade(
&quot;DEM.tif&quot;,
&quot;hillshade.tif&quot;,
altitude=20.0)
</code></pre>
<p>If I hadn't specified the parameter name for <code>altitude</code>, Python would have assumed that the value 20.0 should be assigned to the third parameter, <code>azimuth</code>.</p>
</blockquote>
<a class="header" href="#handling-tool-output" id="handling-tool-output"><h1>Handling tool output</h1></a>
<p>Tools will frequently print text to the standard output during their execution, including warnings, progress updates and other notifications. Sometimes, when users run many tools in complex workflows and in batch mode, these output messages can be undesirable. Most tools will have their outputs suppressed by setting the <em>verbose</em> mode to <em>False</em> as follows:</p>
<pre><code class="language-python">wbt.set_verbose_mode(False)
# Or, alternatively...
wbt.verbose = False
</code></pre>
<p>Alternatively, it may be helpful to capture the text output of a tool for custom processing. This is achieved by specifying a custom <em>callback</em> function to the tool's convenience function:</p>
<pre><code class="language-python"># This callback function suppresses printing progress updates,
# which always use the '%' character. The callback function
# approach is flexible and allows for any level of complex
# interaction with tool outputs.
def my_callback(value):
    if not &quot;%&quot; in value:
        print(value)

wbt.slope('DEM.tif', 'slope_raster.tif', callback=my_callback)
</code></pre>
<p>Every convienience function takes an optional callback as the last parameter. The default callback simply prints tool outputs to the standard output without any additional processing. The default callback itself can be overridden, instead of having to set callbacks in convienience functions individually:</p>
<pre><code class="language-python">wbt.default_callback = my_callback
</code></pre>
<p>Callback functions can serve as a means of cancelling operations:</p>
<pre><code class="language-python">def my_callback(value):
    if user_selected_cancel_btn: # Assumes a 'Cancel' button on a GUI
        print('Cancelling operation...')
        wbt.cancel_op = True
    else:
        print(value)

wbt.breach_depressions('DEM.tif', 'DEM_breached.tif', callback=my_callback)
</code></pre>
<a class="header" href="#additional-functions-in-whitebox_toolspy" id="additional-functions-in-whitebox_toolspy"><h1>Additional functions in <em>whitebox_tools.py</em></h1></a>
<p>The <em>whitebox_tools.py</em> script provides several other functions for interacting with the <em>WhiteboxTools</em> library, including:</p>
<pre><code class="language-python"># Print the WhiteboxTools help...a listing of available commands
print(wbt.help())

# Print the WhiteboxTools license
print(wbt.license())

# Print the WhiteboxTools version
print(&quot;Version information: {}&quot;.format(wbt.version()))

# Get the toolbox associated with a tool
tb = wbt.toolbox('lidar_info')

# Retrieve a JSON object of a tool's parameters.
tp = tool_parameters('raster_histogram')

# Opens a browser and navigates to a tool's source code in the
# WhiteboxTools GitHub repository
wbt.view_code('watershed')
</code></pre>
<p>For a working example of how to call functions and run tools from Python, see the <em>whitebox_example.py</em> Python script, which is distributed with the <em>WhiteboxTools</em> library.</p>
<blockquote>
<p>Additional resources for using <em>WhiteboxTools</em>' Python interface can be found on the <a href="https://www.uoguelph.ca/%7Ehydrogeo/WhiteboxTools/tutorials.html">Tutorials</a> site of the <em>WhiteboxTools</em> home page. This site contains in-depth tutorials on topics such as, '<em>Interpolating LiDAR data</em>'.</p>
</blockquote>
<a class="header" href="#an-example-python-project" id="an-example-python-project"><h1>An Example Python Project</h1></a>
<p>In this section, we will create a Python project that utilizes the <em>WhiteboxTools</em> library to interpolate a LiDAR point-cloud, to process the resulting digital elevation model (DEM) to make it suitable for hydrological applications, and to perform a simple flow-accumulation operation. I suggest using an advanced coding text editor, such as <em>Visual Studio Code</em> or <em>Atom</em>, for this tutorial, but Python code can be written using any basic text editor.</p>
<p>Begin by creating a dedicated project directory called <em>FlowAccumExample</em> and copy <em>WhiteboxTools</em> binary file (i.e. the compressed file downloaded from the <em>Geomorphometry &amp; Hydrogeomatics Research Group</em> website) into this folder. Using the decompression software on your computer, decompress (i.e. an operation sometimes called <em>unzipping</em>) the file into the newly created <em>FlowAccumExample</em> directory. You will find the compressed file contains a folder with contents similar to the following:</p>
<!-- ![Folder contents of *WhiteboxTools* compressed download file](../img/wbt_directory.png) -->
<img src="../img/wbt_directory.png" alt="Folder contents of *WhiteboxTools* compressed download file" width="450" align="center"/>
<p>The folder contains a number of files, including the <em>WhiteboxTools</em> executable file, the <em>whitebox_tools.py</em> python script, the WhiteboxTools Runner (<em>wb_runner.py</em>; see below), and this user manual. It is likely that the folder has a name that reflects the operating system and architecture that the binary file was compiled for (e.g. WhiteboxTools_darwin_amd64). Rename this directory to <em>WBT</em>. Also note, depending on your decompression software, it may be the case that the contents of the <em>WBT</em> folder itself contains a sub-directory that actually holds these files. If this is the case, be sure to move the contents of the sub-directory into the <em>WBT</em> parent directory.</p>
<p>Using your text editor, create a new Python script file, called <em>FlowAccumulation.py</em> within the <em>FlowAccumExample</em> directory. We will begin by importing the <em>WhiteboxTools</em> class from the <em>whitebox_tools.py</em> script contained within the <em>WBT</em> sub-directory. Unfortunately, Python's module system is only able to import classes and function definitions declared in external Python scripts <em>if these external files are contained somewhere on the Python path or in the directory containing the script file into which you are importing</em>. This is important because based on the project structure that we have established, the <em>whitebox_tools.py</em> script is actually contained within a sub-directory of the <em>FlowAccumExample</em> directory and is therefore not directly accessible, unless you have previously installed the script on the Python path. Another, perhaps easier solution to this problem is to create a file named <em>__init__.py</em> (those are two leading and trailing underscore characters) within the <em>FlowAccumExample</em> directory. The presence of this empty file will make Python treat the <em>WBT</em> directory as containing packages, in this case, the <em>whitebox_tools</em> package. For more information, see the Python documentation on <a href="https://docs.python.org/3/tutorial/modules.html#packages">modules and packages</a>.</p>
<p>At this stage, you should have a project directory structure like the following:</p>
<!-- ![Example project set-up](../img/example_project_dir.png) -->
<img src="../img/example_project_dir.png" alt="Example project set-up" width="580" align="center"/>
<p>Many operating systems will disallow the execution of files that are downloaded directly from the Internet. As such, it is possible that you will need to explicitly give the <em>whitebox_tools.exe</em> permission to execute on your computer (Note: here we are referring to the compiled <em>WhiteboxTools</em> binary file and not the similarly named Python script <em>whitebox_tools.py</em> also contained in the folder). The procedure for doing this depends on your specific operating system. On MacOS, for example, this is usually achieved using the 'Security &amp; Privacy' tab under 'System Preferences'. To test whether <em>whitebox_tools.exe</em> has permission to run on your system, double-click the file. If the file is configured to execute, a command terminal will automatically open and the <em>WhiteboxTools</em> help documentation and a listing of the available tools will be printed. If this does not occur, you likely need to give the file permission to execute.</p>
<p>Using your text editor, you may now add the following lines to the <em>FlowAccumulation.py</em> file.</p>
<pre><code class="language-python">from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
</code></pre>
<p>In the import statement, <code>WBT</code> is a reference to the package folder containing the <em>WhiteboxTools</em> files; <code>whitebox_tools</code> is a reference to the <em>whitebox_tools.py</em> script contained with this package folder; and <code>WhiteboxTools</code> is a reference to the <em>WhiteboxTools</em> class contained within this script file. Please note that if you named your directory containing the <em>WhiteboxTools</em> files something other than <em>WBT</em>, you would need to alter the import statement accordingly.</p>
<p>Visit the <a href="http://www.uoguelph.ca/%7Ehydrogeo/software.shtml#WhiteboxTools"><em>Geomorphometry and Hydrogeomatics Research Group</em></a> website and download the St. Elis Mountains and Gulf of Alaska sample data set (<em>StElisAk.las</em>) from the <em>WhiteboxTools</em> section of the site. This file contains a LiDAR point cloud that has been previously filtered to remove points associated with non-ground returns, mainly trees (Figure 4). Create a sub-directory within the project folder called 'data' and copy <em>StElisAk.las</em> into the folder.</p>
<p><img src="../img/StElis_point_cloud.png" alt="St. Elis Mountains LiDAR point cloud, visualized using the plas.io software" /></p>
<p>Now we can complete our flow accumulation analysis with the following code:</p>
<pre><code class="language-python">import os
from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()

# Set the working directory, i.e. the folder containing the data,
# to the 'data' sub-directory.
wbt.work_dir = os.path.dirname(os.path.abspath(__file__)) + &quot;/data/&quot;

# When you're running mulitple tools, the outputs can be a tad
# chatty. In this case, you may want to suppress the output by
# setting the verbose mode to False.
# wbt.verbose = False

# Interpolate the LiDAR data using an inverse-distance weighting
# (IDW) scheme.
print(&quot;Interpolating DEM...&quot;)
wbt.lidar_idw_interpolation(
i=&quot;StElisAk.las&quot;,
output=&quot;raw_dem.tif&quot;,
parameter=&quot;elevation&quot;,
returns=&quot;last&quot;,
resolution=1.0,
weight=1.0,
radius=2.5
)

# The resulting DEM will contain NoData gaps. We need to fill
# these in by interpolating across the gap.
print(&quot;Filling missing data...&quot;)
wbt.fill_missing_data(
i=&quot;raw_dem.tif&quot;,
output=&quot;dem_nodata_filled.tif&quot;,
filter=11
)

# This DEM will contain grid cells that have no lower neighbours.
# This condition is unsuited for flow-path modelling applications
# because these operations assume that each interior cell in the
# DEM has at least one downslope neighour. We'll use an operation
# called depression breaching to 'fix' the elevations within the
# DEM to enforce continuous flow.
print(&quot;Performing flow enforcement...&quot;)
wbt.breach_depressions(
dem=&quot;dem_nodata_filled.tif&quot;,
output=&quot;dem_hydro_enforced.tif&quot;
)

# Lastly, perform the flow accumulation operation using the
# D-infinity flow algorithm.
print(&quot;Performing flow accumulation...&quot;)
wbt.d_inf_flow_accumulation(
dem=&quot;dem_hydro_enforced.tif&quot;,
output=&quot;flow_accum.tif&quot;,
log=True
)

print(&quot;Complete!&quot;)
</code></pre>
<p>To run the above script, open a terminal (command prompt), <em>cd</em> to the script containing folder, and run the following command:</p>
<pre><code>&gt;&gt;python FlowAccumulation.py
</code></pre>
<p>If Python 3 is not your default Python version, substitute <code>python3</code> for <code>python</code> in the above command line. The final D-infinity flow accumulation raster can be displayed in any GIS software of choice and should look similar to Figure 5.</p>
<p><img src="../img/flow_accum.png" alt="Output of the flow accumulation script for the St. Elis Mountains data set." /></p>
<a class="header" href="#whiteboxtools-runner" id="whiteboxtools-runner"><h1><em>WhiteboxTools Runner</em></h1></a>
<p>There is a Python script contained within the <em>WhiteboxTools</em> directory called '<em>wb_runner.py</em>'. This script is intended to provide a very basic user-interface, <em>WhiteboxTools Runner</em>, for running the tools contained within the <em>WhiteboxTools</em> library. The user-interface uses Python's TkInter GUI library and is cross-platform. The user interface is currently experimental and is under heavy testing. Please report any issues that you experience in using it.</p>
<p><img src="img/WBRunner.png" alt="The WhiteboxTools Runner user-interface" /></p>
<p>The <em>WhiteboxTools Runner</em> does not rely on the <em>Whitebox GAT</em> user interface at all and can therefore be used indepedent of the larger project. The script must be run from a directory that also contains the '<em>whitebox_tools.py</em>' Python script and the '<em>whitebox_tools</em>' executable file. There are plans to link tool help documentation in <em>WhiteboxTools Runner</em> and to incorporate toolbox information, rather than one large listing of available tools.</p>
<p>To launch the runner application, open a terminal and issue issue the following commands:</p>
<pre><code>cd /path/to/whitebox_runner/script/
python3 wb_runner.py
</code></pre>
<a class="header" href="#whiteboxtools-qgis-plugin" id="whiteboxtools-qgis-plugin"><h1>WhiteboxTools QGIS Plugin</h1></a>
<p><em>WhiteboxTools</em> functionality can also be accessed conveniently through the popular open-source geospatial software QGIS. QGIS developer <a href="https://wiki.osgeo.org/wiki/User:Alexbruy">Alexander Bruy</a> maintains a plugin for the toolbox called <a href="https://plugins.bruy.me/processing-whitebox.html">Whitebox For Processing</a>. Detailed instructions for installing and setting-up the QGIS toolbox can be found on the site.</p>
<p><img src="img/QGIS_plugin.png" alt="The WhiteboxTools QGIS plugin toolbox" /></p>
<blockquote>
<p>Note that the QGIS plugin does not come with a copy of the <em>WhiteboxTools</em> executable and so you will still need to download <em>WhiteboxTools</em> from the Geomorphometry and Hydrogeomatics Research Group website prior to running the plugin.</p>
</blockquote>
<a class="header" href="#tools-reference" id="tools-reference"><h1>Tools Reference</h1></a>
<p>The <em>WhiteboxTools</em> library currently contains approximately 400 tools, which are organized into themed toolboxes, including:</p>
<ul>
<li><a href="./data_tools.html">Data Tools</a></li>
<li><a href="./geomorphometric_analysis.html">Geomorphometric Analysis</a></li>
<li><a href="./gis_analysis.html">GIS Analysis</a></li>
<li><a href="./hydrological_analysis.html">Hydrological Analysis</a></li>
<li><a href="./image_processing_tools.html">Image Analysis</a></li>
<li><a href="./lidar_tools.html">LiDAR Analysis</a></li>
<li><a href="./mathand_stats_tools.html">Mathematical and Statistical Analysis</a></li>
<li><a href="./stream_network_analysis.html">Stream Network Analysis </a></li>
</ul>
<p>To retrieve detailed information about a tool's input arguments and example usage, either use the <code>toolhelp</code> command from the terminal, or the <code>wbt.tool_help('tool_name')</code> function from the <em>whitebox_tools.py</em> script. The following is a complete listing of available tools, with brief descriptions, tool parameter, and example usage.</p>
<p>The <a href="../tool_index.html">Tool Index</a> located at the end of the user manual contains a complete alphabetical listing of the available tools.</p>
<a class="header" href="#data-tools" id="data-tools"><h1>Data Tools</h1></a>
<ul>
<li><a href="#AddPointCoordinatesToTable">AddPointCoordinatesToTable</a></li>
<li><a href="#ConvertNodataToZero">ConvertNodataToZero</a></li>
<li><a href="#ConvertRasterFormat">ConvertRasterFormat</a></li>
<li><a href="#ExportTableToCsv">ExportTableToCsv</a></li>
<li><a href="#JoinTables">JoinTables</a></li>
<li><a href="#LinesToPolygons">LinesToPolygons</a></li>
<li><a href="#MergeTableWithCsv">MergeTableWithCsv</a></li>
<li><a href="#MergeVectors">MergeVectors</a></li>
<li><a href="#MultiPartToSinglePart">MultiPartToSinglePart</a></li>
<li><a href="#NewRasterFromBase">NewRasterFromBase</a></li>
<li><a href="#PolygonsToLines">PolygonsToLines</a></li>
<li><a href="#PrintGeoTiffTags">PrintGeoTiffTags</a></li>
<li><a href="#RasterToVectorLines">RasterToVectorLines</a></li>
<li><a href="#RasterToVectorPoints">RasterToVectorPoints</a></li>
<li><a href="#ReinitializeAttributeTable">ReinitializeAttributeTable</a></li>
<li><a href="#RemovePolygonHoles">RemovePolygonHoles</a></li>
<li><a href="#SetNodataValue">SetNodataValue</a></li>
<li><a href="#SinglePartToMultiPart">SinglePartToMultiPart</a></li>
<li><a href="#VectorLinesToRaster">VectorLinesToRaster</a></li>
<li><a href="#VectorPointsToRaster">VectorPointsToRaster</a></li>
<li><a href="#VectorPolygonsToRaster">VectorPolygonsToRaster</a></li>
</ul>
<p><a name="AddPointCoordinatesToTable"></a></p>
<a class="header" href="#addpointcoordinatestotable" id="addpointcoordinatestotable"><h1>AddPointCoordinatesToTable</h1></a>
<p>This tool modifies the attribute table of a vector of POINT ShapeType by adding two fields,
XCOORD and YCOORD, containing each point's X and Y coordinates respectively.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector Points file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">add_point_coordinates_to_table(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AddPointCoordinatesToTable -v ^
--wd=&quot;/path/to/data/&quot; --input=points.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/add_point_coordinates_to_table.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/09/2018</p>
<p><a name="ConvertNodataToZero"></a></p>
<a class="header" href="#convertnodatatozero" id="convertnodatatozero"><h1>ConvertNodataToZero</h1></a>
<p>Converts nodata values in a raster to zero.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">convert_nodata_to_zero(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ConvertNodataToZero -v ^
--wd=&quot;/path/to/data/&quot; --input=in.tif -o=NewRaster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/convert_nodata_to_zero.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="ConvertRasterFormat"></a></p>
<a class="header" href="#convertrasterformat" id="convertrasterformat"><h1>ConvertRasterFormat</h1></a>
<p>This tool converts raster data from one format to another. It determines input/output raster
formats based on extensions, but due to file extension naming collisions, it would be good to
add user hints. For example, the extension 'grd' could belong to a SurferAscii or a Surfer7Binary.
This is more important for distinguishing output files since input files can be read and
distiguishing features idenfitied from the file structure. At the moment, this tool does not
support user hints however.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">convert_raster_format(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ConvertRasterFormat -v ^
--wd=&quot;/path/to/data/&quot; --input=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/convert_raster_format.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="ExportTableToCsv"></a></p>
<a class="header" href="#exporttabletocsv" id="exporttabletocsv"><h1>ExportTableToCsv</h1></a>
<p>This tool can be used to export a vector's attribute table to a comma separated values (CSV)
file. CSV files stores tabular data (numbers and text) in plain-text form such that each row
corresponds to a record and each column to a field. Fields are typically separated by commas
within records. The user must specify the name of the vector (and associated attribute file),
the name of the output CSV file, and whether or not to include the field names as a header
column in the output CSV file.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#mergetablewithcsv"><strong>MergeTableWithCsv</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--headers          </td><td> Export field names as file header?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">export_table_to_csv(
    i, 
    output, 
    headers=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ExportTableToCsv -v ^
--wd=&quot;/path/to/data/&quot; -i=lines.shp -o=output.csv --headers 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/export_table_to_csv.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 24/04/2018</p>
<p><a name="JoinTables"></a></p>
<a class="header" href="#jointables" id="jointables"><h1>JoinTables</h1></a>
<p>This tool can be used to join (i.e. merge) a vector's attribute table with a second table. The
user must specify the name of the vector file (and associated attribute file) as well as the
<em>primary key</em> within the table. The <em>primary key</em> (<code>--pkey</code> flag) is the field
within the table that is being appended to that serves as the identifier. Additionally, the user
must specify the name of a second vector from which the data appended into the first table will be
derived. The <em>foreign key</em> (<code>--fkey</code> flag), the identifying field within the
second table that corresponds with the data contained within the primary key in the table, must be
specified. Both the primary and foreign keys should either be strings (text) or integer values.
<em>Fields containing decimal values are not good candidates for keys.</em> Lastly, the names of the field
within the second file to include in the merge operation can also be input (<code>--import_field</code>). If the
<code>--import_field</code> field is not input, all fields in the attribute table of the second file, that are not
the foreign key nor FID, will be imported to the first table.</p>
<p>Merging works for one-to-one and many-to-one database relations. A <em>one-to-one</em> relations exists when
each record in the attribute table corresponds to one record in the second table and each primary
key is unique. Since each record in the attribute table is associated with a geospatial feature in
the vector, an example of a one-to-one relation may be where the second file contains AREA and
PERIMETER fields for each polygon feature in the vector. This is the most basic type of relation.
A many-to-one relation would exist when each record in the first attribute table corresponds to one
record in the second file and the primary key is NOT unique. Consider as an example a vector and
attribute table associated with a world map of countries. Each country has one or more more polygon
features in the shapefile, e.g. Canada has its mainland and many hundred large islands. You may want
to append a table containing data about the population and area of each country. In this case, the
COUNTRY columns in the attribute table and the second file serve as the primary and foreign keys
respectively. While there may be many duplicate primary keys (all of those Canadian polygons) each
will correspond to only one foreign key containing the population and area data. This is a
<em>many-to-one</em> relation. The <a href="./data_tools.html#jointables"><strong>JoinTables</strong></a> tool does not support one-to-many nor many-to-many relations.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#mergetablewithcsv"><strong>MergeTableWithCsv</strong></a>, <a href="./data_tools.html#reinitializeattributetable"><strong>ReinitializeAttributeTable</strong></a>, <a href="./data_tools.html#exporttabletocsv"><strong>ExportTableToCsv</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input primary vector file (i.e. the table to be modified)</td></tr>
<tr><td>--pkey             </td><td> Primary key field</td></tr>
<tr><td>--i2, --input2    </td><td> Input foreign vector file (i.e. source of data to be imported)</td></tr>
<tr><td>--fkey             </td><td> Foreign key field</td></tr>
<tr><td>--import_field     </td><td> Imported field (all fields will be imported if not specified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">join_tables(
    input1, 
    pkey, 
    input2, 
    fkey, 
    import_field, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=JoinTables -v --wd=&quot;/path/to/data/&quot; ^
--i1=properties.shp --pkey=TYPE --i2=land_class.shp ^
--fkey=VALUE --import_field=NEW_VALUE 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/join_tables.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Prof. John Lindsay</p>
<p><em>Created</em>: 07/10/2018</p>
<p><a name="LinesToPolygons"></a></p>
<a class="header" href="#linestopolygons" id="linestopolygons"><h1>LinesToPolygons</h1></a>
<p>This tool converts vector polylines into polygons. Note that this tool will close polygons
that are open and will ensure that the first part of an input line is interpreted
as the polygon hull and subsequent parts are considered holes. The tool does not
examine input lines for line crossings (self intersections), which are topological
errors.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#polygonstolines"><strong>PolygonsToLines</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector line file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lines_to_polygons(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LinesToPolygons -v ^
--wd=&quot;/path/to/data/&quot; -i=input.shp -o=output.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/lines_to_polygons.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/09/2018</p>
<p><a name="MergeTableWithCsv"></a></p>
<a class="header" href="#mergetablewithcsv" id="mergetablewithcsv"><h1>MergeTableWithCsv</h1></a>
<p>This tool can be used to merge a vector's attribute table with data contained within a comma
separated values (CSV) text file. CSV files stores tabular data (numbers and text) in plain-text
form such that each row is a record and each column a field. Fields are typically separated by
commas although the tool will also support seimi-colon, tab, and space delimited files. The user
must specify the name of the vector (and associated attribute file) as well as the <em>primary key</em>
within the table. The <em>primary key</em> (<code>--pkey</code> flag) is the field within the
table that is being appended to that serves as the unique identifier. Additionally, the user must
specify the name of a CSV text file with either a *.csv or *.txt extension. The file must possess a
header row, i.e. the first row must contain information about the names of the various fields. The
<em>foreign key</em> (<code>--fkey</code> flag), that is the identifying field within the
CSV file that corresponds with the data contained within the <em>primary key</em> in the table, must also
be specified. Both the primary and foreign keys should either be strings (text) or integer values.
<em>Fields containing decimal values are not good candidates for keys.</em> Lastly, the user may optionally
specify the name of a field within the CSV file to import in the merge operation (<code>--import_field</code> flag).
If this flag is not specified, all of the fields within the CSV, with the exception of the foreign
key, will be appended to the attribute table.</p>
<p>Merging works for one-to-one and many-to-one database relations. A <em>one-to-one</em> relations exists when
each record in the attribute table corresponds to one record in the second table and each primary
key is unique. Since each record in the attribute table is associated with a geospatial feature in
the vector, an example of a one-to-one relation may be where the second file contains AREA and
PERIMETER fields for each polygon feature in the vector. This is the most basic type of relation.
A many-to-one relation would exist when each record in the first attribute table corresponds to one
record in the second file and the primary key is NOT unique. Consider as an example a vector and
attribute table associated with a world map of countries. Each country has one or more more polygon
features in the shapefile, e.g. Canada has its mainland and many hundred large islands. You may want
to append a table containing data about the population and area of each country. In this case, the
COUNTRY columns in the attribute table and the second file serve as the primary and foreign keys
respectively. While there may be many duplicate primary keys (all of those Canadian polygons) each
will correspond to only one foreign key containing the population and area data. This is a
<em>many-to-one</em> relation. The <a href="./data_tools.html#jointables"><strong>JoinTables</strong></a> tool does not support one-to-many nor many-to-many relations.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#jointables"><strong>JoinTables</strong></a>, <a href="./data_tools.html#reinitializeattributetable"><strong>ReinitializeAttributeTable</strong></a>, <a href="./data_tools.html#exporttabletocsv"><strong>ExportTableToCsv</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input primary vector file (i.e. the table to be modified)</td></tr>
<tr><td>--pkey             </td><td> Primary key field</td></tr>
<tr><td>--csv              </td><td> Input CSV file (i.e. source of data to be imported)</td></tr>
<tr><td>--fkey             </td><td> Foreign key field</td></tr>
<tr><td>--import_field     </td><td> Imported field (all fields will be imported if not specified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">merge_table_with_csv(
    i, 
    pkey, 
    csv, 
    fkey, 
    import_field=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MergeTableWithCsv -v ^
--wd=&quot;/path/to/data/&quot; -i=properties.shp --pkey=TYPE ^
--csv=land_class.csv --fkey=VALUE ^
--import_field=NEW_VALUE
&gt;&gt;./whitebox_tools -r=MergeTableWithCsv ^
-v --wd=&quot;/path/to/data/&quot; -i=properties.shp --pkey=TYPE ^
--csv=land_class.csv --fkey=VALUE 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/merge_table_with_csv.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Prof. John Lindsay</p>
<p><em>Created</em>: 11/10/2018</p>
<p><a name="MergeVectors"></a></p>
<a class="header" href="#mergevectors" id="mergevectors"><h1>MergeVectors</h1></a>
<p>Combines two or more input vectors of the same ShapeType creating a single, new output
vector. Importantly, the attribute table of the output vector will contain the ubiquitous
file-specific FID, the parent file name, the parent FID, and the list of attribute fields
that are shared among each of the input files. For a field to be considered common
between tables, it must have the same <code>name</code> and <code>field_type</code> (i.e. data type and
precision).</p>
<p>Overlapping features will not be identified nor handled in the merging. If you have
significant areas of overlap, it is advisable to use one of the vector overlay tools
instead.</p>
<p>The difference between <a href="./data_tools.html#mergevectors"><strong>MergeVectors</strong></a> and the <code>Append</code> tool is that merging takes two
or more files and creates one new file containing the features of all inputs, and
<code>Append</code> places the features of a single vector into another existing (appended) vector.</p>
<p>This tool only operates on vector files. Use the <a href="./image_processing_tools.html#mosaic"><strong>Mosaic</strong></a> tool to combine raster data.</p>
<p><em>See Also</em>:</p>
<p><code>Append</code>, <a href="./image_processing_tools.html#mosaic"><strong>Mosaic</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input vector files</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">merge_vectors(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MergeVectors -v --wd=&quot;/path/to/data/&quot; ^
-i='polys1.shp;polys2.shp;polys3.shp' -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/merge_vectors.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 01/10/2018</p>
<p><a name="MultiPartToSinglePart"></a></p>
<a class="header" href="#multiparttosinglepart" id="multiparttosinglepart"><h1>MultiPartToSinglePart</h1></a>
<p>This tool can be used to convert a vector file containing multi-part features into a vector
containing only single-part features. Any multi-part polygons or lines within the input
vector file will be split into seperate features in the output file, each possessing their
own entry in the associated attribute file. For polygon-type vectors, the user may optionally
choose to exclude hole-parts from being separated from their containing polygons. That is,
with the <code>--exclude_holes</code> flag, hole parts in the input vector will continue to belong to
their enclosing polygon in the output vector.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#singleparttomultipart"><strong>SinglePartToMultiPart</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector line or polygon file</td></tr>
<tr><td>-o, --output       </td><td> Output vector line or polygon file</td></tr>
<tr><td>--exclude_holes    </td><td> Exclude hole parts from the feature splitting? (holes will continue to belong to their features in output.)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">multi_part_to_single_part(
    i, 
    output, 
    exclude_holes=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MultiPartToSinglePart -v ^
--wd=&quot;/path/to/data/&quot; -i=input.shp -o=output.shp ^
--exclude_holes 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/multipart_to_singlepart.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/09/2018</p>
<p><a name="NewRasterFromBase"></a></p>
<a class="header" href="#newrasterfrombase" id="newrasterfrombase"><h1>NewRasterFromBase</h1></a>
<p>This tool can be used to create a new raster with the same coordinates and dimensions
(i.e. rows and columns) as an existing base image. The user must specify the name of the
base image (<code>--base</code>), the value that the new grid will be filled with (<code>--value</code> flag;
default of NoData), and the data type (<code>--data_type</code> flag; options include 'double',
'float', and 'integer'). Notice that the functionality of this tool is the same as
multiplying the base image by zero and adding the constant value.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#rastercellassignment"><strong>RasterCellAssignment</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --base         </td><td> Input base raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--value            </td><td> Constant value to fill raster with; either 'nodata' or numeric value</td></tr>
<tr><td>--data_type        </td><td> Output raster data type; options include 'double' (64-bit), 'float' (32-bit), and 'integer' (signed 16-bit) (default is 'float')</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">new_raster_from_base(
    base, 
    output, 
    value=&quot;nodata&quot;, 
    data_type=&quot;float&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NewRasterFromBase -v ^
--wd=&quot;/path/to/data/&quot; --base=base.tif -o=NewRaster.tif ^
--value=0.0 --data_type=integer
&gt;&gt;./whitebox_tools ^
-r=NewRasterFromBase -v --wd=&quot;/path/to/data/&quot; --base=base.tif ^
-o=NewRaster.tif --value=nodata 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/new_raster.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="PolygonsToLines"></a></p>
<a class="header" href="#polygonstolines" id="polygonstolines"><h1>PolygonsToLines</h1></a>
<p>This tool converts vector polygons into polylines, simply by modifying the Shapefile geometry type.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#linestopolygons"><strong>LinesToPolygons</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
<tr><td>-o, --output       </td><td> Output vector lines file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">polygons_to_lines(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PolygonsToLines -v ^
--wd=&quot;/path/to/data/&quot; -i=input.shp -o=output.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/polygons_to_lines.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 04/09/2018</p>
<p><a name="PrintGeoTiffTags"></a></p>
<a class="header" href="#printgeotifftags" id="printgeotifftags"><h1>PrintGeoTiffTags</h1></a>
<p>This tool can be used to view the tags contained within a GeoTiff file. Viewing
the tags of a GeoTiff file can be useful when trying to import the GeoTiff to
different software environments. The user must specify the name of a GeoTiff file
and the tag information will be output to the StdOut output stream (e.g. console).
Note that tags that contain greater than 100 values will be truncated in the output.
GeoKeys will also be interpreted as per the GeoTIFF specification.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input GeoTIFF file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">print_geo_tiff_tags(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PrintGeoTiffTags -v ^
--wd=&quot;/path/to/data/&quot; --input=DEM.tiff 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/print_geotiff_tags.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: March 2, 2018</p>
<p><a name="RasterToVectorLines"></a></p>
<a class="header" href="#rastertovectorlines" id="rastertovectorlines"><h1>RasterToVectorLines</h1></a>
<p>This tool converts raster lines features into a vector of the POLYLINE ShapeType.
Grid cells associated with line features will contain non-zero, non-NoData cell
values. The algorithm requires three passes of the raster. The first pass counts
the number of line neighbours of each line cell; the second pass traces line
segments starting from line ends (i.e. line cells with only one neighbouring line
cell); lastly, the final pass traces any remaining line segments, which are likely
forming closed loops (and therefore do not have line ends).</p>
<p>If the line raster contains streams, it is preferable to use the <a href="./stream_network_analysis.html#rasterstreamstovector"><strong>RasterStreamsToVector</strong></a>
instead. This tool will use knowledge of flow directions to ensure connections
between stream segments at confluence sites, whereas <a href="./data_tools.html#rastertovectorlines"><strong>RasterToVectorLines</strong></a> will not.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#rastertovectorpoints"><strong>RasterToVectorPoints</strong></a>, <a href="./stream_network_analysis.html#rasterstreamstovector"><strong>RasterStreamsToVector</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster lines file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">raster_to_vector_lines(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RasterToVectorLines -v ^
--wd=&quot;/path/to/data/&quot; -i=lines.tif -o=lines.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/raster_to_vector_lines.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 09/10/2018</p>
<p><a name="RasterToVectorPoints"></a></p>
<a class="header" href="#rastertovectorpoints" id="rastertovectorpoints"><h1>RasterToVectorPoints</h1></a>
<p>Converts a raster dataset to a vector of the POINT shapetype. The user must specify
the name of a raster file and the name of the output vector. Points will correspond
with grid cell centre points. All grid cells containing non-zero, non-NoData values
will be considered a point. The vector's attribute table will contain a field called
'VALUE' that will contain the cell value for each point feature.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output vector points file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">raster_to_vector_points(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RasterToVectorPoints -v ^
--wd=&quot;/path/to/data/&quot; --input=points.tif -o=out.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/raster_to_vector_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/09/2018</p>
<p><a name="ReinitializeAttributeTable"></a></p>
<a class="header" href="#reinitializeattributetable" id="reinitializeattributetable"><h1>ReinitializeAttributeTable</h1></a>
<p>Reinitializes a vector's attribute table deleting all fields but the feature ID (FID).
Caution: this tool overwrites the input file's attribute table.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">reinitialize_attribute_table(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ReinitializeAttributeTable -v ^
--wd=&quot;/path/to/data/&quot; -i=input.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/reinitialize_attribute_table.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 04/09/2018</p>
<p><a name="RemovePolygonHoles"></a></p>
<a class="header" href="#removepolygonholes" id="removepolygonholes"><h1>RemovePolygonHoles</h1></a>
<p>This tool can be used to remove holes from the features within a vector
polygon file. The user must specify the name of the input vector file,
which must be of a polygon shapetype, and the name of the output file.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">remove_polygon_holes(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RemovePolygonHoles -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp ^
--output=no_holes.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/remove_polygon_holes.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 26/09/2018</p>
<p><a name="SetNodataValue"></a></p>
<a class="header" href="#setnodatavalue" id="setnodatavalue"><h1>SetNodataValue</h1></a>
<p>This tool will re-assign a user-defined background value in an input raster image the <strong>NoData</strong> value.
More precisely, the NoData value will be changed to the specified background value and any existing
grid cells containing the previous NoData value, if it had been defined, will be changed to this new
value. Most WhiteboxTools tools recognize NoData grid cells and treat them specially. NoData grid cells
are also often displayed transparently by GIS software. The user must specify the names of the input and
output rasters and the background value. The default background value is zero, although any numeric
value is possible.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--back_value       </td><td> Background value to set to nodata</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">set_nodata_value(
    i, 
    output, 
    back_value=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SetNodataValue -v --wd=&quot;/path/to/data/&quot; ^
-i=in.tif -o=newRaster.tif --back_value=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/set_nodata_value.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 10/09/2017</p>
<p><a name="SinglePartToMultiPart"></a></p>
<a class="header" href="#singleparttomultipart" id="singleparttomultipart"><h1>SinglePartToMultiPart</h1></a>
<p>This tool can be used to convert a vector file containing single-part features into a vector
containing multi-part features. The user has the option to either group features based on an
ID Field (<code>--field</code> flag), which is a categorical field within the vector's attribute table.
The ID Field should either be of String (text) or Integer type. Fields containing decimal values
are not good candidates for the ID Field. <strong>If no <code>--field</code> flag is specified, all features will
be grouped together into one large multi-part vector</strong>.</p>
<p>This tool works for vectors containing either point, line, or polygon features.
Since vectors of a POINT ShapeType cannot represent multi-part features, the ShapeType of the
output file will be modified to a MULTIPOINT ShapeType if the input file is of a POINT ShapeType.
If the input vector is of a POLYGON ShapeType, the user can optionally set the algorithm to search
for polygons that should be represented as hole parts. In the case of grouping based on an ID Field,
hole parts are polygon features contained within larger polygons of the same ID Field value. Please
note that searching for polygon holes may significantly increase processing time for larger polygon
coverages.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#multiparttosinglepart"><strong>MultiPartToSinglePart</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector line or polygon file</td></tr>
<tr><td>--field            </td><td> Grouping ID field name in attribute table</td></tr>
<tr><td>-o, --output       </td><td> Output vector line or polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">single_part_to_multi_part(
    i, 
    output, 
    field=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SinglePartToMultiPart -v ^
--wd=&quot;/path/to/data/&quot; -i=input.shp -o=output.shp ^
--field='COUNTRY' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/singlepart_to_multipart.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/09/2018</p>
<p><a name="VectorLinesToRaster"></a></p>
<a class="header" href="#vectorlinestoraster" id="vectorlinestoraster"><h1>VectorLinesToRaster</h1></a>
<p>Converts a vector containing polylines into a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector lines file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--nodata           </td><td> Background value to set to NoData. Without this flag, it will be set to 0.0</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
<tr><td>--base             </td><td> Optionally specified input base raster file. Not used when a cell size is specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">vector_lines_to_raster(
    i, 
    output, 
    field=&quot;FID&quot;, 
    nodata=True, 
    cell_size=None, 
    base=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=VectorLinesToRaster -v ^
--wd=&quot;/path/to/data/&quot; -i=lines.shp --field=ELEV -o=output.tif ^
--nodata --cell_size=10.0
        &gt;&gt;./whitebox_tools ^
-r=VectorLinesToRaster -v --wd=&quot;/path/to/data/&quot; -i=lines.shp ^
--field=FID -o=output.tif --base=existing_raster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/vector_lines_to_raster.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 18/04/2018</p>
<p><a name="VectorPointsToRaster"></a></p>
<a class="header" href="#vectorpointstoraster" id="vectorpointstoraster"><h1>VectorPointsToRaster</h1></a>
<p>This tool can be used to convert a vector points file into a raster grid. The user must
specify the name of the input vector and the output raster file. The field name (<code>--field</code>)
is the field from the attributes table from which the tool will retrieve the information to
assign to grid cells in the output raster. The field must contain numerical data. If the user does not
supply a field name parameter, each feature in the raster will be assigned the record number
of the feature. The assignment operation determines how the situation of multiple points
contained within the same grid cell is handled. The background value is zero by default
but can be set to <code>NoData</code> optionally using the <code>--nodata</code> value.</p>
<p>If the user optionally specifies the grid cell size parameter (<code>--cell_size</code>) then the coordinates
will be determined by the input vector (i.e. the bounding box) and the specified cell size. This
will also determine the number of rows and columns in the output raster. If the user instead
specifies the optional base raster file parameter (<code>--base</code>), the output raster's coordinates (i.e.
north, south, east, west) and row and column count will be the same as the base file.</p>
<p>In the case that multiple points are contained within a single grid cell, the output can be
assigned (<code>--assign</code>) the first, last (default), min, max, or sum of the contained points.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#vectorpolygonstoraster"><strong>VectorPolygonsToRaster</strong></a>, <a href="./data_tools.html#vectorlinestoraster"><strong>VectorLinesToRaster</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector Points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--assign           </td><td> Assignment operation, where multiple points are in the same grid cell; options include 'first', 'last' (default), 'min', 'max', 'sum'</td></tr>
<tr><td>--nodata           </td><td> Background value to set to NoData. Without this flag, it will be set to 0.0</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
<tr><td>--base             </td><td> Optionally specified input base raster file. Not used when a cell size is specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">vector_points_to_raster(
    i, 
    output, 
    field=&quot;FID&quot;, 
    assign=&quot;last&quot;, 
    nodata=True, 
    cell_size=None, 
    base=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=VectorPointsToRaster -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --field=ELEV -o=output.tif ^
--assign=min --nodata ^
--cell_size=10.0
        &gt;&gt;./whitebox_tools ^
-r=VectorPointsToRaster -v --wd=&quot;/path/to/data/&quot; -i=points.shp ^
--field=FID -o=output.tif --assign=last ^
--base=existing_raster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/vector_points_to_raster.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 19/04/2018</p>
<p><a name="VectorPolygonsToRaster"></a></p>
<a class="header" href="#vectorpolygonstoraster" id="vectorpolygonstoraster"><h1>VectorPolygonsToRaster</h1></a>
<p>Converts a vector containing polygons into a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygons file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--nodata           </td><td> Background value to set to NoData. Without this flag, it will be set to 0.0</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
<tr><td>--base             </td><td> Optionally specified input base raster file. Not used when a cell size is specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">vector_polygons_to_raster(
    i, 
    output, 
    field=&quot;FID&quot;, 
    nodata=True, 
    cell_size=None, 
    base=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=VectorPolygonsToRaster -v ^
--wd=&quot;/path/to/data/&quot; -i=lakes.shp --field=ELEV -o=output.tif ^
--nodata --cell_size=10.0
        &gt;&gt;./whitebox_tools ^
-r=VectorPolygonsToRaster -v --wd=&quot;/path/to/data/&quot; ^
-i=lakes.shp --field=ELEV -o=output.tif ^
--base=existing_raster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/data_tools/vector_polygons_to_raster.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 17/04/2018</p>
<a class="header" href="#geomorphometric-analysis" id="geomorphometric-analysis"><h1>Geomorphometric Analysis</h1></a>
<ul>
<li><a href="#Aspect">Aspect</a></li>
<li><a href="#DevFromMeanElev">DevFromMeanElev</a></li>
<li><a href="#DiffFromMeanElev">DiffFromMeanElev</a></li>
<li><a href="#DirectionalRelief">DirectionalRelief</a></li>
<li><a href="#DownslopeIndex">DownslopeIndex</a></li>
<li><a href="#DrainagePreservingSmoothing">DrainagePreservingSmoothing</a></li>
<li><a href="#ElevAbovePit">ElevAbovePit</a></li>
<li><a href="#ElevPercentile">ElevPercentile</a></li>
<li><a href="#ElevRelativeToMinMax">ElevRelativeToMinMax</a></li>
<li><a href="#ElevRelativeToWatershedMinMax">ElevRelativeToWatershedMinMax</a></li>
<li><a href="#FeaturePreservingDenoise">FeaturePreservingDenoise</a></li>
<li><a href="#FetchAnalysis">FetchAnalysis</a></li>
<li><a href="#FillMissingData">FillMissingData</a></li>
<li><a href="#FindRidges">FindRidges</a></li>
<li><a href="#Hillshade">Hillshade</a></li>
<li><a href="#HorizonAngle">HorizonAngle</a></li>
<li><a href="#HypsometricAnalysis">HypsometricAnalysis</a></li>
<li><a href="#MaxAnisotropyDev">MaxAnisotropyDev</a></li>
<li><a href="#MaxAnisotropyDevSignature">MaxAnisotropyDevSignature</a></li>
<li><a href="#MaxBranchLength">MaxBranchLength</a></li>
<li><a href="#MaxDifferenceFromMean">MaxDifferenceFromMean</a></li>
<li><a href="#MaxDownslopeElevChange">MaxDownslopeElevChange</a></li>
<li><a href="#MaxElevDevSignature">MaxElevDevSignature</a></li>
<li><a href="#MaxElevationDeviation">MaxElevationDeviation</a></li>
<li><a href="#MinDownslopeElevChange">MinDownslopeElevChange</a></li>
<li><a href="#MultiscaleRoughness">MultiscaleRoughness</a></li>
<li><a href="#MultiscaleRoughnessSignature">MultiscaleRoughnessSignature</a></li>
<li><a href="#MultiscaleTopographicPositionImage">MultiscaleTopographicPositionImage</a></li>
<li><a href="#NumDownslopeNeighbours">NumDownslopeNeighbours</a></li>
<li><a href="#NumUpslopeNeighbours">NumUpslopeNeighbours</a></li>
<li><a href="#PennockLandformClass">PennockLandformClass</a></li>
<li><a href="#PercentElevRange">PercentElevRange</a></li>
<li><a href="#PlanCurvature">PlanCurvature</a></li>
<li><a href="#Profile">Profile</a></li>
<li><a href="#ProfileCurvature">ProfileCurvature</a></li>
<li><a href="#RelativeAspect">RelativeAspect</a></li>
<li><a href="#RelativeStreamPowerIndex">RelativeStreamPowerIndex</a></li>
<li><a href="#RelativeTopographicPosition">RelativeTopographicPosition</a></li>
<li><a href="#RemoveOffTerrainObjects">RemoveOffTerrainObjects</a></li>
<li><a href="#RuggednessIndex">RuggednessIndex</a></li>
<li><a href="#SedimentTransportIndex">SedimentTransportIndex</a></li>
<li><a href="#Slope">Slope</a></li>
<li><a href="#SlopeVsElevationPlot">SlopeVsElevationPlot</a></li>
<li><a href="#StandardDeviationOfSlope">StandardDeviationOfSlope</a></li>
<li><a href="#TangentialCurvature">TangentialCurvature</a></li>
<li><a href="#TotalCurvature">TotalCurvature</a></li>
<li><a href="#Viewshed">Viewshed</a></li>
<li><a href="#VisibilityIndex">VisibilityIndex</a></li>
<li><a href="#WetnessIndex">WetnessIndex</a></li>
</ul>
<p><a name="Aspect"></a></p>
<a class="header" href="#aspect" id="aspect"><h1>Aspect</h1></a>
<p>Calculates an aspect raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">aspect(
    dem, 
    output, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Aspect -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/aspect.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="DevFromMeanElev"></a></p>
<a class="header" href="#devfrommeanelev" id="devfrommeanelev"><h1>DevFromMeanElev</h1></a>
<p>This tool can be used to calculate the difference between the elevation of each grid
cell and the mean elevation of the centering local neighbourhood, normalized by standard
deviation. Therefore, this index of topographic residual is essentially equivalent to a
local z-score. This attribute measures the relative topographic position as a fraction of
local relief, and so is normalized to the local surface roughness.</p>
<p>The user must specify the name (<code>--dem</code>) of the input digital elevation model (DEM), the
name of the output file (<code>--output</code>), and the size of the neighbourhood in the <em>x</em> and <em>y</em>
directions (<code>--filterx</code> and <code>--filtery</code>), measured in grid size.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input, --dem</td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">dev_from_mean_elev(
    dem, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DevFromMeanElev -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/dev_from_mean_elev.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 21, 2017</p>
<p><a name="DiffFromMeanElev"></a></p>
<a class="header" href="#difffrommeanelev" id="difffrommeanelev"><h1>DiffFromMeanElev</h1></a>
<p>Calculates difference from mean elevation (equivalent to a high-pass filter).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input, --dem</td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">diff_from_mean_elev(
    dem, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DiffFromMeanElev -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/diff_from_mean_elev.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="DirectionalRelief"></a></p>
<a class="header" href="#directionalrelief" id="directionalrelief"><h1>DirectionalRelief</h1></a>
<p>Calculates relief for cells in an input DEM for a specified direction.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--azimuth          </td><td> Wind azimuth in degrees</td></tr>
<tr><td>--max_dist         </td><td> Optional maximum search distance (unspecified if none; in xy units)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">directional_relief(
    dem, 
    output, 
    azimuth=0.0, 
    max_dist=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DirectionalRelief -v ^
--wd=&quot;/path/to/data/&quot; -i='input.tif' -o=output.tif ^
--azimuth=315.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/directional_relief.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 7, 2017</p>
<p><a name="DownslopeIndex"></a></p>
<a class="header" href="#downslopeindex" id="downslopeindex"><h1>DownslopeIndex</h1></a>
<p>This tool can be used to calculate the downslope index described by Hjerdt et al. (2004).
The downslope index is a measure of the slope gradient between a grid cell and some
downslope location (along the flowpath passing through the upslope grid cell) that
represents a specified vertical drop (i.e. a potential head drop). The index has been
shown to be useful for hydrological, geomorphological, and biogeochemical applications.</p>
<p>The user must specify the name of a digital elevaton model (DEM) raster. This DEM
should be have been pre-processed to remove artifact topographic depressions and flat
areas. The user must also specify the head potential drop (d), and the output type. The
output type can be either '<code>tangent</code>', '<code>degrees</code>', '<code>radians</code>', or '<code>distance</code>'. If
'<code>distance</code>' is selected as the output type, the output grid actually represents the
downslope flowpath length required to drop d meters from each grid cell. Linear
interpolation is used when the specified drop value is encountered between two adjacent
grid cells along a flowpath traverse.</p>
<p>Notice that this algorithm is affected by edge contamination. That is, for some grid cells,
the edge of the grid will be encountered along a flowpath traverse before the specified
vertical drop occurs. In these cases, the value of the downslope index is approximated by
replacing d with the actual elevation drop observed along the flowpath. To avoid this problem,
the entire watershed containing an area of interest should be contained in the DEM.</p>
<p>Grid cells containing NoData values in any of the input images are assigned the NoData
value in the output raster. The output raster is of the float data type and continuous
data scale.</p>
<p><em>Reference</em>:</p>
<p>Hjerdt, K.N., McDonnell, J.J., Seibert, J. Rodhe, A. (2004) <em>A new topographic index to
quantify downslope controls on local drainage</em>, <strong>Water Resources Research</strong>, 40, W05602,
doi:10.1029/2004WR003130.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--drop             </td><td> Vertical drop value (default is 2.0)</td></tr>
<tr><td>--out_type         </td><td> Output type, options include 'tangent', 'degrees', 'radians', 'distance' (default is 'tangent')</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">downslope_index(
    dem, 
    output, 
    drop=2.0, 
    out_type=&quot;tangent&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DownslopeIndex -v --wd=&quot;/path/to/data/&quot; ^
--dem=pointer.tif -o=dsi.tif --drop=5.0 --out_type=distance 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/downslope_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 17, 2017</p>
<p><a name="DrainagePreservingSmoothing"></a></p>
<a class="header" href="#drainagepreservingsmoothing" id="drainagepreservingsmoothing"><h1>DrainagePreservingSmoothing</h1></a>
<p>This tool implements a modified form of the algorithm described by
Sun, Rosin, Martin, and Langbein (2007) '<em>Fast and effective feature-preserving
mesh denoising</em>'. This implimentation varies the threshold angle between
neighbouring grid cell normal vectors, used during the smoothing operation. The
threshold is varied as a function of how low-lying a site is. This varying
smoothing level better preserves small drainage features, such as ditches,
rills, gullies, etc., which would otherwise be smoothed over.</p>
<p><em>See Also</em>:</p>
<p><a href="./geomorphometric_analysis.html#featurepreservingdenoise"><strong>FeaturePreservingDenoise</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filter           </td><td> Size of the filter kernel</td></tr>
<tr><td>--norm_diff        </td><td> Maximum difference in normal vectors, in degrees</td></tr>
<tr><td>--num_iter         </td><td> Number of iterations</td></tr>
<tr><td>--max_diff         </td><td> Maximum allowable absolute elevation change (optional)</td></tr>
<tr><td>--reduction        </td><td> Maximum Amount to reduce the threshold angle by (0 = full smoothing; 100 = no smoothing)</td></tr>
<tr><td>--dfm              </td><td> Difference from median threshold (in z-units), determines when a location is low-lying</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">drainage_preserving_smoothing(
    dem, 
    output, 
    filter=11, 
    norm_diff=15.0, 
    num_iter=10, 
    max_diff=2.0, 
    reduction=80.0, 
    dfm=0.15, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DrainagePreservingSmoothing -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif --filter=15 ^
--norm_diff=20.0 --num_iter=4 --reduction=90.0 --dfm=0.15 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/drainage_preserving_smoothing.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay and Anthony Francioni</p>
<p><em>Created</em>: 06/09/2018</p>
<p><a name="ElevAbovePit"></a></p>
<a class="header" href="#elevabovepit" id="elevabovepit"><h1>ElevAbovePit</h1></a>
<p>Calculate the elevation of each grid cell above the nearest downstream pit cell or grid edge cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">elev_above_pit(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ElevAbovePit -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/elev_above_pit.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="ElevPercentile"></a></p>
<a class="header" href="#elevpercentile" id="elevpercentile"><h1>ElevPercentile</h1></a>
<p>Calculates the elevation percentile raster from a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input, --dem</td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>--sig_digits       </td><td> Number of significant digits</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">elev_percentile(
    dem, 
    output, 
    filterx=11, 
    filtery=11, 
    sig_digits=2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ElevPercentile -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/elev_percentile.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="ElevRelativeToMinMax"></a></p>
<a class="header" href="#elevrelativetominmax" id="elevrelativetominmax"><h1>ElevRelativeToMinMax</h1></a>
<p>Calculates the elevation of a location relative to the minimum and maximum elevations in a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">elev_relative_to_min_max(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ElevRelativeToMinMax -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/elev_relative_to_min_max.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 12, 2017</p>
<p><a name="ElevRelativeToWatershedMinMax"></a></p>
<a class="header" href="#elevrelativetowatershedminmax" id="elevrelativetowatershedminmax"><h1>ElevRelativeToWatershedMinMax</h1></a>
<p>Calculates the elevation of a location relative to the minimum and maximum elevations in a watershed.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--watersheds       </td><td> Input raster watersheds file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">elev_relative_to_watershed_min_max(
    dem, 
    watersheds, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ElevRelativeToWatershedMinMax -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --watersheds=watershed.tif ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/elev_relative_to_watershed_min_max.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 12, 2017</p>
<p><a name="FeaturePreservingDenoise"></a></p>
<a class="header" href="#featurepreservingdenoise" id="featurepreservingdenoise"><h1>FeaturePreservingDenoise</h1></a>
<p>This tool implements a highly modified form of the DEM de-noising algorithm described
by Sun et al. (2007). It is very effective at removing surface roughness from digital
elevation models (DEMs), without significantly altering breaks-in-slope. As such,
this tool should be used for smoothing DEMs rather than either smoothing with
low-pass filters (e.g. mean, median, Gaussian filters) or grid size coarsening
by resampling. The algorithm works by 1) calculating the surface normal 3D vector
of each grid cell in the DEM, 2) smoothing the normal vector field using a
filtering scheme that applies more weight to neighbours with lower angular difference
in surface normal vectors, and 3) uses the smoothed normal vector field to update
the elevations in the input DEM.</p>
<p>Sun et al.'s (2007) original method was intended to work on input point clouds and
fitted triangular irregular networks (TINs). The algorithm has been modified to
work with input raster DEMs instead. In so doing, this algorithm calculates surface
normal vectors from the planes fitted to 3 x 3 neighbourhoods surrounding each
grid cell, rather than the triangular facet. The normal vector field smoothing and
elevation updating procedures are also based on raster filtering operations. These
modifications make this tool more efficient than Sun's original method, but will
also result in a slightly different output than what would be achieved with Sun's
method.</p>
<p><em>Reference</em>:</p>
<p>Sun, Rosin, Martin, and Langbein (2007) Fast and effective feature-preserving
mesh denoising.</p>
<p><em>See Also</em>:</p>
<p><a href="./geomorphometric_analysis.html#drainagepreservingsmoothing"><strong>DrainagePreservingSmoothing</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filter           </td><td> Size of the filter kernel</td></tr>
<tr><td>--norm_diff        </td><td> Maximum difference in normal vectors, in degrees</td></tr>
<tr><td>--num_iter         </td><td> Number of iterations</td></tr>
<tr><td>--max_diff         </td><td> Maximum allowable absolute elevation change (optional)</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">feature_preserving_denoise(
    dem, 
    output, 
    filter=11, 
    norm_diff=15.0, 
    num_iter=10, 
    max_diff=2.0, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FeaturePreservingDenoise -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif --filter=15 ^
--norm_diff=20.0 --num_iter=4 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/feature_preserving_denoise.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: November 23, 2017</p>
<p><a name="FetchAnalysis"></a></p>
<a class="header" href="#fetchanalysis" id="fetchanalysis"><h1>FetchAnalysis</h1></a>
<p>Performs an analysis of fetch or upwind distance to an obstacle.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--azimuth          </td><td> Wind azimuth in degrees in degrees</td></tr>
<tr><td>--hgt_inc          </td><td> Height increment value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fetch_analysis(
    dem, 
    output, 
    azimuth=0.0, 
    hgt_inc=0.05, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FetchAnalysis -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif --azimuth=315.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/fetch_analysis.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 7, 2017</p>
<p><a name="FillMissingData"></a></p>
<a class="header" href="#fillmissingdata" id="fillmissingdata"><h1>FillMissingData</h1></a>
<p>Fills nodata holes in a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filter           </td><td> Filter size (cells)</td></tr>
<tr><td>--weight           </td><td> IDW weight value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fill_missing_data(
    i, 
    output, 
    filter=11, 
    weight=2.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FillMissingData -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif --filter=25 ^
--weight=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/fill_missing_data.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 14, 2017</p>
<p><a name="FindRidges"></a></p>
<a class="header" href="#findridges" id="findridges"><h1>FindRidges</h1></a>
<p>Identifies potential ridge and peak grid cells.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--line_thin        </td><td> Optional flag indicating whether post-processing line-thinning should be performed</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">find_ridges(
    dem, 
    output, 
    line_thin=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FindRidges -v --wd=&quot;/path/to/data/&quot; ^
--dem=pointer.tif -o=out.tif --line_thin 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/find_ridges.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 04/12/2017</p>
<p><a name="Hillshade"></a></p>
<a class="header" href="#hillshade" id="hillshade"><h1>Hillshade</h1></a>
<p>Calculates a hillshade raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--azimuth          </td><td> Illumination source azimuth in degrees</td></tr>
<tr><td>--altitude         </td><td> Illumination source altitude in degrees</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">hillshade(
    dem, 
    output, 
    azimuth=315.0, 
    altitude=30.0, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Hillshade -v --wd=&quot;/path/to/data/&quot; ^
-i=DEM.tif -o=output.tif --azimuth=315.0 --altitude=30.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/hillshade.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="HorizonAngle"></a></p>
<a class="header" href="#horizonangle" id="horizonangle"><h1>HorizonAngle</h1></a>
<p>Calculates horizon angle (maximum upwind slope) for each grid cell in an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--azimuth          </td><td> Wind azimuth in degrees</td></tr>
<tr><td>--max_dist         </td><td> Optional maximum search distance (unspecified if none; in xy units)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">horizon_angle(
    dem, 
    output, 
    azimuth=0.0, 
    max_dist=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HorizonAngle -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif --azimuth=315.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/horizon_angle.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 7, 2017</p>
<p><a name="HypsometricAnalysis"></a></p>
<a class="header" href="#hypsometricanalysis" id="hypsometricanalysis"><h1>HypsometricAnalysis</h1></a>
<p>Calculates a hypsometric curve for one or more DEMs.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input DEM files</td></tr>
<tr><td>--watershed        </td><td> Input watershed files (optional)</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">hypsometric_analysis(
    inputs, 
    output, 
    watershed=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HypsometricAnalysis -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;DEM1.tif;DEM2.tif&quot; ^
--watershed=&quot;ws1.tif;ws2.tif&quot; -o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/hypsometric_analysis.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: January 30, 2018</p>
<p><a name="MaxAnisotropyDev"></a></p>
<a class="header" href="#maxanisotropydev" id="maxanisotropydev"><h1>MaxAnisotropyDev</h1></a>
<p>Calculates the maximum anisotropy (directionality) in elevation deviation over a range of spatial scales.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--out_mag          </td><td> Output raster DEVmax magnitude file</td></tr>
<tr><td>--out_scale        </td><td> Output raster DEVmax scale file</td></tr>
<tr><td>--min_scale        </td><td> Minimum search neighbourhood radius in grid cells</td></tr>
<tr><td>--max_scale        </td><td> Maximum search neighbourhood radius in grid cells</td></tr>
<tr><td>--step             </td><td> Step size as any positive non-zero integer</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_anisotropy_dev(
    dem, 
    out_mag, 
    out_scale, 
    max_scale, 
    min_scale=3, 
    step=2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxAnisotropyDev -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --out_mag=DEVmax_mag.tif ^
--out_scale=DEVmax_scale.tif --min_scale=1 --max_scale=1000 ^
--step=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/max_anisotropy_dev.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dan Newman and John Lindsay</p>
<p><em>Created</em>: January 26, 2018</p>
<p><a name="MaxAnisotropyDevSignature"></a></p>
<a class="header" href="#maxanisotropydevsignature" id="maxanisotropydevsignature"><h1>MaxAnisotropyDevSignature</h1></a>
<p>Calculates the anisotropy in deviation from mean for points over a range of spatial scales.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--points           </td><td> Input vector points file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
<tr><td>--min_scale        </td><td> Minimum search neighbourhood radius in grid cells</td></tr>
<tr><td>--max_scale        </td><td> Maximum search neighbourhood radius in grid cells</td></tr>
<tr><td>--step             </td><td> Step size as any positive non-zero integer</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_anisotropy_dev_signature(
    dem, 
    points, 
    output, 
    max_scale, 
    min_scale=1, 
    step=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxAnisotropyDevSignature -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --points=sites.shp ^
--output=roughness.html --min_scale=1 --max_scale=1000 ^
--step=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/max_anisotropy_dev_signature.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dan Newman and John Lindsay</p>
<p><em>Created</em>: 27/03/2018</p>
<p><a name="MaxBranchLength"></a></p>
<a class="header" href="#maxbranchlength" id="maxbranchlength"><h1>MaxBranchLength</h1></a>
<p>Maximum branch length (<code>Bmax</code>) is the longest branch length between a grid cell's flowpath
and the flowpaths initiated at each of its neighbours. It can be conceptualized as the
downslope distance that a volume of water that is split into two portions by a drainage
divide would travel before reuniting.</p>
<p>If the two flowpaths of neighbouring grid cells do not intersect, <code>Bmax</code> is simply the
flowpath length from the starting cell to its terminus at the edge of the grid or a cell
with undefined flow direction (i.e. a pit cell either in a topographic depression or at
the edge of a major body of water).</p>
<p>The pattern of <code>Bmax</code> derived from a DEM should be familiar to anyone who has interpreted
upslope contributing area images. In fact, <code>Bmax</code> can be thought of as the complement of
upslope contributing area. Whereas contributing area is greatest along valley bottoms and lowest at
drainage divides, <code>Bmax</code> is greatest at divides and lowest along channels. The two topographic
attributes are also distinguished by their units of measurements; <code>Bmax</code> is a length rather
than an area. The presence of a major drainage divide between neighbouring grid cells is apparent in
a <code>Bmax</code> image as a linear feature, often two grid cells wide, of relatively high values. This
property makes <code>Bmax</code> a useful land surface parameter for mapping ridges and divides.</p>
<p><code>Bmax</code> is useful in the study of landscape structure, particularly with respect to drainage patterns.
The index gives the relative significance of a specific location along a divide, with respect to the
dispersion of materials across the landscape, in much the same way that stream ordering can be used
to assess stream size.</p>
<p><em>See Also</em>:</p>
<p><a href="./hydrological_analysis.html#flowlengthdiff"><strong>FlowLengthDiff</strong></a></p>
<p><em>Reference</em>:</p>
<p>Lindsay JB, Seibert J. 2013. Measuring the significance of a divide to local drainage patterns.
International Journal of Geographical Information Science, 27: 1453-1468. DOI:
10.1080/13658816.2012.705289</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--log              </td><td> Optional flag to request the output be log-transformed</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_branch_length(
    dem, 
    output, 
    log=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxBranchLength -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/max_branch_length.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 9, 2017</p>
<p><a name="MaxDifferenceFromMean"></a></p>
<a class="header" href="#maxdifferencefrommean" id="maxdifferencefrommean"><h1>MaxDifferenceFromMean</h1></a>
<p>Calculates the maximum difference from mean elevation over a range of spatial scales.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--out_mag          </td><td> Output raster DIFFmax magnitude file</td></tr>
<tr><td>--out_scale        </td><td> Output raster DIFFmax scale file</td></tr>
<tr><td>--min_scale        </td><td> Minimum search neighbourhood radius in grid cells</td></tr>
<tr><td>--max_scale        </td><td> Maximum search neighbourhood radius in grid cells</td></tr>
<tr><td>--step             </td><td> Step size as any positive non-zero integer</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_difference_from_mean(
    dem, 
    out_mag, 
    out_scale, 
    min_scale, 
    max_scale, 
    step=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxDifferenceFromMean -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --out_mag=DEVmax_mag.tif ^
--out_scale=DEVmax_scale.tif --min_scale=1 --max_scale=1000 ^
--step=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/max_diff_from_mean.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 28/08/2018</p>
<p><a name="MaxDownslopeElevChange"></a></p>
<a class="header" href="#maxdownslopeelevchange" id="maxdownslopeelevchange"><h1>MaxDownslopeElevChange</h1></a>
<p>Calculates the maximum downslope change in elevation between a grid cell and its eight downslope neighbors.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_downslope_elev_change(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxDownslopeElevChange -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=out.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/max_downslope_elev_change.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="MaxElevDevSignature"></a></p>
<a class="header" href="#maxelevdevsignature" id="maxelevdevsignature"><h1>MaxElevDevSignature</h1></a>
<p>Calculates the maximum elevation deviation over a range of spatial scales and for a set of points.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--points           </td><td> Input vector points file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
<tr><td>--min_scale        </td><td> Minimum search neighbourhood radius in grid cells</td></tr>
<tr><td>--max_scale        </td><td> Maximum search neighbourhood radius in grid cells</td></tr>
<tr><td>--step             </td><td> Step size as any positive non-zero integer</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_elev_dev_signature(
    dem, 
    points, 
    output, 
    min_scale, 
    max_scale, 
    step=10, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxElevDevSignature -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --points=sites.tif ^
--output=topo_position.html --min_scale=1 --max_scale=1000 ^
--step=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/max_elev_dev_signature.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: March 1, 2018</p>
<p><a name="MaxElevationDeviation"></a></p>
<a class="header" href="#maxelevationdeviation" id="maxelevationdeviation"><h1>MaxElevationDeviation</h1></a>
<p>Calculates the maximum elevation deviation over a range of spatial scales.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--out_mag          </td><td> Output raster DEVmax magnitude file</td></tr>
<tr><td>--out_scale        </td><td> Output raster DEVmax scale file</td></tr>
<tr><td>--min_scale        </td><td> Minimum search neighbourhood radius in grid cells</td></tr>
<tr><td>--max_scale        </td><td> Maximum search neighbourhood radius in grid cells</td></tr>
<tr><td>--step             </td><td> Step size as any positive non-zero integer</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_elevation_deviation(
    dem, 
    out_mag, 
    out_scale, 
    min_scale, 
    max_scale, 
    step=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxElevationDeviation -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --out_mag=DEVmax_mag.tif ^
--out_scale=DEVmax_scale.tif --min_scale=1 --max_scale=1000 ^
--step=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/max_elev_deviation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 20, 2017</p>
<p><a name="MinDownslopeElevChange"></a></p>
<a class="header" href="#mindownslopeelevchange" id="mindownslopeelevchange"><h1>MinDownslopeElevChange</h1></a>
<p>Calculates the minimum downslope change in elevation between a grid cell and its eight downslope neighbors.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">min_downslope_elev_change(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinDownslopeElevChange -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=out.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/min_downslope_elev_change.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="MultiscaleRoughness"></a></p>
<a class="header" href="#multiscaleroughness" id="multiscaleroughness"><h1>MultiscaleRoughness</h1></a>
<p>Calculates surface roughness over a range of spatial scales.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--out_mag          </td><td> Output raster roughness magnitude file</td></tr>
<tr><td>--out_scale        </td><td> Output raster roughness scale file</td></tr>
<tr><td>--min_scale        </td><td> Minimum search neighbourhood radius in grid cells</td></tr>
<tr><td>--max_scale        </td><td> Maximum search neighbourhood radius in grid cells</td></tr>
<tr><td>--step             </td><td> Step size as any positive non-zero integer</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">multiscale_roughness(
    dem, 
    out_mag, 
    out_scale, 
    max_scale, 
    min_scale=1, 
    step=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MultiscaleRoughness -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --out_mag=roughness_mag.tif ^
--out_scale=roughness_scale.tif --min_scale=1 --max_scale=1000 ^
--step=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/multiscale_roughness.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 26, 2018</p>
<p><a name="MultiscaleRoughnessSignature"></a></p>
<a class="header" href="#multiscaleroughnesssignature" id="multiscaleroughnesssignature"><h1>MultiscaleRoughnessSignature</h1></a>
<p>Calculates the surface roughness for points over a range of spatial scales.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--points           </td><td> Input vector points file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
<tr><td>--min_scale        </td><td> Minimum search neighbourhood radius in grid cells</td></tr>
<tr><td>--max_scale        </td><td> Maximum search neighbourhood radius in grid cells</td></tr>
<tr><td>--step             </td><td> Step size as any positive non-zero integer</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">multiscale_roughness_signature(
    dem, 
    points, 
    output, 
    max_scale, 
    min_scale=1, 
    step=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MultiscaleRoughnessSignature -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif --points=sites.shp ^
--output=roughness.html --min_scale=1 --max_scale=1000 ^
--step=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/multiscale_roughness_signature.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 27, 2018</p>
<p><a name="MultiscaleTopographicPositionImage"></a></p>
<a class="header" href="#multiscaletopographicpositionimage" id="multiscaletopographicpositionimage"><h1>MultiscaleTopographicPositionImage</h1></a>
<p>Creates a multiscale topographic position image from three DEVmax rasters of differing spatial scale ranges.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--local            </td><td> Input local-scale topographic position (DEVmax) raster file</td></tr>
<tr><td>--meso             </td><td> Input meso-scale topographic position (DEVmax) raster file</td></tr>
<tr><td>--broad            </td><td> Input broad-scale topographic position (DEVmax) raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--lightness        </td><td> Image lightness value (default is 1.2)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">multiscale_topographic_position_image(
    local, 
    meso, 
    broad, 
    output, 
    lightness=1.2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MultiscaleTopographicPositionImage -v ^
--wd=&quot;/path/to/data/&quot; --local=DEV_local.tif --meso=DEV_meso.tif ^
--broad=DEV_broad.tif -o=output.tif --lightness=1.5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/multiscale_topographic_position_image.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 19, 2017</p>
<p><a name="NumDownslopeNeighbours"></a></p>
<a class="header" href="#numdownslopeneighbours" id="numdownslopeneighbours"><h1>NumDownslopeNeighbours</h1></a>
<p>Calculates the number of downslope neighbours to each grid cell in a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">num_downslope_neighbours(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NumDownslopeNeighbours -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/num_downslope_neighbours.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="NumUpslopeNeighbours"></a></p>
<a class="header" href="#numupslopeneighbours" id="numupslopeneighbours"><h1>NumUpslopeNeighbours</h1></a>
<p>Calculates the number of upslope neighbours to each grid cell in a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">num_upslope_neighbours(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NumUpslopeNeighbours -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/num_upslope_neighbours.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="PennockLandformClass"></a></p>
<a class="header" href="#pennocklandformclass" id="pennocklandformclass"><h1>PennockLandformClass</h1></a>
<p>Tool can be used to perform a simple landform classification based on measures of slope gradient
and curvature derived from a user-specified digital elevation model (DEM). The classification
scheme is based on the method proposed by Pennock, Zebarth, and DeJong (1987). The scheme divides
a landscape into seven element types, including: convergent footslopes (CFS), divergent footslopes
(DFS), convergent shoulders (CSH), divergent shoulders (DSH), convergent backslopes (CBS), divergent
backslopes (DBS), and level terrain (L). The output raster image will record each of these base element
types as:</p>
<table><thead><tr><th>Element Type  </th><th>  Code</th></tr></thead><tbody>
<tr><td>CFS           </td><td>  1</td></tr>
<tr><td>DFS           </td><td>  2</td></tr>
<tr><td>CSH           </td><td>  3</td></tr>
<tr><td>DSH           </td><td>  4</td></tr>
<tr><td>CBS           </td><td>  5</td></tr>
<tr><td>DBS           </td><td>  6</td></tr>
<tr><td>L             </td><td>  7</td></tr>
</tbody></table>
<p>The definition of each of the elements, based on the original Pennock et al. (1987) paper, is
as follows:</p>
<table><thead><tr><th align="left">    PROFILE             </th><th align="left">   GRADIENT    </th><th>   PLAN         </th><th>  Element </th></tr></thead><tbody>
<tr><td align="left"> Concave ( -0.10)       </td><td align="left">  High &gt;3.0    </td><td align="left"> Concave 0.0    </td><td align="left">  CFS     </td></tr>
<tr><td align="left"> Concave ( -0.10)       </td><td align="left">  High &gt;3.0    </td><td align="left"> Convex &gt;0.0    </td><td align="left">  DFS     </td></tr>
<tr><td align="left"> Convex (&gt;0.10)         </td><td align="left">  High &gt;3.0    </td><td align="left"> Concave 0.0    </td><td align="left">  CSH     </td></tr>
<tr><td align="left"> Convex (&gt;0.10)         </td><td align="left">  High &gt;3.0    </td><td align="left"> Convex &gt;0.0    </td><td align="left">  DSH     </td></tr>
<tr><td align="left"> Linear (-0.10...0.10)  </td><td align="left">  High &gt;3.0    </td><td align="left"> Concave 0.0    </td><td align="left">  CBS     </td></tr>
<tr><td align="left"> Linear (-0.10...0.10)  </td><td align="left">  High &gt;3.0    </td><td align="left"> Convex &gt;0.0    </td><td align="left">  DBS     </td></tr>
<tr><td align="left"> --                     </td><td align="left">  Low 3.0      </td><td align="left"> --             </td><td align="left">  L       </td></tr>
</tbody></table>
<p>Where PROFILE is profile curvature, GRADIENT is the slope gradient, and PLAN is the plan curvature.
Note that these values are likely landscape and data specific and can be adjusted by the user.
Landscape classification schemes that are based on terrain attributes are highly sensitive to
short-range topographic variability (i.e. roughness) and can benefit from pre-processing the DEM
with a smoothing filter to reduce the effect of surface roughness and emphasize the longer-range
topographic signal. The <a href="./geomorphometric_analysis.html#featurepreservingdenoise"><strong>FeaturePreservingDenoise</strong></a> and <a href="./geomorphometric_analysis.html#drainagepreservingsmoothing"><strong>DrainagePreservingSmoothing</strong></a> tools
offer excellent performance in smoothing DEMs without removing the sharpness of breaks-in-slope.</p>
<p><em>Reference</em>:</p>
<p>Pennock, D.J., Zebarth, B.J., and DeJong, E. (1987) Landform classification and soil distribution
in hummocky terrain, Saskatchewan, Canada. Geoderma, 40: 297-315.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--slope            </td><td> Slope threshold value, in degrees (default is 3.0)</td></tr>
<tr><td>--prof             </td><td> Profile curvature threshold value (default is 0.1)</td></tr>
<tr><td>--plan             </td><td> Plan curvature threshold value (default is 0.0)</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">pennock_landform_class(
    dem, 
    output, 
    slope=3.0, 
    prof=0.1, 
    plan=0.0, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PennockLandformClass -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif --slope=3.0 ^
--prof=0.1 --plan=0.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/pennock_landform_class.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 12, 2017</p>
<p><a name="PercentElevRange"></a></p>
<a class="header" href="#percentelevrange" id="percentelevrange"><h1>PercentElevRange</h1></a>
<p>Calculates percent of elevation range from a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input, --dem</td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">percent_elev_range(
    dem, 
    output, 
    filterx=3, 
    filtery=3, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PercentElevRange -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/percent_elev_range.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="PlanCurvature"></a></p>
<a class="header" href="#plancurvature" id="plancurvature"><h1>PlanCurvature</h1></a>
<p>This tool calculates the plan curvature (i.e. contour curvature), or the rate of change in
aspect along a contour line, from a digital elevation model (DEM). Curvature is the second
derivative of the topographic surface defined by a DEM. Plan curvature characterizes the
degree of flow convergence or divergence within the landscape (Gallant and Wilson, 2000).
The user must specify the name of the input DEM (<code>--dem</code>) and the output raster image.
WhiteboxTools reports curvature in degrees multiplied by 100 for easier interpretation. The
<em>Z conversion factor</em> (<code>--zfactor</code>) is only important when the vertical and horizontal units
are not the same in the DEM. When this is the case, the algorithm will multiply each
elevation in the DEM by the Z Conversion Factor. If the DEM is in the geographic coordinate
system (latitude and longitude), the following equation is used:</p>
<pre><code>zfactor = 1.0 / (113200.0 x cos(mid_lat))
</code></pre>
<p>where <code>mid_lat</code> is the latitude of the centre of the raster, in radians.</p>
<p>The algorithm uses the same formula for the calculation of plan curvature as Gallant and
Wilson (2000). Plan curvature is negative for diverging flow along ridges and positive for
convergent areas, e.g. along valley bottoms.</p>
<p><em>Reference</em>:</p>
<p>Gallant, J. C., and J. P. Wilson, 2000, Primary topographic attributes, in Terrain Analysis: Principles
and Applications, edited by J. P. Wilson and J. C. Gallant pp. 51-86, John Wiley, Hoboken, N.J.</p>
<p><em>See Also</em>:</p>
<p><a href="./geomorphometric_analysis.html#profilecurvature"><strong>ProfileCurvature</strong></a>, <a href="./geomorphometric_analysis.html#tangentialcurvature"><strong>TangentialCurvature</strong></a>, <a href="./geomorphometric_analysis.html#totalcurvature"><strong>TotalCurvature</strong></a>, <a href="./geomorphometric_analysis.html#slope"><strong>Slope</strong></a>, <a href="./geomorphometric_analysis.html#aspect"><strong>Aspect</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">plan_curvature(
    dem, 
    output, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PlanCurvature -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/plan_curvature.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 1, 2017</p>
<p><a name="Profile"></a></p>
<a class="header" href="#profile" id="profile"><h1>Profile</h1></a>
<p>Plots profiles from digital surface models.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--lines            </td><td> Input vector line file</td></tr>
<tr><td>--surface          </td><td> Input raster surface file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">profile(
    lines, 
    surface, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Profile -v --wd=&quot;/path/to/data/&quot; ^
--lines=profile.shp --surface=dem.tif -o=profile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/profile.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 21, 2018</p>
<p><a name="ProfileCurvature"></a></p>
<a class="header" href="#profilecurvature" id="profilecurvature"><h1>ProfileCurvature</h1></a>
<p>Calculates a profile curvature raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">profile_curvature(
    dem, 
    output, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ProfileCurvature -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/prof_curvature.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="RelativeAspect"></a></p>
<a class="header" href="#relativeaspect" id="relativeaspect"><h1>RelativeAspect</h1></a>
<p>Calculates relative aspect (relative to a user-specified direction) from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--azimuth          </td><td> Illumination source azimuth</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">relative_aspect(
    dem, 
    output, 
    azimuth=0.0, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RelativeAspect -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif --azimuth=180.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/relative_aspect.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 17, 2017</p>
<p><a name="RelativeStreamPowerIndex"></a></p>
<a class="header" href="#relativestreampowerindex" id="relativestreampowerindex"><h1>RelativeStreamPowerIndex</h1></a>
<p>Calculates the relative stream power index.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--sca              </td><td> Input raster specific contributing area (SCA) file</td></tr>
<tr><td>--slope            </td><td> Input raster slope file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--exponent         </td><td> SCA exponent value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">relative_stream_power_index(
    sca, 
    slope, 
    output, 
    exponent=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RelativeStreamPowerIndex -v ^
--wd=&quot;/path/to/data/&quot; --sca='flow_accum.tif' ^
--slope='slope.tif' -o=output.tif --exponent=1.1 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/relative_stream_power_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<p><a name="RelativeTopographicPosition"></a></p>
<a class="header" href="#relativetopographicposition" id="relativetopographicposition"><h1>RelativeTopographicPosition</h1></a>
<p>Calculates the relative topographic position index from a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">relative_topographic_position(
    dem, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RelativeTopographicPosition -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/relative_topographic_position.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 6, 2017</p>
<p><a name="RemoveOffTerrainObjects"></a></p>
<a class="header" href="#removeoffterrainobjects" id="removeoffterrainobjects"><h1>RemoveOffTerrainObjects</h1></a>
<p>Removes off-terrain objects from a raster digital elevation model (DEM).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input, --dem</td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filter           </td><td> Filter size (cells)</td></tr>
<tr><td>--slope            </td><td> Slope threshold value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">remove_off_terrain_objects(
    dem, 
    output, 
    filter=11, 
    slope=15.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RemoveOffTerrainObjects -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=bare_earth_DEM.tif ^
--filter=25 --slope=10.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/remove_off_terrain_objects.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 6, 2017</p>
<p><a name="RuggednessIndex"></a></p>
<a class="header" href="#ruggednessindex" id="ruggednessindex"><h1>RuggednessIndex</h1></a>
<p>Calculates the Riley et al.'s (1999) terrain ruggedness index from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">ruggedness_index(
    dem, 
    output, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RuggednessIndex -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/ruggedness_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="SedimentTransportIndex"></a></p>
<a class="header" href="#sedimenttransportindex" id="sedimenttransportindex"><h1>SedimentTransportIndex</h1></a>
<p>Calculates the sediment transport index.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--sca              </td><td> Input raster specific contributing area (SCA) file</td></tr>
<tr><td>--slope            </td><td> Input raster slope file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--sca_exponent     </td><td> SCA exponent value</td></tr>
<tr><td>--slope_exponent   </td><td> Slope exponent value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">sediment_transport_index(
    sca, 
    slope, 
    output, 
    sca_exponent=0.4, 
    slope_exponent=1.3, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SedimentTransportIndex -v ^
--wd=&quot;/path/to/data/&quot; --sca='flow_accum.tif' ^
--slope='slope.tif' -o=output.tif --sca_exponent=0.5 ^
--slope_exponent=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/sediment_transport_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<p><a name="Slope"></a></p>
<a class="header" href="#slope" id="slope"><h1>Slope</h1></a>
<p>This tool calculates slope gradient (i.e. slope steepness in degrees) for each grid cell
in an input digital elevation model (DEM). The user must specify the name of the input
DEM (<code>--dem</code>) and the output raster image. The <em>Z conversion factor</em> is only important
when the vertical and horizontal units are not the same in the DEM. When this is the case,
the algorithm will multiply each elevation in the DEM by the Z conversion factor. If the
DEM is in the geographic coordinate system (latitude and longitude), the following equation
is used:</p>
<pre><code>zfactor = 1.0 / (113200.0 x cos(mid_lat))
</code></pre>
<p>where <code>mid_lat</code> is the latitude of the centre of the raster, in radians.</p>
<p>The tool uses Horn's (1981) 3rd-order finite difference method to estimate slope. Given
the following clock-type grid cell numbering scheme (Gallant and Wilson, 2000),</p>
<p>|  7  |  8  |  1  | <br />
|  6  |  9  |  2  | <br />
|  5  |  4  |  3  |</p>
<pre><code>slope = arctan(fx2 + fy2)^0.5
</code></pre>
<p>where,</p>
<pre><code>fx = (z3 - z5 + 2(z2 - z6) + z1 - z7)
</code></pre>
<p>and,</p>
<pre><code>fy = (z7 - z5 + 2(z8 - z4) + z1 - z3)
</code></pre>
<p><em>Reference</em>:</p>
<p>Gallant, J. C., and J. P. Wilson, 2000, Primary topographic attributes, in Terrain Analysis: Principles
and Applications, edited by J. P. Wilson and J. C. Gallant pp. 51-86, John Wiley, Hoboken, N.J.</p>
<p><em>See Also</em>:</p>
<p><a href="./geomorphometric_analysis.html#aspect"><strong>Aspect</strong></a>, <a href="./geomorphometric_analysis.html#plancurvature"><strong>PlanCurvature</strong></a>, <a href="./geomorphometric_analysis.html#profilecurvature"><strong>ProfileCurvature</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">slope(
    dem, 
    output, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Slope -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/slope.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="SlopeVsElevationPlot"></a></p>
<a class="header" href="#slopevselevationplot" id="slopevselevationplot"><h1>SlopeVsElevationPlot</h1></a>
<p>Creates a slope vs. elevation plot for one or more DEMs.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input DEM files</td></tr>
<tr><td>--watershed        </td><td> Input watershed files (optional)</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">slope_vs_elevation_plot(
    inputs, 
    output, 
    watershed=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SlopeVsElevationPlot -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;DEM1.tif;DEM2.tif&quot; ^
--watershed=&quot;ws1.tif;ws2.tif&quot; -o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/slope_vs_elev_plot.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 1, 2018</p>
<p><a name="StandardDeviationOfSlope"></a></p>
<a class="header" href="#standarddeviationofslope" id="standarddeviationofslope"><h1>StandardDeviationOfSlope</h1></a>
<p>Calculates the standard deviation of slope from an input DEM, a metric of
roughness described by Grohmann et al., (2011).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster DEM file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">standard_deviation_of_slope(
    i, 
    output, 
    zfactor=1.0, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StandardDeviationOfSlope -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/standard_deviation_of_slope.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Anthony Francioni</p>
<p><em>Created</em>: May 26, 2018</p>
<p><a name="TangentialCurvature"></a></p>
<a class="header" href="#tangentialcurvature" id="tangentialcurvature"><h1>TangentialCurvature</h1></a>
<p>Calculates a tangential curvature raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">tangential_curvature(
    dem, 
    output, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TangentialCurvature -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/tan_curvature.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="TotalCurvature"></a></p>
<a class="header" href="#totalcurvature" id="totalcurvature"><h1>TotalCurvature</h1></a>
<p>Calculates a total curvature raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zfactor          </td><td> Optional multiplier for when the vertical and horizontal units are not the same</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">total_curvature(
    dem, 
    output, 
    zfactor=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TotalCurvature -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/total_curvature.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="Viewshed"></a></p>
<a class="header" href="#viewshed" id="viewshed"><h1>Viewshed</h1></a>
<p>Identifies the viewshed for a point or set of points.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>--stations         </td><td> Input viewing station vector file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--height           </td><td> Viewing station height, in z units</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">viewshed(
    dem, 
    stations, 
    output, 
    height=2.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Viewshed -v --wd=&quot;/path/to/data/&quot; ^
--dem='dem.tif' --stations='stations.shp' -o=output.tif ^
--height=10.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/viewshed.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: January 10, 2018</p>
<p><a name="VisibilityIndex"></a></p>
<a class="header" href="#visibilityindex" id="visibilityindex"><h1>VisibilityIndex</h1></a>
<p>Estimates the relative visibility of sites in a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--height           </td><td> Viewing station height, in z units</td></tr>
<tr><td>--res_factor       </td><td> The resolution factor determines the density of measured viewsheds</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">visibility_index(
    dem, 
    output, 
    height=2.0, 
    res_factor=2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=VisibilityIndex -v ^
--wd=&quot;/path/to/data/&quot; --dem=dem.tif -o=output.tif ^
--height=10.0 --res_factor=4 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/visibility_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 07/04/2018</p>
<p><a name="WetnessIndex"></a></p>
<a class="header" href="#wetnessindex" id="wetnessindex"><h1>WetnessIndex</h1></a>
<p>Calculates the topographic wetness index, Ln(A / tan(slope)).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--sca              </td><td> Input raster specific contributing area (SCA) file</td></tr>
<tr><td>--slope            </td><td> Input raster slope file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">wetness_index(
    sca, 
    slope, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=WetnessIndex -v --wd=&quot;/path/to/data/&quot; ^
--sca='flow_accum.tif' --slope='slope.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/terrain_analysis/wetness_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<a class="header" href="#gis-analysis" id="gis-analysis"><h1>GIS Analysis</h1></a>
<ul>
<li><a href="#AggregateRaster">AggregateRaster</a></li>
<li><a href="#BlockMaximumGridding">BlockMaximumGridding</a></li>
<li><a href="#BlockMinimumGridding">BlockMinimumGridding</a></li>
<li><a href="#Centroid">Centroid</a></li>
<li><a href="#CentroidVector">CentroidVector</a></li>
<li><a href="#Clump">Clump</a></li>
<li><a href="#ConstructVectorTin">ConstructVectorTin</a></li>
<li><a href="#CreateHexagonalVectorGrid">CreateHexagonalVectorGrid</a></li>
<li><a href="#CreatePlane">CreatePlane</a></li>
<li><a href="#CreateRectangularVectorGrid">CreateRectangularVectorGrid</a></li>
<li><a href="#Dissolve">Dissolve</a></li>
<li><a href="#EliminateCoincidentPoints">EliminateCoincidentPoints</a></li>
<li><a href="#ExtendVectorLines">ExtendVectorLines</a></li>
<li><a href="#ExtractNodes">ExtractNodes</a></li>
<li><a href="#ExtractRasterValuesAtPoints">ExtractRasterValuesAtPoints</a></li>
<li><a href="#FindLowestOrHighestPoints">FindLowestOrHighestPoints</a></li>
<li><a href="#IdwInterpolation">IdwInterpolation</a></li>
<li><a href="#LayerFootprint">LayerFootprint</a></li>
<li><a href="#Medoid">Medoid</a></li>
<li><a href="#MinimumBoundingBox">MinimumBoundingBox</a></li>
<li><a href="#MinimumBoundingCircle">MinimumBoundingCircle</a></li>
<li><a href="#MinimumBoundingEnvelope">MinimumBoundingEnvelope</a></li>
<li><a href="#MinimumConvexHull">MinimumConvexHull</a></li>
<li><a href="#NearestNeighbourGridding">NearestNeighbourGridding</a></li>
<li><a href="#PolygonArea">PolygonArea</a></li>
<li><a href="#PolygonLongAxis">PolygonLongAxis</a></li>
<li><a href="#PolygonPerimeter">PolygonPerimeter</a></li>
<li><a href="#PolygonShortAxis">PolygonShortAxis</a></li>
<li><a href="#RasterCellAssignment">RasterCellAssignment</a></li>
<li><a href="#Reclass">Reclass</a></li>
<li><a href="#ReclassEqualInterval">ReclassEqualInterval</a></li>
<li><a href="#ReclassFromFile">ReclassFromFile</a></li>
<li><a href="#SmoothVectors">SmoothVectors</a></li>
<li><a href="#TinGridding">TinGridding</a></li>
<li><a href="#VectorHexBinning">VectorHexBinning</a></li>
<li><a href="#VoronoiDiagram">VoronoiDiagram</a></li>
</ul>
<p><a name="AggregateRaster"></a></p>
<a class="header" href="#aggregateraster" id="aggregateraster"><h1>AggregateRaster</h1></a>
<p>Aggregates a raster to a lower resolution.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--agg_factor       </td><td> Aggregation factor, in pixels</td></tr>
<tr><td>--type             </td><td> Statistic used to fill output pixels</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">aggregate_raster(
    i, 
    output, 
    agg_factor=2, 
    type=&quot;mean&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AggregateRaster -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif ^
--output_text 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/aggregate_raster.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: December 13, 2017</p>
<p><a name="BlockMaximumGridding"></a></p>
<a class="header" href="#blockmaximumgridding" id="blockmaximumgridding"><h1>BlockMaximumGridding</h1></a>
<p>Creates a raster grid based on a set of vector points and assigns grid values using a block maximum scheme.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector Points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>--use_z            </td><td> Use z-coordinate instead of field?</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
<tr><td>--base             </td><td> Optionally specified input base raster file. Not used when a cell size is specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">block_maximum_gridding(
    i, 
    field, 
    output, 
    use_z=False, 
    cell_size=None, 
    base=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=BlockMaximumGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --field=ELEV -o=output.tif ^
--cell_size=1.0
&gt;&gt;./whitebox_tools -r=BlockMaximumGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --use_z -o=output.tif ^
--base=existing_raster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/block_maximum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 09/10/2018</p>
<p><a name="BlockMinimumGridding"></a></p>
<a class="header" href="#blockminimumgridding" id="blockminimumgridding"><h1>BlockMinimumGridding</h1></a>
<p>Creates a raster grid based on a set of vector points and assigns grid values using a block minimum scheme.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector Points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>--use_z            </td><td> Use z-coordinate instead of field?</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
<tr><td>--base             </td><td> Optionally specified input base raster file. Not used when a cell size is specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">block_minimum_gridding(
    i, 
    field, 
    output, 
    use_z=False, 
    cell_size=None, 
    base=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=BlockMinimumGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --field=ELEV -o=output.tif ^
--cell_size=1.0
&gt;&gt;./whitebox_tools -r=BlockMinimumGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --use_z -o=output.tif ^
--base=existing_raster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/block_minimum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 09/10/2018</p>
<p><a name="Centroid"></a></p>
<a class="header" href="#centroid" id="centroid"><h1>Centroid</h1></a>
<p>This tool calculates the centroid, or average location, of raster polygon objects.
For vector features, use the <a href="./gis_analysis.html#centroidvector"><strong>CentroidVector</strong></a> tool instead.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#centroidvector"><strong>CentroidVector</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--text_output      </td><td> Optional text output</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">centroid(
    i, 
    output, 
    text_output=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Centroid -v --wd=&quot;/path/to/data/&quot; ^
-i=polygons.tif -o=output.tif
&gt;&gt;./whitebox_tools -r=Centroid ^
-v --wd=&quot;/path/to/data/&quot; -i=polygons.tif -o=output.tif ^
--text_output 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/centroid.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 22 2017</p>
<p><a name="CentroidVector"></a></p>
<a class="header" href="#centroidvector" id="centroidvector"><h1>CentroidVector</h1></a>
<p>This can be used to identify the centroid point of a vector polyline or polygon feature or a group of
vector points. The output is a vector shapefile of points. For multi-part polyline or polygon features,
the user can optionally specify whether to identify the centroid of each part. The default is to treat
multi-part features a single entity.</p>
<p>For raster features, use the <a href="./gis_analysis.html#centroid"><strong>Centroid</strong></a> tool instead.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#centroid"><strong>Centroid</strong></a>, <a href="./gis_analysis.html#medoid"><strong>Medoid</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">centroid_vector(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CentroidVector -v --wd=&quot;/path/to/data/&quot; ^
-i=in_file.shp -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/centroid_vector.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 20/09/2018</p>
<p><a name="Clump"></a></p>
<a class="header" href="#clump" id="clump"><h1>Clump</h1></a>
<p>Groups cells that form physically discrete areas, assigning them unique identifiers.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--diag             </td><td> Flag indicating whether diagonal connections should be considered</td></tr>
<tr><td>--zero_back        </td><td> Flag indicating whether zero values should be treated as a background</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">clump(
    i, 
    output, 
    diag=True, 
    zero_back=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Clump -v --wd=&quot;/path/to/data/&quot; ^
-i=input.tif -o=output.tif --diag 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/clump.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 22/06/2017</p>
<p><a name="ConstructVectorTin"></a></p>
<a class="header" href="#constructvectortin" id="constructvectortin"><h1>ConstructVectorTin</h1></a>
<p>This tool creates a vector triangular irregular network (TIN) for a set of vector points.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>--use_z            </td><td> Use the 'z' dimension of the Shapefile's geometry instead of an attribute field?</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">construct_vector_tin(
    i, 
    output, 
    field=None, 
    use_z=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ConstructVectorTIN -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --field=HEIGHT ^
-o=tin.shp
&gt;&gt;./whitebox_tools -r=ConstructVectorTIN -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --use_z -o=tin.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/construct_vector_tin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 21/09/2018</p>
<p><a name="CreateHexagonalVectorGrid"></a></p>
<a class="header" href="#createhexagonalvectorgrid" id="createhexagonalvectorgrid"><h1>CreateHexagonalVectorGrid</h1></a>
<p>This tool can be used to create a hexagonal vector grid. The extent of the hexagonal
grid is based on the extent of a user-specified base file (any supported raster format,
shapefiles, or LAS files). The user must also specify the origin of the grid (x and y
coordinates) and the hexagonal cell width.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input base file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--width            </td><td> The grid cell width</td></tr>
<tr><td>--orientation      </td><td> Grid Orientation, 'horizontal' or 'vertical'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">create_hexagonal_vector_grid(
    i, 
    output, 
    width, 
    orientation=&quot;horizontal&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CreateHexagonalVectorGrid -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp --width=10.0 ^
--orientation=vertical 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/create_hexagonal_vector_grid.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 15/09/2018</p>
<p><a name="CreatePlane"></a></p>
<a class="header" href="#createplane" id="createplane"><h1>CreatePlane</h1></a>
<p>Creates a raster image based on the equation for a simple plane.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--base             </td><td> Input base raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--gradient         </td><td> Slope gradient in degrees (-85.0 to 85.0)</td></tr>
<tr><td>--aspect           </td><td> Aspect (direction) in degrees clockwise from north (0.0-360.0)</td></tr>
<tr><td>--constant         </td><td> Constant value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">create_plane(
    base, 
    output, 
    gradient=15.0, 
    aspect=90.0, 
    constant=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CreatePlane -v --wd=&quot;/path/to/data/&quot; ^
--base=base.tif -o=NewRaster.tif --gradient=15.0 ^
--aspect=315.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/create_plane.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="CreateRectangularVectorGrid"></a></p>
<a class="header" href="#createrectangularvectorgrid" id="createrectangularvectorgrid"><h1>CreateRectangularVectorGrid</h1></a>
<p>This tool can be used to create a rectangular vector grid. The extent of the rectangular
grid is based on the extent of a user-specified base file (any supported raster format,
shapefiles, or LAS files). The user must also specify the origin of the grid (x and y
coordinates) and the grid cell width and height.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input base file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--width            </td><td> The grid cell width</td></tr>
<tr><td>--height           </td><td> The grid cell height</td></tr>
<tr><td>--xorig            </td><td> The grid origin x-coordinate</td></tr>
<tr><td>--yorig            </td><td> The grid origin y-coordinate</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">create_rectangular_vector_grid(
    i, 
    output, 
    width, 
    height, 
    xorig=0, 
    yorig=0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CreateRectangularVectorGrid -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp --width=10.0 ^
--height=10.0 --xorig=0.0 --yorig=0.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/create_rectangular_vector_grid.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 15/09/2018</p>
<p><a name="Dissolve"></a></p>
<a class="header" href="#dissolve" id="dissolve"><h1>Dissolve</h1></a>
<p>This tool can be used to remove the interior, or shared, boundaries within a vector
polygon coverage. You can either dissolve all interior boundaries or dissolve those
boundaries along polygons with the same value of a user-specified attribute within
the vector's attribute table. It may be desirable to use the <code>VectorCleaning</code> tool
to correct any topological errors resulting from the slight misalignment of nodes
along shared boundaries in the vector coverage before performing the <a href="./gis_analysis.html#dissolve"><strong>Dissolve</strong></a> operation.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a>, <a href="./gis_analysis_overlay_tools.html#erase"><strong>Erase</strong></a>, <a href="./gis_analysis_overlay_tools.html#polygonize"><strong>Polygonize</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>--field            </td><td> Dissolve field attribute (optional)</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--snap             </td><td> Snap tolerance</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">dissolve(
    i, 
    output, 
    field=None, 
    snap=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Dissolve -v --wd=&quot;/path/to/data/&quot; ^
-input=layer1.shp --field=SIZE -o=out_file.shp ^
--snap=0.0000001 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/dissolve.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 13/11/2018</p>
<p><a name="EliminateCoincidentPoints"></a></p>
<a class="header" href="#eliminatecoincidentpoints" id="eliminatecoincidentpoints"><h1>EliminateCoincidentPoints</h1></a>
<p>This tool can be used to remove any coincident, or nearly coincident, points
from a vector points file. The user must specify the name of the input file,
which must be of a POINTS ShapeType, the output file name, and the tolerance
distance. All points that are within the specified tolerance distance will be
eliminated from the output file. A tolerance distance of 0.0 indicates that
points must be exactly coincident to be removed.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--tolerance        </td><td> The distance tolerance for points</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">eliminate_coincident_points(
    i, 
    output, 
    tolerance, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=EliminateCoincidentPoints -v ^
--wd=&quot;/path/to/data/&quot; -i=input_file.shp -o=out_file.shp ^
--tolerance=0.01 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/eliminate_coincident_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 16/09/2018</p>
<p><a name="ExtendVectorLines"></a></p>
<a class="header" href="#extendvectorlines" id="extendvectorlines"><h1>ExtendVectorLines</h1></a>
<p>This tool can be used to extend vector lines by a specified distance. The user must
input the names of the input and output shapefiles, the distance to extend features
by, and whether to extend both ends, line starts, or line ends. The input shapefile
must be of a POLYLINE base shape type and should be in a projected coordinate system.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polyline file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polyline file</td></tr>
<tr><td>--dist             </td><td> The distance to extend</td></tr>
<tr><td>--extend           </td><td> Extend direction, 'both ends' (default), 'line start', 'line end'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">extend_vector_lines(
    i, 
    output, 
    dist, 
    extend=&quot;both ends&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ExtendVectorLines -v ^
--wd=&quot;/path/to/data/&quot; -i=in_file.shp -o=out_file.shp ^
--dist=10.0 --extend='both ends' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/extend_vector_lines.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 20/09/2018</p>
<p><a name="ExtractNodes"></a></p>
<a class="header" href="#extractnodes" id="extractnodes"><h1>ExtractNodes</h1></a>
<p>Converts vector lines or polygons into vertex points.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector lines or polygon file</td></tr>
<tr><td>-o, --output       </td><td> Output vector points file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">extract_nodes(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ExtractNodes -v --wd=&quot;/path/to/data/&quot; ^
-i=file.shp -o=outfile.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/extract_nodes.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 04/09/2018</p>
<p><a name="ExtractRasterValuesAtPoints"></a></p>
<a class="header" href="#extractrastervaluesatpoints" id="extractrastervaluesatpoints"><h1>ExtractRasterValuesAtPoints</h1></a>
<p>Extracts the values of raster(s) at vector point locations.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--points           </td><td> Input vector points file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">extract_raster_values_at_points(
    inputs, 
    points, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ExtractRasterValuesAtPoints -v ^
--wd=&quot;/path/to/data/&quot; -i='image1.tif;image2.tif;image3.tif' ^
-points=points.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/extract_raster_values_at_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 17/06/2018</p>
<p><a name="FindLowestOrHighestPoints"></a></p>
<a class="header" href="#findlowestorhighestpoints" id="findlowestorhighestpoints"><h1>FindLowestOrHighestPoints</h1></a>
<p>Locates the lowest and/or highest valued cells in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output vector points file</td></tr>
<tr><td>--out_type         </td><td> Output type; one of 'area' (default) and 'volume'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">find_lowest_or_highest_points(
    i, 
    output, 
    out_type=&quot;lowest&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FindLowestOrHighestPoints -v ^
--wd=&quot;/path/to/data/&quot; --input=DEM.tif -o=out.shp ^
--out_type=highest 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/find_lowest_or_highest_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 12/06/2018</p>
<p><a name="IdwInterpolation"></a></p>
<a class="header" href="#idwinterpolation" id="idwinterpolation"><h1>IdwInterpolation</h1></a>
<p>points or a fixed neighbourhood size. This tool is currently configured to perform the later
only, using a FixedRadiusSearch structure. Using a fixed number of neighbours will require
use of a KD-tree structure. I've been testing one Rust KD-tree library but its performance
does not appear to be satisfactory compared to the FixedRadiusSearch. I will need to explore
other options here.</p>
<p>Another change that will need to be implemented is the use of a nodal function. The original
Whitebox GAT tool allows for use of a constant or a quadratic. This tool only allows the
former.
This tool interpolates vector points into a raster surface using an inverse-distance weighted scheme.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector Points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>--use_z            </td><td> Use z-coordinate instead of field?</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--weight           </td><td> IDW weight value</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
<tr><td>--min_points       </td><td> Minimum number of points</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
<tr><td>--base             </td><td> Optionally specified input base raster file. Not used when a cell size is specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">idw_interpolation(
    i, 
    field, 
    output, 
    use_z=False, 
    weight=2.0, 
    radius=None, 
    min_points=None, 
    cell_size=None, 
    base=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=IdwInterpolation -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --field=ELEV -o=output.tif ^
--weight=2.0 --radius=4.0 --min_points=3 ^
--cell_size=1.0
&gt;&gt;./whitebox_tools -r=IdwInterpolation -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --use_z -o=output.tif ^
--weight=2.0 --radius=4.0 --min_points=3 ^
--base=existing_raster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/idw_interpolation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 10/05/2018</p>
<p><a name="LayerFootprint"></a></p>
<a class="header" href="#layerfootprint" id="layerfootprint"><h1>LayerFootprint</h1></a>
<p>This tool creates a vector polygon footprint of the area covered by a raster grid or vector
layer. It will create a vector rectangle corresponding to the bounding box. The user must
specify the name of the input file, which may be either a Whitebox raster or a vector, and
the name of the output file.</p>
<p>If an input raster grid is specified which has an irregular shape, i.e. it contains NoData
values at the edges, the resulting vector will still correspond to the full grid extent,
ignoring the irregular boundary. If this is not the desired effect, you should reclass the
grid such that all cells containing valid values are assigned some positive, non-zero value,
and then use the <code>RasterToVectorPolygons</code> tool to vectorize the irregular-shaped extent
boundary.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#minimumboundingenvelope"><strong>MinimumBoundingEnvelope</strong></a>, <code>RasterToVectorPolygons</code></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster or vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">layer_footprint(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LayerFootprint -v --wd=&quot;/path/to/data/&quot; ^
-i=file.shp -o=outfile.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/layer_footprint.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 31/09/2018</p>
<p><a name="Medoid"></a></p>
<a class="header" href="#medoid" id="medoid"><h1>Medoid</h1></a>
<p>This tool calculates the medoid for a series of vector features contained in a shapefile. The medoid
of a two-dimensional feature is conceptually similar its centroid, or mean position, but the medoid
is always a members of the input feature data set. Thus, the medoid is a measure of central tendency
that is robust in the presence of outliers. If the input vector is of a POLYLINE or POLYGON ShapeType,
the nodes of each feature will be used to estimate the feature medoid. If the input vector is of a
POINT base ShapeType, the medoid will be calculated for the collection of points. While there are
more than one competing method of calculating the medoid, this tool uses an algorithm that works as follows:</p>
<ol>
<li>The x-coordinate and y-coordinate of each point/node are placed into two arrays.</li>
<li>The x- and y-coordinate arrays are then sorted and the median x-coordinate (Med X) and median
y-coordinate (Med Y) are calculated.</li>
<li>The point/node in the dataset that is nearest the point (Med X, Med Y) is identified as the medoid.</li>
</ol>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#centroidvector"><strong>CentroidVector</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">medoid(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Medoid -v --wd=&quot;/path/to/data/&quot; ^
-i=in_file.shp -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/medoid.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 20/09/2018</p>
<p><a name="MinimumBoundingBox"></a></p>
<a class="header" href="#minimumboundingbox" id="minimumboundingbox"><h1>MinimumBoundingBox</h1></a>
<p>This tool delineates the minimum bounding box (MBB) for a group of vectors. The MBB is the smallest box to
completely enclose a feature. The algorithm works by rotating the feature, calculating the axis-aligned
bounding box for each rotation, and finding the box with the smallest area, length, width, or perimeter. The
MBB is needed to compute several shape indices, such as the Elongation Ratio. The <code>MinimumBoundingEnvelop</code>
tool can be used to calculate the axis-aligned bounding rectangle around each feature in a vector file.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#minimumboundingcircle"><strong>MinimumBoundingCircle</strong></a>, <a href="./gis_analysis.html#minimumboundingenvelope"><strong>MinimumBoundingEnvelope</strong></a>, <a href="./gis_analysis.html#minimumconvexhull"><strong>MinimumConvexHull</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--criterion        </td><td> Minimization criterion; options include 'area' (default), 'length', 'width', and 'perimeter'</td></tr>
<tr><td>--features         </td><td> Find the minimum bounding rectangles around each individual vector feature</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">minimum_bounding_box(
    i, 
    output, 
    criterion=&quot;area&quot;, 
    features=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinimumBoundingBox -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp ^
--criterion=length --features 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/minimum_bounding_box.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 14/09/2018</p>
<p><a name="MinimumBoundingCircle"></a></p>
<a class="header" href="#minimumboundingcircle" id="minimumboundingcircle"><h1>MinimumBoundingCircle</h1></a>
<p>This tool delineates the minimum bounding circle (MBC) for a group of vectors. The MBC is the smallest enclosing
circle to completely enclose a feature.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#minimumboundingbox"><strong>MinimumBoundingBox</strong></a>, <a href="./gis_analysis.html#minimumboundingenvelope"><strong>MinimumBoundingEnvelope</strong></a>, <a href="./gis_analysis.html#minimumconvexhull"><strong>MinimumConvexHull</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--features         </td><td> Find the minimum bounding circle around each individual vector feature</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">minimum_bounding_circle(
    i, 
    output, 
    features=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinimumBoundingCircle -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp --features 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/minimum_bounding_circle.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 14/09/2018</p>
<p><a name="MinimumBoundingEnvelope"></a></p>
<a class="header" href="#minimumboundingenvelope" id="minimumboundingenvelope"><h1>MinimumBoundingEnvelope</h1></a>
<p>This tool delineates the minimum bounding axis-aligned box for a group of vector features. The is the smallest
rectangle to completely enclose a feature, in which the sides of the envelope are aligned with the x and y
axis of the coordinate system. The <a href="./gis_analysis.html#minimumboundingbox"><strong>MinimumBoundingBox</strong></a> can be used instead to find the smallest possible
non-axis aligned rectangular envelope.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#minimumboundingbox"><strong>MinimumBoundingBox</strong></a>, <a href="./gis_analysis.html#minimumboundingcircle"><strong>MinimumBoundingCircle</strong></a>, <a href="./gis_analysis.html#minimumconvexhull"><strong>MinimumConvexHull</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--features         </td><td> Find the minimum bounding envelop around each individual vector feature</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">minimum_bounding_envelope(
    i, 
    output, 
    features=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinimumBoundingEnvelope -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp --features 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/minimum_bounding_envelope.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 31/09/2018</p>
<p><a name="MinimumConvexHull"></a></p>
<a class="header" href="#minimumconvexhull" id="minimumconvexhull"><h1>MinimumConvexHull</h1></a>
<p>This tool creates a vector convex polygon around vector features. The convex hull
is a convex closure of a set of points or polygon verticies and can be may be
conceptualized as the shape enclosed by a rubber band stretched around the point
set. The convex hull has many applications and is most notably used in various
shape indices. The Delaunay triangulation of a point set and its dual, the
Voronoi diagram, are mathematically related to convex hulls.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#minimumboundingbox"><strong>MinimumBoundingBox</strong></a>, <a href="./gis_analysis.html#minimumboundingcircle"><strong>MinimumBoundingCircle</strong></a>, <a href="./gis_analysis.html#minimumboundingenvelope"><strong>MinimumBoundingEnvelope</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--features         </td><td> Find the hulls around each vector feature</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">minimum_convex_hull(
    i, 
    output, 
    features=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinimumConvexHull -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp --features 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/minimum_convex_hull.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 03/09/2018</p>
<p><a name="NearestNeighbourGridding"></a></p>
<a class="header" href="#nearestneighbourgridding" id="nearestneighbourgridding"><h1>NearestNeighbourGridding</h1></a>
<p>Creates a raster grid based on a set of vector points and assigns grid values using the nearest neighbour.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector Points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>--use_z            </td><td> Use z-coordinate instead of field?</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
<tr><td>--base             </td><td> Optionally specified input base raster file. Not used when a cell size is specified</td></tr>
<tr><td>--max_dist         </td><td> Maximum search distance (optional)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">nearest_neighbour_gridding(
    i, 
    field, 
    output, 
    use_z=False, 
    cell_size=None, 
    base=None, 
    max_dist=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NearestNeighbourGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --field=ELEV -o=output.tif ^
--cell_size=1.0
&gt;&gt;./whitebox_tools -r=NearestNeighbourGridding ^
-v --wd=&quot;/path/to/data/&quot; -i=points.shp --use_z -o=output.tif ^
--base=existing_raster.tif --max_dist=5.5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/nearest_neighbour_gridding.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 09/10/2018</p>
<p><a name="PolygonArea"></a></p>
<a class="header" href="#polygonarea" id="polygonarea"><h1>PolygonArea</h1></a>
<p>This tool calculates the area of vector polygons, adding the result to the
vector's attribute table (AREA field). The area calculation will account
for any holes contained within polygons. The vector should be in a
projected coordinate system.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">polygon_area(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PolygonArea -v --wd=&quot;/path/to/data/&quot; ^
--input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/polygon_area.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/09/2018</p>
<p><a name="PolygonLongAxis"></a></p>
<a class="header" href="#polygonlongaxis" id="polygonlongaxis"><h1>PolygonLongAxis</h1></a>
<p>This tool can be used to map the long axis of polygon features. The long axis is the
longer of the two primary axes of the minimum bounding box (MBB), i.e. the smallest box
to completely enclose a feature. The long axis is drawn for each polygon in the input
vector file such that it passes through the centre point of the MBB. The output file is
therefore a vector of simple two-point polylines forming a vector field.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygons file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">polygon_long_axis(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PolygonLongAxis -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/polygon_long_axis.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 14/09/2018</p>
<p><a name="PolygonPerimeter"></a></p>
<a class="header" href="#polygonperimeter" id="polygonperimeter"><h1>PolygonPerimeter</h1></a>
<p>This tool calculates the perimeter of vector polygons, adding the result
to the vector's attribute table (PERIMETER field). The area calculation will
account for any holes contained within polygons. The vector should be in a
a projected coordinate system.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">polygon_perimeter(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PolygonPerimeter -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/polygon_perimeter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/09/2018</p>
<p><a name="PolygonShortAxis"></a></p>
<a class="header" href="#polygonshortaxis" id="polygonshortaxis"><h1>PolygonShortAxis</h1></a>
<p>This tool can be used to map the short axis of polygon features. The short axis is the
shorter of the two primary axes of the minimum bounding box (MBB), i.e. the smallest box
to completely enclose a feature. The short axis is drawn for each polygon in the input
vector file such that it passes through the centre point of the MBB. The output file is
therefore a vector of simple two-point polylines forming a vector field.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygons file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">polygon_short_axis(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PolygonShortAxis -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/polygon_short_axis.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 15/09/2018</p>
<p><a name="RasterCellAssignment"></a></p>
<a class="header" href="#rastercellassignment" id="rastercellassignment"><h1>RasterCellAssignment</h1></a>
<p>This tool can be used to create a new raster with the same coordinates and dimensions
(i.e. rows and columns) as an existing base image. Grid cells in the new raster will be
assigned either the row or column number or the x- or y-coordinate, depending on the
selected option (<code>--assign</code> flag). The user must also specify the name of the base
image (<code>--input</code>).</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#newrasterfrombase"><strong>NewRasterFromBase</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>-a, --assign       </td><td> Which variable would you like to assign to grid cells? Options include 'column', 'row', 'x', and 'y'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">raster_cell_assignment(
    i, 
    output, 
    assign=&quot;column&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RasterCellAssignment -v ^
--wd=&quot;/path/to/data/&quot; -i='input.tif' -o=output.tif ^
--assign='column' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/raster_cell_assignment.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Daniel Newman</p>
<p><em>Created</em>: August 10, 2017</p>
<p><a name="Reclass"></a></p>
<a class="header" href="#reclass" id="reclass"><h1>Reclass</h1></a>
<p>Reclassifies the values in a raster image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--reclass_vals     </td><td> Reclassification triplet values (new value; from value; to less than), e.g. '0.0;0.0;1.0;1.0;1.0;2.0'</td></tr>
<tr><td>--assign_mode      </td><td> Optional Boolean flag indicating whether to operate in assign mode, reclass_vals values are interpreted as new value; old value pairs</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">reclass(
    i, 
    output, 
    reclass_vals, 
    assign_mode=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Reclass -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif ^
--reclass_vals='0.0;0.0;1.0;1.0;1.0;2.0'
&gt;&gt;./whitebox_tools ^
-r=Reclass -v --wd=&quot;/path/to/data/&quot; -i='input.tif' ^
-o=output.tif --reclass_vals='10;1;20;2;30;3;40;4' ^
--assign_mode 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/reclass.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 9, 2017</p>
<p><a name="ReclassEqualInterval"></a></p>
<a class="header" href="#reclassequalinterval" id="reclassequalinterval"><h1>ReclassEqualInterval</h1></a>
<p>Reclassifies the values in a raster image based on equal-ranges.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--interval         </td><td> Class interval size</td></tr>
<tr><td>--start_val        </td><td> Optional starting value (default is input minimum value)</td></tr>
<tr><td>--end_val          </td><td> Optional ending value (default is input maximum value)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">reclass_equal_interval(
    i, 
    output, 
    interval=10.0, 
    start_val=None, 
    end_val=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ReclassEqualInterval -v ^
--wd=&quot;/path/to/data/&quot; -i='input.tif' -o=output.tif ^
--interval=10.0 --start_val=0.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/reclass_equal_interval.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="ReclassFromFile"></a></p>
<a class="header" href="#reclassfromfile" id="reclassfromfile"><h1>ReclassFromFile</h1></a>
<p>Reclassifies the values in a raster image using reclass ranges in a text file.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--reclass_file     </td><td> Input text file containing reclass ranges</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">reclass_from_file(
    i, 
    reclass_file, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ReclassFromFile -v ^
--wd=&quot;/path/to/data/&quot; -i='input.tif' ^
--reclass_file='reclass.txt' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/reclass_from_file.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 10, 2017</p>
<p><a name="SmoothVectors"></a></p>
<a class="header" href="#smoothvectors" id="smoothvectors"><h1>SmoothVectors</h1></a>
<p>This tool smooths a vector coverage of either a POLYLINE or POLYGON base ShapeType. The algorithm
uses a simple moving average method for smoothing, where the size of the averaging window is specified
by the user. The default filter size is 3 and can be any odd integer larger than or equal to 3. The
larger the averaging window, the greater the degree of line smoothing.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector POLYLINE or POLYGON file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--filter           </td><td> The filter size, any odd integer greater than or equal to 3; default is 3</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">smooth_vectors(
    i, 
    output, 
    filter=3, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SmoothVectors -v --wd=&quot;/path/to/data/&quot; ^
-i=in_file.shp -o=out_file.shp --filter=9 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/smooth_vectors.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 01/10/2018</p>
<p><a name="TinGridding"></a></p>
<a class="header" href="#tingridding" id="tingridding"><h1>TinGridding</h1></a>
<p>Creates a raster grid based on a triangular irregular network (TIN) fitted to vector points
and linear interpolation within each triangular-shaped plane.</p>
<p><em>See Also</em>:</p>
<p><a href="./lidar_tools.html#lidartingridding"><strong>LidarTINGridding</strong></a>, <a href="./gis_analysis.html#constructvectortin"><strong>ConstructVectorTIN</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>--use_z            </td><td> Use the 'z' dimension of the Shapefile's geometry instead of an attribute field?</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">tin_gridding(
    i, 
    output, 
    resolution, 
    field=None, 
    use_z=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TINGridding -v --wd=&quot;/path/to/data/&quot; ^
-i=points.shp --field=HEIGHT -o=tin.shp ^
--resolution=10.0
&gt;&gt;./whitebox_tools -r=TINGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=points.shp --use_z -o=tin.shp ^
--resolution=5.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/tin_gridding.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 23/09/2018</p>
<p><a name="VectorHexBinning"></a></p>
<a class="header" href="#vectorhexbinning" id="vectorhexbinning"><h1>VectorHexBinning</h1></a>
<p>The practice of binning point data to form a type of 2D histogram, density plot,
or what is sometimes called a heatmap, is quite useful as an alternative for the
cartographic display of of very dense points sets. This is particularly the case
when the points experience significant overlap at the displayed scale. The
<code>PointDensity</code> tool can be used to perform binning based on a regular grid (raster
output). This tool, by comparison, bases the binning on a hexagonal grid.</p>
<p>The tool is similar to the <a href="./gis_analysis.html#createhexagonalvectorgrid"><strong>CreateHexagonalVectorGrid</strong></a> tool, however instead will
create an output hexagonal grid in which each hexagonal cell possesses a <code>COUNT</code>
attribute which specifies the number of points from an input points file (Shapefile
vector) that are contained within the hexagonal cell.</p>
<p>In addition to the names of the input points file and the output Shapefile, the user
must also specify the desired hexagon width (w), which is the distance between opposing
sides of each hexagon. The size (s) each side of the hexagon can then be calculated as,
s = w / [2 x cos(PI / 6)]. The area of each hexagon (A) is, A = 3s(w / 2). The user must
also specify the orientation of the grid with options of horizontal (pointy side up) and
vertical (flat side up).</p>
<p><em>See Also</em>:</p>
<p><a href="./lidar_tools.html#lidarhexbinning"><strong>LidarHexBinning</strong></a>, <code>PointDensity</code>, <a href="./gis_analysis.html#createhexagonalvectorgrid"><strong>CreateHexagonalVectorGrid</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input base file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--width            </td><td> The grid cell width</td></tr>
<tr><td>--orientation      </td><td> Grid Orientation, 'horizontal' or 'vertical'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">vector_hex_binning(
    i, 
    output, 
    width, 
    orientation=&quot;horizontal&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=VectorHexBinning -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.shp --width=10.0 ^
--orientation=vertical 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/vector_hex_bin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 16/09/2018</p>
<p><a name="VoronoiDiagram"></a></p>
<a class="header" href="#voronoidiagram" id="voronoidiagram"><h1>VoronoiDiagram</h1></a>
<p>This tool creates a vector Voronoi diagram for a set of vector points. The
Voronoi diagram is the dual graph of the Delaunay triangulation. The tool
operates by first constructing the Delaunay triangulation and then
connecting the circumcenters of each triangle. Each Voronoi cell contains
one point of the input vector points. All locations within the cell are
nearer to the contained point than any other input point.</p>
<p>A dense frame of 'ghost' (hidden) points is inserted around the input point
set to limit the spatial extent of the diagram. The frame is set back from
the bounding box of the input points by 2 x the average point  spacing. The
polygons of these ghost points are not output, however, points that are
situated along the edges of the data will have somewhat rounded (paraboloic)
exterior boundaries as a result of this edge condition. If this property is
unacceptable for application, clipping the Voronoi diagram to the convex
hull may be a better alternative.</p>
<p>This tool works on vector input data only. If a Voronoi diagram is needed
to tesselate regions associated with a set of raster points, use the
<a href="./gis_analysis_distance_tools.html#euclideanallocation"><strong>EuclideanAllocation</strong></a> tool instead. To use Voronoi diagrams for gridding
data (i.e. raster interpolation), use the <a href="./gis_analysis.html#nearestneighbourgridding"><strong>NearestNeighbourGridding</strong></a> tool.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#constructvectortin"><strong>ConstructVectorTIN</strong></a>, <a href="./gis_analysis_distance_tools.html#euclideanallocation"><strong>EuclideanAllocation</strong></a>, <a href="./gis_analysis.html#nearestneighbourgridding"><strong>NearestNeighbourGridding</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector points file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">voronoi_diagram(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=VoronoiDiagram -v --wd=&quot;/path/to/data/&quot; ^
-i=points.shp -o=tin.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/voronoi_diagram.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 03/10/2018</p>
<a class="header" href="#gis-analysis--distance-tools" id="gis-analysis--distance-tools"><h1>GIS Analysis → Distance Tools</h1></a>
<ul>
<li><a href="#BufferRaster">BufferRaster</a></li>
<li><a href="#CostAllocation">CostAllocation</a></li>
<li><a href="#CostDistance">CostDistance</a></li>
<li><a href="#CostPathway">CostPathway</a></li>
<li><a href="#EuclideanAllocation">EuclideanAllocation</a></li>
<li><a href="#EuclideanDistance">EuclideanDistance</a></li>
</ul>
<p><a name="BufferRaster"></a></p>
<a class="header" href="#bufferraster" id="bufferraster"><h1>BufferRaster</h1></a>
<p>This tool can be used to identify an area of interest within a specified distance of
features of interest in a raster data set.</p>
<p>The Euclidean distance (i.e. straight-line distance) is calculated between each grid
cell and the nearest 'target cell' in the input image. Distance is calcualted using the
efficient method of Shih and Wu (2004). Target cells are all non-zero,
non-NoData grid cells. Because NoData values in the input image are assigned the NoData
value in the output image, the only valid background value in the input image is zero.</p>
<p>The user must specify the input and output image names, the desired buffer size (<code>--size</code>), and,
optionally, whether the distance units are measured in grid cells (i.e. <code>--gridcells</code> flag).
If the <code>--gridcells</code> flag is not specified, the linear units of the raster's coordinate
reference system will be used.</p>
<p><em>Reference</em>:</p>
<p>Shih FY and Wu Y-T (2004), Fast Euclidean distance transformation in two scans using a 3 x 3
neighborhood, <em>Computer Vision and Image Understanding</em>, 93: 195-205.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_distance_tools.html#euclideandistance"><strong>EuclideanDistance</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--size             </td><td> Buffer size</td></tr>
<tr><td>--gridcells        </td><td> Optional flag to indicate that the 'size' threshold should be measured in grid cells instead of the default map units</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">buffer_raster(
    i, 
    output, 
    size, 
    gridcells=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=BufferRaster -v --wd=&quot;/path/to/data/&quot; ^
-i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/buffer_raster.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 22/06/2017</p>
<p><a name="CostAllocation"></a></p>
<a class="header" href="#costallocation" id="costallocation"><h1>CostAllocation</h1></a>
<p>This tool can be used to identify the 'catchment area' of each source grid cell in a
cost-distance analysis. The user must specify the names of the input <em>source</em> and
<em>back-link</em> raster files. Source cells (i.e. starting points for the cost-distance or
least-cost path analysis) are designated as all positive, non-zero valued grid cells
in the <em>source</em> raster. A <em>back-link</em> raster file can be created using the <a href="./gis_analysis_distance_tools.html#costdistance"><strong>CostDistance</strong></a>
tool and is conceptually similar to the D8 flow-direction pointer raster grid in that
it describes the connectivity between neighbouring cells on the accumulated cost surface.</p>
<p>NoData values in the input <em>back-link</em> image are assigned NoData values in the output
image.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_distance_tools.html#costdistance"><strong>CostDistance</strong></a>, <a href="./gis_analysis_distance_tools.html#costpathway"><strong>CostPathway</strong></a>, <a href="./gis_analysis_distance_tools.html#euclideanallocation"><strong>EuclideanAllocation</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--source           </td><td> Input source raster file</td></tr>
<tr><td>--backlink         </td><td> Input backlink raster file generated by the cost-distance tool</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">cost_allocation(
    source, 
    backlink, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CostAllocation -v --wd=&quot;/path/to/data/&quot; ^
--source='source.tif' --backlink='backlink.tif' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/cost_allocation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 4, 2017</p>
<p><a name="CostDistance"></a></p>
<a class="header" href="#costdistance" id="costdistance"><h1>CostDistance</h1></a>
<p>This tool can be used to perform cost-distance or least-cost pathway analyses. Specifically,
this tool can be used to calculate the accumulated cost of traveling from the 'source grid
cell' to each other grid cell in a raster dataset. It is based on the costs associated with
traveling through each cell along a pathway represented in a cost (or friction) surface. If
there are multiple source grid cells, each cell in the resulting cost-accumulation surface
will reflect the accumulated cost to the source cell that is connected by the minimum accumulated
cost-path. The user must specify the names of the raster file containing the source cells
(<code>--source</code>), the raster file containing the cost surface information (<code>--cost</code>), the output
cost-accumulation surface raster (<code>--out_accum</code>), and the output back-link raster (<code>--out_backlink</code>).
Source cells are designated as all positive, non-zero valued grid cells in the source raster.
The cost (friction) raster can be created by combining the various cost factors associated with
the specific problem (e.g. slope gradient, visibility, etc.) using a raster calculator or the
<a href="./gis_analysis_overlay_tools.html#weightedoverlay"><strong>WeightedOverlay</strong></a> tool.</p>
<p>While the cost-accumulation surface raster can be helpful for visualizing
the three-dimensional characteristics of the 'cost landscape', it is actually the back-link raster
that is used as inputs to the other two cost-distance tools, <a href="./gis_analysis_distance_tools.html#costallocation"><strong>CostAllocation</strong></a> and <a href="./gis_analysis_distance_tools.html#costpathway"><strong>CostPathway</strong></a>, to
determine the least-cost linkages among neighbouring grid cells on the cost surface. If the
accumulated cost surface is analogous to a digital elevation model (DEM) then the back-link raster
is equivalent to the D8 flow-direction pointer. In fact, it is created in a similar way and uses
the same convention for designating 'flow directions' between neighbouring grid cells. The algorithm
for the cost distance accumulation operation uses a type of priority-flood method similar to
what is used for depression filling and flow accumulation operations.</p>
<p>NoData values in the input cost surface image are ignored during processing and assigned NoData values
in the outputs. The output cost accumulation raster is of the float data type and continuous data scale.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_distance_tools.html#costallocation"><strong>CostAllocation</strong></a>, <a href="./gis_analysis_distance_tools.html#costpathway"><strong>CostPathway</strong></a>, <a href="./gis_analysis_overlay_tools.html#weightedoverlay"><strong>WeightedOverlay</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--source           </td><td> Input source raster file</td></tr>
<tr><td>--cost             </td><td> Input cost (friction) raster file</td></tr>
<tr><td>--out_accum        </td><td> Output cost accumulation raster file</td></tr>
<tr><td>--out_backlink     </td><td> Output backlink raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">cost_distance(
    source, 
    cost, 
    out_accum, 
    out_backlink, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CostDistance -v --wd=&quot;/path/to/data/&quot; ^
--source=src.tif --cost=cost.tif --out_accum=accum.tif ^
--out_backlink=backlink.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/cost_distance.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 4, 2017</p>
<p><a name="CostPathway"></a></p>
<a class="header" href="#costpathway" id="costpathway"><h1>CostPathway</h1></a>
<p>This tool can be used to map the least-cost pathway connecting each destination grid cell
in a cost-distance analysis to a source cell. The user must specify the names of the input
<em>destination</em> and <em>back-link</em> raster files. Destination cells (i.e. end points for the
least-cost path analysis) are designated as all positive, non-zero valued grid cells in the
<em>destination</em> raster. A <em>back-link</em> raster file can be created using the <a href="./gis_analysis_distance_tools.html#costdistance"><strong>CostDistance</strong></a>
tool and is conceptually similar to the D8 flow-direction pointer raster grid in that it
describes the connectivity between neighbouring cells on the accumulated cost surface. All
background grid cells in the output image are assigned the NoData value.</p>
<p>NoData values in the input <em>back-link</em> image are assigned NoData values in the output image.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_distance_tools.html#costdistance"><strong>CostDistance</strong></a>, <a href="./gis_analysis_distance_tools.html#costallocation"><strong>CostAllocation</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--destination      </td><td> Input destination raster file</td></tr>
<tr><td>--backlink         </td><td> Input backlink raster file generated by the cost-distance tool</td></tr>
<tr><td>-o, --output       </td><td> Output cost pathway raster file</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether zero values should be treated as a background</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">cost_pathway(
    destination, 
    backlink, 
    output, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CostPathway -v --wd=&quot;/path/to/data/&quot; ^
--destination=dst.tif --backlink=backlink.tif ^
--output=cost_path.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/cost_pathway.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 4, 2017</p>
<p><a name="EuclideanAllocation"></a></p>
<a class="header" href="#euclideanallocation" id="euclideanallocation"><h1>EuclideanAllocation</h1></a>
<p>This tool assigns grid cells in the output image the value of the nearest target cell in
the input image, measured by the Euclidean distance (i.e. straight-line distance). Thus,
<a href="./gis_analysis_distance_tools.html#euclideanallocation"><strong>EuclideanAllocation</strong></a> essentially creates the Voronoi diagram for a set of target cells.
Target cells are all non-zero, non-NoData grid cells in the input image. Distances are
calculated using the same efficient algorithm (Shih and Wu, 2003) as the <a href="./gis_analysis_distance_tools.html#euclideandistance"><strong>EuclideanDistance</strong></a>
tool.</p>
<p><em>Reference</em>:</p>
<p>Shih FY and Wu Y-T (2004), Fast Euclidean distance transformation in two scans using a 3 x 3
neighborhood, <em>Computer Vision and Image Understanding</em>, 93: 195-205.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_distance_tools.html#euclideandistance"><strong>EuclideanDistance</strong></a>, <a href="./gis_analysis.html#voronoidiagram"><strong>VoronoiDiagram</strong></a>, <a href="./gis_analysis_distance_tools.html#costallocation"><strong>CostAllocation</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">euclidean_allocation(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=EuclideanAllocation -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/euclidean_allocation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="EuclideanDistance"></a></p>
<a class="header" href="#euclideandistance" id="euclideandistance"><h1>EuclideanDistance</h1></a>
<p>This tool will estimate the Euclidean distance (i.e. straight-line distance) between each
grid cell and the nearest 'target cell' in the input image. Target cells are all non-zero,
non-NoData grid cells. Distance in the output image is measured in the same units as the
horizontal units of the input image.</p>
<p><em>Algorithm Description</em>:</p>
<p>The algorithm is based on the highly efficient distance transform of Shih and Wu (2003).
It makes four passes of the image; the first pass initializes the output image; the second
and third passes calculate the minimum squared Euclidean distance by examining the 3 x 3
neighbourhood surrounding each cell; the last pass takes the square root of cell values,
transforming them into true Euclidean distances, and deals with NoData values that may be
present. All NoData value grid cells in the input image will contain NoData values in the
output image. As such, NoData is not a suitable background value for non-target cells.
Background areas should be designated with zero values.</p>
<p><em>Reference</em>:</p>
<p>Shih FY and Wu Y-T (2004), Fast Euclidean distance transformation in two scans using a 3 x 3
neighborhood, <em>Computer Vision and Image Understanding</em>, 93: 195-205.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_distance_tools.html#euclideanallocation"><strong>EuclideanAllocation</strong></a>, <a href="./gis_analysis_distance_tools.html#costdistance"><strong>CostDistance</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">euclidean_distance(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=EuclideanDistance -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/euclidean_distance.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<a class="header" href="#gis-analysis--overlay-tools" id="gis-analysis--overlay-tools"><h1>GIS Analysis → Overlay Tools</h1></a>
<ul>
<li><a href="#AverageOverlay">AverageOverlay</a></li>
<li><a href="#Clip">Clip</a></li>
<li><a href="#ClipRasterToPolygon">ClipRasterToPolygon</a></li>
<li><a href="#CountIf">CountIf</a></li>
<li><a href="#Difference">Difference</a></li>
<li><a href="#Erase">Erase</a></li>
<li><a href="#ErasePolygonFromRaster">ErasePolygonFromRaster</a></li>
<li><a href="#HighestPosition">HighestPosition</a></li>
<li><a href="#Intersect">Intersect</a></li>
<li><a href="#LineIntersections">LineIntersections</a></li>
<li><a href="#LowestPosition">LowestPosition</a></li>
<li><a href="#MaxAbsoluteOverlay">MaxAbsoluteOverlay</a></li>
<li><a href="#MaxOverlay">MaxOverlay</a></li>
<li><a href="#MinAbsoluteOverlay">MinAbsoluteOverlay</a></li>
<li><a href="#MinOverlay">MinOverlay</a></li>
<li><a href="#PercentEqualTo">PercentEqualTo</a></li>
<li><a href="#PercentGreaterThan">PercentGreaterThan</a></li>
<li><a href="#PercentLessThan">PercentLessThan</a></li>
<li><a href="#PickFromList">PickFromList</a></li>
<li><a href="#Polygonize">Polygonize</a></li>
<li><a href="#SplitWithLines">SplitWithLines</a></li>
<li><a href="#SumOverlay">SumOverlay</a></li>
<li><a href="#SymmetricalDifference">SymmetricalDifference</a></li>
<li><a href="#Union">Union</a></li>
<li><a href="#WeightedOverlay">WeightedOverlay</a></li>
<li><a href="#WeightedSum">WeightedSum</a></li>
</ul>
<p><a name="AverageOverlay"></a></p>
<a class="header" href="#averageoverlay" id="averageoverlay"><h1>AverageOverlay</h1></a>
<p>Calculates the average for each grid cell from a group of raster images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">average_overlay(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AverageOverlay -v --wd='/path/to/data/' ^
-i='image1.dep;image2.dep;image3.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/average_overlay.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="Clip"></a></p>
<a class="header" href="#clip" id="clip"><h1>Clip</h1></a>
<p>This tool will extract all the features, or parts of features, that overlap with the features
of the clip vector file. The clipping operation is one of the most common vector overlay
operations in GIS and effectively imposes the boundary of the clip layer on a set of input
vector features, or target features. The operation is sometimes likened to a 'cookie-cutter'.
The input vector file can be of any feature type (i.e. points, lines, polygons), however, the
clip vector must consist of polygons.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#erase"><strong>Erase</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>--clip             </td><td> Input clip polygon vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">clip(
    i, 
    clip, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Clip -v --wd=&quot;/path/to/data/&quot; ^
-i=lines1.shp --clip=clip_poly.shp -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/clip.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 28/10/2018</p>
<p><a name="ClipRasterToPolygon"></a></p>
<a class="header" href="#cliprastertopolygon" id="cliprastertopolygon"><h1>ClipRasterToPolygon</h1></a>
<p>Clips a raster to a vector polygon.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--polygons         </td><td> Input vector polygons file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--maintain_dimensions</td><td> Maintain input raster dimensions?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">clip_raster_to_polygon(
    i, 
    polygons, 
    output, 
    maintain_dimensions=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ClipRasterToPolygon -v ^
--wd=&quot;/path/to/data/&quot; -i=raster.tif --polygons=poly.shp ^
-o=output.tif --maintain_dimensions 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/clip_raster_to_polygon.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/04/2018</p>
<p><a name="CountIf"></a></p>
<a class="header" href="#countif" id="countif"><h1>CountIf</h1></a>
<p>Counts the number of occurrences of a specified value in a cell-stack of rasters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--value            </td><td> Search value (e.g. countif value = 5.0)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">count_if(
    inputs, 
    output, 
    value, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CountIf -v --wd='/path/to/data/' ^
-i='image1.dep;image2.dep;image3.tif' -o=output.tif ^
--value=5.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/count_if.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 14/04/2018</p>
<p><a name="Difference"></a></p>
<a class="header" href="#difference" id="difference"><h1>Difference</h1></a>
<p>This tool will remove all the overlapping features, or parts of overlapping features, between
input and overlay vector files, outputting only the features that occur in one of the two
inputs but not both. The <em>Symmetrical Difference</em> is related to the Boolean
exclusive-or (<strong>XOR</strong>) operation in  set theory and is one of the common vector overlay
operations in GIS. The user must specify  the names of the input and overlay vector files
as well as the output vector file name. The tool operates on vector points,
lines, or polygon, but both the input and overlay files must contain the same ShapeType.</p>
<p>The <em>Symmetrical Difference</em> can also be derived using a combination of other vector
overlay operations, as either <code>(A union B) difference (A intersect B)</code>, or
<code>(A difference B) union (B difference A)</code>.</p>
<p>The attributes of the two input vectors will be merged in the output attribute table.
Fields that are duplicated between the inputs will share a single attribute in the
output. Fields that only exist in one of the two inputs will be populated by <code>null</code>
in the output table. Multipoint ShapeTypes however will simply contain a single
ouptut feature indentifier (<code>FID</code>) attribute. Also, note that depending on the
ShapeType (polylines and polygons), <code>Measure</code> and <code>Z</code> ShapeDimension data will not
be transfered to the output geometries. If the input attribute table contains fields
that measure the geometric properties of their associated features (e.g. length or area),
these fields will not be updated to reflect changes in geometry shape and size
resulting from the overlay operation.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#intersect"><strong>Intersect</strong></a>, <a href="./gis_analysis_overlay_tools.html#difference"><strong>Difference</strong></a>, <a href="./gis_analysis_overlay_tools.html#union"><strong>Union</strong></a>, <a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a>, <a href="./gis_analysis_overlay_tools.html#erase"><strong>Erase</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>--overlay          </td><td> Input overlay vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">difference(
    i, 
    overlay, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Difference -v --wd=&quot;/path/to/data/&quot; ^
-input=layer1.shp --overlay=layer2.shp -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/difference.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 8/11/2018</p>
<p><a name="Erase"></a></p>
<a class="header" href="#erase" id="erase"><h1>Erase</h1></a>
<p>This tool will remove all the features, or parts of features, that overlap with the features
of the erase vector file. The erasing operation is one of the most common vector overlay
operations in GIS and effectively imposes the boundary of the erase layer on a set of input
vector features, or target features.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>--erase            </td><td> Input erase polygon vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">erase(
    i, 
    erase, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Erase -v --wd=&quot;/path/to/data/&quot; ^
-i=lines1.shp --erase=erase_poly.shp -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/erase.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 4/11/2018</p>
<p><a name="ErasePolygonFromRaster"></a></p>
<a class="header" href="#erasepolygonfromraster" id="erasepolygonfromraster"><h1>ErasePolygonFromRaster</h1></a>
<p>Erases (cuts out) a vector polygon from a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--polygons         </td><td> Input vector polygons file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">erase_polygon_from_raster(
    i, 
    polygons, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ErasePolygonFromRaster -v ^
--wd=&quot;/path/to/data/&quot; -i='DEM.tif' --polygons='lakes.shp' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/erase_polygon_from_raster.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/03/2018</p>
<p><a name="HighestPosition"></a></p>
<a class="header" href="#highestposition" id="highestposition"><h1>HighestPosition</h1></a>
<p>Identifies the stack position of the maximum value within a raster stack on a cell-by-cell basis.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">highest_position(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HighestPosition -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/highest_pos.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="Intersect"></a></p>
<a class="header" href="#intersect" id="intersect"><h1>Intersect</h1></a>
<p>The result of the <a href="./gis_analysis_overlay_tools.html#intersect"><strong>Intersect</strong></a> vector overlay operation includes all the feature parts that
occur in both input layers, excluding all other parts. It is analogous to the <strong>OR</strong>
logical operator and multiplication in arithmetic. This tool is one of the common vector overlay
operations in GIS. The user must specify the names of the input and overlay vector files
as well as the output vector file name. The tool operates on vector points,
lines, or polygon, but both the input and overlay files must contain the same ShapeType.</p>
<p>The <a href="./gis_analysis_overlay_tools.html#intersect"><strong>Intersect</strong></a> tool is similar to the <a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a> tool. The difference is that the overlay vector
layer in a <a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a> operation must always be polygons, regardless of whether the input layer
consists of points or polylines.</p>
<p>The attributes of the two input vectors will be merged in the output attribute table.
Note, duplicate fields should not exist between the inputs layers, as they will share a
single attribute in the output (assigned from the first layer). Multipoint ShapeTypes will
simply contain a single  ouptut feature indentifier (<code>FID</code>) attribute. Also, note that depending
on the ShapeType (polylines and polygons), <code>Measure</code> and <code>Z</code> ShapeDimension data will not
be transfered to the output geometries. If the input attribute table contains fields
that measure the geometric properties of their associated features (e.g. length or area),
these fields will not be updated to reflect changes in geometry shape and size
resulting from the overlay operation.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#difference"><strong>Difference</strong></a>, <a href="./gis_analysis_overlay_tools.html#union"><strong>Union</strong></a>, <a href="./gis_analysis_overlay_tools.html#symmetricaldifference"><strong>SymmetricalDifference</strong></a>, <a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a>, <a href="./gis_analysis_overlay_tools.html#erase"><strong>Erase</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>--overlay          </td><td> Input overlay vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--snap             </td><td> Snap tolerance</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">intersect(
    i, 
    overlay, 
    output, 
    snap=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Intersect -v --wd=&quot;/path/to/data/&quot; ^
-input=layer1.shp --overlay=layer2.shp -o=out_file.shp ^
--snap=0.0000001 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/intersect.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 8/11/2018</p>
<p><a name="LineIntersections"></a></p>
<a class="header" href="#lineintersections" id="lineintersections"><h1>LineIntersections</h1></a>
<p>This tool identifies points where the features of two vector line/polygon layers
intersect. The user must specify the names of two input vector line files and the
output file. The output file will be a vector of POINT ShapeType. If the input
vectors intersect at a line segment, the beginning and end vertices of the segment
will be present in the output file. A warning is issued if intersection line segments
are identified during analysis. If no intersections are found between the input line
files, the output file will not be saved and a warning will be issued.</p>
<p>Each intersection point will contain <code>PARENT1</code> and <code>PARENT2</code> attribute fields,
identifying the instersecting features in the first and second input line files
respectively. Additionally, the output attribute table will contain all of the
attributes (excluding <code>FID</code>s) of the two parent line features.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input vector polyline file</td></tr>
<tr><td>--i2, --input2    </td><td> Input vector polyline file</td></tr>
<tr><td>-o, --output       </td><td> Output vector point file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">line_intersections(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LineIntersections -v ^
--wd=&quot;/path/to/data/&quot; --i1=lines1.shp --i2=lines2.shp ^
-o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/line_intersections.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 16/10/2018</p>
<p><a name="LowestPosition"></a></p>
<a class="header" href="#lowestposition" id="lowestposition"><h1>LowestPosition</h1></a>
<p>Identifies the stack position of the minimum value within a raster stack on a cell-by-cell basis.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lowest_position(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LowestPosition -v --wd='/path/to/data/' ^
-i='image1.tif;image2.tif;image3.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/lowest_pos.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 4, 2017</p>
<p><a name="MaxAbsoluteOverlay"></a></p>
<a class="header" href="#maxabsoluteoverlay" id="maxabsoluteoverlay"><h1>MaxAbsoluteOverlay</h1></a>
<p>Evaluates the maximum absolute value for each grid cell from a stack of input rasters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_absolute_overlay(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxAbsoluteOverlay -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/max_abs_overlay.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="MaxOverlay"></a></p>
<a class="header" href="#maxoverlay" id="maxoverlay"><h1>MaxOverlay</h1></a>
<p>Evaluates the maximum value for each grid cell from a stack of input rasters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_overlay(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxOverlay -v --wd='/path/to/data/' ^
-i='image1.tif;image2.tif;image3.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/max_overlay.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="MinAbsoluteOverlay"></a></p>
<a class="header" href="#minabsoluteoverlay" id="minabsoluteoverlay"><h1>MinAbsoluteOverlay</h1></a>
<p>Evaluates the minimum absolute value for each grid cell from a stack of input rasters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">min_absolute_overlay(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinAbsoluteOverlay -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/min_abs_overlay.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="MinOverlay"></a></p>
<a class="header" href="#minoverlay" id="minoverlay"><h1>MinOverlay</h1></a>
<p>Evaluates the minimum value for each grid cell from a stack of input rasters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">min_overlay(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinOverlay -v --wd='/path/to/data/' ^
-i='image1.tif;image2.tif;image3.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/min_overlay.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="PercentEqualTo"></a></p>
<a class="header" href="#percentequalto" id="percentequalto"><h1>PercentEqualTo</h1></a>
<p>Calculates the percentage of a raster stack that have cell values equal to an input on a cell-by-cell basis.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--comparison       </td><td> Input comparison raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">percent_equal_to(
    inputs, 
    comparison, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PercentEqualTo -v --wd='/path/to/data/' ^
-i='image1.tif;image2.tif;image3.tif' --comparison='comp.tif' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/percent_equal_to.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="PercentGreaterThan"></a></p>
<a class="header" href="#percentgreaterthan" id="percentgreaterthan"><h1>PercentGreaterThan</h1></a>
<p>Calculates the percentage of a raster stack that have cell values greather than an input on a cell-by-cell basis.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--comparison       </td><td> Input comparison raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">percent_greater_than(
    inputs, 
    comparison, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PercentGreaterThan -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
--comparison='comp.tif' -o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/percent_greater_than.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="PercentLessThan"></a></p>
<a class="header" href="#percentlessthan" id="percentlessthan"><h1>PercentLessThan</h1></a>
<p>Calculates the percentage of a raster stack that have cell values less than an input on a cell-by-cell basis.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--comparison       </td><td> Input comparison raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">percent_less_than(
    inputs, 
    comparison, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PercentLessThan -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
--comparison='comp.tif' -o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/percent_less_than.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="PickFromList"></a></p>
<a class="header" href="#pickfromlist" id="pickfromlist"><h1>PickFromList</h1></a>
<p>Outputs the value from a raster stack specified by a position raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--pos_input        </td><td> Input position raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">pick_from_list(
    inputs, 
    pos_input, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PickFromList -v --wd='/path/to/data/' ^
--pos_input=position.tif -i='image1.tif;image2.tif;image3.tif' ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/pick_from_list.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22 2017</p>
<p><a name="Polygonize"></a></p>
<a class="header" href="#polygonize" id="polygonize"><h1>Polygonize</h1></a>
<p>This tool outputs a vector polygon layer from two or more intersecting line features
contained in one or more input vector line files. Each space enclosed by the intersecting
line set is converted to polygon added to the output layer. This tool should not be
confused with the <a href="./data_tools.html#linestopolygons"><strong>LinesToPolygons</strong></a> tool, which can be used to convert a vector file of
polylines into a set of polygons, simply by closing each line feature. The <a href="./data_tools.html#linestopolygons"><strong>LinesToPolygons</strong></a>
tool does not deal with line intersection in the same way that the <a href="./gis_analysis_overlay_tools.html#polygonize"><strong>Polygonize</strong></a> tool does.</p>
<p><em>See Also</em>:</p>
<p><a href="./data_tools.html#linestopolygons"><strong>LinesToPolygons</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input vector polyline file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">polygonize(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Polygonize -v --wd=&quot;/path/to/data/&quot; ^
-i='lines1.shp;lines2.shp;lines3.shp' -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/polygonize.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 19/10/2018</p>
<p><a name="SplitWithLines"></a></p>
<a class="header" href="#splitwithlines" id="splitwithlines"><h1>SplitWithLines</h1></a>
<p>This tool splits the lines or polygons in one layer using the lines in another layer
to define the breaking points. Intersection points between geometries in both layers
are considered as split points. The input layer (<code>--input</code>) can be of either
POLYLINE or POLYGON ShapeType and the output file will share this geometry type.
The user must also specify an split layer (<code>--split</code>), of POLYLINE ShapeType, used
to bisect the input geometries.</p>
<p>Each split geometry's attribute record will contain <code>FID</code> and <code>PARENT_FID</code> values
and all of the attributes (excluding <code>FID</code>'s) of the input layer.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector line or polygon file</td></tr>
<tr><td>--split            </td><td> Input vector polyline file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">split_with_lines(
    i, 
    split, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SplitWithLines -v --wd=&quot;/path/to/data/&quot; ^
--input=polygons.shp --split=lines.shp -o=out_file.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/split_with_lines.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 17/10/2018</p>
<p><a name="SumOverlay"></a></p>
<a class="header" href="#sumoverlay" id="sumoverlay"><h1>SumOverlay</h1></a>
<p>This tool calculates the sum for each grid cell from a group of raster images.</p>
<p><em>Warning</em>:</p>
<p>Each of the input rasters must have the same spatial extent and number of rows
and columns.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#weightedsum"><strong>WeightedSum</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">sum_overlay(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SumOverlay -v --wd='/path/to/data/' ^
-i='image1.dep;image2.dep;image3.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/sum_overlay.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 26/09/2018</p>
<p><a name="SymmetricalDifference"></a></p>
<a class="header" href="#symmetricaldifference" id="symmetricaldifference"><h1>SymmetricalDifference</h1></a>
<p>This tool will remove all the overlapping features, or parts of overlapping features, between
input and overlay vector files, outputting only the features that occur in one of the two
inputs but not both. The <em>Symmetrical Difference</em> is related to the Boolean
exclusive-or (<strong>XOR</strong>) operation in  set theory and is one of the common vector overlay
operations in GIS. The user must specify  the names of the input and overlay vector files
as well as the output vector file name. The tool operates on vector points,
lines, or polygon, but both the input and overlay files must contain the same ShapeType.</p>
<p>The <em>Symmetrical Difference</em> can also be derived using a combination of other vector
overlay operations, as either <code>(A union B) difference (A intersect B)</code>, or
<code>(A difference B) union (B difference A)</code>.</p>
<p>The attributes of the two input vectors will be merged in the output attribute table.
Fields that are duplicated between the inputs will share a single attribute in the
output. Fields that only exist in one of the two inputs will be populated by <code>null</code>
in the output table. Multipoint ShapeTypes however will simply contain a single
ouptut feature indentifier (<code>FID</code>) attribute. Also, note that depending on the
ShapeType (polylines and polygons), <code>Measure</code> and <code>Z</code> ShapeDimension data will not
be transfered to the output geometries. If the input attribute table contains fields
that measure the geometric properties of their associated features (e.g. length or area),
these fields will not be updated to reflect changes in geometry shape and size
resulting from the overlay operation.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#intersect"><strong>Intersect</strong></a>, <a href="./gis_analysis_overlay_tools.html#difference"><strong>Difference</strong></a>, <a href="./gis_analysis_overlay_tools.html#union"><strong>Union</strong></a>, <a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a>, <a href="./gis_analysis_overlay_tools.html#erase"><strong>Erase</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>--overlay          </td><td> Input overlay vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--snap             </td><td> Snap tolerance</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">symmetrical_difference(
    i, 
    overlay, 
    output, 
    snap=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SymmetricalDifference -v ^
--wd=&quot;/path/to/data/&quot; -input=layer1.shp --overlay=layer2.shp ^
-o=out_file.shp --snap=0.0000001 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/symmetrical_difference.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 5/11/2018</p>
<p><a name="Union"></a></p>
<a class="header" href="#union" id="union"><h1>Union</h1></a>
<p>This tool splits vector layers at their overlaps, creating a layer containing all the portions from both
input and overlay layers. The <em>Union</em> is related to the Boolean
<strong>OR</strong> operation in  set theory and is one of the common vector overlay
operations in GIS. The user must specify  the names of the input and overlay vector files
as well as the output vector file name. The tool operates on vector points,
lines, or polygon, but both the input and overlay files must contain the same ShapeType.</p>
<p>The attributes of the two input vectors will be merged in the output attribute table.
Fields that are duplicated between the inputs will share a single attribute in the
output. Fields that only exist in one of the two inputs will be populated by <code>null</code>
in the output table. Multipoint ShapeTypes however will simply contain a single
ouptut feature indentifier (<code>FID</code>) attribute. Also, note that depending on the
ShapeType (polylines and polygons), <code>Measure</code> and <code>Z</code> ShapeDimension data will not
be transfered to the output geometries. If the input attribute table contains fields
that measure the geometric properties of their associated features (e.g. length or area),
these fields will not be updated to reflect changes in geometry shape and size
resulting from the overlay operation.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#intersect"><strong>Intersect</strong></a>, <a href="./gis_analysis_overlay_tools.html#difference"><strong>Difference</strong></a>, <a href="./gis_analysis_overlay_tools.html#symmetricaldifference"><strong>SymmetricalDifference</strong></a>, <a href="./gis_analysis_overlay_tools.html#clip"><strong>Clip</strong></a>, <a href="./gis_analysis_overlay_tools.html#erase"><strong>Erase</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector file</td></tr>
<tr><td>--overlay          </td><td> Input overlay vector file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--snap             </td><td> Snap tolerance</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">union(
    i, 
    overlay, 
    output, 
    snap=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Union -v --wd=&quot;/path/to/data/&quot; ^
-input=layer1.shp --overlay=layer2.shp -o=out_file.shp ^
--snap=0.0000001 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/union.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 5/11/2018</p>
<p><a name="WeightedOverlay"></a></p>
<a class="header" href="#weightedoverlay" id="weightedoverlay"><h1>WeightedOverlay</h1></a>
<p>This tool performs a weighted overlay on multiple input images. It can be used to
combine multiple factors with varying levels of weight or relative importance. The
WeightedOverlay tool is similar to the WeightedSum tool but is more powerful because
it automatically converts the input factors to a common user-defined scale and allows
the user to specify benefit factors and cost factors. A benefit factor is a factor
for which higher values are more suitable. A cost factor is a factor for which higher
values are less suitable. By default, WeightedOverlay assumes that input images are
benefit factors, unless a cost value of 'true' is entered in the cost array.
Constraints are absolute restriction with values of 0 (unsuitable) and 1 (suitable).
This tool is particularly useful for performing multi-criteria evaluations (MCE).</p>
<p>Notice that the algorithm will convert the user-defined factor weights internally such
that the sum of the weights is always equal to one. As such, the user can specify the
relative weights as decimals, percentages, or relative weightings (e.g. slope is 2 times
more important than elevation, in which case the weights may not sum to 1 or 100).</p>
<p>NoData valued grid cells in any of the input images will be assigned NoData values in
the output image. The output raster is of the float data type and continuous data scale.</p>
<p><em>Warning</em>:</p>
<p>Each of the input rasters must have the same spatial extent and number of rows
and columns.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--factors          </td><td> Input factor raster files</td></tr>
<tr><td>-w, --weights      </td><td> Weight values, contained in quotes and separated by commas or semicolons. Must have the same number as factors</td></tr>
<tr><td>--cost             </td><td> Weight values, contained in quotes and separated by commas or semicolons. Must have the same number as factors</td></tr>
<tr><td>--constraints      </td><td> Input constraints raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--scale_max        </td><td> Suitability scale maximum value (common values are 1.0, 100.0, and 255.0)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">weighted_overlay(
    factors, 
    weights, 
    output, 
    cost=None, 
    constraints=None, 
    scale_max=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=WeightedOverlay -v ^
--wd='/path/to/data/' ^
--factors='image1.tif;image2.tif;image3.tif' ^
--weights='0.3;0.2;0.5' --cost='false;false;true' -o=output.tif ^
--scale_max=100.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/weighted_overlay.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 07/05/2018</p>
<p><a name="WeightedSum"></a></p>
<a class="header" href="#weightedsum" id="weightedsum"><h1>WeightedSum</h1></a>
<p>This tool performs a weighted-sum overlay on multiple input raster images.
If you have a stack of rasters that you would like to sum, each with an
equal weighting (1.0), then use the <a href="./gis_analysis_overlay_tools.html#sumoverlay"><strong>SumOverlay</strong></a> tool instead.</p>
<p><em>Warning</em>:</p>
<p>Each of the input rasters must have the same spatial extent and number of rows
and columns.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_overlay_tools.html#sumoverlay"><strong>SumOverlay</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-w, --weights      </td><td> Weight values, contained in quotes and separated by commas or semicolons</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">weighted_sum(
    inputs, 
    weights, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=WeightedSum -v --wd='/path/to/data/' ^
-i='image1.tif;image2.tif;image3.tif' --weights='0.3;0.2;0.5' ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/weighted_sum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 22/06/2017</p>
<a class="header" href="#gis-analysis--patch-shape-tools" id="gis-analysis--patch-shape-tools"><h1>GIS Analysis → Patch Shape Tools</h1></a>
<ul>
<li><a href="#CompactnessRatio">CompactnessRatio</a></li>
<li><a href="#EdgeProportion">EdgeProportion</a></li>
<li><a href="#ElongationRatio">ElongationRatio</a></li>
<li><a href="#FindPatchOrClassEdgeCells">FindPatchOrClassEdgeCells</a></li>
<li><a href="#HoleProportion">HoleProportion</a></li>
<li><a href="#LinearityIndex">LinearityIndex</a></li>
<li><a href="#PatchOrientation">PatchOrientation</a></li>
<li><a href="#PerimeterAreaRatio">PerimeterAreaRatio</a></li>
<li><a href="#RadiusOfGyration">RadiusOfGyration</a></li>
<li><a href="#RelatedCircumscribingCircle">RelatedCircumscribingCircle</a></li>
<li><a href="#ShapeComplexityIndex">ShapeComplexityIndex</a></li>
</ul>
<p><a name="CompactnessRatio"></a></p>
<a class="header" href="#compactnessratio" id="compactnessratio"><h1>CompactnessRatio</h1></a>
<p>The compactness ratio is an indicator of polygon shape complexity. The compactness
ratio is defined as the polygon area divided by its perimeter. Unlike some other
shape parameters (e.g. <a href="./gis_analysis_patch_shape_tools.html#shapecomplexityindex"><strong>ShapeComplexityIndex</strong></a>), compactness ratio does not standardize
to a simple Euclidean shape. Although widely used for landscape analysis, compactness
ratio, like its inverse, the <a href="./gis_analysis_patch_shape_tools.html#perimeterarearatio"><strong>PerimeterAreaRatio</strong></a>, exhibits the undesirable property
of polygon size dependence (Mcgarigal et al. 2002). That is, holding shape constant,
an increase in polygon size will cause a change in the compactness ratio.</p>
<p>The output data will be contained in the input vector's attribute table as a new field
(COMPACT).</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_patch_shape_tools.html#perimeterarearatio"><strong>PerimeterAreaRatio</strong></a>, <a href="./gis_analysis_patch_shape_tools.html#shapecomplexityindex"><strong>ShapeComplexityIndex</strong></a>, <a href="./gis_analysis_patch_shape_tools.html#relatedcircumscribingcircle"><strong>RelatedCircumscribingCircle</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">compactness_ratio(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CompactnessRatio -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/compactness_ratio.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/09/2018</p>
<p><a name="EdgeProportion"></a></p>
<a class="header" href="#edgeproportion" id="edgeproportion"><h1>EdgeProportion</h1></a>
<p>Calculate the proportion of cells in a raster polygon that are edge cells.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--output_text      </td><td> flag indicating whether a text report should also be output</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">edge_proportion(
    i, 
    output, 
    output_text=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=EdgeProportion -v --wd=&quot;/path/to/data/&quot; ^
-i=input.tif -o=output.tif --output_text 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/edge_proportion.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="ElongationRatio"></a></p>
<a class="header" href="#elongationratio" id="elongationratio"><h1>ElongationRatio</h1></a>
<p>This tool can be used to calculate the elongation ratio for vector polygons. The
elongation ratio values calculated for each vector polygon feature will be placed
in the accompanying database file (.dbf) as an elongation field (ELONGATION).</p>
<p>The elongation ratio (<code>E</code>) is:</p>
<p><code>E = 1 - S / L</code></p>
<p>Where <code>S</code> is the short-axis length, and <code>L</code> is the long-axis length. Axes
lengths are determined by estimating the minimum bounding box.</p>
<p>The elongation ratio provides similar information as the Linearity Index. The
ratio is not an adequate measure of overall polygon narrowness, because a highly
sinuous but narrow polygon will have a low linearity (elongation) owing to the
compact nature of these polygon.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">elongation_ratio(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ElongationRatio -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/elongation_ratio.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/09/2018</p>
<p><a name="FindPatchOrClassEdgeCells"></a></p>
<a class="header" href="#findpatchorclassedgecells" id="findpatchorclassedgecells"><h1>FindPatchOrClassEdgeCells</h1></a>
<p>Finds all cells located on the edge of patch or class features.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">find_patch_or_class_edge_cells(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FindPatchOrClassEdgeCells -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/find_patch_edge_cells.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 14, 2017</p>
<p><a name="HoleProportion"></a></p>
<a class="header" href="#holeproportion" id="holeproportion"><h1>HoleProportion</h1></a>
<p>This calculates the proportion of the total area of a polygon's holes (i.e. islands)
relative to the area of the polygon's hull. It can be a useful measure of shape
complexity, or how discontinuous a patch is. The user must specify the name of the
input vector file and the output data will be contained within the input vector's
database file as a new field (HOLE_PROP).</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_patch_shape_tools.html#shapecomplexityindex"><strong>ShapeComplexityIndex</strong></a>, <a href="./gis_analysis_patch_shape_tools.html#elongationratio"><strong>ElongationRatio</strong></a>, <a href="./gis_analysis_patch_shape_tools.html#perimeterarearatio"><strong>PerimeterAreaRatio</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">hole_proportion(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HoleProportion -v --wd=&quot;/path/to/data/&quot; ^
--input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/hole_proportion.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 26/09/2018</p>
<p><a name="LinearityIndex"></a></p>
<a class="header" href="#linearityindex" id="linearityindex"><h1>LinearityIndex</h1></a>
<p>This tool calculates the linearity index of polygon features based on a regression analysis.
The index is simply the coefficient of determination (r-squared) calculated from a regression
analysis of the x and y coordinates of the exterior hull nodes of a vector polygon. Linearity
index is a measure of how well a polygon can be described by a straight line. It is a related
index to the <a href="./gis_analysis_patch_shape_tools.html#elongationratio"><strong>ElongationRatio</strong></a>, but is more efficient to calculate as it does not require
finding the minimum bounding box. The Pearson correlation coefficient between linearity index
and the elongation ratio for a large data set of lake polygons in northern Canada was found
to be 0.656, suggesting a moderate level of association between the two measures of polygon
linearity. Note that this index is not useful for identifying narrow yet sinuous polygons, such
as meandering rivers.</p>
<p>The only required input is the name of the file. The linearity values calculated for each vector
polygon feature will be placed in the accompanying attribute table as a new field (LINEARITY).</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_patch_shape_tools.html#elongationratio"><strong>ElongationRatio</strong></a>, <a href="./gis_analysis_patch_shape_tools.html#patchorientation"><strong>PatchOrientation</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">linearity_index(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LinearityIndex -v --wd=&quot;/path/to/data/&quot; ^
--input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/linearity_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 01/10/2018</p>
<p><a name="PatchOrientation"></a></p>
<a class="header" href="#patchorientation" id="patchorientation"><h1>PatchOrientation</h1></a>
<p>This tool calculates the orientation of polygon features based on the slope of a reduced major
axis (RMA) regression line. The regression analysis use the vertices of the exterior hull nodes
of a vector polygon. The only required input is the name of the vector polygon file. The
orientation values, measured in degrees from north, will be placed in the accompanying attribute
table as a new field (ORIENT). The value of the orientation measure for any polygon will
depend on how elongated the feature is.</p>
<p>Note that the output values are polygon orientations and not true directions. While directions
may take values ranging from 0-360, orientation is expressed as an angle between 0 and 180 degrees
clockwise from north. Lastly, the orientation measure may become unstable when polygons are
oriented nearly vertical or horizontal.</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis_patch_shape_tools.html#linearityindex"><strong>LinearityIndex</strong></a>, <a href="./gis_analysis_patch_shape_tools.html#elongationratio"><strong>ElongationRatio</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">patch_orientation(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PatchOrientation -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/patch_orientation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 16/10/2018</p>
<p><a name="PerimeterAreaRatio"></a></p>
<a class="header" href="#perimeterarearatio" id="perimeterarearatio"><h1>PerimeterAreaRatio</h1></a>
<p>The perimeter-area ratio is an indicator of polygon shape complexity. Unlike some
other shape parameters (e.g. shape complexity index), perimeter-area ratio does not
standardize to a simple Euclidean shape. Although widely used for landscape analysis,
perimeter-area ratio exhibits the undesirable property of polygon size dependence
(Mcgarigal et al. 2002). That is, holding shape constant, an increase in polygon
size will cause a decrease in the perimeter-area ratio. The perimeter-area ratio is
the inverse of the compactness ratio.</p>
<p>The output data will be displayed as a new field (P-A_RATIO) in the input vector's
database file.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">perimeter_area_ratio(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PerimeterAreaRatio -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/perimeter_area_ratio.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 26/09/2018</p>
<p><a name="RadiusOfGyration"></a></p>
<a class="header" href="#radiusofgyration" id="radiusofgyration"><h1>RadiusOfGyration</h1></a>
<p>This can be used to calculate the radius of gyration (RoG) for the polygon
features within a raster image. RoG measures how far across the landscape a polygon
extends its reach on average, given by the mean distance between cells in a patch
(Mcgarigal et al. 2002). The radius of gyration can be considered a measure of the
average distance an organism can move within a patch before encountering the patch
boundary from a random starting point (Mcgarigal et al. 2002). The input raster grid
should contain polygons with unique identifiers greater than zero. The user must also
specify the name of the output raster file (where the radius of gyration will be
assigned to each feature in the input file) and the specified option of outputting text
data.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--text_output      </td><td> Optional text output</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">radius_of_gyration(
    i, 
    output, 
    text_output=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RadiusOfGyration -v ^
--wd=&quot;/path/to/data/&quot; -i=polygons.tif -o=output.tif ^
--text_output 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/radius_of_gyration.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: December 31 2017</p>
<p><a name="RelatedCircumscribingCircle"></a></p>
<a class="header" href="#relatedcircumscribingcircle" id="relatedcircumscribingcircle"><h1>RelatedCircumscribingCircle</h1></a>
<p>This tool can be used to calculate the related circumscribing circle (Mcgarigal et al. 2002)
for vector polygon features. The related circumscribing circle values calculated for each
vector polygon feature will be placed in the accompanying attribute table as a new field
(RC_CIRCLE).</p>
<p>Related circumscribing circle (RCC) is defined as:</p>
<p><code>RCC = 1 - A / Ac</code></p>
<p>Where <code>A</code> is the polygon's area and <code>Ac</code> the area of the smallest circumscribing circle.</p>
<p>Theoretically, <a href="./gis_analysis_patch_shape_tools.html#relatedcircumscribingcircle"><strong>RelatedCircumscribingCircle</strong></a> ranges from 0 to 1, where a value of 0 indicates
a circular polygon and a value of 1 indicates a highly elongated shape. The circumscribing
circle provides a measure of polygon elongation. Unlike the <a href="./gis_analysis_patch_shape_tools.html#elongationratio"><strong>ElongationRatio</strong></a>, however, it
does not provide a measure of polygon direction in addition to overall elongation. Like the
<a href="./gis_analysis_patch_shape_tools.html#elongationratio"><strong>ElongationRatio</strong></a> and <a href="./gis_analysis_patch_shape_tools.html#linearityindex"><strong>LinearityIndex</strong></a>, <a href="./gis_analysis_patch_shape_tools.html#relatedcircumscribingcircle"><strong>RelatedCircumscribingCircle</strong></a> is not an adequate
measure of overall polygon narrowness, because a highly sinuous but narrow patch will have
a low related circumscribing circle index owing to the compact nature of these polygon.</p>
<p>Note: Holes are excluded from the area calculation of polygons.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">related_circumscribing_circle(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RelatedCircumscribingCircle -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/related_circumscribing_circle.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/09/2018</p>
<p><a name="ShapeComplexityIndex"></a></p>
<a class="header" href="#shapecomplexityindex" id="shapecomplexityindex"><h1>ShapeComplexityIndex</h1></a>
<p>This tool provides a measure of overall polygon shape complexity, or irregularity,
for vector polygons. Several shape indices have been created to compare a polygon's
shape to simple Euclidean shapes (e.g. circles, squares, etc.). One of the problems
with this approach is that it inherently convolves the characteristics of polygon
complexity and elongation. The Shape Complexity Index (SCI) was developed as a
parameter for assessing the complexity of a polygon that is independent of its
elongation.</p>
<p>SCI relates a polygon's shape to that of an encompassing convex hull. It is
defined as:</p>
<p><code>SCI = 1 - A / Ah</code></p>
<p>Where <code>A</code> is the polygon's area and <code>Ah</code> is the area of the convex hull containing
the polygon. Convex polygons, i.e. those that do not contain concavities or holes,
have a value of 0. As the shape of the polygon becomes more complex, the SCI
approaches 1. Note that polygon shape complexity also increases with the greater
number of holes (i.e. islands), since holes have the effect of reducing the lake
area.</p>
<p>The SCI values calculated for each vector polygon feature will be placed in the
accompanying database file (.dbf) as a complexity field (COMPLEXITY).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">shape_complexity_index(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ShapeComplexityIndex -v ^
--wd=&quot;/path/to/data/&quot; --input=polygons.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/gis_analysis/shape_complexity_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/09/2018</p>
<a class="header" href="#hydrological-analysis" id="hydrological-analysis"><h1>Hydrological Analysis</h1></a>
<ul>
<li><a href="#AverageFlowpathSlope">AverageFlowpathSlope</a></li>
<li><a href="#AverageUpslopeFlowpathLength">AverageUpslopeFlowpathLength</a></li>
<li><a href="#Basins">Basins</a></li>
<li><a href="#BreachDepressions">BreachDepressions</a></li>
<li><a href="#BreachSingleCellPits">BreachSingleCellPits</a></li>
<li><a href="#D8FlowAccumulation">D8FlowAccumulation</a></li>
<li><a href="#D8MassFlux">D8MassFlux</a></li>
<li><a href="#D8Pointer">D8Pointer</a></li>
<li><a href="#DInfFlowAccumulation">DInfFlowAccumulation</a></li>
<li><a href="#DInfMassFlux">DInfMassFlux</a></li>
<li><a href="#DInfPointer">DInfPointer</a></li>
<li><a href="#DepthInSink">DepthInSink</a></li>
<li><a href="#DownslopeDistanceToStream">DownslopeDistanceToStream</a></li>
<li><a href="#DownslopeFlowpathLength">DownslopeFlowpathLength</a></li>
<li><a href="#ElevationAboveStream">ElevationAboveStream</a></li>
<li><a href="#ElevationAboveStreamEuclidean">ElevationAboveStreamEuclidean</a></li>
<li><a href="#Fd8FlowAccumulation">Fd8FlowAccumulation</a></li>
<li><a href="#Fd8Pointer">Fd8Pointer</a></li>
<li><a href="#FillBurn">FillBurn</a></li>
<li><a href="#FillDepressions">FillDepressions</a></li>
<li><a href="#FillSingleCellPits">FillSingleCellPits</a></li>
<li><a href="#FindNoFlowCells">FindNoFlowCells</a></li>
<li><a href="#FindParallelFlow">FindParallelFlow</a></li>
<li><a href="#FlattenLakes">FlattenLakes</a></li>
<li><a href="#FloodOrder">FloodOrder</a></li>
<li><a href="#FlowAccumulationFullWorkflow">FlowAccumulationFullWorkflow</a></li>
<li><a href="#FlowLengthDiff">FlowLengthDiff</a></li>
<li><a href="#Hillslopes">Hillslopes</a></li>
<li><a href="#ImpoundmentIndex">ImpoundmentIndex</a></li>
<li><a href="#Isobasins">Isobasins</a></li>
<li><a href="#JensonSnapPourPoints">JensonSnapPourPoints</a></li>
<li><a href="#LongestFlowpath">LongestFlowpath</a></li>
<li><a href="#MaxUpslopeFlowpathLength">MaxUpslopeFlowpathLength</a></li>
<li><a href="#NumInflowingNeighbours">NumInflowingNeighbours</a></li>
<li><a href="#RaiseWalls">RaiseWalls</a></li>
<li><a href="#Rho8Pointer">Rho8Pointer</a></li>
<li><a href="#Sink">Sink</a></li>
<li><a href="#SnapPourPoints">SnapPourPoints</a></li>
<li><a href="#StochasticDepressionAnalysis">StochasticDepressionAnalysis</a></li>
<li><a href="#StrahlerOrderBasins">StrahlerOrderBasins</a></li>
<li><a href="#Subbasins">Subbasins</a></li>
<li><a href="#TraceDownslopeFlowpaths">TraceDownslopeFlowpaths</a></li>
<li><a href="#UnnestBasins">UnnestBasins</a></li>
<li><a href="#Watershed">Watershed</a></li>
</ul>
<p><a name="AverageFlowpathSlope"></a></p>
<a class="header" href="#averageflowpathslope" id="averageflowpathslope"><h1>AverageFlowpathSlope</h1></a>
<p>Measures the average slope gradient from each grid cell to all upslope divide cells.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">average_flowpath_slope(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AverageFlowpathSlope -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/average_flowpath_slope.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 22, 2017</p>
<p><a name="AverageUpslopeFlowpathLength"></a></p>
<a class="header" href="#averageupslopeflowpathlength" id="averageupslopeflowpathlength"><h1>AverageUpslopeFlowpathLength</h1></a>
<p>Measures the average length of all upslope flowpaths draining each grid cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">average_upslope_flowpath_length(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AverageUpslopeFlowpathLength -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/average_upslope_flowpath_length.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="Basins"></a></p>
<a class="header" href="#basins" id="basins"><h1>Basins</h1></a>
<p>Identifies drainage basins that drain to the DEM edge.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">basins(
    d8_pntr, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Basins -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr='d8pntr.tif' -o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/basins.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="BreachDepressions"></a></p>
<a class="header" href="#breachdepressions" id="breachdepressions"><h1>BreachDepressions</h1></a>
<p>This tool can be used to remove the depressions in a digital elevation model, a
common requirement of spatial hydrological operations such as flow accumulation
and watershed modelling. The algorithm based on:</p>
<blockquote>
<p>Lindsay JB. 2016. <em>Efficient hybrid breaching-filling sink removal methods for
flow path enforcement in digital elevation models.</em> <strong>Hydrological Processes</strong>,
30(6): 846–857. DOI: 10.1002/hyp.10648</p>
</blockquote>
<p>It uses a breach-first, fill-second approach.</p>
<p><em>See Also</em>:</p>
<p><a href="./hydrological_analysis.html#filldepressions"><strong>FillDepressions</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--max_depth        </td><td> Optional maximum breach depth (default is Inf)</td></tr>
<tr><td>--max_length       </td><td> Optional maximum breach channel length (in grid cells; default is Inf)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">breach_depressions(
    dem, 
    output, 
    max_depth=None, 
    max_length=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=BreachDepressions -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/breach_depressions.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="BreachSingleCellPits"></a></p>
<a class="header" href="#breachsinglecellpits" id="breachsinglecellpits"><h1>BreachSingleCellPits</h1></a>
<p>Removes single-cell pits from an input DEM by breaching.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">breach_single_cell_pits(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=BreachSingleCellPits -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/breach_pits.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="D8FlowAccumulation"></a></p>
<a class="header" href="#d8flowaccumulation" id="d8flowaccumulation"><h1>D8FlowAccumulation</h1></a>
<p>This tool is used to generate a flow accumulation grid (i.e. catchment area) using the
D8 (O'Callaghan and Mark, 1984) algorithm. This algorithm is an example of single-flow-direction
(SFD) method because the flow entering each grid cell is routed to only one downslope neighbour,
i.e. flow divergence is not permitted. The user must specify the name of the input digital
elevation model (DEM). The DEM must have been hydrologically corrected to remove all spurious
depressions and flat areas. DEM pre-processing is usually achieved using the <a href="./hydrological_analysis.html#breachdepressions"><strong>BreachDepressions</strong></a> or
<a href="./hydrological_analysis.html#filldepressions"><strong>FillDepressions</strong></a> tools.</p>
<p>In addition to the input DEM, the user must specify the output type. The output flow-accumulation
can be 1) <code>cells</code> (i.e. the number of inflowing grid cells), <code>catchment area</code> (i.e. the upslope area),
or <code>specific contributing area</code> (i.e. the catchment area divided by the flow width. The default value
is <code>cells</code>. The user must also specify whether the output flow-accumulation grid should be
log-tranformed, i.e. the output, if this option is selected, will be the natural-logarithm of the
accumulated flow value. This is a transformation that is often performed to better visualize the
contributing area distribution. Because contributing areas tend to be very high along valley bottoms
and relatively low on hillslopes, when a flow-accumulation image is displayed, the distribution of
values on hillslopes tends to be 'washed out' because the palette is stretched out to represent the
highest values. Log-transformation provides a means of compensating for this phenomenon. Importantly,
however, log-transformed flow-accumulation grids must not be used to estimate other secondary terrain
indices, such as the wetness index, or relative stream power index.</p>
<p>Grid cells possessing the <strong>NoData</strong> value in the input flow-pointer grid are assigned the <strong>NoData</strong>
value in the output flow-accumulation image.</p>
<p><em>See Also</em>:</p>
<p><a href="./hydrological_analysis.html#dinfflowaccumulation"><strong>DInfFlowAccumulation</strong></a>, <a href="./hydrological_analysis.html#breachdepressions"><strong>BreachDepressions</strong></a>, <a href="./hydrological_analysis.html#filldepressions"><strong>FillDepressions</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--out_type         </td><td> Output type; one of 'cells' (default), 'catchment area', and 'specific contributing area'</td></tr>
<tr><td>--log              </td><td> Optional flag to request the output be log-transformed</td></tr>
<tr><td>--clip             </td><td> Optional flag to request clipping the display max by 1%</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">d8_flow_accumulation(
    dem, 
    output, 
    out_type=&quot;cells&quot;, 
    log=False, 
    clip=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=D8FlowAccumulation -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--out_type='cells'
&gt;&gt;./whitebox_tools -r=D8FlowAccumulation -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--out_type='specific catchment area' --log --clip 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/d8_flow_accum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="D8MassFlux"></a></p>
<a class="header" href="#d8massflux" id="d8massflux"><h1>D8MassFlux</h1></a>
<p>Performs a D8 mass flux calculation.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>--loading          </td><td> Input loading raster file</td></tr>
<tr><td>--efficiency       </td><td> Input efficiency raster file</td></tr>
<tr><td>--absorption       </td><td> Input absorption raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">d8_mass_flux(
    dem, 
    loading, 
    efficiency, 
    absorption, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=D8MassFlux -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif --loading=load.tif --efficiency=eff.tif ^
--absorption=abs.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/d8_mass_flux.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 29, 2017</p>
<p><a name="D8Pointer"></a></p>
<a class="header" href="#d8pointer" id="d8pointer"><h1>D8Pointer</h1></a>
<p>Calculates a D8 flow pointer raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">d8_pointer(
    dem, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=D8Pointer -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/d8_pointer.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 16, 2017</p>
<p><a name="DInfFlowAccumulation"></a></p>
<a class="header" href="#dinfflowaccumulation" id="dinfflowaccumulation"><h1>DInfFlowAccumulation</h1></a>
<p>Calculates a D-infinity flow accumulation raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--out_type         </td><td> Output type; one of 'cells', 'sca' (default), and 'ca'</td></tr>
<tr><td>--threshold        </td><td> Optional convergence threshold parameter, in grid cells; default is inifinity</td></tr>
<tr><td>--log              </td><td> Optional flag to request the output be log-transformed</td></tr>
<tr><td>--clip             </td><td> Optional flag to request clipping the display max by 1%</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">d_inf_flow_accumulation(
    dem, 
    output, 
    out_type=&quot;Specific Contributing Area&quot;, 
    threshold=None, 
    log=False, 
    clip=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DInfFlowAccumulation -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--out_type=sca
&gt;&gt;./whitebox_tools -r=DInfFlowAccumulation -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--out_type=sca --threshold=10000 --log --clip 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/dinf_flow_accum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 24, 2017</p>
<p><a name="DInfMassFlux"></a></p>
<a class="header" href="#dinfmassflux" id="dinfmassflux"><h1>DInfMassFlux</h1></a>
<p>Performs a D-infinity mass flux calculation.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>--loading          </td><td> Input loading raster file</td></tr>
<tr><td>--efficiency       </td><td> Input efficiency raster file</td></tr>
<tr><td>--absorption       </td><td> Input absorption raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">d_inf_mass_flux(
    dem, 
    loading, 
    efficiency, 
    absorption, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DInfMassFlux -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif --loading=load.tif --efficiency=eff.tif ^
--absorption=abs.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/dinf_mass_flux.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 29, 2017</p>
<p><a name="DInfPointer"></a></p>
<a class="header" href="#dinfpointer" id="dinfpointer"><h1>DInfPointer</h1></a>
<p>Calculates a D-infinity flow pointer (flow direction) raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">d_inf_pointer(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DInfPointer -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/dinf_pointer.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="DepthInSink"></a></p>
<a class="header" href="#depthinsink" id="depthinsink"><h1>DepthInSink</h1></a>
<p>Measures the depth of sinks (depressions) in a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether the background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">depth_in_sink(
    dem, 
    output, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DepthInSink -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/depth_in_sink.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="DownslopeDistanceToStream"></a></p>
<a class="header" href="#downslopedistancetostream" id="downslopedistancetostream"><h1>DownslopeDistanceToStream</h1></a>
<p>Measures distance to the nearest downslope stream cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">downslope_distance_to_stream(
    dem, 
    streams, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DownslopeDistanceToStream -v ^
--wd=&quot;/path/to/data/&quot; --dem='dem.tif' --streams='streams.tif' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/downslope_distance_to_stream.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 9, 2017</p>
<p><a name="DownslopeFlowpathLength"></a></p>
<a class="header" href="#downslopeflowpathlength" id="downslopeflowpathlength"><h1>DownslopeFlowpathLength</h1></a>
<p>Calculates the downslope flowpath length from each cell to basin outlet.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input D8 pointer raster file</td></tr>
<tr><td>--watersheds       </td><td> Optional input watershed raster file</td></tr>
<tr><td>--weights          </td><td> Optional input weights raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">downslope_flowpath_length(
    d8_pntr, 
    output, 
    watersheds=None, 
    weights=None, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DownslopeFlowpathLength -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=pointer.tif ^
-o=flowpath_len.tif
&gt;&gt;./whitebox_tools ^
-r=DownslopeFlowpathLength -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr=pointer.tif --watersheds=basin.tif ^
--weights=weights.tif -o=flowpath_len.tif --esri_pntr 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/downslope_flowpath_length.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 8, 2017</p>
<p><a name="ElevationAboveStream"></a></p>
<a class="header" href="#elevationabovestream" id="elevationabovestream"><h1>ElevationAboveStream</h1></a>
<p>Calculates the elevation of cells above the nearest downslope stream cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">elevation_above_stream(
    dem, 
    streams, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ElevationAboveStream -v ^
--wd=&quot;/path/to/data/&quot; --dem='dem.tif' --streams='streams.tif' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/elevation_above_stream.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 9, 2017</p>
<p><a name="ElevationAboveStreamEuclidean"></a></p>
<a class="header" href="#elevationabovestreameuclidean" id="elevationabovestreameuclidean"><h1>ElevationAboveStreamEuclidean</h1></a>
<p>Calculates the elevation of cells above the nearest (Euclidean distance) stream cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">elevation_above_stream_euclidean(
    dem, 
    streams, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ElevationAboveStreamEuclidean -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif --streams=streams.tif ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/elevation_above_stream_euclidean.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: March 11, 2018</p>
<p><a name="Fd8FlowAccumulation"></a></p>
<a class="header" href="#fd8flowaccumulation" id="fd8flowaccumulation"><h1>Fd8FlowAccumulation</h1></a>
<p>Calculates an FD8 flow accumulation raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--out_type         </td><td> Output type; one of 'cells', 'specific contributing area' (default), and 'catchment area'</td></tr>
<tr><td>--exponent         </td><td> Optional exponent parameter; default is 1.1</td></tr>
<tr><td>--threshold        </td><td> Optional convergence threshold parameter, in grid cells; default is inifinity</td></tr>
<tr><td>--log              </td><td> Optional flag to request the output be log-transformed</td></tr>
<tr><td>--clip             </td><td> Optional flag to request clipping the display max by 1%</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fd8_flow_accumulation(
    dem, 
    output, 
    out_type=&quot;specific contributing area&quot;, 
    exponent=1.1, 
    threshold=None, 
    log=False, 
    clip=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FD8FlowAccumulation -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--out_type='cells'
&gt;&gt;./whitebox_tools -r=FD8FlowAccumulation -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--out_type='catchment area' --exponent=1.5 --threshold=10000 ^
--log --clip 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/fd8_flow_accum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="Fd8Pointer"></a></p>
<a class="header" href="#fd8pointer" id="fd8pointer"><h1>Fd8Pointer</h1></a>
<p>Calculates an FD8 flow pointer raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fd8_pointer(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FD8Pointer -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/fd8_pointer.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="FillBurn"></a></p>
<a class="header" href="#fillburn" id="fillburn"><h1>FillBurn</h1></a>
<p>Burns streams into a DEM using the FillBurn (Saunders, 1999) method. This tool uses the
algorithm described in:</p>
<p>Lindsay JB. 2016. The practice of DEM stream burning revisited. Earth Surface Processes
and Landforms, 41(5): 658-668. DOI: 10.1002/esp.3888</p>
<p>And:</p>
<p>Saunders, W. 1999. Preparation of DEMs for use in environmental modeling analysis, in: ESRI User
Conference. pp. 24-30.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>--streams          </td><td> Input vector streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fill_burn(
    dem, 
    streams, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FillBurn -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif --streams=streams.shp -o=dem_burned.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/fill_burn.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 01/04/2018</p>
<p><a name="FillDepressions"></a></p>
<a class="header" href="#filldepressions" id="filldepressions"><h1>FillDepressions</h1></a>
<p>Fills all of the depressions in a DEM. Depression breaching should be preferred in most cases.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--fix_flats        </td><td> Optional flag indicating whether flat areas should have a small gradient applied</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fill_depressions(
    dem, 
    output, 
    fix_flats=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FillDepressions -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=output.tif ^
--fix_flats 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/fill_depressions.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="FillSingleCellPits"></a></p>
<a class="header" href="#fillsinglecellpits" id="fillsinglecellpits"><h1>FillSingleCellPits</h1></a>
<p>Raises pit cells to the elevation of their lowest neighbour.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fill_single_cell_pits(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FillSingleCellPits -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=NewRaster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/fill_pits.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="FindNoFlowCells"></a></p>
<a class="header" href="#findnoflowcells" id="findnoflowcells"><h1>FindNoFlowCells</h1></a>
<p>Finds grid cells with no downslope neighbours.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">find_no_flow_cells(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FindNoFlowCells -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=NewRaster.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/find_noflow_cells.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="FindParallelFlow"></a></p>
<a class="header" href="#findparallelflow" id="findparallelflow"><h1>FindParallelFlow</h1></a>
<p>Finds areas of parallel flow in D8 flow direction rasters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input D8 pointer raster file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">find_parallel_flow(
    d8_pntr, 
    streams, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FindParallelFlow -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=pointer.tif ^
-o=out.tif
&gt;&gt;./whitebox_tools -r=FindParallelFlow -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=pointer.tif -o=out.tif ^
--streams='streams.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/find_parallel_flow.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="FlattenLakes"></a></p>
<a class="header" href="#flattenlakes" id="flattenlakes"><h1>FlattenLakes</h1></a>
<p>Flattens lake polygons in a raster DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--lakes            </td><td> Input lakes vector polygons file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">flatten_lakes(
    dem, 
    lakes, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FlattenLakes -v --wd=&quot;/path/to/data/&quot; ^
--dem='DEM.tif' --lakes='lakes.shp' -o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/flatten_lakes.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 29/03/2018</p>
<p><a name="FloodOrder"></a></p>
<a class="header" href="#floodorder" id="floodorder"><h1>FloodOrder</h1></a>
<p>Assigns each DEM grid cell its order in the sequence of inundations that are encountered during a search starting from the edges, moving inward at increasing elevations.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">flood_order(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FloodOrder -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/flood_order.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 12, 2017</p>
<p><a name="FlowAccumulationFullWorkflow"></a></p>
<a class="header" href="#flowaccumulationfullworkflow" id="flowaccumulationfullworkflow"><h1>FlowAccumulationFullWorkflow</h1></a>
<p>Resolves all of the depressions in a DEM, outputting a breached DEM, an aspect-aligned non-divergent flow pointer, and a flow accumulation raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--out_dem          </td><td> Output raster DEM file</td></tr>
<tr><td>--out_pntr         </td><td> Output raster flow pointer file</td></tr>
<tr><td>--out_accum        </td><td> Output raster flow accumulation file</td></tr>
<tr><td>--out_type         </td><td> Output type; one of 'cells', 'sca' (default), and 'ca'</td></tr>
<tr><td>--log              </td><td> Optional flag to request the output be log-transformed</td></tr>
<tr><td>--clip             </td><td> Optional flag to request clipping the display max by 1%</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">flow_accumulation_full_workflow(
    dem, 
    out_dem, 
    out_pntr, 
    out_accum, 
    out_type=&quot;Specific Contributing Area&quot;, 
    log=False, 
    clip=False, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FlowAccumulationFullWorkflow -v ^
--wd=&quot;/path/to/data/&quot; --dem='DEM.tif' ^
--out_dem='DEM_filled.tif' --out_pntr='pointer.tif' ^
--out_accum='accum.tif' --out_type=sca --log --clip 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/flow_accum_full_workflow.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="FlowLengthDiff"></a></p>
<a class="header" href="#flowlengthdiff" id="flowlengthdiff"><h1>FlowLengthDiff</h1></a>
<p><a href="./hydrological_analysis.html#flowlengthdiff"><strong>FlowLengthDiff</strong></a> calculates the local maximum absolute difference in downslope flowpath length,
which is useful in mapping drainage divides and ridges.</p>
<p><em>See Also</em>:</p>
<p><a href="./geomorphometric_analysis.html#maxbranchlength"><strong>MaxBranchLength</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input D8 pointer raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">flow_length_diff(
    d8_pntr, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FlowLengthDiff -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr=pointer.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/flow_length_diff.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 8, 2017</p>
<p><a name="Hillslopes"></a></p>
<a class="header" href="#hillslopes" id="hillslopes"><h1>Hillslopes</h1></a>
<p>Identifies the individual hillslopes draining to each link in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">hillslopes(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Hillslopes -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr='d8pntr.tif' --streams='streams.tif' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/hillslopes.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 16, 2017</p>
<p><a name="ImpoundmentIndex"></a></p>
<a class="header" href="#impoundmentindex" id="impoundmentindex"><h1>ImpoundmentIndex</h1></a>
<p>Calculates the impoundment size resulting from damming a DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--out_type         </td><td> Output type; one of 'depth' (default), 'volume', and 'area'</td></tr>
<tr><td>--damlength        </td><td> Maximum length of thr dam</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">impoundment_index(
    dem, 
    output, 
    damlength, 
    out_type=&quot;depth&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ImpoundmentIndex -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=out.tif ^
--out_type=depth --damlength=11 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/impoundment_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 28/05/2018</p>
<p><a name="Isobasins"></a></p>
<a class="header" href="#isobasins" id="isobasins"><h1>Isobasins</h1></a>
<p>Divides a landscape into nearly equal sized drainage basins (i.e. watersheds).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--size             </td><td> Target basin size, in grid cells</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">isobasins(
    dem, 
    output, 
    size, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Isobasins -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif --size=1000 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/isobasins.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: December 3, 2017</p>
<p><a name="JensonSnapPourPoints"></a></p>
<a class="header" href="#jensonsnappourpoints" id="jensonsnappourpoints"><h1>JensonSnapPourPoints</h1></a>
<p>Moves outlet points used to specify points of interest in a watershedding operation to the nearest stream cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--pour_pts         </td><td> Input vector pour points (outlet) file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--snap_dist        </td><td> Maximum snap distance in map units</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">jenson_snap_pour_points(
    pour_pts, 
    streams, 
    output, 
    snap_dist, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=JensonSnapPourPoints -v ^
--wd=&quot;/path/to/data/&quot; --pour_pts='pour_pts.shp' ^
--streams='streams.tif' -o='output.shp' --snap_dist=15.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/jenson_snap_pour_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="LongestFlowpath"></a></p>
<a class="header" href="#longestflowpath" id="longestflowpath"><h1>LongestFlowpath</h1></a>
<p>This tool delineates the longest flowpaths for a group of subbasins or watersheds.
Flowpaths are initiated along drainage divides and continue along the D8-defined
flow direction until either the subbasin outlet or DEM edge is encountered. Each input
subbasin/watershed will have an associated vector flowpath in the output image. <a href="./hydrological_analysis.html#longestflowpath"><strong>LongestFlowpath</strong></a>
is similar to the <code>r.lfp</code> plugin tool for GRASS GIS. The length of the longest flowpath
draining to an outlet is related to the time of concentration, which is a parameter
used in certain hydrological models.</p>
<p>The user must input the filename of a digital elevation model (DEM), a basins raster, and the
output vector. The DEM must be depressionless and should have been pre-processed using the
<a href="./hydrological_analysis.html#breachdepressions"><strong>BreachDepressions</strong></a> or <a href="./hydrological_analysis.html#filldepressions"><strong>FillDepressions</strong></a> tool. The <em>basins raster</em> must contain features
that are delineated by categorical (integer valued) unique indentifier values. All non-NoData,
non-zero valued grid cells in the basins raster are interpreted as belonging to features.
In practice, this tool is usual run using either a single watershed, a group of contiguous
non-overlapping watersheds, or a series of nested subbasins. These are often derived using
the <a href="./hydrological_analysis.html#watershed"><strong>Watershed</strong></a> tool, based on a series of input outlets, or the <a href="./hydrological_analysis.html#subbasins"><strong>Subbasins</strong></a> tool, based on
an input stream network. If subbasins are input to <a href="./hydrological_analysis.html#longestflowpath"><strong>LongestFlowpath</strong></a>, each traced flowpath
will include only the non-overlapping portions within nested areas. Therefore, this can be a
convienent method of delineating the longest flowpath to each bifurcation in a stream network.</p>
<p>The output vector file will contain fields in the attribute table that identify the associated
basin unique identifier (<em>BASIN</em>), the elevation of the flowpath source point on the divide
(<em>UP_ELEV</em>), the elevation of the outlet point (<em>DN_ELEV</em>), the length of the flowpath (<em>LENGTH</em>),
and finally, the average slope (<em>AVG_SLOPE</em>) along the flowpath, measured as a percent grade.</p>
<p><em>See Also</em>:</p>
<p><code>MaximumUpslopeFlowpath</code>, <a href="./hydrological_analysis.html#breachdepressions"><strong>BreachDepressions</strong></a>, <a href="./hydrological_analysis.html#filldepressions"><strong>FillDepressions</strong></a>, <a href="./hydrological_analysis.html#watershed"><strong>Watershed</strong></a>, <a href="./hydrological_analysis.html#subbasins"><strong>Subbasins</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>--basins           </td><td> Input raster basins file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">longest_flowpath(
    dem, 
    basins, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LongestFlowpath -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif --basins=basins.tif ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/longest_flowpath.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 29/10/2018</p>
<p><a name="MaxUpslopeFlowpathLength"></a></p>
<a class="header" href="#maxupslopeflowpathlength" id="maxupslopeflowpathlength"><h1>MaxUpslopeFlowpathLength</h1></a>
<p>Measures the maximum length of all upslope flowpaths draining each grid cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max_upslope_flowpath_length(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaxUpslopeFlowpathLength -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/max_upslope_flowpath.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="NumInflowingNeighbours"></a></p>
<a class="header" href="#numinflowingneighbours" id="numinflowingneighbours"><h1>NumInflowingNeighbours</h1></a>
<p>Computes the number of inflowing neighbours to each cell in an input DEM based on the D8 algorithm.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">num_inflowing_neighbours(
    dem, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NumInflowingNeighbours -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/num_inflowing_neighbours.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="RaiseWalls"></a></p>
<a class="header" href="#raisewalls" id="raisewalls"><h1>RaiseWalls</h1></a>
<p>Raises walls in a DEM along a line or around a polygon, e.g. a watershed.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector lines or polygons file</td></tr>
<tr><td>--breach           </td><td> Optional input vector breach lines</td></tr>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--height           </td><td> Wall height</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">raise_walls(
    i, 
    dem, 
    output, 
    breach=None, 
    height=100.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RaiseWalls -v --wd=&quot;/path/to/data/&quot; ^
-i=watershed.shp --dem=dem.tif -o=output.tif ^
--height=25.0
&gt;&gt;./whitebox_tools -r=RaiseWalls -v ^
--wd=&quot;/path/to/data/&quot; -i=watershed.shp --breach=outlet.shp ^
--dem=dem.tif -o=output.tif --height=25.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/raise_walls.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 22/04/2018</p>
<p><a name="Rho8Pointer"></a></p>
<a class="header" href="#rho8pointer" id="rho8pointer"><h1>Rho8Pointer</h1></a>
<p>Calculates a stochastic Rho8 flow pointer raster from an input DEM.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">rho8_pointer(
    dem, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Rho8Pointer -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/rho8_pointer.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 16, 2017</p>
<p><a name="Sink"></a></p>
<a class="header" href="#sink" id="sink"><h1>Sink</h1></a>
<p>Identifies the depressions in a DEM, giving each feature a unique identifier.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">sink(
    dem, 
    output, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Sink -v --wd=&quot;/path/to/data/&quot; ^
--dem=DEM.tif -o=output.tif --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/sink.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="SnapPourPoints"></a></p>
<a class="header" href="#snappourpoints" id="snappourpoints"><h1>SnapPourPoints</h1></a>
<p>Moves outlet points used to specify points of interest in a watershedding operation to the cell with the highest flow accumulation in its neighbourhood.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--pour_pts         </td><td> Input vector pour points (outlet) file</td></tr>
<tr><td>--flow_accum       </td><td> Input raster D8 flow accumulation file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--snap_dist        </td><td> Maximum snap distance in map units</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">snap_pour_points(
    pour_pts, 
    flow_accum, 
    output, 
    snap_dist, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SnapPourPoints -v --wd=&quot;/path/to/data/&quot; ^
--pour_pts='pour_pts.shp' --flow_accum='d8accum.tif' ^
-o='output.shp' --snap_dist=15.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/snap_pour_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="StochasticDepressionAnalysis"></a></p>
<a class="header" href="#stochasticdepressionanalysis" id="stochasticdepressionanalysis"><h1>StochasticDepressionAnalysis</h1></a>
<p>Preforms a stochastic analysis of depressions within a DEM, calculating the
probability of each cell belonging to a depression.</p>
<p>This tool differs from the original Whitebox GAT tool in a few significant ways:</p>
<ol>
<li>
<p>The Whitebox GAT tool took an error histogram as an input. In practice people found
it difficult to create this input. Usually they just generated a normal distribution
in a spreadsheet using information about the DEM RMSE. As such, this tool takes a
RMSE input and generates the histogram internally. This is far more convienent for
most applications but loses the flexibility of specifying the error distribution
more completely.</p>
</li>
<li>
<p>The Whitebox GAT tool generated the error fields using the turning bands method.
This tool generates a random Gaussian error field with no spatial autocorrelation
and then applies local spatial averaging using a Gaussian filter (the size of
which depends of the error autocorrelation length input) to increase the level of
autocorrelation. We use the Fast Almost Gaussian Filter of Peter Kovesi (2010),
which uses five repeat passes of a mean filter, based on an integral image. This
filter method is highly efficient. This results in a very significant performance
increase compared with the original tool.</p>
</li>
<li>
<p>The tool operates concurrently, compared with the original tool which calculated
pdep in serial. Again, this parallel processing can significantly improve the
performance, particularly when the tool is applied on hardware with four or
more processors.</p>
</li>
</ol>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--rmse             </td><td> The DEM's root-mean-square-error (RMSE), in z units. This determines error magnitude</td></tr>
<tr><td>--range            </td><td> The error field's correlation length, in xy-units</td></tr>
<tr><td>--iterations       </td><td> The number of iterations</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">stochastic_depression_analysis(
    dem, 
    output, 
    rmse, 
    range, 
    iterations=1000, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StochasticDepressionAnalysis -v ^
--wd=&quot;/path/to/data/&quot; --dem=DEM.tif -o=out.tif --rmse=10.0 ^
--range=850.0 --iterations=2500 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/stochastic_depression_analysis.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 11/05/2018</p>
<p><a name="StrahlerOrderBasins"></a></p>
<a class="header" href="#strahlerorderbasins" id="strahlerorderbasins"><h1>StrahlerOrderBasins</h1></a>
<p>Identifies Strahler-order basins from an input stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">strahler_order_basins(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StrahlerOrderBasins -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr='d8pntr.tif' ^
--streams='streams.tif' -o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/strahler_basins.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="Subbasins"></a></p>
<a class="header" href="#subbasins" id="subbasins"><h1>Subbasins</h1></a>
<p>Identifies the catchments, or sub-basin, draining to each link in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input D8 pointer raster file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">subbasins(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Subbasins -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr='d8pntr.tif' --streams='streams.tif' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/subbasins.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="TraceDownslopeFlowpaths"></a></p>
<a class="header" href="#tracedownslopeflowpaths" id="tracedownslopeflowpaths"><h1>TraceDownslopeFlowpaths</h1></a>
<p>Traces downslope flowpaths from one or more target sites (i.e. seed points).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--seed_pts         </td><td> Input vector seed points file</td></tr>
<tr><td>--d8_pntr          </td><td> Input D8 pointer raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">trace_downslope_flowpaths(
    seed_pts, 
    d8_pntr, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TraceDownslopeFlowpaths -v ^
--wd=&quot;/path/to/data/&quot; --seed_pts=seeds.shp ^
--flow_dir=flow_directions.tif --output=flow_paths.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/trace_downslope_flowpaths.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 4, 2017</p>
<p><a name="UnnestBasins"></a></p>
<a class="header" href="#unnestbasins" id="unnestbasins"><h1>UnnestBasins</h1></a>
<p>In some applications it is necessary to relate a measured variable for a group of
hydrometric stations (e.g. characteristics of flow timing and duration or water
chemistry) to some characteristics of each outlet's catchment (e.g. mean slope,
area of wetlands, etc.). When the group of outlets are nested, i.e. some stations
are located downstream of others, then performing a watershed operation will
result in inappropriate watershed delineation. In particular, the delineated
watersheds of each nested outlet will not include the catchment areas of upstream
outlets. This creates a serious problem for this type of application.</p>
<p>The Unnest Basin tool can be used to perform a watershedding operation based on a
group of specified pour points, i.e. outlets or target cells, such that each
complete watershed is delineated. The user must specify the name of a flow pointer
(flow direction) raster, a pour point raster, and the name of the output rasters.
Multiple numbered outputs will be created, one for each nesting level. Pour point,
or target, cells are denoted in the input pour-point image as any non-zero,
non-NoData value. The flow pointer raster should be generated using the D8
algorithm.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input D8 pointer raster file</td></tr>
<tr><td>--pour_pts         </td><td> Input vector pour points (outlet) file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">unnest_basins(
    d8_pntr, 
    pour_pts, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=UnnestBasins -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr='d8pntr.tif' --pour_pts='pour_pts.shp' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/unnest_basins.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/04/2018</p>
<p><a name="Watershed"></a></p>
<a class="header" href="#watershed" id="watershed"><h1>Watershed</h1></a>
<p>Identifies the watershed, or drainage basin, draining to a set of target cells.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input D8 pointer raster file</td></tr>
<tr><td>--pour_pts         </td><td> Input vector pour points (outlet) file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">watershed(
    d8_pntr, 
    pour_pts, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Watershed -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr='d8pntr.tif' --pour_pts='pour_pts.shp' ^
-o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/hydro_analysis/watershed.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<a class="header" href="#image-processing-tools" id="image-processing-tools"><h1>Image Processing Tools</h1></a>
<ul>
<li><a href="#ChangeVectorAnalysis">ChangeVectorAnalysis</a></li>
<li><a href="#Closing">Closing</a></li>
<li><a href="#CreateColourComposite">CreateColourComposite</a></li>
<li><a href="#FlipImage">FlipImage</a></li>
<li><a href="#IhsToRgb">IhsToRgb</a></li>
<li><a href="#ImageStackProfile">ImageStackProfile</a></li>
<li><a href="#IntegralImage">IntegralImage</a></li>
<li><a href="#KMeansClustering">KMeansClustering</a></li>
<li><a href="#LineThinning">LineThinning</a></li>
<li><a href="#ModifiedKMeansClustering">ModifiedKMeansClustering</a></li>
<li><a href="#Mosaic">Mosaic</a></li>
<li><a href="#MosaicWithFeathering">MosaicWithFeathering</a></li>
<li><a href="#NormalizedDifferenceVegetationIndex">NormalizedDifferenceVegetationIndex</a></li>
<li><a href="#Opening">Opening</a></li>
<li><a href="#RemoveSpurs">RemoveSpurs</a></li>
<li><a href="#Resample">Resample</a></li>
<li><a href="#RgbToIhs">RgbToIhs</a></li>
<li><a href="#SplitColourComposite">SplitColourComposite</a></li>
<li><a href="#ThickenRasterLine">ThickenRasterLine</a></li>
<li><a href="#TophatTransform">TophatTransform</a></li>
<li><a href="#WriteFunctionMemoryInsertion">WriteFunctionMemoryInsertion</a></li>
</ul>
<p><a name="ChangeVectorAnalysis"></a></p>
<a class="header" href="#changevectoranalysis" id="changevectoranalysis"><h1>ChangeVectorAnalysis</h1></a>
<p>Change Vector Analysis (CVA) is a change detection method that characterizes the
magnitude and change direction in spectral space between two times. A change vector
is the difference vector between two vectors in n-dimensional feature space defined
for two observations of the same geographical location (i.e. corresponding pixels)
during two dates. The CVA inputs include the set of raster images corresponding to
the multispectral data for each date. Note that there must be the same number of
image files (bands) for the two dates and they must be entered in the same order,
i.e. if three bands, red, green, and blue are entered for date one, these same
bands must be entered in the same order for date two.</p>
<p>CVA outputs two image files. The first image contains the change vector length,
i.e. magnitude, for each pixel in the multi-spectral dataset. The second image
contains information about the direction of the change event in spectral feature
space, which is related to the type of change event, e.g. deforestation will likely
have a different change direction than say crop growth. The vector magnitude is a
continuous numerical variable. The change vector direction is presented in the form
of a code, referring to the multi-dimensional sector in which the change vector
occurs. A text output will be produced to provide a key describing sector codes,
relating the change vector to positive or negative shifts in n-dimensional feature
space.</p>
<p>It is common to apply a simple thresholding operation on the magnitude data to
determine 'actual' change (i.e. change above some assumed level of error). The type
of change (qualitatively) is then defined according to the corresponding sector code.
Jensen (2005) provides a useful description of this approach to change detection.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--date1            </td><td> Input raster files for the earlier date</td></tr>
<tr><td>--date2            </td><td> Input raster files for the later date</td></tr>
<tr><td>--magnitude        </td><td> Output vector magnitude raster file</td></tr>
<tr><td>--direction        </td><td> Output vector Direction raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">change_vector_analysis(
    date1, 
    date2, 
    magnitude, 
    direction, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ChangeVectorAnalysis -v ^
--wd=&quot;/path/to/data/&quot; ^
--date1='d1_band1.tif;d1_band2.tif;d1_band3.tif' ^
--date2='d2_band1.tif;d2_band2.tif;d2_band3.tif' ^
--magnitude=mag_out.tif --direction=dir_out.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/change_vector_analysis.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 29/04/2018</p>
<p><a name="Closing"></a></p>
<a class="header" href="#closing" id="closing"><h1>Closing</h1></a>
<p>A closing is a mathematical morphology operation involving an erosion (min filter) of a dilation (max filter) set.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">closing(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Closing -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/closing.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="CreateColourComposite"></a></p>
<a class="header" href="#createcolourcomposite" id="createcolourcomposite"><h1>CreateColourComposite</h1></a>
<p>This tool can be used to create a colour-composite image from three bands of multi-spectral imagery.
The user must specify the names of the input images to enter into the red, green, and blue channels
of the resulting composite image. The output image uses the 32-bit aRGB colour model, and therefore,
in addition to red, green and blue bands, the user may optionally specify a fourth image that will
be used to determine pixel opacity (the 'a' channel). If no opacity image is specified, each pixel
will be opaque. This can be useful for cropping an image to an irregular-shaped boundary. The opacity
channel can also be used to create transparent gradients in the composite image.</p>
<p>A balance contrast enchancment (BCE) can optionally be performed on the bands prior to creation of
the colour composite. While this operation will add to the runtime of <a href="./image_processing_tools.html#createcolourcomposite"><strong>CreateColourComposite</strong></a>, if
the individual input bands have not already had contrast enchancements, then it is advisable that
the BCE option be used to improve the quality of the resulting colour composite image.</p>
<p>NoData values in any of the input images are assigned NoData values in the output image and are not
taken into account when performing the BCE operation. Please note, not all images have NoData values
identified. When this is the case, and when the background value is 0 (often the case with
multispectral imagery), then the <a href="./image_processing_tools.html#createcolourcomposite"><strong>CreateColourComposite</strong></a> tool can be told to ignore zero values using
the <code>--zeros</code> flag.</p>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools_image_enhancement.html#balancecontrastenhancement"><strong>BalanceContrastEnhancement</strong></a>, <a href="./image_processing_tools.html#splitcolourcomposite"><strong>SplitColourComposite</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--red              </td><td> Input red band image file</td></tr>
<tr><td>--green            </td><td> Input green band image file</td></tr>
<tr><td>--blue             </td><td> Input blue band image file</td></tr>
<tr><td>--opacity          </td><td> Input opacity band image file (optional)</td></tr>
<tr><td>-o, --output       </td><td> Output colour composite file</td></tr>
<tr><td>--enhance          </td><td> Optional flag indicating whether a balance contrast enhancement is performed</td></tr>
<tr><td>--zeros            </td><td> Optional flag to indicate if zeros are nodata values</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">create_colour_composite(
    red, 
    green, 
    blue, 
    output, 
    opacity=None, 
    enhance=True, 
    zeros=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CreateColourComposite -v ^
--wd=&quot;/path/to/data/&quot; --red=band3.tif --green=band2.tif ^
--blue=band1.tif -o=output.tif
&gt;&gt;./whitebox_tools ^
-r=CreateColourComposite -v --wd=&quot;/path/to/data/&quot; ^
--red=band3.tif --green=band2.tif --blue=band1.tif ^
--opacity=a.tif -o=output.tif --enhance --zeros 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/create_colour_composite.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 19, 2017</p>
<p><a name="FlipImage"></a></p>
<a class="header" href="#flipimage" id="flipimage"><h1>FlipImage</h1></a>
<p>Reflects an image in the vertical or horizontal axis.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--direction        </td><td> Direction of reflection; options include 'v' (vertical), 'h' (horizontal), and 'b' (both)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">flip_image(
    i, 
    output, 
    direction=&quot;vertical&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FlipImage -v --wd=&quot;/path/to/data/&quot; ^
--input=in.tif -o=out.tif --direction=h 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/flip_image.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="IhsToRgb"></a></p>
<a class="header" href="#ihstorgb" id="ihstorgb"><h1>IhsToRgb</h1></a>
<p>Converts intensity, hue, and saturation (IHS) images into red, green, and blue (RGB) images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--intensity        </td><td> Input intensity file</td></tr>
<tr><td>--hue              </td><td> Input hue file</td></tr>
<tr><td>--saturation       </td><td> Input saturation file</td></tr>
<tr><td>--red              </td><td> Output red band file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--green            </td><td> Output green band file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--blue             </td><td> Output blue band file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>-o, --output       </td><td> Output colour-composite file. Only used if individual bands are not specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">ihs_to_rgb(
    intensity, 
    hue, 
    saturation, 
    red=None, 
    green=None, 
    blue=None, 
    output=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=IhsToRgb -v --wd=&quot;/path/to/data/&quot; ^
--intensity=intensity.tif --hue=hue.tif ^
--saturation=saturation.tif --red=band3.tif --green=band2.tif ^
--blue=band1.tif
&gt;&gt;./whitebox_tools -r=IhsToRgb -v ^
--wd=&quot;/path/to/data/&quot; --intensity=intensity.tif --hue=hue.tif ^
--saturation=saturation.tif --composite=image.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/ihs_to_rgb.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 25, 2017</p>
<p><a name="ImageStackProfile"></a></p>
<a class="header" href="#imagestackprofile" id="imagestackprofile"><h1>ImageStackProfile</h1></a>
<p>Plots an image stack profile (i.e. signature) for a set of points and multispectral images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input multispectral image files</td></tr>
<tr><td>--points           </td><td> Input vector points file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">image_stack_profile(
    inputs, 
    points, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ImageStackProfile -v ^
--wd=&quot;/path/to/data/&quot; -i='image1.tif;image2.tif;image3.tif' ^
--points=pts.shp -o=output.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/image_stack_profile.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: March 15, 2018</p>
<p><a name="IntegralImage"></a></p>
<a class="header" href="#integralimage" id="integralimage"><h1>IntegralImage</h1></a>
<p>Transforms an input image (summed area table) into its integral image equivalent.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">integral_image(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=IntegralImage -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/integral_image.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="KMeansClustering"></a></p>
<a class="header" href="#kmeansclustering" id="kmeansclustering"><h1>KMeansClustering</h1></a>
<p>Performs a k-means clustering operation on a multi-spectral dataset.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--out_html         </td><td> Output HTML report file</td></tr>
<tr><td>--classes          </td><td> Number of classes</td></tr>
<tr><td>--max_iterations   </td><td> Maximum number of iterations</td></tr>
<tr><td>--class_change     </td><td> Minimum percent of cells changed between iterations before completion</td></tr>
<tr><td>--initialize       </td><td> How to initialize cluster centres?</td></tr>
<tr><td>--min_class_size   </td><td> Minimum class size, in pixels</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">k_means_clustering(
    inputs, 
    output, 
    classes, 
    out_html=None, 
    max_iterations=10, 
    class_change=2.0, 
    initialize=&quot;diagonal&quot;, 
    min_class_size=10, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=KMeansClustering -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
-o=output.tif --out_html=report.html --classes=15 ^
--max_iterations=25 --class_change=1.5 --initialize='random' ^
--min_class_size=500 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/k_means_clustering.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 27, 2017</p>
<p><a name="LineThinning"></a></p>
<a class="header" href="#linethinning" id="linethinning"><h1>LineThinning</h1></a>
<p>Performs line thinning a on Boolean raster image; intended to be used with the RemoveSpurs tool.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">line_thinning(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LineThinning -v --wd=&quot;/path/to/data/&quot; ^
--input=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/line_thin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="ModifiedKMeansClustering"></a></p>
<a class="header" href="#modifiedkmeansclustering" id="modifiedkmeansclustering"><h1>ModifiedKMeansClustering</h1></a>
<p>This modified k-means algorithm is similar to that described by Mather (2004).
The main difference between the traditional k-means and this technique is that the user
does not need to specify the desired number of classes/clusters prior to running the
tool. Instead, the algorithm initializes with a very liberal overestimate of the number
of classes and then merges classes that have cluster centres that are separated by less
than a user-defined threshold. The main difference between this algorithm and the ISODATA
technique is that clusters can not be broken apart into two smaller clusters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--out_html         </td><td> Output HTML report file</td></tr>
<tr><td>--start_clusters   </td><td> Initial number of clusters</td></tr>
<tr><td>--merger_dist      </td><td> Cluster merger distance</td></tr>
<tr><td>--max_iterations   </td><td> Maximum number of iterations</td></tr>
<tr><td>--class_change     </td><td> Minimum percent of cells changed between iterations before completion</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">modified_k_means_clustering(
    inputs, 
    output, 
    out_html=None, 
    start_clusters=1000, 
    merger_dist=None, 
    max_iterations=10, 
    class_change=2.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ModifiedKMeansClustering -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
-o=output.tif --out_html=report.html --start_clusters=100 ^
--merger_dist=30.0 --max_iterations=25 --class_change=1.5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/modified_k_means_clustering.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 30, 2017</p>
<p><a name="Mosaic"></a></p>
<a class="header" href="#mosaic" id="mosaic"><h1>Mosaic</h1></a>
<p>This tool will create an image mosaic from one or more input image files using
one of three resampling methods including, nearest neighbour, bilinear interpolation,
and cubic convolution. The order of the input source image files is important. Grid
cells in the output image will be assigned the corresponding value determined from the
first image found in the list to possess an overlapping coordinate.</p>
<p>This is the preferred mosaicing tool to use when appending multiple images with
little to no overlapping areas, e.g. tiled data. When images have significant overlap
areas, users are advised to use the <a href="./image_processing_tools.html#mosaicwithfeathering"><strong>MosaicWithFeathering</strong></a> tool instead.</p>
<p>Resample is very similar in operation to the Mosaic tool. The Resample tool should be
used when there is an existing image into which you would like to dump information from
one or more source images. If the source images are more extensive than the destination
image, i.e. there are areas that extend beyond the destination image boundaries, these
areas will not be represented in the updated image. Grid cells in the destination image
that are not overlapping with any of the input source images will not be updated, i.e.
they will possess the same value as before the resampling operation. The Mosaic tool is
used when there is no existing destination image. In this case, a new image is created
that represents the bounding rectangle of each of the two or more input images. Grid
cells in the output image that do not overlap with any of the input images will be
assigned the NoData value.</p>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools.html#mosaicwithfeathering"><strong>MosaicWithFeathering</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--method           </td><td> Resampling method; options include 'nn' (nearest neighbour), 'bilinear', and 'cc' (cubic convolution)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">mosaic(
    inputs, 
    output, 
    method=&quot;cc&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Mosaic -v --wd='/path/to/data/' ^
-i='image1.tif;image2.tif;image3.tif' -o=dest.tif ^
--method='cc' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/mosaic.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: January 2 2018</p>
<p><a name="MosaicWithFeathering"></a></p>
<a class="header" href="#mosaicwithfeathering" id="mosaicwithfeathering"><h1>MosaicWithFeathering</h1></a>
<p>This tool will create a mosaic from two input images. It is similar in operation to the <a href="./image_processing_tools.html#mosaic"><strong>Mosaic</strong></a> tool,
however, this tool is the preferred method of mosaicing images when there is significant overlap between
the images. For areas of overlap, the feathering method will calculate the output value as a weighted
combination of the two input values, where the weights are derived from the squared distance of the
pixel to the edge of the data in each of the input raster files. Therefore, less weight is assigned to
an image's pixel value where the pixel is very near the edge of the image. Note that the distance is
actually calculated to the edge of the grid and not necessarily the edge of the data, which can differ
if the image has been rotated during registration.  The result of this feathering method is that the
output mosaic image should have very little evidence of the original image edges within the overlapping
area.</p>
<p>Unlike the Mosaic tool, which can take multiple input images, this tool only accepts two input images.
Mosaic is therefore useful when there are many, adjacent or only slightly overlapping images, e.g. for
tiled data sets.</p>
<p>Users may want to use the <a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a> tool prior to mosaicing if the two input images differ
significantly in their radiometric properties. i.e. if image contrast differences exist.</p>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools.html#mosaic"><strong>Mosaic</strong></a>, <a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input raster file to modify</td></tr>
<tr><td>--i2, --input2    </td><td> Input reference raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--method           </td><td> Resampling method; options include 'nn' (nearest neighbour), 'bilinear', and 'cc' (cubic convolution)</td></tr>
<tr><td>--weight           </td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">mosaic_with_feathering(
    input1, 
    input2, 
    output, 
    method=&quot;cc&quot;, 
    weight=4.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MosaicWithFeathering -v ^
--wd='/path/to/data/' --input1='image1.tif' ^
--input2='image2.tif' -o='output.tif' --method='cc' ^
--weight=4.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/mosaic_with_feathering.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 29/12/2018</p>
<p><a name="NormalizedDifferenceVegetationIndex"></a></p>
<a class="header" href="#normalizeddifferencevegetationindex" id="normalizeddifferencevegetationindex"><h1>NormalizedDifferenceVegetationIndex</h1></a>
<p>Calculates the normalized difference vegetation index (NDVI) from near-infrared and red imagery.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--nir              </td><td> Input near-infrared band image</td></tr>
<tr><td>--red              </td><td> Input red band image</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
<tr><td>--osavi            </td><td> Optional flag indicating whether the optimized soil-adjusted veg index (OSAVI) should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">normalized_difference_vegetation_index(
    nir, 
    red, 
    output, 
    clip=0.0, 
    osavi=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NormalizedDifferenceVegetationIndex -v ^
--wd=&quot;/path/to/data/&quot; --nir=band4.tif --red=band3.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools ^
-r=NormalizedDifferenceVegetationIndex -v --wd=&quot;/path/to/data/&quot; ^
--nir=band4.tif --red=band3.tif -o=output.tif --clip=1.0 ^
--osavi 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/ndvi.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="Opening"></a></p>
<a class="header" href="#opening" id="opening"><h1>Opening</h1></a>
<p>An opening is a mathematical morphology operation involving a dilation (max filter) of an erosion (min filter) set.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">opening(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Opening -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/opening.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="RemoveSpurs"></a></p>
<a class="header" href="#removespurs" id="removespurs"><h1>RemoveSpurs</h1></a>
<p>Removes the spurs (pruning operation) from a Boolean line image.; intended to be used on the output of the LineThinning tool.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--iterations       </td><td> Maximum number of iterations</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">remove_spurs(
    i, 
    output, 
    iterations=10, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RemoveSpurs -v --wd=&quot;/path/to/data/&quot; ^
--input=DEM.tif -o=output.tif --iterations=10 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/remove_spurs.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Resample"></a></p>
<a class="header" href="#resample" id="resample"><h1>Resample</h1></a>
<p>Resample is very similar in operation to the Mosaic tool. The Resample tool should
be used when there is an existing image into which you would like to dump information
from one or more source images. If the source images are more extensive than the
destination image, i.e. there are areas that extend beyond the destination image
boundaries, these areas will not be represented in the updated image. Grid cells in the
destination image that are not overlapping with any of the input source images will not
be updated, i.e. they will possess the same value as before the resampling operation. The
Mosaic tool is used when there is no existing destination image. In this case, a new
image is created that represents the bounding rectangle of each of the two or more input
images. Grid cells in the output image that do not overlap with any of the input images
will be assigned the NoData value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--destination      </td><td> Destination raster file</td></tr>
<tr><td>--method           </td><td> Resampling method; options include 'nn' (nearest neighbour), 'bilinear', and 'cc' (cubic convolution)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">resample(
    inputs, 
    destination, 
    method=&quot;cc&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Resample -v --wd='/path/to/data/' ^
-i='image1.tif;image2.tif;image3.tif' --destination=dest.tif ^
--method='cc 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/resample.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: January 1 2018</p>
<p><a name="RgbToIhs"></a></p>
<a class="header" href="#rgbtoihs" id="rgbtoihs"><h1>RgbToIhs</h1></a>
<p>Converts red, green, and blue (RGB) images into intensity, hue, and saturation (IHS) images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--red              </td><td> Input red band image file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--green            </td><td> Input green band image file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--blue             </td><td> Input blue band image file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--composite        </td><td> Input colour-composite image file. Only used if individual bands are not specified</td></tr>
<tr><td>--intensity        </td><td> Output intensity raster file</td></tr>
<tr><td>--hue              </td><td> Output hue raster file</td></tr>
<tr><td>--saturation       </td><td> Output saturation raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">rgb_to_ihs(
    intensity, 
    hue, 
    saturation, 
    red=None, 
    green=None, 
    blue=None, 
    composite=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RgbToIhs -v --wd=&quot;/path/to/data/&quot; ^
--red=band3.tif --green=band2.tif --blue=band1.tif ^
--intensity=intensity.tif --hue=hue.tif ^
--saturation=saturation.tif
&gt;&gt;./whitebox_tools -r=RgbToIhs -v ^
--wd=&quot;/path/to/data/&quot; --composite=image.tif ^
--intensity=intensity.tif --hue=hue.tif ^
--saturation=saturation.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/rgb_to_ihs.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 25, 2017</p>
<p><a name="SplitColourComposite"></a></p>
<a class="header" href="#splitcolourcomposite" id="splitcolourcomposite"><h1>SplitColourComposite</h1></a>
<p>This tool splits an RGB colour composite image into seperate multispectral images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input colour composite image file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file (suffixes of '_r', '_g', and '_b' will be appended)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">split_colour_composite(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SplitColourComposite -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/split_colour_composite.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 15, 2017</p>
<p><a name="ThickenRasterLine"></a></p>
<a class="header" href="#thickenrasterline" id="thickenrasterline"><h1>ThickenRasterLine</h1></a>
<p>Thickens single-cell wide lines within a raster image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">thicken_raster_line(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ThickenRasterLine -v ^
--wd=&quot;/path/to/data/&quot; --input=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/thicken_line.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 4, 2017</p>
<p><a name="TophatTransform"></a></p>
<a class="header" href="#tophattransform" id="tophattransform"><h1>TophatTransform</h1></a>
<p>Performs either a white or black top-hat transform on an input image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>--variant          </td><td> Optional variant value. Options include 'white' and 'black'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">tophat_transform(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    variant=&quot;white&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TophatTransform -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/tophat.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="WriteFunctionMemoryInsertion"></a></p>
<a class="header" href="#writefunctionmemoryinsertion" id="writefunctionmemoryinsertion"><h1>WriteFunctionMemoryInsertion</h1></a>
<p>Performs a write function memory insertion for single-band multi-date change detection.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input raster file associated with the first date</td></tr>
<tr><td>--i2, --input2    </td><td> Input raster file associated with the second date</td></tr>
<tr><td>--i3, --input3    </td><td> Optional input raster file associated with the third date</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">write_function_memory_insertion(
    input1, 
    input2, 
    output, 
    input3=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=WriteFunctionMemoryInsertion -v ^
--wd=&quot;/path/to/data/&quot; -i1=input1.tif -i2=input2.tif ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/write_func_memory_insertion.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 18, 2017</p>
<a class="header" href="#image-processing-tools--filters" id="image-processing-tools--filters"><h1>Image Processing Tools → Filters</h1></a>
<ul>
<li><a href="#AdaptiveFilter">AdaptiveFilter</a></li>
<li><a href="#BilateralFilter">BilateralFilter</a></li>
<li><a href="#ConservativeSmoothingFilter">ConservativeSmoothingFilter</a></li>
<li><a href="#CornerDetection">CornerDetection</a></li>
<li><a href="#DiffOfGaussianFilter">DiffOfGaussianFilter</a></li>
<li><a href="#DiversityFilter">DiversityFilter</a></li>
<li><a href="#EdgePreservingMeanFilter">EdgePreservingMeanFilter</a></li>
<li><a href="#EmbossFilter">EmbossFilter</a></li>
<li><a href="#FastAlmostGaussianFilter">FastAlmostGaussianFilter</a></li>
<li><a href="#GaussianFilter">GaussianFilter</a></li>
<li><a href="#HighPassFilter">HighPassFilter</a></li>
<li><a href="#HighPassMedianFilter">HighPassMedianFilter</a></li>
<li><a href="#KNearestMeanFilter">KNearestMeanFilter</a></li>
<li><a href="#LaplacianFilter">LaplacianFilter</a></li>
<li><a href="#LaplacianOfGaussianFilter">LaplacianOfGaussianFilter</a></li>
<li><a href="#LeeFilter">LeeFilter</a></li>
<li><a href="#LineDetectionFilter">LineDetectionFilter</a></li>
<li><a href="#MajorityFilter">MajorityFilter</a></li>
<li><a href="#MaximumFilter">MaximumFilter</a></li>
<li><a href="#MeanFilter">MeanFilter</a></li>
<li><a href="#MedianFilter">MedianFilter</a></li>
<li><a href="#MinimumFilter">MinimumFilter</a></li>
<li><a href="#OlympicFilter">OlympicFilter</a></li>
<li><a href="#PercentileFilter">PercentileFilter</a></li>
<li><a href="#PrewittFilter">PrewittFilter</a></li>
<li><a href="#RangeFilter">RangeFilter</a></li>
<li><a href="#RobertsCrossFilter">RobertsCrossFilter</a></li>
<li><a href="#ScharrFilter">ScharrFilter</a></li>
<li><a href="#SobelFilter">SobelFilter</a></li>
<li><a href="#StandardDeviationFilter">StandardDeviationFilter</a></li>
<li><a href="#TotalFilter">TotalFilter</a></li>
<li><a href="#UnsharpMasking">UnsharpMasking</a></li>
<li><a href="#UserDefinedWeightsFilter">UserDefinedWeightsFilter</a></li>
</ul>
<p><a name="AdaptiveFilter"></a></p>
<a class="header" href="#adaptivefilter" id="adaptivefilter"><h1>AdaptiveFilter</h1></a>
<p>Performs an adaptive filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>--threshold        </td><td> Difference from mean threshold, in standard deviations</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">adaptive_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    threshold=2.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AdaptiveFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=DEM.tif -o=output.tif --filter=25 --threshold = 2.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/adaptive_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="BilateralFilter"></a></p>
<a class="header" href="#bilateralfilter" id="bilateralfilter"><h1>BilateralFilter</h1></a>
<p>A bilateral filter is an edge-preserving smoothing filter introduced by Tomasi and Manduchi (1998).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--sigma_dist       </td><td> Standard deviation in distance in pixels</td></tr>
<tr><td>--sigma_int        </td><td> Standard deviation in intensity in pixels</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">bilateral_filter(
    i, 
    output, 
    sigma_dist=0.75, 
    sigma_int=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=BilateralFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif ^
--sigma_dist=2.5 --sigma_int=4.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/bilateral_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="ConservativeSmoothingFilter"></a></p>
<a class="header" href="#conservativesmoothingfilter" id="conservativesmoothingfilter"><h1>ConservativeSmoothingFilter</h1></a>
<p>Performs a conservative-smoothing filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">conservative_smoothing_filter(
    i, 
    output, 
    filterx=3, 
    filtery=3, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ConservativeSmoothingFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/conservative_smoothing_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="CornerDetection"></a></p>
<a class="header" href="#cornerdetection" id="cornerdetection"><h1>CornerDetection</h1></a>
<p>Identifies corner patterns in boolean images using hit-and-miss pattern mattching.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input boolean image</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">corner_detection(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CornerDetection -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --sigma=2.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/corner_detection.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Simon Gudim</p>
<p><em>Created</em>: 04/05/2017</p>
<p><a name="DiffOfGaussianFilter"></a></p>
<a class="header" href="#diffofgaussianfilter" id="diffofgaussianfilter"><h1>DiffOfGaussianFilter</h1></a>
<p>Performs a Difference of Gaussian (DoG) filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--sigma1           </td><td> Standard deviation distance in pixels</td></tr>
<tr><td>--sigma2           </td><td> Standard deviation distance in pixels</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">diff_of_gaussian_filter(
    i, 
    output, 
    sigma1=2.0, 
    sigma2=4.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DiffOfGaussianFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --sigma1=2.0 ^
--sigma2=4.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/dog_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="DiversityFilter"></a></p>
<a class="header" href="#diversityfilter" id="diversityfilter"><h1>DiversityFilter</h1></a>
<p>Assigns each cell in the output grid the number of different values in a moving window centred on each grid cell in the input raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">diversity_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DiversityFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/diversity_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="EdgePreservingMeanFilter"></a></p>
<a class="header" href="#edgepreservingmeanfilter" id="edgepreservingmeanfilter"><h1>EdgePreservingMeanFilter</h1></a>
<p>Performs a simple edge-preserving mean filter on an input image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filter           </td><td> Size of the filter kernel</td></tr>
<tr><td>--threshold        </td><td> Maximum difference in values</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">edge_preserving_mean_filter(
    i, 
    output, 
    threshold, 
    filter=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=EdgePreservingMeanFilter -v ^
--wd=&quot;/path/to/data/&quot; --input=image.tif -o=output.tif ^
--filter=5 --threshold=20 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/edge_preserving_mean_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 24/03/2018</p>
<p><a name="EmbossFilter"></a></p>
<a class="header" href="#embossfilter" id="embossfilter"><h1>EmbossFilter</h1></a>
<p>Performs an emboss filter on an image, similar to a hillshade operation.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--direction        </td><td> Direction of reflection; options include 'n', 's', 'e', 'w', 'ne', 'se', 'nw', 'sw'</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">emboss_filter(
    i, 
    output, 
    direction=&quot;n&quot;, 
    clip=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=EmbossFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --direction='s' --clip=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/emboss_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="FastAlmostGaussianFilter"></a></p>
<a class="header" href="#fastalmostgaussianfilter" id="fastalmostgaussianfilter"><h1>FastAlmostGaussianFilter</h1></a>
<p>The tool is somewhat modified from Dr. Kovesi's original Matlab code in that it
works with both greyscale and RGB images (decomposes to HSI and uses the intensity
data) and it handles the case of rasters that contain NoData values. This adds
complexity to the original 20 additions and 5 multiplications assertion of the
original paper.</p>
<p>Also note, for small values of sigma (&lt; 1.8), you should probably just use the
regular GaussianFilter tool.</p>
<p><em>Reference</em>:</p>
<p>P. Kovesi 2010 Fast Almost-Gaussian Filtering, Digital Image Computing:
Techniques and Applications (DICTA), 2010 International Conference on.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--sigma            </td><td> Standard deviation distance in pixels</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">fast_almost_gaussian_filter(
    i, 
    output, 
    sigma=1.8, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FastAlmostGaussianFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --sigma=2.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/fast_almost_gaussian_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 19/05/2019</p>
<p><a name="GaussianFilter"></a></p>
<a class="header" href="#gaussianfilter" id="gaussianfilter"><h1>GaussianFilter</h1></a>
<p>Performs a Gaussian filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--sigma            </td><td> Standard deviation distance in pixels</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">gaussian_filter(
    i, 
    output, 
    sigma=0.75, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=GaussianFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --sigma=2.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/gaussian_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="HighPassFilter"></a></p>
<a class="header" href="#highpassfilter" id="highpassfilter"><h1>HighPassFilter</h1></a>
<p>Performs a high-pass filter on an input image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">high_pass_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HighPassFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/highpass_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="HighPassMedianFilter"></a></p>
<a class="header" href="#highpassmedianfilter" id="highpassmedianfilter"><h1>HighPassMedianFilter</h1></a>
<p>Performs a high pass filter based on a median filter.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>--sig_digits       </td><td> Number of significant digits</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">high_pass_median_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    sig_digits=2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HighPassMedianFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/highpass_median_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 10/09/2018</p>
<p><a name="KNearestMeanFilter"></a></p>
<a class="header" href="#knearestmeanfilter" id="knearestmeanfilter"><h1>KNearestMeanFilter</h1></a>
<p>A k-nearest mean filter is a type of edge-preserving smoothing filter.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>-k                  </td><td> k-value in pixels; this is the number of nearest-valued neighbours to use</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">k_nearest_mean_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    k=5, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=KNearestMeanFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --filter=9 ^
-k=5
&gt;&gt;./whitebox_tools -r=KNearestMeanFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --filtery=7 ^
--filtery=9 -k=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/k_nearest_mean_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="LaplacianFilter"></a></p>
<a class="header" href="#laplacianfilter" id="laplacianfilter"><h1>LaplacianFilter</h1></a>
<p>Performs a Laplacian filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--variant          </td><td> Optional variant value. Options include 3x3(1), 3x3(2), 3x3(3), 3x3(4), 5x5(1), and 5x5(2) (default is 3x3(1))</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">laplacian_filter(
    i, 
    output, 
    variant=&quot;3x3(1)&quot;, 
    clip=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LaplacianFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif ^
--variant='3x3(1)' --clip=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/laplacian_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="LaplacianOfGaussianFilter"></a></p>
<a class="header" href="#laplacianofgaussianfilter" id="laplacianofgaussianfilter"><h1>LaplacianOfGaussianFilter</h1></a>
<p>Performs a Laplacian-of-Gaussian (LoG) filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--sigma            </td><td> Standard deviation in pixels</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">laplacian_of_gaussian_filter(
    i, 
    output, 
    sigma=0.75, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LaplacianOfGaussianFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --sigma=2.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/log_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="LeeFilter"></a></p>
<a class="header" href="#leefilter" id="leefilter"><h1>LeeFilter</h1></a>
<p>Performs a Lee (Sigma) smoothing filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>--sigma            </td><td> Sigma value should be related to the standarad deviation of the distribution of image speckle noise</td></tr>
<tr><td>-m                  </td><td> M-threshold value the minimum allowable number of pixels within the intensity range</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lee_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    sigma=10.0, 
    m=5.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LeeFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=9 --sigma=10.0 ^
-m=5
&gt;&gt;./whitebox_tools -r=LeeFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filtery=7 --filtery=9 ^
--sigma=10.0 -m=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/lee_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="LineDetectionFilter"></a></p>
<a class="header" href="#linedetectionfilter" id="linedetectionfilter"><h1>LineDetectionFilter</h1></a>
<p>Performs a line-detection filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--variant          </td><td> Optional variant value. Options include 'v' (vertical), 'h' (horizontal), '45', and '135' (default is 'v')</td></tr>
<tr><td>--absvals          </td><td> Optional flag indicating whether outputs should be absolute values</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">line_detection_filter(
    i, 
    output, 
    variant=&quot;vertical&quot;, 
    absvals=False, 
    clip=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LineDetectionFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --variant=h ^
--clip=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/line_detection_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="MajorityFilter"></a></p>
<a class="header" href="#majorityfilter" id="majorityfilter"><h1>MajorityFilter</h1></a>
<p>Assigns each cell in the output grid the most frequently occurring value (mode) in a moving window centred on each grid cell in the input raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">majority_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MajorityFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/majority_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="MaximumFilter"></a></p>
<a class="header" href="#maximumfilter" id="maximumfilter"><h1>MaximumFilter</h1></a>
<p>Assigns each cell in the output grid the maximum value in a moving window centred on each grid cell in the input raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">maximum_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MaximumFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/max_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="MeanFilter"></a></p>
<a class="header" href="#meanfilter" id="meanfilter"><h1>MeanFilter</h1></a>
<p>Performs a mean filter (low-pass filter) on an input image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">mean_filter(
    i, 
    output, 
    filterx=3, 
    filtery=3, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MeanFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filterx=25 --filtery=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/mean_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="MedianFilter"></a></p>
<a class="header" href="#medianfilter" id="medianfilter"><h1>MedianFilter</h1></a>
<p>Performs an efficient median filter based on Huang, Yang, and Tang's (1979) method.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>--sig_digits       </td><td> Number of significant digits</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">median_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    sig_digits=2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MedianFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=input.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/median_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 15, 2017</p>
<p><a name="MinimumFilter"></a></p>
<a class="header" href="#minimumfilter" id="minimumfilter"><h1>MinimumFilter</h1></a>
<p>Assigns each cell in the output grid the minimum value in a moving window centred on each grid cell in the input raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">minimum_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinimumFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/min_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="OlympicFilter"></a></p>
<a class="header" href="#olympicfilter" id="olympicfilter"><h1>OlympicFilter</h1></a>
<p>Performs an olympic smoothing filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">olympic_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=OlympicFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/olympic_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="PercentileFilter"></a></p>
<a class="header" href="#percentilefilter" id="percentilefilter"><h1>PercentileFilter</h1></a>
<p>Performs a percentile filter on an input image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
<tr><td>--sig_digits       </td><td> Number of significant digits</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">percentile_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    sig_digits=2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PercentileFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/percentile_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 22, 2017</p>
<p><a name="PrewittFilter"></a></p>
<a class="header" href="#prewittfilter" id="prewittfilter"><h1>PrewittFilter</h1></a>
<p>Performs a Prewitt edge-detection filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">prewitt_filter(
    i, 
    output, 
    clip=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PrewittFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --clip=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/prewitt_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="RangeFilter"></a></p>
<a class="header" href="#rangefilter" id="rangefilter"><h1>RangeFilter</h1></a>
<p>Assigns each cell in the output grid the range of values in a moving window centred on each grid cell in the input raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">range_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RangeFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/range_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="RobertsCrossFilter"></a></p>
<a class="header" href="#robertscrossfilter" id="robertscrossfilter"><h1>RobertsCrossFilter</h1></a>
<p>Performs a Robert's cross edge-detection filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">roberts_cross_filter(
    i, 
    output, 
    clip=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RobertsCrossFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --clip=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/roberts_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="ScharrFilter"></a></p>
<a class="header" href="#scharrfilter" id="scharrfilter"><h1>ScharrFilter</h1></a>
<p>Performs a Scharr edge-detection filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">scharr_filter(
    i, 
    output, 
    clip=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ScharrFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --clip=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/scharr_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="SobelFilter"></a></p>
<a class="header" href="#sobelfilter" id="sobelfilter"><h1>SobelFilter</h1></a>
<p>Performs a Sobel edge-detection filter on an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--variant          </td><td> Optional variant value. Options include 3x3 and 5x5 (default is 3x3)</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent (default is 0.0)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">sobel_filter(
    i, 
    output, 
    variant=&quot;3x3&quot;, 
    clip=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SobelFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --variant=5x5 --clip=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/sobel_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="StandardDeviationFilter"></a></p>
<a class="header" href="#standarddeviationfilter" id="standarddeviationfilter"><h1>StandardDeviationFilter</h1></a>
<p>Assigns each cell in the output grid the standard deviation of values in a moving window centred on each grid cell in the input raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">standard_deviation_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StandardDeviationFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/stdev_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="TotalFilter"></a></p>
<a class="header" href="#totalfilter" id="totalfilter"><h1>TotalFilter</h1></a>
<p>Performs a total filter on an input image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--filterx          </td><td> Size of the filter kernel in the x-direction</td></tr>
<tr><td>--filtery          </td><td> Size of the filter kernel in the y-direction</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">total_filter(
    i, 
    output, 
    filterx=11, 
    filtery=11, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TotalFilter -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --filter=25 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/total_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="UnsharpMasking"></a></p>
<a class="header" href="#unsharpmasking" id="unsharpmasking"><h1>UnsharpMasking</h1></a>
<p>An image sharpening technique that enhances edges.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--sigma            </td><td> Standard deviation distance in pixels</td></tr>
<tr><td>--amount           </td><td> A percentage and controls the magnitude of each overshoot</td></tr>
<tr><td>--threshold        </td><td> Controls the minimal brightness change that will be sharpened</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">unsharp_masking(
    i, 
    output, 
    sigma=0.75, 
    amount=100.0, 
    threshold=0.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=UnsharpMasking -v --wd=&quot;/path/to/data/&quot; ^
-i=image.tif -o=output.tif --sigma=2.0 --amount=50.0 ^
--threshold=10.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/unsharp_masking.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 02/05/2018</p>
<p><a name="UserDefinedWeightsFilter"></a></p>
<a class="header" href="#userdefinedweightsfilter" id="userdefinedweightsfilter"><h1>UserDefinedWeightsFilter</h1></a>
<p>NoData values in the input image are ignored during the convolution operation.
This can lead to unexpected behavior at the edges of images (since the default behavior
is to return NoData when addressing cells beyond the grid edge) and where the grid
contains interior areas of NoData values. Normalization of kernel weights can be useful
for handling the edge effects associated with interior areas of NoData values. When the
normalization option is selected, the sum of the cell value-weight product is divided
by the sum of the weights on a cell-by-cell basis. Therefore, if the kernel at a
particular grid cell contains neighboring cells of NoData values, normalization
effectively re-adjusts the weighting to account for the missing data values. Normalization
also ensures that the output image will possess values within the range of the input
image and allows the user to specify integer value weights in the kernel. However, note
that this implies that the sum of weights should equal one. In some cases, alternative
sums (e.g. zero) are more appropriate, and as such normalization should not be applied
in these cases.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--weights          </td><td> Input weights file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--center           </td><td> Kernel center cell; options include 'center', 'upper-left', 'upper-right', 'lower-left', 'lower-right'</td></tr>
<tr><td>--normalize        </td><td> Normalize kernel weights? This can reduce edge effects and lessen the impact of data gaps (nodata) but is not suited when the kernel weights sum to zero</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">user_defined_weights_filter(
    i, 
    weights, 
    output, 
    center=&quot;center&quot;, 
    normalize=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=UserDefinedWeightsFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=image.tif --weights=weights.txt ^
-o=output.tif --center=center --normalize 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/user_defined_weights_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 26/04/2018</p>
<a class="header" href="#image-processing-tools--image-enhancement" id="image-processing-tools--image-enhancement"><h1>Image Processing Tools → Image Enhancement</h1></a>
<ul>
<li><a href="#BalanceContrastEnhancement">BalanceContrastEnhancement</a></li>
<li><a href="#CorrectVignetting">CorrectVignetting</a></li>
<li><a href="#DirectDecorrelationStretch">DirectDecorrelationStretch</a></li>
<li><a href="#GammaCorrection">GammaCorrection</a></li>
<li><a href="#GaussianContrastStretch">GaussianContrastStretch</a></li>
<li><a href="#HistogramEqualization">HistogramEqualization</a></li>
<li><a href="#HistogramMatching">HistogramMatching</a></li>
<li><a href="#HistogramMatchingTwoImages">HistogramMatchingTwoImages</a></li>
<li><a href="#MinMaxContrastStretch">MinMaxContrastStretch</a></li>
<li><a href="#PanchromaticSharpening">PanchromaticSharpening</a></li>
<li><a href="#PercentageContrastStretch">PercentageContrastStretch</a></li>
<li><a href="#SigmoidalContrastStretch">SigmoidalContrastStretch</a></li>
<li><a href="#StandardDeviationContrastStretch">StandardDeviationContrastStretch</a></li>
</ul>
<p><a name="BalanceContrastEnhancement"></a></p>
<a class="header" href="#balancecontrastenhancement" id="balancecontrastenhancement"><h1>BalanceContrastEnhancement</h1></a>
<p>This tool can be used to reduce colour bias in a colour composite image based on the
technique described by Liu (1991). Colour bias is a common phenomena with colour images
derived from multispectral imagery, whereby a higher average brightness value in one
band results in over-representation of that band in the colour composite. The tool
essentially applies a parabolic stretch to each of the three bands in a user specified
RGB colour composite, forcing the histograms of each band to have the same minimum,
maximum, and average values while maintaining their overall histogram shape. For greater
detail on the operation of the tool, please see Liu (1991). Aside from the names of the
input and output colour composite images, the user must also set the value of E, the
desired output band mean, where 20 &lt; E &lt; 235.</p>
<p><em>Reference</em>:</p>
<p>Liu, J.G. (1991) Balance contrast enhancement technique and its application in image
colour composition. <em>International Journal of Remote Sensing</em>, 12:10.</p>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools_image_enhancement.html#directdecorrelationstretch"><strong>DirectDecorrelationStretch</strong></a>, <a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a>, <a href="./image_processing_tools_image_enhancement.html#histogrammatchingtwoimages"><strong>HistogramMatchingTwoImages</strong></a>, <a href="./image_processing_tools_image_enhancement.html#histogramequalization"><strong>HistogramEqualization</strong></a>, <a href="./image_processing_tools_image_enhancement.html#gaussiancontraststretch"><strong>GaussianContrastStretch</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input colour composite image file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--band_mean        </td><td> Band mean value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">balance_contrast_enhancement(
    i, 
    output, 
    band_mean=100.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=BalanceContrastEnhancement -v ^
--wd=&quot;/path/to/data/&quot; --input=image.tif -o=output.tif ^
--band_mean=120 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/balance_contrast_enhancement.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 19, 2017</p>
<p><a name="CorrectVignetting"></a></p>
<a class="header" href="#correctvignetting" id="correctvignetting"><h1>CorrectVignetting</h1></a>
<p>This tool can be used to reduce vignetting within an image. Vignetting refers to the
reducuction of image brightness away from the image centre (i.e. the principal point).
Vignetting is a radiometric distortion resulting from lens characteristics. The
algorithm calculates the brightness value in the output image (BVout) as:</p>
<p>BVout = BVin / [cos^n(arctan(d / f))]</p>
<p>Where d is the photo-distance from the principal point in millimetres, f is the focal
length of the camera, in millimeters, and n is a user-specified parameter. Pixel
distances are converted to photo-distances (in millimetres) using the specified
image width, i.e. distance between left and right edges (mm). For many cameras, 4.0
is an appropriate value of the n parameter. A second pass of the image is used to
rescale the output image so that it possesses the same minimum and maximum values as
the input image.</p>
<p>If an RGB image is input, the analysis will be performed on the intensity component
of the HSI transform.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--pp               </td><td> Input principal point file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--focal_length     </td><td> Camera focal length, in millimeters</td></tr>
<tr><td>--image_width      </td><td> Distance between photograph edges, in millimeters</td></tr>
<tr><td>-n                  </td><td> The 'n' parameter</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">correct_vignetting(
    i, 
    pp, 
    output, 
    focal_length=304.8, 
    image_width=228.6, 
    n=4.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CorrectVignetting -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif --pp=princ_pt.shp ^
-o=output.tif --focal_length=304.8 --image_width=228.6 ^
-n=4.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/correct_vignetting.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 24/04/2018</p>
<p><a name="DirectDecorrelationStretch"></a></p>
<a class="header" href="#directdecorrelationstretch" id="directdecorrelationstretch"><h1>DirectDecorrelationStretch</h1></a>
<p>The Direct Decorrelation Stretch (DDS) is a simple type of saturation stretch. The stretch is
applied to a colour composite image and is used to improve the saturation, or colourfulness,
of the image. The DDS operates by reducing the achromatic (grey) component of a pixel's colour
by a scale factor (<em>k</em>), such that the red (r), green (g), and blue (b) components of the output
colour are defined as:</p>
<p>r<sub><em>k</em></sub> = r - <em>k</em> min(r, g, b)</p>
<p>g<sub><em>k</em></sub> = g - <em>k</em> min(r, g, b)</p>
<p>b<sub><em>k</em></sub> = b - <em>k</em> min(r, g, b)</p>
<p>The achromatic factor (<em>k</em>) can range between 0 (no effect) and 1 (full saturation stretch),
although typical values range from 0.3 to 0.7. A linear stretch is used afterwards to adjust
overall image brightness. Liu and Moore (1996) recommend applying a colour balance stretch,
such as <a href="./image_processing_tools_image_enhancement.html#balancecontrastenhancement"><strong>BalanceContrastEnhancement</strong></a> before using the DDS.</p>
<p><em>Reference</em>:</p>
<p>Liu, J.G., and Moore, J. (1996) Direct decorrelation stretch technique for RGB colour composition.
International Journal of Remote Sensing, 17:5, 1005-1018.</p>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools.html#createcolourcomposite"><strong>CreateColourComposite</strong></a>, <a href="./image_processing_tools_image_enhancement.html#balancecontrastenhancement"><strong>BalanceContrastEnhancement</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input colour composite image file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>-k                  </td><td> Achromatic factor (k) ranges between 0 (no effect) and 1 (full saturation stretch), although typical values range from 0.3 to 0.7</td></tr>
<tr><td>--clip             </td><td> Optional percent to clip the upper tail by during the stretch</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">direct_decorrelation_stretch(
    i, 
    output, 
    k=0.5, 
    clip=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DirectDecorrelationStretch -v ^
--wd=&quot;/path/to/data/&quot; --input=image.tif -o=output.tif -k=0.4 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/direct_decorrelation_stretch.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 21, 2017</p>
<p><a name="GammaCorrection"></a></p>
<a class="header" href="#gammacorrection" id="gammacorrection"><h1>GammaCorrection</h1></a>
<p>Performs a sigmoidal contrast stretch on input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--gamma            </td><td> Gamma value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">gamma_correction(
    i, 
    output, 
    gamma=0.5, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=GammaCorrection -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif --gamma=0.5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/gamma_correction.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="GaussianContrastStretch"></a></p>
<a class="header" href="#gaussiancontraststretch" id="gaussiancontraststretch"><h1>GaussianContrastStretch</h1></a>
<p>Performs a Gaussian contrast stretch on input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--num_tones        </td><td> Number of tones in the output image</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">gaussian_contrast_stretch(
    i, 
    output, 
    num_tones=256, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=GaussianContrastStretch -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif ^
--num_tones=1024 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/gaussian_contrast_stretch.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 21/05/2018</p>
<p><a name="HistogramEqualization"></a></p>
<a class="header" href="#histogramequalization" id="histogramequalization"><h1>HistogramEqualization</h1></a>
<p>This tool alters the cumulative distribution function (CDF) of a raster image to match,
as closely as possible, the CDF of a uniform distribution. Histogram equalization works
by first calculating the histogram of the input image. This input histogram is then
converted into a CDF. Each grid cell value in the input image is then mapped to the
corresponding value in the uniform distribution's CDF that has an equivalent (or as close
as possible) cumulative probability value. Histogram equalization provides a very effective
means of performing image contrast adjustment in an efficient manner with little need for
human input.</p>
<p>The user must specify the name of the input image to perform histogram equalization on.
The user must also specify the number of tones, corresponding to the number
of histogram bins used in the analysis.</p>
<p><a href="./image_processing_tools_image_enhancement.html#histogramequalization"><strong>HistogramEqualization</strong></a> is related to the <a href="./image_processing_tools_image_enhancement.html#histogrammatchingtwoimages"><strong>HistogramMatchingTwoImages</strong></a> tool (used when an image's
CDF is to be matched to a reference CDF derived from a reference image). Similarly, <a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a>,
and <a href="./image_processing_tools_image_enhancement.html#gaussiancontraststretch"><strong>GaussianContrastStretch</strong></a> are similarly related tools frequently used for image contrast
adjustment, where the reference CDFs are uniform and Gaussian (normal) respectively.</p>
<p><strong>Notes</strong>:</p>
<ul>
<li>The algorithm can introduces gaps in the histograms (steps in the CDF). This is to be expected because
the histogram is being distorted. This is more prevalent for integer-level images.</li>
<li>Histogram equalization is not appropriate for images containing categorical (class) data.</li>
</ul>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a>, <a href="./image_processing_tools_image_enhancement.html#histogrammatchingtwoimages"><strong>HistogramMatchingTwoImages</strong></a>, <a href="./image_processing_tools_image_enhancement.html#gaussiancontraststretch"><strong>GaussianContrastStretch</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--num_tones        </td><td> Number of tones in the output image</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">histogram_equalization(
    i, 
    output, 
    num_tones=256, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HistogramEqualization -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif ^
--num_tones=1024 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/histogram_equalization.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: August 26, 2017</p>
<p><a name="HistogramMatching"></a></p>
<a class="header" href="#histogrammatching" id="histogrammatching"><h1>HistogramMatching</h1></a>
<p>This tool alters the cumulative distribution function (CDF) of a raster image to match,
as closely as possible, the CDF of a reference histogram. Histogram matching works by
first calculating the histogram of the input image. This input histogram and reference
histograms are each then converted into CDFs. Each grid cell value in the input image
is then mapped to the corresponding value in the reference CDF that has an equivalent
(or as close as possible) cumulative probability value. Histogram matching provides
the most flexible means of performing image contrast adjustment.</p>
<p>The reference histogram must be specified to the tool in the form of a text file (.txt),
provided using the <code>--histo_file</code> flag. This file must contain two columns (delimited by
a tab, space, comma, colon, or semicolon) where the first column contains the x value
(i.e. the values that will be assigned to the grid cells in the output image) and the second
column contains the frequency or probability. Note that 1) the file must not contain a
header row, 2) each x value/frequency pair must be on a separate row, and 3) the
frequency/probability must not be cumulative (i.e. the file must contain the histogram and
not the CDF). The CDF will be computed for the reference histogram automatically by the tool.
It is possible to create this type of histogram using the wide range of distribution tools
available in most spreadsheet programs (e.g. Excel or LibreOffice's Calc program). You must
save the file as a text-only (ASCII) file.</p>
<p><a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a> is related to the <a href="./image_processing_tools_image_enhancement.html#histogrammatchingtwoimages"><strong>HistogramMatchingTwoImages</strong></a> tool, which can be used
when a reference CDF can be derived from a reference image. <a href="./image_processing_tools_image_enhancement.html#histogramequalization"><strong>HistogramEqualization</strong></a> and
<a href="./image_processing_tools_image_enhancement.html#gaussiancontraststretch"><strong>GaussianContrastStretch</strong></a> are similarly related tools frequently used for image contrast
adjustment, where the reference CDFs are uniform and Gaussian (normal) respectively.</p>
<p><strong>Notes:</strong></p>
<ul>
<li>The algorithm can introduces gaps in the histograms (steps in the CDF). This is to be expected
because the histogram is being distorted. This is more prevalent for integer-level images.</li>
<li>Histogram matching is not appropriate for images containing categorical (class) data.</li>
<li>This tool is not intended for images containing RGB data. If this is the case, the colour
channels should be split using the <a href="./image_processing_tools.html#splitcolourcomposite"><strong>SplitColourComposite</strong></a> tool.</li>
</ul>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools_image_enhancement.html#histogrammatchingtwoimages"><strong>HistogramMatchingTwoImages</strong></a>, <a href="./image_processing_tools_image_enhancement.html#histogramequalization"><strong>HistogramEqualization</strong></a>, <a href="./image_processing_tools_image_enhancement.html#gaussiancontraststretch"><strong>GaussianContrastStretch</strong></a>, <a href="./image_processing_tools.html#splitcolourcomposite"><strong>SplitColourComposite</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--histo_file       </td><td> Input reference probability distribution function (pdf) text file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">histogram_matching(
    i, 
    histo_file, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HistogramMatching -v ^
--wd=&quot;/path/to/data/&quot; -i=input1.tif --histo_file=histo.txt ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/histogram_matching.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 14, 2017</p>
<p><a name="HistogramMatchingTwoImages"></a></p>
<a class="header" href="#histogrammatchingtwoimages" id="histogrammatchingtwoimages"><h1>HistogramMatchingTwoImages</h1></a>
<p>This tool alters the cumulative distribution function (CDF) of a raster image to match, as closely
as possible, the CDF of a reference image. Histogram matching works by first calculating the
histograms of the input image (i.e. the image to be adjusted) and the reference image. These
histograms are then converted into CDFs. Each grid cell value in the input image is then mapped
to the corresponding value in the reference CDF that has the an equivalent (or as close as
possible) cumulative probability value. A common application of this is to match the images from
two sensors with slightly different responses, or images from the same sensor, but the sensor's
response is known to change over time.The size of the two images (rows and columns) do not need
to be the same, nor do they need to be geographically overlapping.</p>
<p><a href="./image_processing_tools_image_enhancement.html#histogrammatchingtwoimages"><strong>HistogramMatchingTwoImages</strong></a> is related to the <a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a> tool, which can be used
when a reference CDF is used directly rather than deriving it from a reference image.
<a href="./image_processing_tools_image_enhancement.html#histogramequalization"><strong>HistogramEqualization</strong></a> and <a href="./image_processing_tools_image_enhancement.html#gaussiancontraststretch"><strong>GaussianContrastStretch</strong></a> are similarly related tools, where the
reference CDFs are uniform and Gaussian (normal) respectively.</p>
<p>The algorithm may introduces gaps in the histograms (steps in the CDF). This is to be expected
because the histograms are being distorted. This is more prevalent for integer-level images.
Histogram matching is not appropriate for images containing categorical (class) data. It is also
not intended for images containing RGB data, in which case, the colour channels should be split
using the <a href="./image_processing_tools.html#splitcolourcomposite"><strong>SplitColourComposite</strong></a> tool.</p>
<p><em>See Also</em>:</p>
<p><a href="./image_processing_tools_image_enhancement.html#histogrammatching"><strong>HistogramMatching</strong></a>, <a href="./image_processing_tools_image_enhancement.html#histogramequalization"><strong>HistogramEqualization</strong></a>, <a href="./image_processing_tools_image_enhancement.html#gaussiancontraststretch"><strong>GaussianContrastStretch</strong></a>, <a href="./image_processing_tools.html#splitcolourcomposite"><strong>SplitColourComposite</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input raster file to modify</td></tr>
<tr><td>--i2, --input2    </td><td> Input reference raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">histogram_matching_two_images(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HistogramMatchingTwoImages -v ^
--wd=&quot;/path/to/data/&quot; --i1=input1.tif --i2=input2.tif ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/histogram_matching_two_images.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: August 31, 2017</p>
<p><a name="MinMaxContrastStretch"></a></p>
<a class="header" href="#minmaxcontraststretch" id="minmaxcontraststretch"><h1>MinMaxContrastStretch</h1></a>
<p>Performs a min-max contrast stretch on an input greytone image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--min_val          </td><td> Lower tail clip value</td></tr>
<tr><td>--max_val          </td><td> Upper tail clip value</td></tr>
<tr><td>--num_tones        </td><td> Number of tones in the output image</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">min_max_contrast_stretch(
    i, 
    output, 
    min_val, 
    max_val, 
    num_tones=256, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=MinMaxContrastStretch -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif ^
--min_val=45.0 --max_val=200.0 --num_tones=1024 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/min_max_contrast_stretch.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="PanchromaticSharpening"></a></p>
<a class="header" href="#panchromaticsharpening" id="panchromaticsharpening"><h1>PanchromaticSharpening</h1></a>
<p>Increases the spatial resolution of image data by combining multispectral bands with panchromatic data.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--red              </td><td> Input red band image file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--green            </td><td> Input green band image file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--blue             </td><td> Input blue band image file. Optionally specified if colour-composite not specified</td></tr>
<tr><td>--composite        </td><td> Input colour-composite image file. Only used if individual bands are not specified</td></tr>
<tr><td>--pan              </td><td> Input panchromatic band file</td></tr>
<tr><td>-o, --output       </td><td> Output colour composite file</td></tr>
<tr><td>--method           </td><td> Options include 'brovey' (default) and 'ihs'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">panchromatic_sharpening(
    pan, 
    output, 
    red=None, 
    green=None, 
    blue=None, 
    composite=None, 
    method=&quot;brovey&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PanchromaticSharpening -v ^
--wd=&quot;/path/to/data/&quot; --red=red.tif --green=green.tif ^
--blue=blue.tif --pan=pan.tif --output=pan_sharp.tif ^
--method='brovey'
&gt;&gt;./whitebox_tools -r=PanchromaticSharpening ^
-v --wd=&quot;/path/to/data/&quot; --composite=image.tif --pan=pan.tif ^
--output=pan_sharp.tif --method='ihs' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/pan_sharpening.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 27, 2017</p>
<p><a name="PercentageContrastStretch"></a></p>
<a class="header" href="#percentagecontraststretch" id="percentagecontraststretch"><h1>PercentageContrastStretch</h1></a>
<p>Performs a percentage linear contrast stretch on input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--clip             </td><td> Optional amount to clip the distribution tails by, in percent</td></tr>
<tr><td>--tail             </td><td> Specified which tails to clip; options include 'upper', 'lower', and 'both' (default is 'both')</td></tr>
<tr><td>--num_tones        </td><td> Number of tones in the output image</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">percentage_contrast_stretch(
    i, 
    output, 
    clip=0.0, 
    tail=&quot;both&quot;, 
    num_tones=256, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PercentageContrastStretch -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif --clip=2.0 ^
--tail='both' --num_tones=1024 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/percentage_contrast_stretch.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="SigmoidalContrastStretch"></a></p>
<a class="header" href="#sigmoidalcontraststretch" id="sigmoidalcontraststretch"><h1>SigmoidalContrastStretch</h1></a>
<p>Performs a sigmoidal contrast stretch on input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--cutoff           </td><td> Cutoff value between 0.0 and 0.95</td></tr>
<tr><td>--gain             </td><td> Gain value</td></tr>
<tr><td>--num_tones        </td><td> Number of tones in the output image</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">sigmoidal_contrast_stretch(
    i, 
    output, 
    cutoff=0.0, 
    gain=1.0, 
    num_tones=256, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SigmoidalContrastStretch -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif --cutoff=0.1 ^
--gain=2.0 --num_tones=1024 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/sigmoidal_contrast_stretch.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="StandardDeviationContrastStretch"></a></p>
<a class="header" href="#standarddeviationcontraststretch" id="standarddeviationcontraststretch"><h1>StandardDeviationContrastStretch</h1></a>
<p>Performs a standard-deviation contrast stretch on input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--clip, --stdev   </td><td> Standard deviation clip value</td></tr>
<tr><td>--num_tones        </td><td> Number of tones in the output image</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">standard_deviation_contrast_stretch(
    i, 
    output, 
    stdev=2.0, 
    num_tones=256, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StandardDeviationContrastStretch -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif --stdev=2.0 ^
--num_tones=1024 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/image_analysis/stdev_contrast_stretch.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<a class="header" href="#lidar-tools" id="lidar-tools"><h1>LiDAR Tools</h1></a>
<ul>
<li><a href="#ClassifyOverlapPoints">ClassifyOverlapPoints</a></li>
<li><a href="#ClipLidarToPolygon">ClipLidarToPolygon</a></li>
<li><a href="#ErasePolygonFromLidar">ErasePolygonFromLidar</a></li>
<li><a href="#FilterLidarScanAngles">FilterLidarScanAngles</a></li>
<li><a href="#FindFlightlineEdgePoints">FindFlightlineEdgePoints</a></li>
<li><a href="#FlightlineOverlap">FlightlineOverlap</a></li>
<li><a href="#LasToAscii">LasToAscii</a></li>
<li><a href="#LasToMultipointShapefile">LasToMultipointShapefile</a></li>
<li><a href="#LasToShapefile">LasToShapefile</a></li>
<li><a href="#LidarBlockMaximum">LidarBlockMaximum</a></li>
<li><a href="#LidarBlockMinimum">LidarBlockMinimum</a></li>
<li><a href="#LidarClassifySubset">LidarClassifySubset</a></li>
<li><a href="#LidarColourize">LidarColourize</a></li>
<li><a href="#LidarConstructVectorTin">LidarConstructVectorTin</a></li>
<li><a href="#LidarElevationSlice">LidarElevationSlice</a></li>
<li><a href="#LidarGroundPointFilter">LidarGroundPointFilter</a></li>
<li><a href="#LidarHexBinning">LidarHexBinning</a></li>
<li><a href="#LidarHillshade">LidarHillshade</a></li>
<li><a href="#LidarHistogram">LidarHistogram</a></li>
<li><a href="#LidarIdwInterpolation">LidarIdwInterpolation</a></li>
<li><a href="#LidarInfo">LidarInfo</a></li>
<li><a href="#LidarJoin">LidarJoin</a></li>
<li><a href="#LidarKappaIndex">LidarKappaIndex</a></li>
<li><a href="#LidarNearestNeighbourGridding">LidarNearestNeighbourGridding</a></li>
<li><a href="#LidarPointDensity">LidarPointDensity</a></li>
<li><a href="#LidarPointStats">LidarPointStats</a></li>
<li><a href="#LidarRemoveDuplicates">LidarRemoveDuplicates</a></li>
<li><a href="#LidarRemoveOutliers">LidarRemoveOutliers</a></li>
<li><a href="#LidarSegmentation">LidarSegmentation</a></li>
<li><a href="#LidarSegmentationBasedFilter">LidarSegmentationBasedFilter</a></li>
<li><a href="#LidarThin">LidarThin</a></li>
<li><a href="#LidarThinHighDensity">LidarThinHighDensity</a></li>
<li><a href="#LidarTile">LidarTile</a></li>
<li><a href="#LidarTileFootprint">LidarTileFootprint</a></li>
<li><a href="#LidarTinGridding">LidarTinGridding</a></li>
<li><a href="#LidarTophatTransform">LidarTophatTransform</a></li>
<li><a href="#NormalVectors">NormalVectors</a></li>
<li><a href="#SelectTilesByPolygon">SelectTilesByPolygon</a></li>
</ul>
<p><a name="ClassifyOverlapPoints"></a></p>
<a class="header" href="#classifyoverlappoints" id="classifyoverlappoints"><h1>ClassifyOverlapPoints</h1></a>
<p>Classifies or filters LAS points in regions of overlapping flight lines.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--resolution       </td><td> The size of the square area used to evaluate nearby points in the LiDAR data</td></tr>
<tr><td>--filter           </td><td> Filter out points from overlapping flightlines? If false, overlaps will simply be classified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">classify_overlap_points(
    i, 
    output, 
    resolution=2.0, 
    filter=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ClassifyOverlapPoints -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.las ^
--resolution=2.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/classify_overlap_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 27/04/2018</p>
<p><a name="ClipLidarToPolygon"></a></p>
<a class="header" href="#cliplidartopolygon" id="cliplidartopolygon"><h1>ClipLidarToPolygon</h1></a>
<p>Clips a LiDAR point cloud to a vector polygon or polygons.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>--polygons         </td><td> Input vector polygons file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">clip_lidar_to_polygon(
    i, 
    polygons, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ClipLidarToPolygon -v ^
--wd=&quot;/path/to/data/&quot; -i='data.las' --polygons='lakes.shp' ^
-o='output.las' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/clip_lidar_to_polygon.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/04/2018</p>
<p><a name="ErasePolygonFromLidar"></a></p>
<a class="header" href="#erasepolygonfromlidar" id="erasepolygonfromlidar"><h1>ErasePolygonFromLidar</h1></a>
<p>Erases (cuts out) a vector polygon or polygons from a LiDAR point cloud.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>--polygons         </td><td> Input vector polygons file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">erase_polygon_from_lidar(
    i, 
    polygons, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ErasePolygonFromLidar -v ^
--wd=&quot;/path/to/data/&quot; -i='data.las' --polygons='lakes.shp' ^
-o='output.las' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/erase_polygon_from_lidar.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 25/04/2018</p>
<p><a name="FilterLidarScanAngles"></a></p>
<a class="header" href="#filterlidarscanangles" id="filterlidarscanangles"><h1>FilterLidarScanAngles</h1></a>
<p>Removes points in a LAS file with scan angles greater than a threshold.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--threshold        </td><td> Scan angle threshold</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">filter_lidar_scan_angles(
    i, 
    output, 
    threshold, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FilterLidarScanAngles -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--threshold=10.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/filter_lidar_scan_angles.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 17, 2017</p>
<p><a name="FindFlightlineEdgePoints"></a></p>
<a class="header" href="#findflightlineedgepoints" id="findflightlineedgepoints"><h1>FindFlightlineEdgePoints</h1></a>
<p>Identifies points along a flightline's edge in a LAS file.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">find_flightline_edge_points(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FindFlightlineEdgePoints -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/find_flightline_edge_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 14, 2017</p>
<p><a name="FlightlineOverlap"></a></p>
<a class="header" href="#flightlineoverlap" id="flightlineoverlap"><h1>FlightlineOverlap</h1></a>
<p>Reads a LiDAR (LAS) point file and outputs a raster containing the number of overlapping flight lines in each grid cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">flightline_overlap(
    i=None, 
    output=None, 
    resolution=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FlightlineOverlap -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=2.0&quot;
./whitebox_tools -r=FlightlineOverlap -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=5.0 --palette=light_quant.plt 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/flightline_overlap.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 19, 2017</p>
<p><a name="LasToAscii"></a></p>
<a class="header" href="#lastoascii" id="lastoascii"><h1>LasToAscii</h1></a>
<p>Converts one or more LAS files into ASCII text files.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input LiDAR files</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">las_to_ascii(
    inputs, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LasToAscii -v --wd=&quot;/path/to/data/&quot; ^
-i=&quot;file1.las, file2.las, file3.las&quot; -o=outfile.las&quot; 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/las_to_ascii.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 16, 2017</p>
<p><a name="LasToMultipointShapefile"></a></p>
<a class="header" href="#lastomultipointshapefile" id="lastomultipointshapefile"><h1>LasToMultipointShapefile</h1></a>
<p>Converts one or more LAS files into MultipointZ vector Shapefiles. When the input parameter is
not specified, the tool grids all LAS files contained within the working directory.</p>
<p>This tool can be used in place of the <a href="./lidar_tools.html#lastoshapefile"><strong>LasToShapefile</strong></a> tool when the number of points are
relatively high and when the desire is to represent the x,y,z position of points only. The z
values of LAS points will be stored in the z-array of the output Shapefile. Notice that because
the output file stores each point in a single multi-point record, this Shapefile representation,
while unable to represent individual point classes, return numbers, etc, is an efficient means
of converting LAS point positional information.</p>
<p><em>See Also</em>:</p>
<p><a href="./lidar_tools.html#lastoshapefile"><strong>LasToShapefile</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">las_to_multipoint_shapefile(
    i=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LasToMultipointShapefile -v ^
--wd=&quot;/path/to/data/&quot; -i=input.las 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/las_to_multipoint_shapefile.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 04/09/2018</p>
<p><a name="LasToShapefile"></a></p>
<a class="header" href="#lastoshapefile" id="lastoshapefile"><h1>LasToShapefile</h1></a>
<p>This tool converts one or more LAS files into a POINT vector. When the input parameter is
not specified, the tool grids all LAS files contained within the working directory.
The attribute table of the output Shapefile will contain fields for the z-value,
intensity, point class, return number, and number of return.</p>
<p>This tool can be used in place of the <a href="./lidar_tools.html#lastomultipointshapefile"><strong>LasToMultipointShapefile</strong></a> tool when the
number of points are relatively low and when the desire is to represent more than
simply the x,y,z position of points. Notice however that because each point in
the input LAS file will be represented as a separate record in the output
Shapefile, the output file will be many time larger than the equivalent output of
the <a href="./lidar_tools.html#lastomultipointshapefile"><strong>LasToMultipointShapefile</strong></a> tool. There is also a practical limit on the
total number of records that can be held in a single Shapefile and large LAS
files approach this limit. In these cases, the <a href="./lidar_tools.html#lastomultipointshapefile"><strong>LasToMultipointShapefile</strong></a> tool
should be preferred instead.</p>
<p><em>See Also</em>:</p>
<p><a href="./lidar_tools.html#lastomultipointshapefile"><strong>LasToMultipointShapefile</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">las_to_shapefile(
    i=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LasToShapefile -v --wd=&quot;/path/to/data/&quot; ^
-i=input.las 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/las_to_shapefile.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 01/10/2018</p>
<p><a name="LidarBlockMaximum"></a></p>
<a class="header" href="#lidarblockmaximum" id="lidarblockmaximum"><h1>LidarBlockMaximum</h1></a>
<p>Creates a block-maximum raster from an input LAS file. When the input/output parameters are not specified, the tool grids all LAS files contained within the working directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_block_maximum(
    i=None, 
    output=None, 
    resolution=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarBlockMaximum -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=2.0&quot;
./whitebox_tools -r=LidarBlockMaximum -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=5.0 --palette=light_quant.plt 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/block_maximum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<p><a name="LidarBlockMinimum"></a></p>
<a class="header" href="#lidarblockminimum" id="lidarblockminimum"><h1>LidarBlockMinimum</h1></a>
<p>Creates a block-minimum raster from an input LAS file. When the input/output parameters are not specified, the tool grids all LAS files contained within the working directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_block_minimum(
    i=None, 
    output=None, 
    resolution=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarBlockMinimum -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=2.0&quot;
./whitebox_tools -r=LidarBlockMinimum -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=5.0 --palette=light_quant.plt 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/block_minimum.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<p><a name="LidarClassifySubset"></a></p>
<a class="header" href="#lidarclassifysubset" id="lidarclassifysubset"><h1>LidarClassifySubset</h1></a>
<p>This tool classifies points within a user-specified LiDAR point cloud (<code>--base</code>) that correpond
with points in a subset cloud (<code>--subset</code>). The subset point cloud may have been derived by filtering
the original point cloud. The user must specify the names of the two input LAS files (i.e.
the full and subset clouds) and the class value (<code>--subset_class</code>) to assign the matching points. This class
value will be assigned to points in the base cloud, overwriting their input class values in the
output LAS file (<code>--output</code>). Class values
should be numerical (integer valued) and shoud follow the LAS specifications below:</p>
<table><thead><tr><th> Classification Value  </th><th> Meaning                              </th></tr></thead><tbody>
<tr><td align="left"> 0                     </td><td align="left"> Created never classified</td></tr>
<tr><td align="left"> 1                     </td><td align="left"> Unclassified3</td></tr>
<tr><td align="left"> 2                     </td><td align="left"> Ground</td></tr>
<tr><td align="left"> 3                     </td><td align="left"> Low Vegetation</td></tr>
<tr><td align="left"> 4                     </td><td align="left"> Medium Vegetation</td></tr>
<tr><td align="left"> 5                     </td><td align="left"> High Vegetation</td></tr>
<tr><td align="left"> 6                     </td><td align="left"> Building</td></tr>
<tr><td align="left"> 7                     </td><td align="left"> Low Point (noise)</td></tr>
<tr><td align="left"> 8                     </td><td align="left"> Reserved</td></tr>
<tr><td align="left"> 9                     </td><td align="left"> Water</td></tr>
<tr><td align="left"> 10                    </td><td align="left"> Rail</td></tr>
<tr><td align="left"> 11                    </td><td align="left"> Road Surface</td></tr>
<tr><td align="left"> 12                    </td><td align="left"> Reserved</td></tr>
<tr><td align="left"> 13                    </td><td align="left">   Wire – Guard (Shield)</td></tr>
<tr><td align="left"> 14                    </td><td align="left"> Wire – Conductor (Phase)</td></tr>
<tr><td align="left"> 15                    </td><td align="left"> Transmission Tower</td></tr>
<tr><td align="left"> 16                    </td><td align="left"> Wire-structure Connector (e.g. Insulator)</td></tr>
<tr><td align="left"> 17                    </td><td align="left"> Bridge Deck</td></tr>
<tr><td align="left"> 18                    </td><td align="left"> High Noise</td></tr>
</tbody></table>
<p>The user may optionally specify a class value to be assigned to non-subset (i.e. non-matching)
points (<code>--nonsubset_class</code>) in the base file. If this parameter is not specified, output
non-sutset points will have the same class value as the base file.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--base             </td><td> Input base LiDAR file</td></tr>
<tr><td>--subset           </td><td> Input subset LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--subset_class     </td><td> Subset point class value (must be 0-18; see LAS specifications)</td></tr>
<tr><td>--nonsubset_class  </td><td> Non-subset point class value (must be 0-18; see LAS specifications)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_classify_subset(
    base, 
    subset, 
    output, 
    subset_class, 
    nonsubset_class=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarClassifySubset -v ^
--wd=&quot;/path/to/data/&quot; --base=&quot;full_cloud.las&quot; ^
--subset=&quot;filtered_cloud.las&quot; -o=&quot;output.las&quot; ^
--subset_class=2 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_classify_subset.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay and Kevin Roberts</p>
<p><em>Created</em>: 24/10/2018</p>
<p><a name="LidarColourize"></a></p>
<a class="header" href="#lidarcolourize" id="lidarcolourize"><h1>LidarColourize</h1></a>
<p>Adds the red-green-blue colour fields of a LiDAR (LAS) file based on an input image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--in_lidar         </td><td> Input LiDAR file</td></tr>
<tr><td>--in_image         </td><td> Input colour image file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_colourize(
    in_lidar, 
    in_image, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarColourize -v --wd=&quot;/path/to/data/&quot; ^
--in_lidar=&quot;input.las&quot; --in_image=&quot;image.tif&quot; ^
-o=&quot;output.las&quot; 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_colourize.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 18, 2018</p>
<p><a name="LidarConstructVectorTin"></a></p>
<a class="header" href="#lidarconstructvectortin" id="lidarconstructvectortin"><h1>LidarConstructVectorTin</h1></a>
<p>Creates a vector triangular irregular network (TIN) fitted to LiDAR points.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file (including extension)</td></tr>
<tr><td>-o, --output       </td><td> Output raster file (including extension)</td></tr>
<tr><td>--returns          </td><td> Point return types to include; options are 'all' (default), 'last', 'first'</td></tr>
<tr><td>--exclude_cls      </td><td> Optional exclude classes from interpolation; Valid class values range from 0 to 18, based on LAS specifications. Example, --exclude_cls='3,4,5,6,7,18'</td></tr>
<tr><td>--minz             </td><td> Optional minimum elevation for inclusion in interpolation</td></tr>
<tr><td>--maxz             </td><td> Optional maximum elevation for inclusion in interpolation</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_construct_vector_tin(
    i=None, 
    output=None, 
    returns=&quot;all&quot;, 
    exclude_cls=None, 
    minz=None, 
    maxz=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarConstructVectorTIN -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--returns=last --exclude_cls='3,4,5,6,7,18' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_construct_vector_tin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 23/09/2018</p>
<p><a name="LidarElevationSlice"></a></p>
<a class="header" href="#lidarelevationslice" id="lidarelevationslice"><h1>LidarElevationSlice</h1></a>
<p>Outputs all of the points within a LiDAR (LAS) point file that lie between a specified elevation range.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--minz             </td><td> Minimum elevation value (optional)</td></tr>
<tr><td>--maxz             </td><td> Maximum elevation value (optional)</td></tr>
<tr><td>--class            </td><td> Optional boolean flag indicating whether points outside the range should be retained in output but reclassified</td></tr>
<tr><td>--inclassval       </td><td> Optional parameter specifying the class value assigned to points within the slice</td></tr>
<tr><td>--outclassval      </td><td> Optional parameter specifying the class value assigned to points within the slice</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_elevation_slice(
    i, 
    output, 
    minz=None, 
    maxz=None, 
    cls=False, 
    inclassval=2, 
    outclassval=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarElevationSlice -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--minz=100.0 --maxz=250.0
&gt;&gt;./whitebox_tools ^
-r=LidarElevationSlice -v -i=&quot;/path/to/data/input.las&quot; ^
-o=&quot;/path/to/data/output.las&quot; --minz=100.0 --maxz=250.0 ^
--class
&gt;&gt;./whitebox_tools -r=LidarElevationSlice -v ^
-i=&quot;/path/to/data/input.las&quot; -o=&quot;/path/to/data/output.las&quot; ^
--minz=100.0 --maxz=250.0 --inclassval=1 --outclassval=0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_elevation_slice.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 2, 2017</p>
<p><a name="LidarGroundPointFilter"></a></p>
<a class="header" href="#lidargroundpointfilter" id="lidargroundpointfilter"><h1>LidarGroundPointFilter</h1></a>
<p>Identifies ground points within LiDAR dataset using a slope-based method.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
<tr><td>--min_neighbours   </td><td> The minimum number of neighbouring points within search areas. If fewer points than this threshold are idenfied during the fixed-radius search, a subsequent kNN search is performed to identify the k number of neighbours</td></tr>
<tr><td>--slope_threshold  </td><td> Maximum inter-point slope to be considered an off-terrain point</td></tr>
<tr><td>--height_threshold </td><td> Inter-point height difference to be considered an off-terrain point</td></tr>
<tr><td>--classify         </td><td> Classify points as ground (2) or off-ground (1)</td></tr>
<tr><td>--slope_norm       </td><td> Perform initial ground slope normalization?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_ground_point_filter(
    i, 
    output, 
    radius=2.0, 
    min_neighbours=0, 
    slope_threshold=45.0, 
    height_threshold=1.0, 
    classify=True, 
    slope_norm=True, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarGroundPointFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--radius=10.0 --min_neighbours=10 --slope_threshold=30.0 ^
--height_threshold=0.5 --classify --slope_norm 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_ground_point_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 02/06/2017</p>
<p><a name="LidarHexBinning"></a></p>
<a class="header" href="#lidarhexbinning" id="lidarhexbinning"><h1>LidarHexBinning</h1></a>
<p>The practice of binning point data to form a type of 2D histogram, density plot,
or what is sometimes called a heatmap, is quite useful as an alternative for the
cartographic display of of very dense points sets. This is particularly the case
when the points experience significant overlap at the displayed scale. The
<a href="./lidar_tools.html#lidarpointdensity"><strong>LidarPointDensity</strong></a> tool can be used to perform binning based on a regular grid
(raster output). This tool, by comparison, bases the binning on a hexagonal grid.</p>
<p>The tool is similar to the <a href="./gis_analysis.html#createhexagonalvectorgrid"><strong>CreateHexagonalVectorGrid</strong></a> tool, however instead will
create an output hexagonal grid in which each hexagonal cell possesses a <code>COUNT</code>
attribute which specifies the number of points from an input points file (LAS file)
that are contained within the hexagonal cell. The tool will also calculate the
minimum and maximum elevations and intensity values and outputs these data to the
attribute table.</p>
<p>In addition to the names of the input points file and the output Shapefile, the user
must also specify the desired hexagon width (w), which is the distance between opposing
sides of each hexagon. The size (s) each side of the hexagon can then be calculated as,
s = w / [2 x cos(PI / 6)]. The area of each hexagon (A) is, A = 3s(w / 2). The user must
also specify the orientation of the grid with options of horizontal (pointy side up) and
vertical (flat side up).</p>
<p><em>See Also</em>:</p>
<p><a href="./gis_analysis.html#vectorhexbinning"><strong>VectorHexBinning</strong></a>, <a href="./lidar_tools.html#lidarpointdensity"><strong>LidarPointDensity</strong></a>, <a href="./gis_analysis.html#createhexagonalvectorgrid"><strong>CreateHexagonalVectorGrid</strong></a></p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input base file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
<tr><td>--width            </td><td> The grid cell width</td></tr>
<tr><td>--orientation      </td><td> Grid Orientation, 'horizontal' or 'vertical'</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_hex_binning(
    i, 
    output, 
    width, 
    orientation=&quot;horizontal&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarHexBinning -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.shp --width=10.0 ^
--orientation=vertical 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_hex_bin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 16/09/2018</p>
<p><a name="LidarHillshade"></a></p>
<a class="header" href="#lidarhillshade" id="lidarhillshade"><h1>LidarHillshade</h1></a>
<p>Calculates a hillshade value for points within a LAS file and stores these data in the RGB field.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--azimuth          </td><td> Illumination source azimuth in degrees</td></tr>
<tr><td>--altitude         </td><td> Illumination source altitude in degrees</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_hillshade(
    i, 
    output, 
    azimuth=315.0, 
    altitude=30.0, 
    radius=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarHillshade -v --wd=&quot;/path/to/data/&quot; ^
-i=&quot;input.las&quot; -o=&quot;output.las&quot; --radius=10.0
&gt;&gt;./whitebox_tools ^
-r=LidarHillshade -v --wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; ^
-o=&quot;output.las&quot; --azimuth=180.0 --altitude=20.0 --radius=1.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_hillshade.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 14, 2017</p>
<p><a name="LidarHistogram"></a></p>
<a class="header" href="#lidarhistogram" id="lidarhistogram"><h1>LidarHistogram</h1></a>
<p>Creates a histogram from LiDAR data.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
<tr><td>--parameter        </td><td> Parameter; options are 'elevation' (default), 'intensity', 'scan angle', 'class</td></tr>
<tr><td>--clip             </td><td> Amount to clip distribution tails (in percent)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_histogram(
    i, 
    output, 
    parameter=&quot;elevation&quot;, 
    clip=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarHistogram -v --wd=&quot;/path/to/data/&quot; ^
-i=&quot;file1.tif, file2.tif, file3.tif&quot; -o=outfile.htm ^
--contiguity=Bishopsl 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_histogram.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 23, 2017</p>
<p><a name="LidarIdwInterpolation"></a></p>
<a class="header" href="#lidaridwinterpolation" id="lidaridwinterpolation"><h1>LidarIdwInterpolation</h1></a>
<p>Interpolates LAS files using an inverse-distance weighted (IDW) scheme. When the input/output parameters are not specified, the tool interpolates all LAS files contained within the working directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file (including extension)</td></tr>
<tr><td>-o, --output       </td><td> Output raster file (including extension)</td></tr>
<tr><td>--parameter        </td><td> Interpolation parameter; options are 'elevation' (default), 'intensity', 'class', 'scan angle', 'user data'</td></tr>
<tr><td>--returns          </td><td> Point return types to include; options are 'all' (default), 'last', 'first'</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
<tr><td>--weight           </td><td> IDW weight value</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
<tr><td>--exclude_cls      </td><td> Optional exclude classes from interpolation; Valid class values range from 0 to 18, based on LAS specifications. Example, --exclude_cls='3,4,5,6,7,18'</td></tr>
<tr><td>--minz             </td><td> Optional minimum elevation for inclusion in interpolation</td></tr>
<tr><td>--maxz             </td><td> Optional maximum elevation for inclusion in interpolation</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_idw_interpolation(
    i=None, 
    output=None, 
    parameter=&quot;elevation&quot;, 
    returns=&quot;all&quot;, 
    resolution=1.0, 
    weight=1.0, 
    radius=2.5, 
    exclude_cls=None, 
    minz=None, 
    maxz=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarIdwInterpolation -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=2.0 --radius=5.0&quot;
./whitebox_tools ^
-r=LidarIdwInterpolation --wd=&quot;/path/to/data/&quot; -i=file.las ^
-o=outfile.tif --resolution=5.0 --weight=2.0 --radius=2.0 ^
--exclude_cls='3,4,5,6,7,18' --palette=light_quant.plt 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_idw_interpolation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 3, 2017</p>
<p><a name="LidarInfo"></a></p>
<a class="header" href="#lidarinfo" id="lidarinfo"><h1>LidarInfo</h1></a>
<p>This tool can be used to print basic information about the data contained within a LAS file, used to store LiDAR
data. The reported information will include including data on the header, point return frequency, and classification
data and information about the variable length records (VLRs) and geokeys.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file for summary report</td></tr>
<tr><td>--vlr              </td><td> Flag indicating whether or not to print the variable length records (VLRs)</td></tr>
<tr><td>--geokeys          </td><td> Flag indicating whether or not to print the geokeys</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_info(
    i, 
    output=None, 
    vlr=False, 
    geokeys=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarInfo -v --wd=&quot;/path/to/data/&quot; ^
-i=file.las --vlr --geokeys&quot;
./whitebox_tools -r=LidarInfo ^
--wd=&quot;/path/to/data/&quot; -i=file.las 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_info.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 1, 2017</p>
<p><a name="LidarJoin"></a></p>
<a class="header" href="#lidarjoin" id="lidarjoin"><h1>LidarJoin</h1></a>
<p>Joins multiple LiDAR (LAS) files into a single LAS file.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input LiDAR files</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_join(
    inputs, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarJoin -v --wd=&quot;/path/to/data/&quot; ^
-i=&quot;file1.las, file2.las, file3.las&quot; -o=outfile.las&quot; 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_join.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 21, 2017</p>
<p><a name="LidarKappaIndex"></a></p>
<a class="header" href="#lidarkappaindex" id="lidarkappaindex"><h1>LidarKappaIndex</h1></a>
<p>This tool performs a kappa index of agreement (KIA) analysis on the classification values of two LiDAR
(LAS) files. The output report HTML file should be displayed automatically but can also be displayed
afterwards in any web browser. As a measure of overall classification accuracy, the KIA is more robust
than the percent agreement calculation because it takes into account the agreement occurring by random
chance. In addition to the KIA, the tool will output the producer's and user's accuracy, the overall
accuracy, and the error matrix. The KIA is often used as a means of assessing the accuracy of an image
classification analysis; however the <a href="./lidar_tools.html#lidarkappaindex"><strong>LidarKappaIndex</strong></a> tool performs the analysis on a point-to-point
basis, comparing the class values of the points in one input LAS file with the corresponding nearest
points in the second input LAS file.</p>
<p>The user must also specify the name and resolution of an output raster file, which is used to show the
spatial distribution of class accuracy. Each grid cell contains the overall accuracy, i.e. the points
correctly classified divided by the total number of points contained within the cell, expressed as a
percentage.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input LiDAR classification file</td></tr>
<tr><td>--i2, --input2    </td><td> Input LiDAR reference file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
<tr><td>--class_accuracy   </td><td> Output classification accuracy raster file</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_kappa_index(
    input1, 
    input2, 
    output, 
    class_accuracy, 
    resolution=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarKappaIndex -v ^
--wd=&quot;/path/to/data/&quot; --i1=class.las --i2=reference.las ^
-o=kia.html --class_accuracy=new_file.tif --resolution=0.5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_kappa.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 24/09/2017</p>
<p><a name="LidarNearestNeighbourGridding"></a></p>
<a class="header" href="#lidarnearestneighbourgridding" id="lidarnearestneighbourgridding"><h1>LidarNearestNeighbourGridding</h1></a>
<p>Grids LAS files using nearest-neighbour scheme. When the input/output parameters are not specified, the tool grids all LAS files contained within the working directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file (including extension)</td></tr>
<tr><td>-o, --output       </td><td> Output raster file (including extension)</td></tr>
<tr><td>--parameter        </td><td> Interpolation parameter; options are 'elevation' (default), 'intensity', 'class', 'scan angle', 'user data'</td></tr>
<tr><td>--returns          </td><td> Point return types to include; options are 'all' (default), 'last', 'first'</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
<tr><td>--exclude_cls      </td><td> Optional exclude classes from interpolation; Valid class values range from 0 to 18, based on LAS specifications. Example, --exclude_cls='3,4,5,6,7,18'</td></tr>
<tr><td>--minz             </td><td> Optional minimum elevation for inclusion in interpolation</td></tr>
<tr><td>--maxz             </td><td> Optional maximum elevation for inclusion in interpolation</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_nearest_neighbour_gridding(
    i=None, 
    output=None, 
    parameter=&quot;elevation&quot;, 
    returns=&quot;all&quot;, 
    resolution=1.0, 
    radius=2.5, 
    exclude_cls=None, 
    minz=None, 
    maxz=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarNearestNeighbourGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--returns=last --resolution=2.0 --radius=5.0&quot;
./whitebox_tools ^
-r=LidarNearestNeighbourGridding --wd=&quot;/path/to/data/&quot; ^
-i=file.las -o=outfile.tif --resolution=5.0 --radius=2.0 ^
--exclude_cls='3,4,5,6,7,18' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_nn_gridding.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="LidarPointDensity"></a></p>
<a class="header" href="#lidarpointdensity" id="lidarpointdensity"><h1>LidarPointDensity</h1></a>
<p>Calculates the spatial pattern of point density for a LiDAR data set. When the input/output parameters are not specified, the tool grids all LAS files contained within the working directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file (including extension)</td></tr>
<tr><td>-o, --output       </td><td> Output raster file (including extension)</td></tr>
<tr><td>--returns          </td><td> Point return types to include; options are 'all' (default), 'last', 'first'</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
<tr><td>--exclude_cls      </td><td> Optional exclude classes from interpolation; Valid class values range from 0 to 18, based on LAS specifications. Example, --exclude_cls='3,4,5,6,7,18'</td></tr>
<tr><td>--minz             </td><td> Optional minimum elevation for inclusion in interpolation</td></tr>
<tr><td>--maxz             </td><td> Optional maximum elevation for inclusion in interpolation</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_point_density(
    i=None, 
    output=None, 
    returns=&quot;all&quot;, 
    resolution=1.0, 
    radius=2.5, 
    exclude_cls=None, 
    minz=None, 
    maxz=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarPointDensity -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--resolution=2.0 --radius=5.0&quot;
./whitebox_tools ^
-r=LidarPointDensity -v --wd=&quot;/path/to/data/&quot; -i=file.las ^
-o=outfile.tif --resolution=5.0 --radius=2.0 ^
--exclude_cls='3,4,5,6,7,18' --palette=light_quant.plt 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_point_density.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 10, 2017</p>
<p><a name="LidarPointStats"></a></p>
<a class="header" href="#lidarpointstats" id="lidarpointstats"><h1>LidarPointStats</h1></a>
<p>Creates several rasters summarizing the distribution of LAS point data. When the input/output parameters are not specified, the tool works on all LAS files contained within the working directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
<tr><td>--num_points       </td><td> Flag indicating whether or not to output the number of points raster</td></tr>
<tr><td>--num_pulses       </td><td> Flag indicating whether or not to output the number of pulses raster</td></tr>
<tr><td>--z_range          </td><td> Flag indicating whether or not to output the elevation range raster</td></tr>
<tr><td>--intensity_range  </td><td> Flag indicating whether or not to output the intensity range raster</td></tr>
<tr><td>--predom_class     </td><td> Flag indicating whether or not to output the predominant classification raster</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_point_stats(
    i=None, 
    resolution=1.0, 
    num_points=True, 
    num_pulses=False, 
    z_range=False, 
    intensity_range=False, 
    predom_class=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarPointStats -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las --resolution=1.0 ^
--num_points 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_point_stats.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Feb. 18, 2018</p>
<p><a name="LidarRemoveDuplicates"></a></p>
<a class="header" href="#lidarremoveduplicates" id="lidarremoveduplicates"><h1>LidarRemoveDuplicates</h1></a>
<p>Removes duplicate points from a LiDAR data set.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--include_z        </td><td> Include z-values in point comparison?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_remove_duplicates(
    i, 
    output, 
    include_z=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarRemoveDuplicates -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/remove_duplicates.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: March 1, 2018</p>
<p><a name="LidarRemoveOutliers"></a></p>
<a class="header" href="#lidarremoveoutliers" id="lidarremoveoutliers"><h1>LidarRemoveOutliers</h1></a>
<p>Removes outliers (high and low points) in a LiDAR point cloud.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
<tr><td>--elev_diff        </td><td> Max. elevation difference</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_remove_outliers(
    i, 
    output, 
    radius=2.0, 
    elev_diff=50.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarRemoveOutliers -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--radius=10.0 --elev_diff=25.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_outliers.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 6, 2018</p>
<p><a name="LidarSegmentation"></a></p>
<a class="header" href="#lidarsegmentation" id="lidarsegmentation"><h1>LidarSegmentation</h1></a>
<p>Segments a LiDAR point cloud based on normal vectors.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--dist, --radius  </td><td> Search Radius</td></tr>
<tr><td>--norm_diff        </td><td> Maximum difference in normal vectors, in degrees</td></tr>
<tr><td>--maxzdiff         </td><td> Maximum difference in elevation (z units) between neighbouring points of the same segment</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_segmentation(
    i, 
    output, 
    radius=5.0, 
    norm_diff=10.0, 
    maxzdiff=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarSegmentation -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--radius=10.0 --norm_diff=2.5 --maxzdiff=0.75 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_segmentation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 5/12/2017</p>
<p><a name="LidarSegmentationBasedFilter"></a></p>
<a class="header" href="#lidarsegmentationbasedfilter" id="lidarsegmentationbasedfilter"><h1>LidarSegmentationBasedFilter</h1></a>
<p>Identifies ground points within LiDAR point clouds using a segmentation based approach.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--dist, --radius  </td><td> Search Radius</td></tr>
<tr><td>--norm_diff        </td><td> Maximum difference in normal vectors, in degrees</td></tr>
<tr><td>--maxzdiff         </td><td> Maximum difference in elevation (z units) between neighbouring points of the same segment</td></tr>
<tr><td>--classify         </td><td> Classify points as ground (2) or off-ground (1)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_segmentation_based_filter(
    i, 
    output, 
    radius=5.0, 
    norm_diff=2.0, 
    maxzdiff=1.0, 
    classify=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarSegmentationBasedFilter -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--radius=10.0 --norm_diff=2.5 --maxzdiff=0.75 --classify 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_segmentation_based_filter.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 5/12/2017, 2017</p>
<p><a name="LidarThin"></a></p>
<a class="header" href="#lidarthin" id="lidarthin"><h1>LidarThin</h1></a>
<p>Thins a LiDAR point cloud, reducing point density.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--resolution       </td><td> The size of the square area used to evaluate nearby points in the LiDAR data</td></tr>
<tr><td>--method           </td><td> Point selection method; options are 'first', 'last', 'lowest' (default), 'highest', 'nearest'</td></tr>
<tr><td>--save_filtered    </td><td> Save filtered points to seperate file?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_thin(
    i, 
    output, 
    resolution=2.0, 
    method=&quot;lowest&quot;, 
    save_filtered=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarThin -v --wd=&quot;/path/to/data/&quot; ^
-i=file.las -o=outfile.las --resolution=2.0, --method=first ^
--save_filtered 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_thin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 06/05/2018</p>
<p><a name="LidarThinHighDensity"></a></p>
<a class="header" href="#lidarthinhighdensity" id="lidarthinhighdensity"><h1>LidarThinHighDensity</h1></a>
<p>Thins points from high density areas within a LiDAR point cloud.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
<tr><td>--density          </td><td> Max. point density (points / m^3)</td></tr>
<tr><td>--save_filtered    </td><td> Save filtered points to seperate file?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_thin_high_density(
    i, 
    output, 
    density, 
    resolution=1.0, 
    save_filtered=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarThinHighDensity -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--resolution=1.0 --density=100.0 --save_filtered 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_thin_high_density.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 6, 2018</p>
<p><a name="LidarTile"></a></p>
<a class="header" href="#lidartile" id="lidartile"><h1>LidarTile</h1></a>
<p>Tiles a LiDAR LAS file into multiple LAS files.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>--width_x          </td><td> Width of tiles in the X dimension; default 1000.0</td></tr>
<tr><td>--width_y          </td><td> Width of tiles in the Y dimension</td></tr>
<tr><td>--origin_x         </td><td> Origin point X coordinate for tile grid</td></tr>
<tr><td>--origin_y         </td><td> Origin point Y coordinate for tile grid</td></tr>
<tr><td>--min_points       </td><td> Minimum number of points contained in a tile for it to be saved</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_tile(
    i, 
    width_x=1000.0, 
    width_y=1000.0, 
    origin_x=0.0, 
    origin_y=0.0, 
    min_points=2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarTile -v -i=/path/to/data/input.las ^
--width_x=1000.0 --width_y=2500.0 -=min_points=100 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_tile.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="LidarTileFootprint"></a></p>
<a class="header" href="#lidartilefootprint" id="lidartilefootprint"><h1>LidarTileFootprint</h1></a>
<p>Creates a vector polygon of the convex hull of a LiDAR point cloud. When the input/output parameters
are not specified, the tool works with all LAS files contained within the working directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output vector polygon file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_tile_footprint(
    output, 
    i=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarTileFootprint -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.shp 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_tile_footprint.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 31/08/2018</p>
<p><a name="LidarTinGridding"></a></p>
<a class="header" href="#lidartingridding" id="lidartingridding"><h1>LidarTinGridding</h1></a>
<p>Creates a raster grid based on a Delaunay triangular irregular network (TIN) fitted to LiDAR points.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file (including extension)</td></tr>
<tr><td>-o, --output       </td><td> Output raster file (including extension)</td></tr>
<tr><td>--parameter        </td><td> Interpolation parameter; options are 'elevation' (default), 'intensity', 'class', 'scan angle', 'user data'</td></tr>
<tr><td>--returns          </td><td> Point return types to include; options are 'all' (default), 'last', 'first'</td></tr>
<tr><td>--resolution       </td><td> Output raster's grid resolution</td></tr>
<tr><td>--exclude_cls      </td><td> Optional exclude classes from interpolation; Valid class values range from 0 to 18, based on LAS specifications. Example, --exclude_cls='3,4,5,6,7,18'</td></tr>
<tr><td>--minz             </td><td> Optional minimum elevation for inclusion in interpolation</td></tr>
<tr><td>--maxz             </td><td> Optional maximum elevation for inclusion in interpolation</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_tin_gridding(
    i=None, 
    output=None, 
    parameter=&quot;elevation&quot;, 
    returns=&quot;all&quot;, 
    resolution=1.0, 
    exclude_cls=None, 
    minz=None, 
    maxz=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarTINGridding -v ^
--wd=&quot;/path/to/data/&quot; -i=file.las -o=outfile.tif ^
--returns=last --resolution=2.0 --exclude_cls='3,4,5,6,7,18' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_tin_gridding.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 21/09/2018</p>
<p><a name="LidarTophatTransform"></a></p>
<a class="header" href="#lidartophattransform" id="lidartophattransform"><h1>LidarTophatTransform</h1></a>
<p>Performs a white top-hat transform on a Lidar dataset; as an estimate of height above ground, this is useful for modelling the vegetation canopy.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">lidar_tophat_transform(
    i, 
    output, 
    radius=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarTophatTransform -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;input.las&quot; -o=&quot;output.las&quot; ^
--radius=10.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/lidar_tophat_transform.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 22/09/2017</p>
<p><a name="NormalVectors"></a></p>
<a class="header" href="#normalvectors" id="normalvectors"><h1>NormalVectors</h1></a>
<p>Calculates normal vectors for points within a LAS file and stores these data (XYZ vector components) in the RGB field.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input LiDAR file</td></tr>
<tr><td>-o, --output       </td><td> Output LiDAR file</td></tr>
<tr><td>--radius           </td><td> Search Radius</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">normal_vectors(
    i, 
    output, 
    radius=1.0, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NormalVectors -v --wd=&quot;/path/to/data/&quot; ^
-i=&quot;input.las&quot; -o=&quot;output.las&quot; --radius=10.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/normal_vectors.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 26, 2017</p>
<p><a name="SelectTilesByPolygon"></a></p>
<a class="header" href="#selecttilesbypolygon" id="selecttilesbypolygon"><h1>SelectTilesByPolygon</h1></a>
<p>Copies LiDAR tiles overlapping with a polygon into an output directory.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--indir            </td><td> Input LAS file source directory</td></tr>
<tr><td>--outdir           </td><td> Output directory into which LAS files within the polygon are copied</td></tr>
<tr><td>--polygons         </td><td> Input vector polygons file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">select_tiles_by_polygon(
    indir, 
    outdir, 
    polygons, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SelectTilesByPolygon -v ^
--indir='/path/to/lidar/' --outdir='/output/path/' ^
--polygons='watershed.shp' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/lidar_analysis/select_tiles_by_polygon.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 01/08/2018</p>
<a class="header" href="#math-and-stats-tools" id="math-and-stats-tools"><h1>Math and Stats Tools</h1></a>
<ul>
<li><a href="#AbsoluteValue">AbsoluteValue</a></li>
<li><a href="#Add">Add</a></li>
<li><a href="#And">And</a></li>
<li><a href="#Anova">Anova</a></li>
<li><a href="#ArcCos">ArcCos</a></li>
<li><a href="#ArcSin">ArcSin</a></li>
<li><a href="#ArcTan">ArcTan</a></li>
<li><a href="#Atan2">Atan2</a></li>
<li><a href="#AttributeCorrelation">AttributeCorrelation</a></li>
<li><a href="#AttributeHistogram">AttributeHistogram</a></li>
<li><a href="#AttributeScattergram">AttributeScattergram</a></li>
<li><a href="#Ceil">Ceil</a></li>
<li><a href="#Cos">Cos</a></li>
<li><a href="#Cosh">Cosh</a></li>
<li><a href="#CrispnessIndex">CrispnessIndex</a></li>
<li><a href="#CrossTabulation">CrossTabulation</a></li>
<li><a href="#CumulativeDistribution">CumulativeDistribution</a></li>
<li><a href="#Decrement">Decrement</a></li>
<li><a href="#Divide">Divide</a></li>
<li><a href="#EqualTo">EqualTo</a></li>
<li><a href="#Exp">Exp</a></li>
<li><a href="#Exp2">Exp2</a></li>
<li><a href="#ExtractRasterStatistics">ExtractRasterStatistics</a></li>
<li><a href="#Floor">Floor</a></li>
<li><a href="#GreaterThan">GreaterThan</a></li>
<li><a href="#ImageAutocorrelation">ImageAutocorrelation</a></li>
<li><a href="#ImageCorrelation">ImageCorrelation</a></li>
<li><a href="#ImageRegression">ImageRegression</a></li>
<li><a href="#InPlaceAdd">InPlaceAdd</a></li>
<li><a href="#InPlaceDivide">InPlaceDivide</a></li>
<li><a href="#InPlaceMultiply">InPlaceMultiply</a></li>
<li><a href="#InPlaceSubtract">InPlaceSubtract</a></li>
<li><a href="#Increment">Increment</a></li>
<li><a href="#IntegerDivision">IntegerDivision</a></li>
<li><a href="#IsNoData">IsNoData</a></li>
<li><a href="#KappaIndex">KappaIndex</a></li>
<li><a href="#KsTestForNormality">KsTestForNormality</a></li>
<li><a href="#LessThan">LessThan</a></li>
<li><a href="#ListUniqueValues">ListUniqueValues</a></li>
<li><a href="#Ln">Ln</a></li>
<li><a href="#Log10">Log10</a></li>
<li><a href="#Log2">Log2</a></li>
<li><a href="#Max">Max</a></li>
<li><a href="#Min">Min</a></li>
<li><a href="#Modulo">Modulo</a></li>
<li><a href="#Multiply">Multiply</a></li>
<li><a href="#Negate">Negate</a></li>
<li><a href="#Not">Not</a></li>
<li><a href="#NotEqualTo">NotEqualTo</a></li>
<li><a href="#Or">Or</a></li>
<li><a href="#Power">Power</a></li>
<li><a href="#PrincipalComponentAnalysis">PrincipalComponentAnalysis</a></li>
<li><a href="#Quantiles">Quantiles</a></li>
<li><a href="#RandomField">RandomField</a></li>
<li><a href="#RandomSample">RandomSample</a></li>
<li><a href="#RasterHistogram">RasterHistogram</a></li>
<li><a href="#RasterSummaryStats">RasterSummaryStats</a></li>
<li><a href="#Reciprocal">Reciprocal</a></li>
<li><a href="#RescaleValueRange">RescaleValueRange</a></li>
<li><a href="#RootMeanSquareError">RootMeanSquareError</a></li>
<li><a href="#Round">Round</a></li>
<li><a href="#Sin">Sin</a></li>
<li><a href="#Sinh">Sinh</a></li>
<li><a href="#Square">Square</a></li>
<li><a href="#SquareRoot">SquareRoot</a></li>
<li><a href="#Subtract">Subtract</a></li>
<li><a href="#Tan">Tan</a></li>
<li><a href="#Tanh">Tanh</a></li>
<li><a href="#ToDegrees">ToDegrees</a></li>
<li><a href="#ToRadians">ToRadians</a></li>
<li><a href="#TrendSurface">TrendSurface</a></li>
<li><a href="#TrendSurfaceVectorPoints">TrendSurfaceVectorPoints</a></li>
<li><a href="#Truncate">Truncate</a></li>
<li><a href="#TurningBandsSimulation">TurningBandsSimulation</a></li>
<li><a href="#Xor">Xor</a></li>
<li><a href="#ZScores">ZScores</a></li>
</ul>
<p><a name="AbsoluteValue"></a></p>
<a class="header" href="#absolutevalue" id="absolutevalue"><h1>AbsoluteValue</h1></a>
<p>Calculates the absolute value of every cell in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">absolute_value(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AbsoluteValue -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/abs.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Add"></a></p>
<a class="header" href="#add" id="add"><h1>Add</h1></a>
<p>Performs an addition operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">add(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Add -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/add.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="And"></a></p>
<a class="header" href="#and" id="and"><h1>And</h1></a>
<p>Performs a logical AND operator on two Boolean raster images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">And(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=And -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/and.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="Anova"></a></p>
<a class="header" href="#anova" id="anova"><h1>Anova</h1></a>
<p>Performs an analysis of variance (ANOVA) test on a raster dataset.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--features         </td><td> Feature definition (or class) raster</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">anova(
    i, 
    features, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Anova -v --wd=&quot;/path/to/data/&quot; ^
-i=data.tif --features=classes.tif -o=anova.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/anova.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 23, 2017</p>
<p><a name="ArcCos"></a></p>
<a class="header" href="#arccos" id="arccos"><h1>ArcCos</h1></a>
<p>Returns the inverse cosine (arccos) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">arc_cos(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ArcCos -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/arccos.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="ArcSin"></a></p>
<a class="header" href="#arcsin" id="arcsin"><h1>ArcSin</h1></a>
<p>Returns the inverse sine (arcsin) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">arc_sin(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ArcSin -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/arcsin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="ArcTan"></a></p>
<a class="header" href="#arctan" id="arctan"><h1>ArcTan</h1></a>
<p>Returns the inverse tangent (arctan) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">arc_tan(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ArcTan -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/arctan.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Atan2"></a></p>
<a class="header" href="#atan2" id="atan2"><h1>Atan2</h1></a>
<p>Returns the 2-argument inverse tangent (atan2).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input_y          </td><td> Input y raster file or constant value (rise)</td></tr>
<tr><td>--input_x          </td><td> Input x raster file or constant value (run)</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">atan2(
    input_y, 
    input_x, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Atan2 -v --wd=&quot;/path/to/data/&quot; ^
--input_y='in1.tif' --input_x='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/atan2.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="AttributeCorrelation"></a></p>
<a class="header" href="#attributecorrelation" id="attributecorrelation"><h1>AttributeCorrelation</h1></a>
<p>Performs a correlation analysis on attribute fields from a vector database.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">attribute_correlation(
    i, 
    output=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AttributeCorrelation -v ^
--wd=&quot;/path/to/data/&quot; -i=file.shp -o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/attribute_correlation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 29/04/2018</p>
<p><a name="AttributeHistogram"></a></p>
<a class="header" href="#attributehistogram" id="attributehistogram"><h1>AttributeHistogram</h1></a>
<p>Creates a histogram for the field values of a vector's attribute table.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">attribute_histogram(
    i, 
    field, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AttributeHistogram -v ^
--wd=&quot;/path/to/data/&quot; -i=lakes.shp --field=HEIGHT ^
-o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/attribute_histogram.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 12/04/2018</p>
<p><a name="AttributeScattergram"></a></p>
<a class="header" href="#attributescattergram" id="attributescattergram"><h1>AttributeScattergram</h1></a>
<p>Creates a scattergram for two field values of a vector's attribute table.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--fieldx           </td><td> Input field name in attribute table for the x-axis</td></tr>
<tr><td>--fieldy           </td><td> Input field name in attribute table for the y-axis</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
<tr><td>--trendline        </td><td> Draw the trendline</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">attribute_scattergram(
    i, 
    fieldx, 
    fieldy, 
    output, 
    trendline=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=AttributeScattergram -v ^
--wd=&quot;/path/to/data/&quot; -i=lakes.shp --fieldx=HEIGHT ^
--fieldy=area -o=outfile.html --trendline 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/attribute_scattergram.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 12/04/2018</p>
<p><a name="Ceil"></a></p>
<a class="header" href="#ceil" id="ceil"><h1>Ceil</h1></a>
<p>Returns the smallest (closest to negative infinity) value that is greater than or equal to the values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">ceil(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Ceil -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/ceil.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Cos"></a></p>
<a class="header" href="#cos" id="cos"><h1>Cos</h1></a>
<p>Returns the cosine (cos) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">cos(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Cos -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/cos.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Cosh"></a></p>
<a class="header" href="#cosh" id="cosh"><h1>Cosh</h1></a>
<p>Returns the hyperbolic cosine (cosh) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">cosh(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Cosh -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/cosh.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="CrispnessIndex"></a></p>
<a class="header" href="#crispnessindex" id="crispnessindex"><h1>CrispnessIndex</h1></a>
<p>This index (C) is taken from Lindsay (2006) Sensitivity of channel mapping techniques to uncertainty in digital
elevation data:</p>
<p>C = SS_mp ∕ SS_B = [∑(pij − p-bar)^2] ∕ [ ∑pij(1 − p-bar)^2 + p2(RC − ∑pij)]</p>
<p>Please note that there is an error in the original published equation. Specifically, the denominator
read:</p>
<p>∑pij(1 - p_bar)^2 + p_bar^2 (RC - ∑pij)</p>
<p>instead of the original:</p>
<p>∑pij(1 - p_bar^2) - p_bar^2 (RC - ∑pij)</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Optional output html file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">crispness_index(
    i, 
    output=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CrispnessIndex -v --wd=&quot;/path/to/data/&quot; ^
-i=input.tif
&gt;&gt;./whitebox_tools -r=CrispnessIndex -v ^
--wd=&quot;/path/to/data/&quot; -o=crispness.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/crispness_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: August 15, 2017</p>
<p><a name="CrossTabulation"></a></p>
<a class="header" href="#crosstabulation" id="crosstabulation"><h1>CrossTabulation</h1></a>
<p>Performs a cross-tabulation on two categorical images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input raster file 1</td></tr>
<tr><td>--i2, --input2    </td><td> Input raster file 1</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">cross_tabulation(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CrossTabulation -v ^
--wd=&quot;/path/to/data/&quot; --i1=&quot;file1.tif&quot; --i2=&quot;file2.tif&quot; ^
-o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/cross_tabulation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 18, 2017</p>
<p><a name="CumulativeDistribution"></a></p>
<a class="header" href="#cumulativedistribution" id="cumulativedistribution"><h1>CumulativeDistribution</h1></a>
<p>Converts a raster image to its cumulative distribution function.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">cumulative_distribution(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=CumulativeDistribution -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/cumulative_dist.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 22 2017</p>
<p><a name="Decrement"></a></p>
<a class="header" href="#decrement" id="decrement"><h1>Decrement</h1></a>
<p>Decreases the values of each grid cell in an input raster by 1.0 (see also InPlaceSubtract).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">decrement(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Decrement -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/decrement.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="Divide"></a></p>
<a class="header" href="#divide" id="divide"><h1>Divide</h1></a>
<p>Performs a division operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">divide(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Divide -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/divide.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="EqualTo"></a></p>
<a class="header" href="#equalto" id="equalto"><h1>EqualTo</h1></a>
<p>Performs a equal-to comparison operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">equal_to(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=EqualTo -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/equal_to.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="Exp"></a></p>
<a class="header" href="#exp" id="exp"><h1>Exp</h1></a>
<p>Returns the exponential (base e) of values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">exp(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Exp -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/exp.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Exp2"></a></p>
<a class="header" href="#exp2" id="exp2"><h1>Exp2</h1></a>
<p>Returns the exponential (base 2) of values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">exp2(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Exp2 -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/exp2.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="ExtractRasterStatistics"></a></p>
<a class="header" href="#extractrasterstatistics" id="extractrasterstatistics"><h1>ExtractRasterStatistics</h1></a>
<p>This tool can be used to extract common descriptive statistics associated with the distribution
of some underlying data raster based on feature units defined by a feature definition raster.
For example, this tool can be used to measure the maximum or average slope gradient (data image)
for each of a group of watersheds (feature definitions). Although the data raster can contain any
type of data, the feature definition raster must be categorical, i.e. it must define area entities
using integer values.</p>
<p>If an output image name is specified, the tool will assign the descriptive statistic value to
each of the spatial entities defined in the feature definition raster. If text output is selected,
an HTML table will be output, which can then be readily copied into a spreadsheet program for
further analysis. This is a very powerful and useful tool for creating numerical summary data from
spatial data which can then be interrogated using statistical analyses. At least one output type
(image or text) must be specified for the tool to operate.</p>
<p>NoData values in either of the two input images are ignored during the calculation of the
descriptive statistic.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input data raster file</td></tr>
<tr><td>--features         </td><td> Input feature definition raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--stat             </td><td> Statistic to extract</td></tr>
<tr><td>--out_table        </td><td> Output HTML Table file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">extract_raster_statistics(
    i, 
    features, 
    output=None, 
    stat=&quot;average&quot;, 
    out_table=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ExtractRasterStatistics -v ^
--wd=&quot;/path/to/data/&quot; -i='input.tif' --features='groups.tif' ^
-o='output.tif' --stat='minimum'
&gt;&gt;./whitebox_tools ^
-r=ExtractRasterStatistics -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' --features='groups.tif' ^
--out_table='output.html' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/extract_statistics.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 15, 2017</p>
<p><a name="Floor"></a></p>
<a class="header" href="#floor" id="floor"><h1>Floor</h1></a>
<p>Returns the largest (closest to positive infinity) value that is less than or equal to the values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">floor(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Floor -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o='output.tif' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/floor.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="GreaterThan"></a></p>
<a class="header" href="#greaterthan" id="greaterthan"><h1>GreaterThan</h1></a>
<p>Performs a greater-than comparison operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--incl_equals      </td><td> Perform a greater-than-or-equal-to operation</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">greater_than(
    input1, 
    input2, 
    output, 
    incl_equals=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=GreaterThan -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif ^
--incl_equals 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/greater_than.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="ImageAutocorrelation"></a></p>
<a class="header" href="#imageautocorrelation" id="imageautocorrelation"><h1>ImageAutocorrelation</h1></a>
<p>Performs Moran's I analysis on two or more input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--contiguity       </td><td> Contiguity type</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">image_autocorrelation(
    inputs, 
    output, 
    contiguity=&quot;Rook&quot;, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ImageAutocorrelation -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;file1.tif, file2.tif, file3.tif&quot; ^
-o=outfile.html --contiguity=Bishops 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/image_autocorrelation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 16, 2017</p>
<p><a name="ImageCorrelation"></a></p>
<a class="header" href="#imagecorrelation" id="imagecorrelation"><h1>ImageCorrelation</h1></a>
<p>Performs image correlation on two or more input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">image_correlation(
    inputs, 
    output=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ImageCorrelation -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;file1.tif, file2.tif, file3.tif&quot; ^
-o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/image_correlation.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 3, 2017</p>
<p><a name="ImageRegression"></a></p>
<a class="header" href="#imageregression" id="imageregression"><h1>ImageRegression</h1></a>
<p>Performs image regression analysis on two input images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input raster file (independent variable, X)</td></tr>
<tr><td>--i2, --input2    </td><td> Input raster file (dependent variable, Y)</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file for regression summary report</td></tr>
<tr><td>--out_residuals    </td><td> Output raster regression resdidual file</td></tr>
<tr><td>--standardize      </td><td> Optional flag indicating whether to standardize the residuals map</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">image_regression(
    input1, 
    input2, 
    output, 
    out_residuals=None, 
    standardize=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ImageRegression -v ^
--wd=&quot;/path/to/data/&quot; --i1='file1.tif' --i2='file2.tif' ^
-o='outfile.html' --out_residuals='residuals.tif' ^
--standardize 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/image_regression.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 27, 2017</p>
<p><a name="InPlaceAdd"></a></p>
<a class="header" href="#inplaceadd" id="inplaceadd"><h1>InPlaceAdd</h1></a>
<p>Performs an in-place addition operation (input1 += input2).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">in_place_add(
    input1, 
    input2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=InPlaceAdd -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif'&quot;
&gt;&gt;./whitebox_tools ^
-r=InPlaceAdd -v --wd=&quot;/path/to/data/&quot; --input1='in1.tif' ^
--input2=10.5' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/inplace_add.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 18/03/2018</p>
<p><a name="InPlaceDivide"></a></p>
<a class="header" href="#inplacedivide" id="inplacedivide"><h1>InPlaceDivide</h1></a>
<p>Performs an in-place division operation (input1 /= input2).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">in_place_divide(
    input1, 
    input2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=InPlaceDivide -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif'&quot;
&gt;&gt;./whitebox_tools ^
-r=InPlaceDivide -v --wd=&quot;/path/to/data/&quot; --input1='in1.tif' ^
--input2=10.5' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/inplace_divide.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 18/03/2018</p>
<p><a name="InPlaceMultiply"></a></p>
<a class="header" href="#inplacemultiply" id="inplacemultiply"><h1>InPlaceMultiply</h1></a>
<p>Performs an in-place multiplication operation (input1 *= input2).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">in_place_multiply(
    input1, 
    input2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=InPlaceMultiply -v ^
--wd=&quot;/path/to/data/&quot; --input1='in1.tif' ^
--input2='in2.tif'&quot;
&gt;&gt;./whitebox_tools -r=InPlaceMultiply -v ^
--wd=&quot;/path/to/data/&quot; --input1='in1.tif' --input2=10.5' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/inplace_multiply.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 18/03/2018</p>
<p><a name="InPlaceSubtract"></a></p>
<a class="header" href="#inplacesubtract" id="inplacesubtract"><h1>InPlaceSubtract</h1></a>
<p>Performs an in-place subtraction operation (input1 -= input2).</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">in_place_subtract(
    input1, 
    input2, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=InPlaceSubtract -v ^
--wd=&quot;/path/to/data/&quot; --input1='in1.tif' ^
--input2='in2.tif'&quot;
&gt;&gt;./whitebox_tools -r=InPlaceSubtract -v ^
--wd=&quot;/path/to/data/&quot; --input1='in1.tif' --input2=10.5' 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/inplace_subtract.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 18/03/2018</p>
<p><a name="Increment"></a></p>
<a class="header" href="#increment" id="increment"><h1>Increment</h1></a>
<p>Increases the values of each grid cell in an input raster by 1.0. (see also InPlaceAdd)</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">increment(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Increment -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/increment.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="IntegerDivision"></a></p>
<a class="header" href="#integerdivision" id="integerdivision"><h1>IntegerDivision</h1></a>
<p>Performs an integer division operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">integer_division(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=IntegerDivision -v ^
--wd=&quot;/path/to/data/&quot; --input1='in1.tif' --input2='in2.tif' ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/integer_division.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="IsNoData"></a></p>
<a class="header" href="#isnodata" id="isnodata"><h1>IsNoData</h1></a>
<p>Identifies NoData valued pixels in an image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">is_no_data(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=IsNoData -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/isnodata.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="KappaIndex"></a></p>
<a class="header" href="#kappaindex" id="kappaindex"><h1>KappaIndex</h1></a>
<p>Performs a kappa index of agreement (KIA) analysis on two categorical raster files.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--i1, --input1    </td><td> Input classification raster file</td></tr>
<tr><td>--i2, --input2    </td><td> Input reference raster file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">kappa_index(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=KappaIndex -v --wd=&quot;/path/to/data/&quot; ^
--i1=class.tif --i2=reference.tif -o=kia.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/kappa_index.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 24, 2017</p>
<p><a name="KsTestForNormality"></a></p>
<a class="header" href="#kstestfornormality" id="kstestfornormality"><h1>KsTestForNormality</h1></a>
<p>This tool will perform a Kolmogorov-Smirnov (K-S) test for normality to evaluate
whether the frequency distribution of values within a raster image are drawn from a
Gaussian (normal) distribution. The user must specify the name of the raster image. The
test can be performed optionally on the entire image or on a random sub-sample of pixel
values of a user-specified size. In evaluating the significance of the test, it is
important to keep in mind that given a sufficiently large sample, extremely small and
non-notable differences can be found to be statistically significant. Furthermore
statistical significance says nothing about the practical significance of a difference.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
<tr><td>--num_samples      </td><td> Number of samples. Leave blank to use whole image</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">ks_test_for_normality(
    i, 
    output, 
    num_samples=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=KSTestForNormality -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.html ^
--num_samples=1000
&gt;&gt;./whitebox_tools -r=KSTestForNormality -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/ks_normality_test.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: January 2, 2018</p>
<p><a name="LessThan"></a></p>
<a class="header" href="#lessthan" id="lessthan"><h1>LessThan</h1></a>
<p>Performs a less-than comparison operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--incl_equals      </td><td> Perform a less-than-or-equal-to operation</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">less_than(
    input1, 
    input2, 
    output, 
    incl_equals=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LessThan -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif ^
--incl_equals 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/less_than.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="ListUniqueValues"></a></p>
<a class="header" href="#listuniquevalues" id="listuniquevalues"><h1>ListUniqueValues</h1></a>
<p>Lists the unique values contained in a field witin a vector's attribute table.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">list_unique_values(
    i, 
    field, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ListUniqueValues -v ^
--wd=&quot;/path/to/data/&quot; -i=lakes.shp --field=HEIGHT ^
-o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/list_unique_values.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 12/04/2018</p>
<p><a name="Ln"></a></p>
<a class="header" href="#ln" id="ln"><h1>Ln</h1></a>
<p>Returns the natural logarithm of values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">ln(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Ln -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/ln.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Log10"></a></p>
<a class="header" href="#log10" id="log10"><h1>Log10</h1></a>
<p>Returns the base-10 logarithm of values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">log10(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Log10 -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/log10.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Log2"></a></p>
<a class="header" href="#log2" id="log2"><h1>Log2</h1></a>
<p>Returns the base-2 logarithm of values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">log2(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Log2 -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/log2.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Max"></a></p>
<a class="header" href="#max" id="max"><h1>Max</h1></a>
<p>Performs a MAX operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">max(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Max -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/max.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Min"></a></p>
<a class="header" href="#min" id="min"><h1>Min</h1></a>
<p>Performs a MIN operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">min(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Min -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/min.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Modulo"></a></p>
<a class="header" href="#modulo" id="modulo"><h1>Modulo</h1></a>
<p>Performs a modulo operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">modulo(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Modulo -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/modulo.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Multiply"></a></p>
<a class="header" href="#multiply" id="multiply"><h1>Multiply</h1></a>
<p>Performs a multiplication operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">multiply(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Multiply -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/multiply.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Negate"></a></p>
<a class="header" href="#negate" id="negate"><h1>Negate</h1></a>
<p>Changes the sign of values in a raster or the 0-1 values of a Boolean raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">negate(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Negate -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/negate.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Not"></a></p>
<a class="header" href="#not" id="not"><h1>Not</h1></a>
<p>Performs a logical NOT operator on two Boolean raster images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">Not(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Not -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/not.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<p><a name="NotEqualTo"></a></p>
<a class="header" href="#notequalto" id="notequalto"><h1>NotEqualTo</h1></a>
<p>Performs a not-equal-to comparison operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">not_equal_to(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=NotEqualTo -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/not_equal_to.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="Or"></a></p>
<a class="header" href="#or" id="or"><h1>Or</h1></a>
<p>Performs a logical OR operator on two Boolean raster images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">Or(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Or -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/or.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<p><a name="Power"></a></p>
<a class="header" href="#power" id="power"><h1>Power</h1></a>
<p>Raises the values in grid cells of one rasters, or a constant value, by values in another raster or constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">power(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Power -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/power.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="PrincipalComponentAnalysis"></a></p>
<a class="header" href="#principalcomponentanalysis" id="principalcomponentanalysis"><h1>PrincipalComponentAnalysis</h1></a>
<p>Performs a principal component analysis (PCA) on a multi-spectral dataset.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --inputs       </td><td> Input raster files</td></tr>
<tr><td>--out_html         </td><td> Output HTML report file</td></tr>
<tr><td>--num_comp         </td><td> Number of component images to output; &lt;= to num. input images</td></tr>
<tr><td>--standardized     </td><td> Perform standardized PCA?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">principal_component_analysis(
    inputs, 
    out_html, 
    num_comp=None, 
    standardized=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=PrincipalComponentAnalysis -v ^
--wd='/path/to/data/' -i='image1.tif;image2.tif;image3.tif' ^
--out_html=report.html --num_comp=3 --standardized 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/principal_component_analysis.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 15/03/2018</p>
<p><a name="Quantiles"></a></p>
<a class="header" href="#quantiles" id="quantiles"><h1>Quantiles</h1></a>
<p>Transforms raster values into quantiles.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--num_quantiles    </td><td> Number of quantiles</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">quantiles(
    i, 
    output, 
    num_quantiles=5, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Quantiles -v --wd=&quot;/path/to/data/&quot; ^
-i=DEM.tif -o=output.tif --num_quantiles=5 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/quantiles.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="RandomField"></a></p>
<a class="header" href="#randomfield" id="randomfield"><h1>RandomField</h1></a>
<p>Creates an image containing random values.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --base         </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">random_field(
    base, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RandomField -v --wd=&quot;/path/to/data/&quot; ^
--base=in.tif -o=out.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/random_field.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="RandomSample"></a></p>
<a class="header" href="#randomsample" id="randomsample"><h1>RandomSample</h1></a>
<p>This tool can be used to create a random sample of grid cells. The user specifies
the base raster file, which is used to determine the grid dimensions and georeference
information for the output raster, and the number of sample random samples (n). The
output grid will contain n non-zero grid cells, randomly distributed throughout the
raster grid, and a background value of zero. This tool is useful when performing
statistical analyses on raster images when you wish to obtain a random sample of data.</p>
<p>Only valid, non-nodata, cells in the base raster will be sampled.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --base         </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--num_samples      </td><td> Number of samples</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">random_sample(
    base, 
    output, 
    num_samples=1000, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RandomSample -v --wd=&quot;/path/to/data/&quot; ^
--base=in.tif -o=out.tif --num_samples=1000 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/random_sample.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: January 2, 2018</p>
<p><a name="RasterHistogram"></a></p>
<a class="header" href="#rasterhistogram" id="rasterhistogram"><h1>RasterHistogram</h1></a>
<p>Creates a histogram from raster values.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file (default name will be based on input file if unspecified)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">raster_histogram(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RasterHistogram -v ^
--wd=&quot;/path/to/data/&quot; -i=&quot;file1.tif&quot; -o=outfile.html 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/raster_histogram.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: Dec. 19, 2017</p>
<p><a name="RasterSummaryStats"></a></p>
<a class="header" href="#rastersummarystats" id="rastersummarystats"><h1>RasterSummaryStats</h1></a>
<p>Measures a rasters average, standard deviation, num. non-nodata cells, and total.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">raster_summary_stats(
    i, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RasterSummaryStats -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/raster_summary_stats.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<p><a name="Reciprocal"></a></p>
<a class="header" href="#reciprocal" id="reciprocal"><h1>Reciprocal</h1></a>
<p>Returns the reciprocal (i.e. 1 / z) of values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">reciprocal(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Reciprocal -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/reciprocal.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="RescaleValueRange"></a></p>
<a class="header" href="#rescalevaluerange" id="rescalevaluerange"><h1>RescaleValueRange</h1></a>
<p>Performs a min-max contrast stretch on an input greytone image.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--out_min_val      </td><td> New minimum value in output image</td></tr>
<tr><td>--out_max_val      </td><td> New maximum value in output image</td></tr>
<tr><td>--clip_min         </td><td> Optional lower tail clip value</td></tr>
<tr><td>--clip_max         </td><td> Optional upper tail clip value</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">rescale_value_range(
    i, 
    output, 
    out_min_val, 
    out_max_val, 
    clip_min=None, 
    clip_max=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RescaleValueRange -v ^
--wd=&quot;/path/to/data/&quot; -i=input.tif -o=output.tif ^
--out_min_val=0.0 --out_max_val=1.0
&gt;&gt;./whitebox_tools ^
-r=RescaleValueRange -v --wd=&quot;/path/to/data/&quot; -i=input.tif ^
-o=output.tif --out_min_val=0.0 --out_max_val=1.0 ^
--clip_min=45.0 --clip_max=200.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/rescale_value_range.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 10, 2017</p>
<p><a name="RootMeanSquareError"></a></p>
<a class="header" href="#rootmeansquareerror" id="rootmeansquareerror"><h1>RootMeanSquareError</h1></a>
<p>This tool calculates the root-mean-square-error (RMSE) or root-mean-square-difference (RMSD) from two
input rasters. If the two input rasters possess the same number of rows and columns, the RMSE is
calucated on a cell-by-cell basis, otherwise bilinear resampling is used. In addition to RMSE,
the tool also reports other common accuracy statistics including the mean verical error, the
95% confidence limit (RMSE x 1.96), and the 90% linear error (LE90), which is the 90% percentile of
the residuals between two raster surfaces. The LE90 is the most robust of the reported accuracy
statistics when the residuals are non-Gaussian. The LE90 requires sorting the residual values, which
can be a relatively slow operation for larger rasters.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>--base             </td><td> Input base raster file used for comparison</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">root_mean_square_error(
    i, 
    base, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RootMeanSquareError -v ^
--wd=&quot;/path/to/data/&quot; -i=DEM.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/root_mean_square_error.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: September 18, 2017</p>
<p><a name="Round"></a></p>
<a class="header" href="#round" id="round"><h1>Round</h1></a>
<p>Rounds the values in an input raster to the nearest integer value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">round(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Round -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/round.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Sin"></a></p>
<a class="header" href="#sin" id="sin"><h1>Sin</h1></a>
<p>Returns the sine (sin) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">sin(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Sin -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/sin.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Sinh"></a></p>
<a class="header" href="#sinh" id="sinh"><h1>Sinh</h1></a>
<p>Returns the hyperbolic sine (sinh) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">sinh(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Sinh -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/sinh.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Square"></a></p>
<a class="header" href="#square" id="square"><h1>Square</h1></a>
<p>Squares the values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">square(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Square -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/square.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="SquareRoot"></a></p>
<a class="header" href="#squareroot" id="squareroot"><h1>SquareRoot</h1></a>
<p>Returns the square root of the values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">square_root(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=SquareRoot -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/sqrt.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="Subtract"></a></p>
<a class="header" href="#subtract" id="subtract"><h1>Subtract</h1></a>
<p>Performs a differencing operation on two rasters or a raster and a constant value.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file or constant value</td></tr>
<tr><td>--input2           </td><td> Input raster file or constant value</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">subtract(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Subtract -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/subtract.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Tan"></a></p>
<a class="header" href="#tan" id="tan"><h1>Tan</h1></a>
<p>Returns the tangent (tan) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">tan(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Tan -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/tan.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 5, 2017</p>
<p><a name="Tanh"></a></p>
<a class="header" href="#tanh" id="tanh"><h1>Tanh</h1></a>
<p>Returns the hyperbolic tangent (tanh) of each values in a raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">tanh(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Tanh -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/tanh.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="ToDegrees"></a></p>
<a class="header" href="#todegrees" id="todegrees"><h1>ToDegrees</h1></a>
<p>Converts a raster from radians to degrees.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">to_degrees(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ToDegrees -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/to_degrees.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="ToRadians"></a></p>
<a class="header" href="#toradians" id="toradians"><h1>ToRadians</h1></a>
<p>Converts a raster from degrees to radians.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">to_radians(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ToRadians -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/to_radians.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="TrendSurface"></a></p>
<a class="header" href="#trendsurface" id="trendsurface"><h1>TrendSurface</h1></a>
<p>This tool can be used to interpolate a trend surface from a raster image. The
technique uses a polynomial, least-squares regression analysis. The user must
specify the name of the input raster file. In addition, the user must specify
the polynomial order (1 to 10) for the analysis. A first-order polynomial is a
planar surface with no curvature. As the polynomial order is increased, greater
flexibility is allowed in the fitted surface. Although polynomial orders as high
as 10 are accepted, numerical instability in the analysis often creates artifacts
in trend surfaces of orders greater than 5. The operation will display a text
report on completion, in addition to the output raster image. The report will
list each of the coefficient values and the r-square value. Note that the entire
raster image must be able to fit into computer memory, limiting the use of this
tool to relatively small rasters. The Trend Surface (Vector Points) tool can be
used instead if the input data is vector points contained in a shapefile.</p>
<p>Numerical stability is enhanced by transforming the x, y, z data by their minimum
values before performing the regression analysis. These transform parameters
are also reported in the output report.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--order            </td><td> Polynomial order (1 to 10)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">trend_surface(
    i, 
    output, 
    order=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TrendSurface -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o='output.tif' --order=2 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/trend_surface.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 30/04/2018</p>
<p><a name="TrendSurfaceVectorPoints"></a></p>
<a class="header" href="#trendsurfacevectorpoints" id="trendsurfacevectorpoints"><h1>TrendSurfaceVectorPoints</h1></a>
<p>This tool can be used to interpolate a trend surface from a vector points file. The
technique uses a polynomial, least-squares regression analysis. The user must specify
the name of the input shapefile, which must be of a 'Points' base ShapeType and select
the attribute in the shapefile's associated attribute table for which to base the trend
surface analysis. The attribute must be numerical. In addition, the user must specify
the polynomial order (1 to 10) for the analysis. A first-order polynomial is a planar
surface with no curvature. As the polynomial order is increased, greater flexibility is
allowed in the fitted surface. Although polynomial orders as high as 10 are accepted,
numerical instability in the analysis often creates artifacts in trend surfaces of orders
greater than 5. The operation will display a text report on completion, in addition to
the output raster image. The report will list each of the coefficient values and the
r-square value. The Trend Surface tool can be used instead if the input data is a raster
image.</p>
<p>Numerical stability is enhanced by transforming the x, y, z data by their minimum
values before performing the regression analysis. These transform parameters
are also reported in the output report.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input vector Points file</td></tr>
<tr><td>--field            </td><td> Input field name in attribute table</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--order            </td><td> Polynomial order (1 to 10)</td></tr>
<tr><td>--cell_size        </td><td> Optionally specified cell size of output raster. Not used when base raster is specified</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">trend_surface_vector_points(
    i, 
    field, 
    output, 
    cell_size, 
    order=1, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TrendSurfaceVectorPoints -v ^
--wd=&quot;/path/to/data/&quot; -i='input.shp' --field=ELEV ^
-o='output.tif' --order=2 --cell_size=10.0 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/trend_surface_vector_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 01/05/2018</p>
<p><a name="Truncate"></a></p>
<a class="header" href="#truncate" id="truncate"><h1>Truncate</h1></a>
<p>Truncates the values in a raster to the desired number of decimal places.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--num_decimals     </td><td> Number of decimals left after truncation (default is zero)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">truncate(
    i, 
    output, 
    num_decimals=None, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Truncate -v --wd=&quot;/path/to/data/&quot; ^
-i='input.tif' -o=output.tif --num_decimals=2 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/truncate.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="TurningBandsSimulation"></a></p>
<a class="header" href="#turningbandssimulation" id="turningbandssimulation"><h1>TurningBandsSimulation</h1></a>
<p>Creates an image containing random values based on a turning-bands simulation.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --base         </td><td> Input base raster file</td></tr>
<tr><td>-o, --output       </td><td> Output file</td></tr>
<tr><td>--range            </td><td> The field's range, in xy-units, related to the extent of spatial autocorrelation</td></tr>
<tr><td>--iterations       </td><td> The number of iterations</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">turning_bands_simulation(
    base, 
    output, 
    range, 
    iterations=1000, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TurningBandsSimulation -v ^
--wd=&quot;/path/to/data/&quot; --base=in.tif -o=out.tif --range=850.0 ^
--iterations=2500 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/turning_bands.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 14, 2017</p>
<p><a name="Xor"></a></p>
<a class="header" href="#xor" id="xor"><h1>Xor</h1></a>
<p>Performs a logical XOR operator on two Boolean raster images.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--input1           </td><td> Input raster file</td></tr>
<tr><td>--input2           </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">xor(
    input1, 
    input2, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=Xor -v --wd=&quot;/path/to/data/&quot; ^
--input1='in1.tif' --input2='in2.tif' -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/xor.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 2, 2017</p>
<p><a name="ZScores"></a></p>
<a class="header" href="#zscores" id="zscores"><h1>ZScores</h1></a>
<p>Standardizes the values in an input raster by converting to z-scores.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --input        </td><td> Input raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">z_scores(
    i, 
    output, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ZScores -v --wd=&quot;/path/to/data/&quot; ^
-i=DEM.tif -o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/math_stat_analysis/zscores.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 1, 2017</p>
<a class="header" href="#stream-network-analysis" id="stream-network-analysis"><h1>Stream Network Analysis</h1></a>
<ul>
<li><a href="#DistanceToOutlet">DistanceToOutlet</a></li>
<li><a href="#ExtractStreams">ExtractStreams</a></li>
<li><a href="#ExtractValleys">ExtractValleys</a></li>
<li><a href="#FarthestChannelHead">FarthestChannelHead</a></li>
<li><a href="#FindMainStem">FindMainStem</a></li>
<li><a href="#HackStreamOrder">HackStreamOrder</a></li>
<li><a href="#HortonStreamOrder">HortonStreamOrder</a></li>
<li><a href="#LengthOfUpstreamChannels">LengthOfUpstreamChannels</a></li>
<li><a href="#LongProfile">LongProfile</a></li>
<li><a href="#LongProfileFromPoints">LongProfileFromPoints</a></li>
<li><a href="#RasterStreamsToVector">RasterStreamsToVector</a></li>
<li><a href="#RasterizeStreams">RasterizeStreams</a></li>
<li><a href="#RemoveShortStreams">RemoveShortStreams</a></li>
<li><a href="#ShreveStreamMagnitude">ShreveStreamMagnitude</a></li>
<li><a href="#StrahlerStreamOrder">StrahlerStreamOrder</a></li>
<li><a href="#StreamLinkClass">StreamLinkClass</a></li>
<li><a href="#StreamLinkIdentifier">StreamLinkIdentifier</a></li>
<li><a href="#StreamLinkLength">StreamLinkLength</a></li>
<li><a href="#StreamLinkSlope">StreamLinkSlope</a></li>
<li><a href="#StreamSlopeContinuous">StreamSlopeContinuous</a></li>
<li><a href="#TopologicalStreamOrder">TopologicalStreamOrder</a></li>
<li><a href="#TributaryIdentifier">TributaryIdentifier</a></li>
</ul>
<p><a name="DistanceToOutlet"></a></p>
<a class="header" href="#distancetooutlet" id="distancetooutlet"><h1>DistanceToOutlet</h1></a>
<p>Calculates the distance of stream grid cells to the channel network outlet cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">distance_to_outlet(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=DistanceToOutlet -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=DistanceToOutlet -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/dist_to_outlet.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 13, 2017</p>
<p><a name="ExtractStreams"></a></p>
<a class="header" href="#extractstreams" id="extractstreams"><h1>ExtractStreams</h1></a>
<p>Extracts stream grid cells from a flow accumulation raster.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--flow_accum       </td><td> Input raster D8 flow accumulation file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--threshold        </td><td> Threshold in flow accumulation values for channelization</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">extract_streams(
    flow_accum, 
    output, 
    threshold, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ExtractStreams -v --wd=&quot;/path/to/data/&quot; ^
--flow_accum='d8accum.tif' -o='output.tif' --threshold=100.0 ^
--zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/extract_streams.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="ExtractValleys"></a></p>
<a class="header" href="#extractvalleys" id="extractvalleys"><h1>ExtractValleys</h1></a>
<p>Identifies potential valley bottom grid cells based on local topolography alone.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--variant          </td><td> Options include 'lq' (lower quartile), 'JandR' (Johnston and Rosenfeld), and 'PandD' (Peucker and Douglas); default is 'lq'</td></tr>
<tr><td>--line_thin        </td><td> Optional flag indicating whether post-processing line-thinning should be performed</td></tr>
<tr><td>--filter           </td><td> Optional argument (only used when variant='lq') providing the filter size, in grid cells, used for lq-filtering (default is 5)</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">extract_valleys(
    dem, 
    output, 
    variant=&quot;Lower Quartile&quot;, 
    line_thin=True, 
    filter=5, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ExtractValleys -v --wd=&quot;/path/to/data/&quot; ^
--dem=pointer.tif -o=out.tif --variant='JandR' ^
--line_thin
&gt;&gt;./whitebox_tools -r=ExtractValleys -v ^
--wd=&quot;/path/to/data/&quot; --dem=pointer.tif -o=out.tif ^
--variant='lq' --filter=7 --line_thin 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/extract_valleys.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 12, 2017</p>
<p><a name="FarthestChannelHead"></a></p>
<a class="header" href="#farthestchannelhead" id="farthestchannelhead"><h1>FarthestChannelHead</h1></a>
<p>Calculates the distance to the furthest upstream channel head for each stream cell.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">farthest_channel_head(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FarthestChannelHead -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=FarthestChannelHead -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/farthest_channel_head.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 12, 2017</p>
<p><a name="FindMainStem"></a></p>
<a class="header" href="#findmainstem" id="findmainstem"><h1>FindMainStem</h1></a>
<p>Finds the main stem, based on stream lengths, of each stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">find_main_stem(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=FindMainStem -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=FindMainStem -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/find_main_stem.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="HackStreamOrder"></a></p>
<a class="header" href="#hackstreamorder" id="hackstreamorder"><h1>HackStreamOrder</h1></a>
<p>Assigns the Hack stream order to each tributary in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">hack_stream_order(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HackStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=HackStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/hack_order.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="HortonStreamOrder"></a></p>
<a class="header" href="#hortonstreamorder" id="hortonstreamorder"><h1>HortonStreamOrder</h1></a>
<p>Assigns the Horton stream order to each tributary in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">horton_stream_order(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=HortonStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=HortonStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/horton_order.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="LengthOfUpstreamChannels"></a></p>
<a class="header" href="#lengthofupstreamchannels" id="lengthofupstreamchannels"><h1>LengthOfUpstreamChannels</h1></a>
<p>Calculates the total length of channels upstream.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">length_of_upstream_channels(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LengthOfUpstreamChannels -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=LengthOfUpstreamChannels -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/total_length_channels.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 11, 2017</p>
<p><a name="LongProfile"></a></p>
<a class="header" href="#longprofile" id="longprofile"><h1>LongProfile</h1></a>
<p>Plots the stream longitudinal profiles for one or more rivers.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">long_profile(
    d8_pntr, 
    streams, 
    dem, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LongProfile -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr=D8.tif --streams=streams.tif --dem=dem.tif ^
-o=output.html --esri_pntr 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/long_profile.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 20, 2018</p>
<p><a name="LongProfileFromPoints"></a></p>
<a class="header" href="#longprofilefrompoints" id="longprofilefrompoints"><h1>LongProfileFromPoints</h1></a>
<p>Plots the longitudinal profiles from flow-paths initiating from a set of vector points.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--points           </td><td> Input vector points file</td></tr>
<tr><td>--dem              </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output HTML file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">long_profile_from_points(
    d8_pntr, 
    points, 
    dem, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LongProfileFromPoints -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --points=stream_head.shp ^
--dem=dem.tif -o=output.html --esri_pntr 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/long_profile_from_points.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: February 21, 2018</p>
<p><a name="RasterStreamsToVector"></a></p>
<a class="header" href="#rasterstreamstovector" id="rasterstreamstovector"><h1>RasterStreamsToVector</h1></a>
<p>This tool converts a raster stream file into a vector file. The user must specify: 1)
the name of the raster streams file, 2) the name of the D8 flow pointer file,
and 3) the name of the output vector file. Streams in the input raster streams
file are denoted by cells containing any positive, non-zero integer. A field in
the vector database file, called STRM_VAL, will correspond to this positive
integer value. The database file will also have a field for the length of each
link in the stream network. The flow pointer file must be calculated from a DEM with
all topographic depressions and flat areas removed and must be calculated using the
D8 flow pointer algorithm. The output vector will contain PolyLine features.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>-o, --output       </td><td> Output vector file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">raster_streams_to_vector(
    streams, 
    d8_pntr, 
    output, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RasterStreamsToVector -v ^
--wd=&quot;/path/to/data/&quot; --streams=streams.tif --d8_pntr=D8.tif ^
-o=output.shp
&gt;&gt;./whitebox_tools -r=RasterStreamsToVector -v ^
--wd=&quot;/path/to/data/&quot; --streams=streams.tif --d8_pntr=D8.tif ^
-o=output.shp --esri_pntr 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/raster_streams_to_vector.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: 24/09/2018</p>
<p><a name="RasterizeStreams"></a></p>
<a class="header" href="#rasterizestreams" id="rasterizestreams"><h1>RasterizeStreams</h1></a>
<p>Rasterizes vector streams based on Lindsay (2016) method.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--streams          </td><td> Input vector streams file</td></tr>
<tr><td>--base             </td><td> Input base raster file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--nodata           </td><td> Use NoData value for background?</td></tr>
<tr><td>--feature_id       </td><td> Use feature number as output value?</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">rasterize_streams(
    streams, 
    base, 
    output, 
    nodata=True, 
    feature_id=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RasterizeStreams -v ^
--wd=&quot;/path/to/data/&quot; --streams=streams.shp --base=raster.tif ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/rasterize_streams.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: March 11, 2018</p>
<p><a name="RemoveShortStreams"></a></p>
<a class="header" href="#removeshortstreams" id="removeshortstreams"><h1>RemoveShortStreams</h1></a>
<p>Removes short first-order streams from a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--min_length       </td><td> Minimum tributary length (in map units) used for network prunning</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">remove_short_streams(
    d8_pntr, 
    streams, 
    output, 
    min_length, 
    esri_pntr=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=RemoveShortStreams -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/remove_short_streams.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="ShreveStreamMagnitude"></a></p>
<a class="header" href="#shrevestreammagnitude" id="shrevestreammagnitude"><h1>ShreveStreamMagnitude</h1></a>
<p>Assigns the Shreve stream magnitude to each link in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">shreve_stream_magnitude(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=ShreveStreamMagnitude -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=ShreveStreamMagnitude -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/shreve_magnitude.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="StrahlerStreamOrder"></a></p>
<a class="header" href="#strahlerstreamorder" id="strahlerstreamorder"><h1>StrahlerStreamOrder</h1></a>
<p>Assigns the Strahler stream order to each link in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">strahler_stream_order(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StrahlerStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=StrahlerStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/strahler_order.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="StreamLinkClass"></a></p>
<a class="header" href="#streamlinkclass" id="streamlinkclass"><h1>StreamLinkClass</h1></a>
<p>Identifies the exterior/interior links and nodes in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">stream_link_class(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StreamLinkClass -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=StreamLinkClass -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/stream_link_class.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 12, 2017</p>
<p><a name="StreamLinkIdentifier"></a></p>
<a class="header" href="#streamlinkidentifier" id="streamlinkidentifier"><h1>StreamLinkIdentifier</h1></a>
<p>Assigns a unique identifier to each link in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">stream_link_identifier(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StreamLinkIdentifier -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=StreamLinkIdentifier -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/stream_link_id.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<p><a name="StreamLinkLength"></a></p>
<a class="header" href="#streamlinklength" id="streamlinklength"><h1>StreamLinkLength</h1></a>
<p>Estimates the length of each link (or tributary) in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--linkid           </td><td> Input raster streams link ID (or tributary ID) file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">stream_link_length(
    d8_pntr, 
    linkid, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StreamLinkLength -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --linkid=streamsID.tif ^
--dem=dem.tif -o=output.tif
&gt;&gt;./whitebox_tools ^
-r=StreamLinkLength -v --wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif ^
--linkid=streamsID.tif --dem=dem.tif -o=output.tif --esri_pntr ^
--zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/stream_link_length.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="StreamLinkSlope"></a></p>
<a class="header" href="#streamlinkslope" id="streamlinkslope"><h1>StreamLinkSlope</h1></a>
<p>Estimates the average slope of each link (or tributary) in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--linkid           </td><td> Input raster streams link ID (or tributary ID) file</td></tr>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">stream_link_slope(
    d8_pntr, 
    linkid, 
    dem, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StreamLinkSlope -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --linkid=streamsID.tif ^
--dem=dem.tif -o=output.tif
&gt;&gt;./whitebox_tools ^
-r=StreamLinkSlope -v --wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif ^
--linkid=streamsID.tif --dem=dem.tif -o=output.tif --esri_pntr ^
--zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/stream_link_slope.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 27, 2017</p>
<p><a name="StreamSlopeContinuous"></a></p>
<a class="header" href="#streamslopecontinuous" id="streamslopecontinuous"><h1>StreamSlopeContinuous</h1></a>
<p>Estimates the slope of each grid cell in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-i, --dem          </td><td> Input raster DEM file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">stream_slope_continuous(
    d8_pntr, 
    streams, 
    dem, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=StreamSlopeContinuous -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --linkid=streamsID.tif ^
--dem=dem.tif -o=output.tif
&gt;&gt;./whitebox_tools ^
-r=StreamSlopeContinuous -v --wd=&quot;/path/to/data/&quot; ^
--d8_pntr=D8.tif --streams=streamsID.tif --dem=dem.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/stream_slope_continuous.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: July 6, 2017</p>
<p><a name="TopologicalStreamOrder"></a></p>
<a class="header" href="#topologicalstreamorder" id="topologicalstreamorder"><h1>TopologicalStreamOrder</h1></a>
<p>Assigns each link in a stream network its topological order.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">topological_stream_order(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TopologicalStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=TopologicalStreamOrder -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/topological_stream_order.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 28, 2017</p>
<p><a name="TributaryIdentifier"></a></p>
<a class="header" href="#tributaryidentifier" id="tributaryidentifier"><h1>TributaryIdentifier</h1></a>
<p>Assigns a unique identifier to each tributary in a stream network.</p>
<p><em>Parameters</em>:</p>
<table><thead><tr><th><strong>Flag</strong>            </th><th>  <strong>Description</strong></th></tr></thead><tbody>
<tr><td>--d8_pntr          </td><td> Input raster D8 pointer file</td></tr>
<tr><td>--streams          </td><td> Input raster streams file</td></tr>
<tr><td>-o, --output       </td><td> Output raster file</td></tr>
<tr><td>--esri_pntr        </td><td> D8 pointer uses the ESRI style scheme</td></tr>
<tr><td>--zero_background  </td><td> Flag indicating whether a background value of zero should be used</td></tr>
</tbody></table>
<p><em>Python function</em>:</p>
<pre><code class="language-python">tributary_identifier(
    d8_pntr, 
    streams, 
    output, 
    esri_pntr=False, 
    zero_background=False, 
    callback=default_callback)
</code></pre>
<p><em>Command-line Interface</em>:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=TributaryIdentifier -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif
&gt;&gt;./whitebox_tools -r=TributaryIdentifier -v ^
--wd=&quot;/path/to/data/&quot; --d8_pntr=D8.tif --streams=streams.tif ^
-o=output.tif --esri_pntr --zero_background 
</code></pre>
<p><a href="https://github.com/jblindsay/whitebox-tools//tree/master/src/tools/stream_network_analysis/tributary_id.rs">Source code on GitHub</a></p>
<p><em>Author</em>: Dr. John Lindsay</p>
<p><em>Created</em>: June 25, 2017</p>
<a class="header" href="#tutorials" id="tutorials"><h1>Tutorials</h1></a>
<ul>
<li><a href="lidar.html">Processing LiDAR data</a></li>
</ul>
<a class="header" href="#processing-lidar-data" id="processing-lidar-data"><h1>Processing LiDAR data</h1></a>
<ul>
<li><a href="#grid_lidar">How do I convert a LAS point cloud into a raster?</a></li>
<li><a href="#classification">How do I exclude points with certain classifications?</a></li>
<li><a href="#folder">I have many LAS files and want to interpolate all of them at once</a></li>
<li><a href="#low_points">What if my data contains anomalously high/low points?</a></li>
<li><a href="#laz">My data are in LAZ format. How do I interpolate them?</a></li>
<li><a href="#intensity">How do I interpolate an image from the intensity data?</a></li>
<li><a href="#resolution">How do I decide on an appropriate grid resolution?</a></li>
<li><a href="#nodata_gaps">My raster contains NoData gaps. How do I remove these?</a></li>
<li><a href="#mosaic">How do I combine many LiDAR tiles into a single raster?</a></li>
<li><a href="#workflow">Is there a complete example LiDAR processing workflow available?</a></li>
</ul>
<p><a name="grid_lidar"></a></p>
<a class="header" href="#how-do-i-convert-a-las-point-cloud-into-a-raster" id="how-do-i-convert-a-las-point-cloud-into-a-raster"><h2>How do I convert a LAS point cloud into a raster?</h2></a>
<p>Converting your LiDAR data into a raster requires an interpolation operation. There are many such interpolation methods. The following is an example of how to interpolate the last-return points of a LAS (LiDAR) file using an inverse distance weighted (IDW) interpolation scheme, with a search window radius of 2.5 m, an exponent of 2.0, and an output grid resolution of 1.5 m.</p>
<pre><code class="language-python">from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
wbt.work_dir = &quot;/path/to/data/&quot;

wbt.lidar_idw_interpolation(
i=&quot;myFile.las&quot;,
output=&quot;myRaster.tif&quot;,
parameter=&quot;elevation&quot;,
returns=&quot;last&quot;,
resolution=1.5,
weight=2.0,
radius=2.5
)
</code></pre>
<p>Other methods for gridding a LAS file include nearest neighbour, Delaunay triangulation (TINing), block minimum, and block maximum gridding schemes.</p>
<p><a name="classification"></a></p>
<a class="header" href="#how-do-i-exclude-points-with-certain-classifications" id="how-do-i-exclude-points-with-certain-classifications"><h2>How do I exclude points with certain classifications?</h2></a></h2>
<p>It is commonly the case that points with certain class values should be excluded from the gridding of LiDAR data. For example, you may wish to exclude points associated with vegetation, buildings, bridges, utility lines, etc. The <em>LidarIdwInterpolation</em> and <em>LidarNearestNeighbourGridding</em> tools allow for excluded point classes using the <em>exclude_cls</em> parameter. The parameter takes a numeric list as input, e.g. <code>exclude_cls='3,4,5,6,7,18'</code>. Class values follow those of the LAS v.1.4 specifications:</p>
<p><em>LAS point classification values</em></p>
<table><thead><tr><th align="left"> Classification Value  </th><th align="left"> Meaning                              </th></tr></thead><tbody>
<tr><td align="left"> 0                     </td><td align="left"> Created never classified</td></tr>
<tr><td align="left"> 1                     </td><td align="left"> Unclassified3</td></tr>
<tr><td align="left"> 2                     </td><td align="left"> Ground</td></tr>
<tr><td align="left"> 3                     </td><td align="left"> Low Vegetation</td></tr>
<tr><td align="left"> 4                     </td><td align="left"> Medium Vegetation</td></tr>
<tr><td align="left"> 5                     </td><td align="left"> High Vegetation</td></tr>
<tr><td align="left"> 6                     </td><td align="left"> Building</td></tr>
<tr><td align="left"> 7                     </td><td align="left"> Low Point (noise)</td></tr>
<tr><td align="left"> 8                     </td><td align="left"> Reserved</td></tr>
<tr><td align="left"> 9                     </td><td align="left"> Water</td></tr>
<tr><td align="left"> 10                    </td><td align="left"> Rail</td></tr>
<tr><td align="left"> 11                    </td><td align="left"> Road Surface</td></tr>
<tr><td align="left"> 12                    </td><td align="left"> Reserved</td></tr>
<tr><td align="left"> 13                    </td><td align="left"> Wire – Guard (Shield)</td></tr>
<tr><td align="left"> 14                    </td><td align="left"> Wire – Conductor (Phase)</td></tr>
<tr><td align="left"> 15                    </td><td align="left"> Transmission Tower</td></tr>
<tr><td align="left"> 16                    </td><td align="left"> Wire-structure Connector (e.g. Insulator)</td></tr>
<tr><td align="left"> 17                    </td><td align="left"> Bridge Deck</td></tr>
<tr><td align="left"> 18                    </td><td align="left"> High Noise</td></tr>
</tbody></table>
<p>Of course, not all LAS files have had point classifications applied and stored. To determine whether your data contains point class
data, you can run the <em>LidarInfo</em> tool before interpolation.</p>
<p><a name="folder"></a></p>
<a class="header" href="#i-have-many-las-files-and-want-to-interpolate-all-of-them-at-once" id="i-have-many-las-files-and-want-to-interpolate-all-of-them-at-once"><h2>I have many LAS files and want to interpolate all of them at once</h2></a>
<p>When you have hundreds, or even thousands, of LAS files you might be inclined to write a Python script that calls the above function for each input file contained within a folder. But that isn't the best way to handle this common situation. Instead, if the input (i) and output parameters are left unspecified, each of WhiteboxTool's LiDAR gridding methods will interpolate all of the LAS files in the working directory, e.g.</p>
<pre><code class="language-python">from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
wbt.work_dir = &quot;/path/to/data/&quot;
wbt.lidar_idw_interpolation(
parameter=&quot;elevation&quot;,
returns=&quot;last&quot;,
resolution=1.0,
weight=1.0,
radius=2.5
)
</code></pre>
<p>Using this approach to folder-based interpolation has some advantages other than a greatly simplified script. WhiteboxTools will be able to parallelize the operation better, greatly improving the overall time required to interpolate the batch of files. Also, the gridding operations will be carried out with a strip of buffered data surrounding each LiDAR tile, i.e. there will be reduced edge-effects. This will reduce the potential for artifacts in the final mosaiced DEM.</p>
<p><a name="low_points"></a></p>
<a class="header" href="#what-if-my-data-contains-anomalously-highlow-points" id="what-if-my-data-contains-anomalously-highlow-points"><h2>What if my data contains anomalously high/low points?</h2></a>
<p>This is a fairly common problem with LiDAR data. If you're fortunate, these points, which often fall hundreds of meters above  or below the terrain surface, will be classified appropriately. When  this is the case, you may simply exclude the points with class values of 7 (low point) and 18 (high point). Alternatively, you may use the optional <code>minz</code> and <code>maxz</code> interpolation parameters to exclude unclassified outlier points. Lastly, you may remove these points from the original point cloud data set using the <em>LidarRemoveOutliers</em> tool.</p>
<p><a name="laz"></a></p>
<a class="header" href="#my-data-are-in-laz-format-how-do-i-interpolate-them" id="my-data-are-in-laz-format-how-do-i-interpolate-them"><h2>My data are in LAZ format. How do I interpolate them?</h2></a>
<p>WhiteboxTools does not currently support the compressed LiDAR format LAZ. To use these data, you will first need to decompress the files to a LAS format. You may wish to use LasTools for this purpose.</p>
<p><a name="intensity"></a></p>
<a class="header" href="#how-do-i-interpolate-an-image-from-the-intensity-data" id="how-do-i-interpolate-an-image-from-the-intensity-data"><h2>How do I interpolate an image from the intensity data?</h2></a>
<p>The <code>parameter</code> argument of the IDW and nearest neighbour interpolator tools allows you to interpolate intensity data (options include 'elevation', 'intensity', 'class', 'scan angle', and 'user data'). Here is an example:</p>
<pre><code class="language-python">from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
wbt.work_dir = &quot;/path/to/data/&quot;

wbt.lidar_nearest_neighbour_gridding(
&quot;in.las&quot;, &quot;out.tif&quot;, parameter=&quot;intensity&quot;)
</code></pre>
<p><a name="resolution"></a></p>
<a class="header" href="#how-do-i-decide-on-an-appropriate-grid-resolution" id="how-do-i-decide-on-an-appropriate-grid-resolution"><h2>How do I decide on an appropriate grid resolution?</h2></a>
<p>You want to choose a grid resolution where the vast majority of grid cells in the area covered by data have at least one return point. If you are interpolating with last-return points only, then this will necessarily reduce the potential resolution. Ultimately, there is not single appropriate value and the range of suitable resolutions will depend on the distribution of point density with the area of coverage. If the specified resolution is too high given the point density of the LiDAR data set, many of the grid cells will either be NoData holes, or represent interpolated values from relatively distant (up to the search radius) points. A higher than necessary grid resolution will also make working with the final mosaiced DEM raster more challenging, due to the computational effort needed to work with massive rasters and increase the storage and memory requirements. It is advisable to experiment with the <em>LidarPointDensity</em> and <em>LidarPointStats</em> tools before deciding upon a grid resolution for interpolation.</p>
<p><a name="nodata_gaps"></a></p>
<a class="header" href="#my-raster-contains-nodata-gaps-how-do-i-remove-these" id="my-raster-contains-nodata-gaps-how-do-i-remove-these"><h2>My raster contains NoData gaps. How do I remove these?</h2></a>
<p>First, we need to distinguish between two common areas of NoData values in the interpolated rasters of LiDAR data sets. Because LiDAR data are often collected for irregularly shaped sites, it is frequently the case that LiDAR DEMs have large NoData areas beyond the area of LiDAR point coverage. These are generally acceptable void areas and should not be altered. The more problemmatic void areas are interior data gaps (so called doughnut holes). These generally arise because the point density in an area of LiDAR coverage is lower than the grid resolution (and search radius) dictate in an area. Sometimes these NoData areas are associated with specific non-reflective surfaces, such as water, or areas of dense vegetation (and therefore the last return point density is far lower than in other areas). If the NoData gaps are extensive and spread throughout he area of coverage, that is a sign that you likely need to interpolate either with a coarser grid resolution or a larger search radius, or quite probably both. If your LiDAR DEM has a small number these void areas, and they are not extensive, then you may interpolate to remove the gaps using the <em>FillMissingData</em> tool:</p>
<pre><code class="language-python">from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
wbt.work_dir = &quot;/path/to/data/&quot;

wbt.fill_missing_data(&quot;dem.tif&quot;, &quot;new_dem.tif&quot;, filter=11)
</code></pre>
<p>The choice of a filter size will depend on the extent of the largest interior void area.</p>
<p><a name="mosaic"></a></p>
<a class="header" href="#how-do-i-combine-many-lidar-tiles-into-a-single-raster" id="how-do-i-combine-many-lidar-tiles-into-a-single-raster"><h2>How do I combine many LiDAR tiles into a single raster?</h2></a>
<p>Often you have many hundred LAS files, which you've intepolated into an equally large number of raster files. To combine these rasters into a single large DEM, use the Mosaic tool.</p>
<pre><code class="language-python">from os import listdir
from WBT.whitebox_tools import WhiteboxTools

wbt = WhiteboxTools()
wbt.work_dir = &quot;/path/to/data/&quot;

# find all GeoTIFFs in the path
input_files = &quot;&quot;
for f in listdir(wbt.work_dir):
    if f.endswith(&quot;.tif&quot;):
        input_files += &quot;;&quot; + f

input_files = input_files[1:]  # strips the first ';'

wbt.mosaic(inputs=input_files, output=&quot;big_DEM.tif&quot;)&lt;/code&gt;&lt;/pre&gt;
</code></pre>
<p><a name="workflow"></a></p>
<a class="header" href="#is-there-a-complete-example-lidar-processing-workflow-available" id="is-there-a-complete-example-lidar-processing-workflow-available"><h2>Is there a complete example LiDAR processing workflow available?</h2></a>
<p>Yes! The following code is an example of some of the common tasks required in processing large LiDAR datasets of many hundreds of LAS files.</p>
<pre><code class="language-python">import os
from os import path
from WBT.whitebox_tools import WhiteboxTools

def main():

    las_files_dir = &quot;/Users/johnlindsay/Documents/data/test_lidar/&quot;
    filtered_las_dir = &quot;/Users/johnlindsay/Documents/data/test_lidar/filtered_LAS/&quot;
    raster_data_dir = &quot;/Users/johnlindsay/Documents/data/test_lidar/interpolated_grids/&quot;

    wbt = WhiteboxTools()
    wbt.work_dir = las_files_dir #set working directory
    wbt.verbose = False
    if not os.path.exists(filtered_las_dir):
        os.makedirs(filtered_las_dir)

    # Sometimes, I like to extract all of the LAS tiles that overlap with a
    # particular area. For example, I might want to interpolate all the files
    # overlapping with a watershed. For this, you can use select_tiles_by_polygon.
    # Uncomment the four lines below if you want to do this.
    # outdir = &quot;/Users/johnlindsay/Documents/data/LAS_files_in_watershed/&quot;
    # polygons = &quot;/Users/johnlindsay/Documents/data/LAS_files_in_watershed/watershed.shp&quot;
    # wbt.select_tiles_by_polygon(las_files_dir, outdir, polygons)
    # las_files_dir = outdir # this way the analysis below works only on the selected tiles.

    ##################################################################################
    # Filter the ground points in the LAS files using lidar_ground_point_filter tool #
    ##################################################################################

    # This one is the SLOWEST part of the workflow and can be avoided if you are
    # confident that you have good point classification data, i.e. that the
    # vegetation and building classes have been properly populated.
    processed_files = []
    num_filtered = 1
    flag = True
    while flag:
        file_names = find_las_files(las_files_dir, processed_files)
        if len(file_names) &gt; 0: # and len(processed_files) &lt; 3000:
            for i in range(len(file_names)):
                in_file = las_files_dir + file_names[i]
                out_file = filtered_las_dir + file_names[i].replace(&quot;.las&quot;, &quot;_filtered.las&quot;)
                print(&quot;Processing LAS {} of {} (total filtered={}) {}&quot;.format(i+1, len(file_names), num_filtered, file_names[i]))
                processed_files.append(file_names[i])
                wbt.lidar_ground_point_filter(in_file, out_file, radius=2.0, slope_threshold=45, height_threshold=1.0)
                num_filtered += 1
        else:
            flag = False


    ##############################
    # Interpolate the LAS files. #
    ##############################

    # If you don't use the above ground point filter the directory below must be
    # updated to point to the original LAS files.
    wbt.work_dir = filtered_las_dir

    # You can use either IDW, nearest neighbour, or TINing (Delaunay triangulation)
    # for the gridding step. TINing option is available as of WhiteboxTools v0.11.
    wbt.lidar_idw_interpolation(parameter=&quot;elevation&quot;, returns=&quot;all&quot;, resolution=2.0, weight=1.0, radius=5.0, exclude_cls='3,4,5,6,7,18')
    # wbt.lidar_nearest_neighbour_gridding(returns=&quot;last&quot;, resolution=1.5, radius=2.5, exclude_cls='3,4,5,6,7,18')
    # wbt.lidar_tin_gridding(parameter=&quot;elevation&quot;, returns=&quot;all&quot;, resolution=2.0, exclude_cls='3,4,5,6,7,18')

    ###############################################################
    # Now mosaic the tiles; this is done using intermediate steps #
    ###############################################################
    if not os.path.exists(raster_data_dir):
        os.makedirs(raster_data_dir)

    wbt.work_dir = filtered_las_dir #set working directory
    wbt.verbose = False
    processed_files = []
    num_mosaiced = 1
    flag = True
    round = 1
    while flag:
        # This will mosaic a maximum of 250 tiles together; these sub-files
        # will subsequently be merged. Mosaicing many hundreds of tiles
        # together at one time is otherwise too intensive.
        file_names = find_tiff_files(filtered_las_dir, processed_files, 2)
        if len(file_names) &gt; 1:
            in_files = &quot;&quot;
            for i in range(len(file_names)):
                if i &lt; len(file_names)-1:
                    in_files += f&quot;{file_names[i]};&quot;
                else:
                    in_files += f&quot;{file_names[i]}&quot;

                processed_files.append(file_names[i])
                num_mosaiced += 1

            out_file = raster_data_dir + f&quot;mosaic{round}.tif&quot;
            wbt.mosaic(inputs=in_files, output=out_file, method=&quot;nn&quot;)
            print(f&quot;Processing mosaic {round}; num. files = {num_mosaiced}&quot;)

            # now clean up the individual tiles
            for i in range(len(file_names)):
                os.remove(filtered_las_dir + file_names[i])

        else:
            flag = False

        round += 1

    wbt.work_dir = raster_data_dir #set working directory
    mosaic_file = raster_data_dir + f&quot;final_mosaic.tif&quot;
    file_names = find_mosaic_files(raster_data_dir)
    if len(file_names) &gt; 1:
        in_files = &quot;&quot;
        for i in range(len(file_names)):
            if i &lt; len(file_names)-1:
                in_files += f&quot;{file_names[i]};&quot;
            else:
                in_files += f&quot;{file_names[i]}&quot;

            num_mosaiced += 1


        wbt.mosaic(inputs=in_files, output=mosaic_file, method=&quot;nn&quot;)

        # now clean up the intermediate mosaics
        for i in range(len(file_names)):
            os.remove(raster_data_dir + file_names[i])



    ##############################################
    # Would you like to fill in the NoData gaps? #
    ##############################################
    dem_nodata_filled = raster_data_dir + f&quot;DEM_gaps_filled.tif&quot;
    wbt.fill_missing_data(mosaic_file, dem_nodata_filled, filter=11)


    ######################################################################
    # I usually remove off-terrain objects, like any remaining buildings #
    ######################################################################
    dem_no_otos = raster_data_dir + f&quot;DEM_no_OTOs.tif&quot;
    wbt.remove_off_terrain_objects(dem_nodata_filled, dem_no_otos, filter=11, slope=15.0)


    #####################################
    # Would you like to smooth the DEM? #
    #####################################
    dem_smoothed = raster_data_dir + f&quot;DEM_smoothed.tif&quot;
    wbt.feature_preserving_denoise(dem_no_otos, dem_smoothed, filter=11, norm_diff=8.0)


    ################################
    # Want to fix the depressions? #
    ################################
    dem_breached = raster_data_dir + f&quot;DEM_breached.tif&quot;
    # Set the maximum breach depth appropriate for the terrain. You can
    # also restrict breaching based on a maximum breach channel length.
    wbt.breach_depressions(dem_smoothed, dem_breached, max_depth=5.0)

    # because we restricted the use of very deep breach channels, there
    # may still be depressions in the DEM. To get rid of these, we can
    # perform a subsequent depression filling operation.
    dem_filled = raster_data_dir + f&quot;DEM_filled.tif&quot;
    wbt.fill_depressions(dem_breached, dem_filled)


    ####################################################################
    # Okay, now we have a good base DEM from which we can extract      #
    # various land-surface parameters. There are really a large        #
    # number of these parameters available, but I'll just showcase     #
    # a few common ones here. See the User Manual for a complete list. #
    ####################################################################

    # slope
    slope_file = raster_data_dir + f&quot;slope.tif&quot;
    wbt.slope(dem_filled, slope_file)

    # plan curvature
    plan_curv_file = raster_data_dir + f&quot;plan_curv.tif&quot;
    wbt.plan_curvature(dem_filled, plan_curv_file)

    # profile curvature; other curvatures are available too.
    profile_curv_file = raster_data_dir + f&quot;profile_curv.tif&quot;
    wbt.profile_curvature(dem_filled, profile_curv_file)

    # hillshade (shaded relief raster)
    hillshade_file = raster_data_dir + f&quot;hillshade.tif&quot;
    wbt.hillshade(dem_filled, hillshade_file)

    # relative topographic position (RTP) index
    rtp_file = raster_data_dir + f&quot;relative_topographic_position.tif&quot;
    wbt.relative_topographic_position(dem_filled, rtp_file, filterx=11, filtery=11)

    # or even better, multiscale topographic position
    dev_max_mag = raster_data_dir + f&quot;multiscale_topo_position_mag.tif&quot;
    dev_max_scale = raster_data_dir + f&quot;multiscale_topo_position_scale.tif&quot;
    wbt.max_elevation_deviation(dem_filled, dev_max_mag, dev_max_scale, min_scale=1, max_scale=100, step=2)

    # ruggedness index
    ruggedness_index_file = raster_data_dir + f&quot;ruggedness_index.tif&quot;
    wbt.ruggedness_index(dem_filled, ruggedness_index_file)

    # or even better, multiscale roughness
    roughness_mag = raster_data_dir + f&quot;multiscale_roughness_mag.tif&quot;
    roughness_scale = raster_data_dir + f&quot;multiscale_roughness_scale.tif&quot;
    wbt.multiscale_roughness(dem_filled, roughness_mag, roughness_scale, min_scale=1, max_scale=100, step=2)

    # D-infinity flow accumulation
    flow_accum_file = raster_data_dir + f&quot;dinf_flow_accum.tif&quot;
    wbt.d_inf_flow_accumulation(dem_filled, flow_accum_file, log=True)

    # There literally hundreds of other useful parameters that could be
    # extracted from our DEM using WhiteboxTools. Take a look at the User Manual.


    print(&quot;Done!&quot;)

def find_las_files(input_dir, processed_files):
    files = os.listdir(input_dir)
    file_names = []
    for f in files:
        if f.endswith(&quot;.las&quot;) and f not in processed_files:
            file_names.append(f)

    return file_names


def find_tiff_files(input_dir, processed_files, max_num=10):
    files = os.listdir(input_dir)
    file_names = []
    for f in files:
        if f.endswith(&quot;.tif&quot;) and f not in processed_files:
            if len(file_names) &lt; max_num:
                file_names.append(f)
            else:
                break

    return file_names

def find_mosaic_files(input_dir):
    files = os.listdir(input_dir)
    file_names = []
    for f in files:
        if &quot;mosaic&quot; in f and (f.endswith(&quot;.tif&quot;) or f.endswith(&quot;.dep&quot;)):
            file_names.append(f)

    return file_names

main()
</code></pre>
<a class="header" href="#supported-data-formats" id="supported-data-formats"><h1>Supported Data Formats</h1></a>
<p>The <em>WhiteboxTools</em> library can currently support reading/writing raster data in GeoTIFF (.tif), <em>Whitebox GAT</em>(.tas and .dep), ESRI (ArcGIS) ASCII (.txt) and binary (.flt and .hdr), GRASS GIS, Idrisi (.rdc and .rst), SAGA GIS (binary--.sdat and .sgrd--and ASCII formats), and Surfer 7 (.grd) data formats. <em>The BigTIFF (64-bit) format is not currently supported.</em> The library is primarily tested using Whitebox raster and GeoTIFF data sets and if you encounter issues when reading/writing data in other formats, you should report the <a href="#reporting-bugs">issue</a>. Please note that there are no plans to incorporate third-party libraries, like <a href="http://www.gdal.org">GDAL</a>, in the project given the design goal of keeping a pure (or as close as possible) Rust codebase without third-party dependencies. This design greatly simplifies installation of the library.</p>
<p>Please note that throughout this manual code examples that manipulate raster files all use the GeoTIFF format (.tif) but any of the supported file extensions can be used in its place.</p>
<p>At present, there is limited ability in <em>WhiteboxTools</em> to work with vector geospatial data. Shapefiles geometries (.shp) and attributes (.dbf) can be read and some tools take vector inputs. There is currently no support for writing vector data although this feature is being actively developed. Other vector data formats may be added in the future.</p>
<p>LiDAR data can be read/written in the common <a href="https://www.asprs.org/committee-general/laser-las-file-format-exchange-activities.html">LAS</a> data format. <em>WhiteboxTools</em> can read and write LAS files that have been compressed (zipped with a .zip extension) using the common DEFLATE algorithm. Note that only LAS file should be contained within a zipped archive file. The compressed LiDAR format LAZ and ESRI LiDAR format are not currently supported by the library. The following is an example of running a LiDAR tool using zipped input/output files:</p>
<pre><code>&gt;&gt;./whitebox_tools -r=LidarTophatTransform -v --wd=&quot;/path/to/data/&quot;
-i=&quot;input.las.zip&quot; -o=&quot;output.las.zip&quot; --radius=10.0
</code></pre>
<p>Note that the double extensions (.las.zip) in the above command are not necessary and are only used for convenience of keeping track of LiDAR data sets (i.e. .zip extensions work too). The extra work of decoding/encoding compressed files does add additional processing time, although the Rust compression library that is used is highly efficient and usually only adds a few seconds to tool run times. Zipping LAS files frequently results 40-60% smaller binary files, making the additional processing time worthwhile for larger LAS file data sets with massive storage requirements.</p>
<a class="header" href="#contributing-and-reporting-bugs" id="contributing-and-reporting-bugs"><h1>Contributing and Reporting Bugs</h1></a>
<p>If you would like to contribute to the project as a developer, follow these instructions to get started:</p>
<ol>
<li>Fork the <em>WhiteboxTools</em> project ( https://github.com/jblindsay/whitebox-tools )</li>
<li>Create your feature branch (git checkout -b my-new-feature)</li>
<li>Commit your changes (git commit -am 'Add some feature')</li>
<li>Push to the branch (git push origin my-new-feature)</li>
<li>Create a new Pull Request</li>
</ol>
<p>Unless explicitly stated otherwise, any contribution intentionally submitted for inclusion in the work shall be licensed under the <a href="./license.html">MIT License</a> without any additional terms or conditions.</p>
<p>If you would like to contribute financial support for the project, please contact <a href="http://www.uoguelph.ca/%7Ehydrogeo/index.html">John Lindsay</a>. We also welcome contributions in the form of media exposure. If you have written an article or blog about <em>WhiteboxTools</em> please let us know about it.</p>
<p>WhiteboxTools is distributed as is and without warranty of suitability for application. If you encounter flaws with the software (i.e. bugs) please report the issue. Providing a detailed description of the conditions under which the bug occurred will help to identify the bug. <em>Use the Issues tracker on GitHub to report issues with the software and to request feature enchancements.</em> Please do not email Dr. Lindsay directly with bugs.</p>
<a class="header" href="#contributors" id="contributors"><h2>Contributors</h2></a>
<p>The following is a list of the developers who have helped to improve <em>WhiteboxTools</em> by contributing code.</p>
<ol>
<li>John Lindsay (jblindsay)</li>
<li>Dharhas Pothina (dharhas)</li>
<li>Qiusheng Wu (giswqs)</li>
<li>Daniel Newman</li>
<li>Anthony Francioni (Afrancioni)</li>
<li>Simon Gudim</li>
<li>Alexander Bruy (alexbruy)</li>
<li>Katrin Leinweber (katrinleinweber)</li>
<li>Yvan Le Bras (yvanlebras)</li>
</ol>
<p>If you feel you're missing from this list, please let us know. Also, the project is grateful to the numerous users that have contributed bug reports that have resulted in improvements to individual tools.</p>
<a class="header" href="#known-issues-and-limitations" id="known-issues-and-limitations"><h1>Known Issues and Limitations</h1></a>
<ul>
<li>The LAZ compressed LiDAR data format is currently unsupported although zipped LAS files (.zip) are.</li>
<li>There is no support for reading waveform data contained within or associated with LAS files.</li>
<li>File directories cannot contain apostrophes (', e.g. /John's data/) as they will be interpreted in the arguments array as single quoted strings.</li>
<li>The Python scripts included with <strong>WhiteboxTools</strong> require Python 3. They will not work with Python 2, which is frequently the default Python version installed on many systems.</li>
<li>Not all types of GeoTIFF files are supported. The GeoTIFF format is very flexible and certain less common varieties may not be read properly by WhiteboxTools. The WhiteboxTools GeoTIFF reader/writer is a custom implementation and does not rely on other library (e.g. GDAL) and therefore there may be difficulties when exchanging GeoTIFF data between WhiteboxTools and GDAL supported software.</li>
</ul>
<a class="header" href="#license" id="license"><h1>License</h1></a>
<p>The <strong>WhiteboxTools</strong> library is distributed under the <a href="LICENSE.txt">MIT license</a>, a permissive open-source (free software) license.</p>
<blockquote>
<p>The MIT License (MIT)</p>
<p>Copyright (c) 2017-2019 John Lindsay</p>
<p>Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the &quot;Software&quot;), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:</p>
<p>The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.</p>
<p>THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.</p>
</blockquote>
<a class="header" href="#frequently-asked-questions" id="frequently-asked-questions"><h1>Frequently Asked Questions</h1></a>
<a class="header" href="#do-i-need-whitebox-gat-to-use-whiteboxtools" id="do-i-need-whitebox-gat-to-use-whiteboxtools"><h2>Do I need Whitebox GAT to use WhiteboxTools?</h2></a>
<p>No you do not. You can call the tools contained within <em>WhiteboxTools</em> completely independent from the <em>Whitebox GAT</em> user interface using a Remote Procedure Call (RPC) approach. In fact, you can interact with the tools using Python scripting or directly, using a terminal application (command prompt). See <a href="#interacting-with-whiteboxtools-from-the-command-prompt"><em>Interacting With <em>WhiteboxTools</em> From the Command Prompt</em></a> for further details.</p>
<a class="header" href="#how-do-i-request-a-tool-be-added" id="how-do-i-request-a-tool-be-added"><h2>How do I request a tool be added?</h2></a>
<p>Eventually most of the tools in <em>Whitebox GAT</em> will be ported over to <em>WhiteboxTools</em> and all new tools will be added to this library as well. Naturally, this will take time. The order by which tools are ported is partly a function of ease of porting, existing infrastructure (i.e. raster and LiDAR tools will be ported first since their is currently no support in the library for vector I/O), and interest. If you are interested in making a tool a higher priority for porting, email <a href="http://www.uoguelph.ca/%7Ehydrogeo/index.html">John Lindsay</a>.</p>
<a class="header" href="#can-whiteboxtools-be-incorporated-into-other-software-and-open-source-gis-projects" id="can-whiteboxtools-be-incorporated-into-other-software-and-open-source-gis-projects"><h2>Can WhiteboxTools be incorporated into other software and open-source GIS projects?</h2></a>
<p><em>WhiteboxTools</em> was developed with the open-source GIS <a href="http://www.uoguelph.ca/%7Ehydrogeo/Whitebox/index.html">Whitebox GAT</a> in mind. That said, the tools can be accessed independently and so long as you abide by the terms of the <a href="#license">MIT license</a>, there is no reason why other software and GIS projects cannot use <em>WhiteboxTools</em> as well. In fact, this was one of the motivating factors for creating the library in the first place. Feel free to use <em>WhiteboxTools</em> as the geospatial analysis engine in your open-source software project.</p>
<a class="header" href="#what-platforms-does-whiteboxtools-support" id="what-platforms-does-whiteboxtools-support"><h2>What platforms does WhiteboxTools support?</h2></a>
<p><em>WhiteboxTools</em> is developed using the Rust programming language, which supports a <a href="https://forge.rust-lang.org/platform-support.html">wide variety of platforms</a> including MS Windows, MacOS, and Linux operating systems and common chip architectures. Interestingly, Rust also supports mobile platforms, and <em>WhiteboxTools</em> should therefore be capable of targeting (although no testing has been completed in this regard to date). Nearly all development and testing of the software is currently carried out on MacOS and we cannot guarantee a bug-free performance on other platforms. In particularly, MS Windows is the most different from the other platforms and is therefore the most likely to encounter platform-specific bugs. If you encounter bugs in the software, please consider reporting an issue using the GitHub support for issue-tracking.</p>
<a class="header" href="#what-are-the-system-requirements" id="what-are-the-system-requirements"><h2>What are the system requirements?</h2></a>
<p>The answer to this question depends strongly on the type of analysis and data that you intend to process. However, generally we find performance to be optimal with a recommended minimum of 8-16GB of memory (RAM), a modern multi-core processor (e.g. 64-bit i5 or i7), and an solid-state-drive (SSD). It is likely that <em>WhiteboxTools</em> will have satisfactory performance on lower-spec systems if smaller datasets are being processed. Because <em>WhiteboxTools</em> reads entire raster datasets into system memory (for optimal performance, and in recognition that modern systems have increasingly larger amounts of fast RAM), this tends to be the limiting factor for the upper-end of data size successfully processed by the library. 64-bit operating systems are recommended and extensive testing has not been carried out on 32-bit OSs. See <a href="#what-platforms-does-whiteboxtools-support"><strong>&quot;What platforms does WhiteboxTools support?&quot;</strong></a> for further details on supported platforms.</p>
<a class="header" href="#are-pre-compiled-executables-of-whiteboxtools-available" id="are-pre-compiled-executables-of-whiteboxtools-available"><h2>Are pre-compiled executables of WhiteboxTools available?</h2></a>
<p>Pre-compiled binaries for <em>WhiteboxTools</em> can be downloaded from the <a href="http://www.uoguelph.ca/%7Ehydrogeo/software.shtml#WhiteboxTools"><em>Geomorphometry and Hydrogeomatics Research Group</em></a> software web site for various supported operating systems. If you need binaries for other operating systems/system architectures, you will need to compile the executable from source files. See <a href="#installation">Installation</a> for details.</p>
<a class="header" href="#why-is-whiteboxtools-programmed-in-rust" id="why-is-whiteboxtools-programmed-in-rust"><h2>Why is WhiteboxTools programmed in Rust?</h2></a>
<p>I spent a long time evaluating potential programming language for future development efforts for the <em>Whitebox GAT</em> project. My most important criterion for a language was that it compile to native code, rather than target the Java virtual machine (JVM). I have been keen to move Whitebox GAT away from Java because of some of the challenges that supporting the JVM has included for many Whitebox users. The language should be fast and productive--Java is already quite fast, but if I am going to change development languages, I would like a performance boost. Furthermore, given that many, though not all, of the algorithms used for geospatial analysis scale well with concurrent (parallel) implementations, I favoured languages that offered easy and safe concurrent programming. Although many would consider C/C++ for this work, I was looking for a modern and safe language. Fortunately, we are living through a renaissance period in programming language development and there are many newer languages that fit the bill nicely. Over the past two years, I considered each of Go, Rust, D, Nim, and Crystal for Whitebox development and ultimately decided on Rust. [See <a href="https://github.com/jblindsay/go-spatial"><em>GoSpatial</em></a> and <a href="https://github.com/jblindsay/lidario"><em>lidario</em></a>.]</p>
<p>Each of the languages I examined has its own advantages of disadvantages, so why Rust? It's a combination of factors that made it a compelling option for this project. Compared with many on the list, Rust is a mature language with a vibrant user community. Like C/C++, it's a high-performance and low-level language that allows for complete control of the system. However, Rust is also one of the safest languages, meaning that I can be confident that <em>WhiteboxTools</em> will not contain common bugs, such as memory use-after-release, memory leaks and race conditions within concurrent code. Importantly, and quite uniquely, this safety is achieved in the Rust language without the use of a garbage collector (automatic memory management). Garbage collectors can be great, but they do generally come with a certain efficiency trade-off that Rust does not have. The other main advantage of Rust's approach to memory management is that it allows for  a level of interaction with scripting languages (e.g. Python) that is quite difficult to do in garbage collected languages. Although <strong>WhiteboxTools</strong> is currently set up to use an automation approach to interacting with Python code that calls it, I like the fact that I have the option to create a <em>WhiteboxTools</em> shared library.</p>
<p>Not everything with Rust is perfect however. It is still a very young language and there are many pieces still missing from its ecosystem. Furthermore, it is not the easiest language to learn, particularly for people who are inexperienced with programming. This may limit my ability to attract other programers to the Whitebox project, which would be unfortunate. However, overall, Rust was the best option for this particular application.</p>
<a class="header" href="#do-i-need-rust-installed-on-my-computer-to-run-whiteboxtools" id="do-i-need-rust-installed-on-my-computer-to-run-whiteboxtools"><h2>Do I need Rust installed on my computer to run WhiteboxTools?</h2></a>
<p>No, you would only need Rust installed if you were compiling the <em>WhiteboxTools</em> codebase from source files.</p>
<a class="header" href="#how-does-whiteboxtools-design-philosophy-differ" id="how-does-whiteboxtools-design-philosophy-differ"><h2>How does WhiteboxTools' design philosophy differ?</h2></a>
<p><em>Whitebox GAT</em> is frequently praised for its consistent design and ease of use. Like <em>Whitebox GAT</em>, <em>WhiteboxTools</em> follows the convention of <em>one tool for one function</em>. For example, in <em>WhiteboxTools</em> assigning the links in a stream channel network their Horton, Strahler, Shreve, or Hack stream ordering numbers requires running separate tools (i.e. <em>HortonStreamOrder</em>, <em>StrahlerStreamOrder</em>, <em>ShreveStreamMagnitude</em>, and <em>HackStreamOrder</em>). By contrast, in GRASS GIS<sup>1</sup> and ArcGIS single tools (i.e. the <em>r.stream.order</em> and <em>Stream Order</em> tools respectively) can be configured to output different channel ordering schemes. The <em>WhiteboxTools</em> design is intended to simplify the user experience and to make it easier to find the right tool for a task. With more specific tool names that are reflective of their specific purposes, users are not as reliant on reading help documentation to identify the tool for the task at hand. Similarly, it is not uncommon for tools in other GIS to have multiple outputs. For example, in GRASS GIS the <em>r.slope.aspect</em> tool can be configured to output slope, aspect, profile curvature, plan curvature, and several other common terrain surface derivatives. Based on the <em>one tool for one function</em> design approach of <em>WhiteboxTools</em>, multiple outputs are indicative that a tool should be split into different, more specific tools. Are you more likely to go to a tool named <em>r.slope.aspect</em> or <em>TangentialCurvature</em> when you want to create a tangential curvature raster from a DEM? If you're new to the software and are unfamiliar with it, probably the later is more obvious. The <em>WhiteboxTools</em> design approach also has the added benefit of simplifying the documentation for tools. The one downside to this design approach, however, is that it results (or will result) in a large number of tools, often with signifcant overlap in function.</p>
<p><sup>1</sup> NOTE: It is not my intent to criticize GRASS GIS, as I deeply respect the work that the GRASS developers have contributed. Rather, I am contrasting the consequences of <em>WhiteboxTools'</em> design philosophy to that of other GIS.</p>
<a class="header" href="#index-of-tools" id="index-of-tools"><h1>Index of Tools</h1></a>
<p><a href="available_tools/mathand_stats_tools.html#AbsoluteValue">AbsoluteValue</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#AdaptiveFilter">AdaptiveFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/data_tools.html#AddPointCoordinatesToTable">AddPointCoordinatesToTable</a> (Data Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Add">Add</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis.html#AggregateRaster">AggregateRaster</a> (GIS Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#And">And</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Anova">Anova</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ArcCos">ArcCos</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ArcSin">ArcSin</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ArcTan">ArcTan</a> (Math and Stats Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#Aspect">Aspect</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Atan2">Atan2</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#AttributeCorrelation">AttributeCorrelation</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#AttributeHistogram">AttributeHistogram</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#AttributeScattergram">AttributeScattergram</a> (Math and Stats Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#AverageFlowpathSlope">AverageFlowpathSlope</a> (Hydrological Analysis)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#AverageOverlay">AverageOverlay</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#AverageUpslopeFlowpathLength">AverageUpslopeFlowpathLength</a> (Hydrological Analysis)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#BalanceContrastEnhancement">BalanceContrastEnhancement</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/hydrological_analysis.html#Basins">Basins</a> (Hydrological Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#BilateralFilter">BilateralFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis.html#BlockMaximumGridding">BlockMaximumGridding</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#BlockMinimumGridding">BlockMinimumGridding</a> (GIS Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#BreachDepressions">BreachDepressions</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#BreachSingleCellPits">BreachSingleCellPits</a> (Hydrological Analysis)</p>
<p><a href="available_tools/gis_analysis_distance_tools.html#BufferRaster">BufferRaster</a> (GIS Analysis → Distance Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Ceil">Ceil</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis.html#CentroidVector">CentroidVector</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#Centroid">Centroid</a> (GIS Analysis)</p>
<p><a href="available_tools/image_processing_tools.html#ChangeVectorAnalysis">ChangeVectorAnalysis</a> (Image Processing Tools)</p>
<p><a href="available_tools/lidar_tools.html#ClassifyOverlapPoints">ClassifyOverlapPoints</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#ClipLidarToPolygon">ClipLidarToPolygon</a> (LiDAR Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#ClipRasterToPolygon">ClipRasterToPolygon</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#Clip">Clip</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/image_processing_tools.html#Closing">Closing</a> (Image Processing Tools)</p>
<p><a href="available_tools/gis_analysis.html#Clump">Clump</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#CompactnessRatio">CompactnessRatio</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#ConservativeSmoothingFilter">ConservativeSmoothingFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis.html#ConstructVectorTin">ConstructVectorTin</a> (GIS Analysis)</p>
<p><a href="available_tools/data_tools.html#ConvertNodataToZero">ConvertNodataToZero</a> (Data Tools)</p>
<p><a href="available_tools/data_tools.html#ConvertRasterFormat">ConvertRasterFormat</a> (Data Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#CornerDetection">CornerDetection</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#CorrectVignetting">CorrectVignetting</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/mathand_stats_tools.html#Cos">Cos</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Cosh">Cosh</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis_distance_tools.html#CostAllocation">CostAllocation</a> (GIS Analysis → Distance Tools)</p>
<p><a href="available_tools/gis_analysis_distance_tools.html#CostDistance">CostDistance</a> (GIS Analysis → Distance Tools)</p>
<p><a href="available_tools/gis_analysis_distance_tools.html#CostPathway">CostPathway</a> (GIS Analysis → Distance Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#CountIf">CountIf</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/image_processing_tools.html#CreateColourComposite">CreateColourComposite</a> (Image Processing Tools)</p>
<p><a href="available_tools/gis_analysis.html#CreateHexagonalVectorGrid">CreateHexagonalVectorGrid</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#CreatePlane">CreatePlane</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#CreateRectangularVectorGrid">CreateRectangularVectorGrid</a> (GIS Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#CrispnessIndex">CrispnessIndex</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#CrossTabulation">CrossTabulation</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#CumulativeDistribution">CumulativeDistribution</a> (Math and Stats Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#D8FlowAccumulation">D8FlowAccumulation</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#D8MassFlux">D8MassFlux</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#D8Pointer">D8Pointer</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#DInfFlowAccumulation">DInfFlowAccumulation</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#DInfMassFlux">DInfMassFlux</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#DInfPointer">DInfPointer</a> (Hydrological Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Decrement">Decrement</a> (Math and Stats Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#DepthInSink">DepthInSink</a> (Hydrological Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#DevFromMeanElev">DevFromMeanElev</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#DiffFromMeanElev">DiffFromMeanElev</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#DiffOfGaussianFilter">DiffOfGaussianFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#Difference">Difference</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#DirectDecorrelationStretch">DirectDecorrelationStretch</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/geomorphometric_analysis.html#DirectionalRelief">DirectionalRelief</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis.html#Dissolve">Dissolve</a> (GIS Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#DistanceToOutlet">DistanceToOutlet</a> (Stream Network Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#DiversityFilter">DiversityFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/mathand_stats_tools.html#Divide">Divide</a> (Math and Stats Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#DownslopeDistanceToStream">DownslopeDistanceToStream</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#DownslopeFlowpathLength">DownslopeFlowpathLength</a> (Hydrological Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#DownslopeIndex">DownslopeIndex</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#DrainagePreservingSmoothing">DrainagePreservingSmoothing</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#EdgePreservingMeanFilter">EdgePreservingMeanFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#EdgeProportion">EdgeProportion</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#ElevAbovePit">ElevAbovePit</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#ElevPercentile">ElevPercentile</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#ElevRelativeToMinMax">ElevRelativeToMinMax</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#ElevRelativeToWatershedMinMax">ElevRelativeToWatershedMinMax</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#ElevationAboveStreamEuclidean">ElevationAboveStreamEuclidean</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#ElevationAboveStream">ElevationAboveStream</a> (Hydrological Analysis)</p>
<p><a href="available_tools/gis_analysis.html#EliminateCoincidentPoints">EliminateCoincidentPoints</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#ElongationRatio">ElongationRatio</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#EmbossFilter">EmbossFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/mathand_stats_tools.html#EqualTo">EqualTo</a> (Math and Stats Tools)</p>
<p><a href="available_tools/lidar_tools.html#ErasePolygonFromLidar">ErasePolygonFromLidar</a> (LiDAR Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#ErasePolygonFromRaster">ErasePolygonFromRaster</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#Erase">Erase</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/gis_analysis_distance_tools.html#EuclideanAllocation">EuclideanAllocation</a> (GIS Analysis → Distance Tools)</p>
<p><a href="available_tools/gis_analysis_distance_tools.html#EuclideanDistance">EuclideanDistance</a> (GIS Analysis → Distance Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Exp2">Exp2</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Exp">Exp</a> (Math and Stats Tools)</p>
<p><a href="available_tools/data_tools.html#ExportTableToCsv">ExportTableToCsv</a> (Data Tools)</p>
<p><a href="available_tools/gis_analysis.html#ExtendVectorLines">ExtendVectorLines</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#ExtractNodes">ExtractNodes</a> (GIS Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#ExtractRasterStatistics">ExtractRasterStatistics</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis.html#ExtractRasterValuesAtPoints">ExtractRasterValuesAtPoints</a> (GIS Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#ExtractStreams">ExtractStreams</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#ExtractValleys">ExtractValleys</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#FarthestChannelHead">FarthestChannelHead</a> (Stream Network Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#FastAlmostGaussianFilter">FastAlmostGaussianFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/hydrological_analysis.html#Fd8FlowAccumulation">Fd8FlowAccumulation</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#Fd8Pointer">Fd8Pointer</a> (Hydrological Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#FeaturePreservingDenoise">FeaturePreservingDenoise</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#FetchAnalysis">FetchAnalysis</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#FillBurn">FillBurn</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#FillDepressions">FillDepressions</a> (Hydrological Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#FillMissingData">FillMissingData</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#FillSingleCellPits">FillSingleCellPits</a> (Hydrological Analysis)</p>
<p><a href="available_tools/lidar_tools.html#FilterLidarScanAngles">FilterLidarScanAngles</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#FindFlightlineEdgePoints">FindFlightlineEdgePoints</a> (LiDAR Tools)</p>
<p><a href="available_tools/gis_analysis.html#FindLowestOrHighestPoints">FindLowestOrHighestPoints</a> (GIS Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#FindMainStem">FindMainStem</a> (Stream Network Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#FindNoFlowCells">FindNoFlowCells</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#FindParallelFlow">FindParallelFlow</a> (Hydrological Analysis)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#FindPatchOrClassEdgeCells">FindPatchOrClassEdgeCells</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#FindRidges">FindRidges</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#FlattenLakes">FlattenLakes</a> (Hydrological Analysis)</p>
<p><a href="available_tools/lidar_tools.html#FlightlineOverlap">FlightlineOverlap</a> (LiDAR Tools)</p>
<p><a href="available_tools/image_processing_tools.html#FlipImage">FlipImage</a> (Image Processing Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#FloodOrder">FloodOrder</a> (Hydrological Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Floor">Floor</a> (Math and Stats Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#FlowAccumulationFullWorkflow">FlowAccumulationFullWorkflow</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#FlowLengthDiff">FlowLengthDiff</a> (Hydrological Analysis)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#GammaCorrection">GammaCorrection</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#GaussianContrastStretch">GaussianContrastStretch</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/image_processing_tools_filters.html#GaussianFilter">GaussianFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/mathand_stats_tools.html#GreaterThan">GreaterThan</a> (Math and Stats Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#HackStreamOrder">HackStreamOrder</a> (Stream Network Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#HighPassFilter">HighPassFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools_filters.html#HighPassMedianFilter">HighPassMedianFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#HighestPosition">HighestPosition</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#Hillshade">Hillshade</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#Hillslopes">Hillslopes</a> (Hydrological Analysis)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#HistogramEqualization">HistogramEqualization</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#HistogramMatchingTwoImages">HistogramMatchingTwoImages</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#HistogramMatching">HistogramMatching</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#HoleProportion">HoleProportion</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#HorizonAngle">HorizonAngle</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#HortonStreamOrder">HortonStreamOrder</a> (Stream Network Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#HypsometricAnalysis">HypsometricAnalysis</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis.html#IdwInterpolation">IdwInterpolation</a> (GIS Analysis)</p>
<p><a href="available_tools/image_processing_tools.html#IhsToRgb">IhsToRgb</a> (Image Processing Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ImageAutocorrelation">ImageAutocorrelation</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ImageCorrelation">ImageCorrelation</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ImageRegression">ImageRegression</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools.html#ImageStackProfile">ImageStackProfile</a> (Image Processing Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#ImpoundmentIndex">ImpoundmentIndex</a> (Hydrological Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#InPlaceAdd">InPlaceAdd</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#InPlaceDivide">InPlaceDivide</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#InPlaceMultiply">InPlaceMultiply</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#InPlaceSubtract">InPlaceSubtract</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Increment">Increment</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#IntegerDivision">IntegerDivision</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools.html#IntegralImage">IntegralImage</a> (Image Processing Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#Intersect">Intersect</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#IsNoData">IsNoData</a> (Math and Stats Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#Isobasins">Isobasins</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#JensonSnapPourPoints">JensonSnapPourPoints</a> (Hydrological Analysis)</p>
<p><a href="available_tools/data_tools.html#JoinTables">JoinTables</a> (Data Tools)</p>
<p><a href="available_tools/image_processing_tools.html#KMeansClustering">KMeansClustering</a> (Image Processing Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#KNearestMeanFilter">KNearestMeanFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/mathand_stats_tools.html#KappaIndex">KappaIndex</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#KsTestForNormality">KsTestForNormality</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#LaplacianFilter">LaplacianFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools_filters.html#LaplacianOfGaussianFilter">LaplacianOfGaussianFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/lidar_tools.html#LasToAscii">LasToAscii</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LasToMultipointShapefile">LasToMultipointShapefile</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LasToShapefile">LasToShapefile</a> (LiDAR Tools)</p>
<p><a href="available_tools/gis_analysis.html#LayerFootprint">LayerFootprint</a> (GIS Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#LeeFilter">LeeFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/stream_network_analysis.html#LengthOfUpstreamChannels">LengthOfUpstreamChannels</a> (Stream Network Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#LessThan">LessThan</a> (Math and Stats Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarBlockMaximum">LidarBlockMaximum</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarBlockMinimum">LidarBlockMinimum</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarClassifySubset">LidarClassifySubset</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarColourize">LidarColourize</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarConstructVectorTin">LidarConstructVectorTin</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarElevationSlice">LidarElevationSlice</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarGroundPointFilter">LidarGroundPointFilter</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarHexBinning">LidarHexBinning</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarHillshade">LidarHillshade</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarHistogram">LidarHistogram</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarIdwInterpolation">LidarIdwInterpolation</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarInfo">LidarInfo</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarJoin">LidarJoin</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarKappaIndex">LidarKappaIndex</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarNearestNeighbourGridding">LidarNearestNeighbourGridding</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarPointDensity">LidarPointDensity</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarPointStats">LidarPointStats</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarRemoveDuplicates">LidarRemoveDuplicates</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarRemoveOutliers">LidarRemoveOutliers</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarSegmentationBasedFilter">LidarSegmentationBasedFilter</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarSegmentation">LidarSegmentation</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarThinHighDensity">LidarThinHighDensity</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarThin">LidarThin</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarTileFootprint">LidarTileFootprint</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarTile">LidarTile</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarTinGridding">LidarTinGridding</a> (LiDAR Tools)</p>
<p><a href="available_tools/lidar_tools.html#LidarTophatTransform">LidarTophatTransform</a> (LiDAR Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#LineDetectionFilter">LineDetectionFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#LineIntersections">LineIntersections</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/image_processing_tools.html#LineThinning">LineThinning</a> (Image Processing Tools)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#LinearityIndex">LinearityIndex</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/data_tools.html#LinesToPolygons">LinesToPolygons</a> (Data Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ListUniqueValues">ListUniqueValues</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Ln">Ln</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Log10">Log10</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Log2">Log2</a> (Math and Stats Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#LongProfileFromPoints">LongProfileFromPoints</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#LongProfile">LongProfile</a> (Stream Network Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#LongestFlowpath">LongestFlowpath</a> (Hydrological Analysis)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#LowestPosition">LowestPosition</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#MajorityFilter">MajorityFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#MaxAbsoluteOverlay">MaxAbsoluteOverlay</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MaxAnisotropyDevSignature">MaxAnisotropyDevSignature</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MaxAnisotropyDev">MaxAnisotropyDev</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MaxBranchLength">MaxBranchLength</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MaxDifferenceFromMean">MaxDifferenceFromMean</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MaxDownslopeElevChange">MaxDownslopeElevChange</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MaxElevDevSignature">MaxElevDevSignature</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MaxElevationDeviation">MaxElevationDeviation</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#MaxOverlay">MaxOverlay</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#MaxUpslopeFlowpathLength">MaxUpslopeFlowpathLength</a> (Hydrological Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Max">Max</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#MaximumFilter">MaximumFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools_filters.html#MeanFilter">MeanFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools_filters.html#MedianFilter">MedianFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis.html#Medoid">Medoid</a> (GIS Analysis)</p>
<p><a href="available_tools/data_tools.html#MergeTableWithCsv">MergeTableWithCsv</a> (Data Tools)</p>
<p><a href="available_tools/data_tools.html#MergeVectors">MergeVectors</a> (Data Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#MinAbsoluteOverlay">MinAbsoluteOverlay</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MinDownslopeElevChange">MinDownslopeElevChange</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#MinMaxContrastStretch">MinMaxContrastStretch</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#MinOverlay">MinOverlay</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Min">Min</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis.html#MinimumBoundingBox">MinimumBoundingBox</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#MinimumBoundingCircle">MinimumBoundingCircle</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#MinimumBoundingEnvelope">MinimumBoundingEnvelope</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#MinimumConvexHull">MinimumConvexHull</a> (GIS Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#MinimumFilter">MinimumFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools.html#ModifiedKMeansClustering">ModifiedKMeansClustering</a> (Image Processing Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Modulo">Modulo</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools.html#MosaicWithFeathering">MosaicWithFeathering</a> (Image Processing Tools)</p>
<p><a href="available_tools/image_processing_tools.html#Mosaic">Mosaic</a> (Image Processing Tools)</p>
<p><a href="available_tools/data_tools.html#MultiPartToSinglePart">MultiPartToSinglePart</a> (Data Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Multiply">Multiply</a> (Math and Stats Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MultiscaleRoughnessSignature">MultiscaleRoughnessSignature</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MultiscaleRoughness">MultiscaleRoughness</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#MultiscaleTopographicPositionImage">MultiscaleTopographicPositionImage</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis.html#NearestNeighbourGridding">NearestNeighbourGridding</a> (GIS Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Negate">Negate</a> (Math and Stats Tools)</p>
<p><a href="available_tools/data_tools.html#NewRasterFromBase">NewRasterFromBase</a> (Data Tools)</p>
<p><a href="available_tools/lidar_tools.html#NormalVectors">NormalVectors</a> (LiDAR Tools)</p>
<p><a href="available_tools/image_processing_tools.html#NormalizedDifferenceVegetationIndex">NormalizedDifferenceVegetationIndex</a> (Image Processing Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#NotEqualTo">NotEqualTo</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Not">Not</a> (Math and Stats Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#NumDownslopeNeighbours">NumDownslopeNeighbours</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#NumInflowingNeighbours">NumInflowingNeighbours</a> (Hydrological Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#NumUpslopeNeighbours">NumUpslopeNeighbours</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#OlympicFilter">OlympicFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools.html#Opening">Opening</a> (Image Processing Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Or">Or</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#PanchromaticSharpening">PanchromaticSharpening</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#PatchOrientation">PatchOrientation</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#PennockLandformClass">PennockLandformClass</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#PercentElevRange">PercentElevRange</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#PercentEqualTo">PercentEqualTo</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#PercentGreaterThan">PercentGreaterThan</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#PercentLessThan">PercentLessThan</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#PercentageContrastStretch">PercentageContrastStretch</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/image_processing_tools_filters.html#PercentileFilter">PercentileFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#PerimeterAreaRatio">PerimeterAreaRatio</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#PickFromList">PickFromList</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#PlanCurvature">PlanCurvature</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis.html#PolygonArea">PolygonArea</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#PolygonLongAxis">PolygonLongAxis</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#PolygonPerimeter">PolygonPerimeter</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#PolygonShortAxis">PolygonShortAxis</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#Polygonize">Polygonize</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/data_tools.html#PolygonsToLines">PolygonsToLines</a> (Data Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Power">Power</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#PrewittFilter">PrewittFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/mathand_stats_tools.html#PrincipalComponentAnalysis">PrincipalComponentAnalysis</a> (Math and Stats Tools)</p>
<p><a href="available_tools/data_tools.html#PrintGeoTiffTags">PrintGeoTiffTags</a> (Data Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#ProfileCurvature">ProfileCurvature</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#Profile">Profile</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Quantiles">Quantiles</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#RadiusOfGyration">RadiusOfGyration</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#RaiseWalls">RaiseWalls</a> (Hydrological Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#RandomField">RandomField</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#RandomSample">RandomSample</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools_filters.html#RangeFilter">RangeFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis.html#RasterCellAssignment">RasterCellAssignment</a> (GIS Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#RasterHistogram">RasterHistogram</a> (Math and Stats Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#RasterStreamsToVector">RasterStreamsToVector</a> (Stream Network Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#RasterSummaryStats">RasterSummaryStats</a> (Math and Stats Tools)</p>
<p><a href="available_tools/data_tools.html#RasterToVectorLines">RasterToVectorLines</a> (Data Tools)</p>
<p><a href="available_tools/data_tools.html#RasterToVectorPoints">RasterToVectorPoints</a> (Data Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#RasterizeStreams">RasterizeStreams</a> (Stream Network Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Reciprocal">Reciprocal</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis.html#ReclassEqualInterval">ReclassEqualInterval</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#ReclassFromFile">ReclassFromFile</a> (GIS Analysis)</p>
<p><a href="available_tools/gis_analysis.html#Reclass">Reclass</a> (GIS Analysis)</p>
<p><a href="available_tools/data_tools.html#ReinitializeAttributeTable">ReinitializeAttributeTable</a> (Data Tools)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#RelatedCircumscribingCircle">RelatedCircumscribingCircle</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#RelativeAspect">RelativeAspect</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#RelativeStreamPowerIndex">RelativeStreamPowerIndex</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#RelativeTopographicPosition">RelativeTopographicPosition</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#RemoveOffTerrainObjects">RemoveOffTerrainObjects</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/data_tools.html#RemovePolygonHoles">RemovePolygonHoles</a> (Data Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#RemoveShortStreams">RemoveShortStreams</a> (Stream Network Analysis)</p>
<p><a href="available_tools/image_processing_tools.html#RemoveSpurs">RemoveSpurs</a> (Image Processing Tools)</p>
<p><a href="available_tools/image_processing_tools.html#Resample">Resample</a> (Image Processing Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#RescaleValueRange">RescaleValueRange</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools.html#RgbToIhs">RgbToIhs</a> (Image Processing Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#Rho8Pointer">Rho8Pointer</a> (Hydrological Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#RobertsCrossFilter">RobertsCrossFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/mathand_stats_tools.html#RootMeanSquareError">RootMeanSquareError</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Round">Round</a> (Math and Stats Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#RuggednessIndex">RuggednessIndex</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#ScharrFilter">ScharrFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/geomorphometric_analysis.html#SedimentTransportIndex">SedimentTransportIndex</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/lidar_tools.html#SelectTilesByPolygon">SelectTilesByPolygon</a> (LiDAR Tools)</p>
<p><a href="available_tools/data_tools.html#SetNodataValue">SetNodataValue</a> (Data Tools)</p>
<p><a href="available_tools/gis_analysis_patch_shape_tools.html#ShapeComplexityIndex">ShapeComplexityIndex</a> (GIS Analysis → Patch Shape Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#ShreveStreamMagnitude">ShreveStreamMagnitude</a> (Stream Network Analysis)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#SigmoidalContrastStretch">SigmoidalContrastStretch</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/mathand_stats_tools.html#Sin">Sin</a> (Math and Stats Tools)</p>
<p><a href="available_tools/data_tools.html#SinglePartToMultiPart">SinglePartToMultiPart</a> (Data Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Sinh">Sinh</a> (Math and Stats Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#Sink">Sink</a> (Hydrological Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#SlopeVsElevationPlot">SlopeVsElevationPlot</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#Slope">Slope</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis.html#SmoothVectors">SmoothVectors</a> (GIS Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#SnapPourPoints">SnapPourPoints</a> (Hydrological Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#SobelFilter">SobelFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools.html#SplitColourComposite">SplitColourComposite</a> (Image Processing Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#SplitWithLines">SplitWithLines</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#SquareRoot">SquareRoot</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Square">Square</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools_image_enhancement.html#StandardDeviationContrastStretch">StandardDeviationContrastStretch</a> (Image Processing Tools → Image Enhancement)</p>
<p><a href="available_tools/image_processing_tools_filters.html#StandardDeviationFilter">StandardDeviationFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/geomorphometric_analysis.html#StandardDeviationOfSlope">StandardDeviationOfSlope</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#StochasticDepressionAnalysis">StochasticDepressionAnalysis</a> (Hydrological Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#StrahlerOrderBasins">StrahlerOrderBasins</a> (Hydrological Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#StrahlerStreamOrder">StrahlerStreamOrder</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#StreamLinkClass">StreamLinkClass</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#StreamLinkIdentifier">StreamLinkIdentifier</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#StreamLinkLength">StreamLinkLength</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#StreamLinkSlope">StreamLinkSlope</a> (Stream Network Analysis)</p>
<p><a href="available_tools/stream_network_analysis.html#StreamSlopeContinuous">StreamSlopeContinuous</a> (Stream Network Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#Subbasins">Subbasins</a> (Hydrological Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Subtract">Subtract</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#SumOverlay">SumOverlay</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#SymmetricalDifference">SymmetricalDifference</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Tan">Tan</a> (Math and Stats Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#TangentialCurvature">TangentialCurvature</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Tanh">Tanh</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools.html#ThickenRasterLine">ThickenRasterLine</a> (Image Processing Tools)</p>
<p><a href="available_tools/gis_analysis.html#TinGridding">TinGridding</a> (GIS Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#ToDegrees">ToDegrees</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ToRadians">ToRadians</a> (Math and Stats Tools)</p>
<p><a href="available_tools/image_processing_tools.html#TophatTransform">TophatTransform</a> (Image Processing Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#TopologicalStreamOrder">TopologicalStreamOrder</a> (Stream Network Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#TotalCurvature">TotalCurvature</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#TotalFilter">TotalFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/hydrological_analysis.html#TraceDownslopeFlowpaths">TraceDownslopeFlowpaths</a> (Hydrological Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#TrendSurfaceVectorPoints">TrendSurfaceVectorPoints</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#TrendSurface">TrendSurface</a> (Math and Stats Tools)</p>
<p><a href="available_tools/stream_network_analysis.html#TributaryIdentifier">TributaryIdentifier</a> (Stream Network Analysis)</p>
<p><a href="available_tools/mathand_stats_tools.html#Truncate">Truncate</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#TurningBandsSimulation">TurningBandsSimulation</a> (Math and Stats Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#Union">Union</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/hydrological_analysis.html#UnnestBasins">UnnestBasins</a> (Hydrological Analysis)</p>
<p><a href="available_tools/image_processing_tools_filters.html#UnsharpMasking">UnsharpMasking</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/image_processing_tools_filters.html#UserDefinedWeightsFilter">UserDefinedWeightsFilter</a> (Image Processing Tools → Filters)</p>
<p><a href="available_tools/gis_analysis.html#VectorHexBinning">VectorHexBinning</a> (GIS Analysis)</p>
<p><a href="available_tools/data_tools.html#VectorLinesToRaster">VectorLinesToRaster</a> (Data Tools)</p>
<p><a href="available_tools/data_tools.html#VectorPointsToRaster">VectorPointsToRaster</a> (Data Tools)</p>
<p><a href="available_tools/data_tools.html#VectorPolygonsToRaster">VectorPolygonsToRaster</a> (Data Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#Viewshed">Viewshed</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/geomorphometric_analysis.html#VisibilityIndex">VisibilityIndex</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/gis_analysis.html#VoronoiDiagram">VoronoiDiagram</a> (GIS Analysis)</p>
<p><a href="available_tools/hydrological_analysis.html#Watershed">Watershed</a> (Hydrological Analysis)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#WeightedOverlay">WeightedOverlay</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/gis_analysis_overlay_tools.html#WeightedSum">WeightedSum</a> (GIS Analysis → Overlay Tools)</p>
<p><a href="available_tools/geomorphometric_analysis.html#WetnessIndex">WetnessIndex</a> (Geomorphometric Analysis)</p>
<p><a href="available_tools/image_processing_tools.html#WriteFunctionMemoryInsertion">WriteFunctionMemoryInsertion</a> (Image Processing Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#Xor">Xor</a> (Math and Stats Tools)</p>
<p><a href="available_tools/mathand_stats_tools.html#ZScores">ZScores</a> (Math and Stats Tools)</p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                        

                        

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                

                
            </nav>

        </div>

        

        

        

        
        <script src="elasticlunr.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="mark.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="searcher.js" type="text/javascript" charset="utf-8"></script>
        

        <script src="clipboard.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="highlight.js" type="text/javascript" charset="utf-8"></script>
        <script src="book.js" type="text/javascript" charset="utf-8"></script>

        <!-- Custom JS scripts -->
        

        
        
        <script type="text/javascript">
        window.addEventListener('load', function() {
            window.setTimeout(window.print, 100);
        });
        </script>
        
        

    </body>
</html>
